*** setup.py~	Wed Jun 22 17:38:46 2016
--- setup.py	Wed Sep 28 09:56:25 2016
***************
*** 74,79 ****
      packages=find_packages(),
      package_data={'pdbfixer': find_package_data()},
      zip_safe=False,
!     install_requires=['numpy', 'openmm >= 7.0'],
      entry_points={'console_scripts': ['pdbfixer = pdbfixer.pdbfixer:main']})
  
--- 74,79 ----
      packages=find_packages(),
      package_data={'pdbfixer': find_package_data()},
      zip_safe=False,
!     install_requires=['numpy'],
      entry_points={'console_scripts': ['pdbfixer = pdbfixer.pdbfixer:main']})
  
