<html>

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Command: alphafold</title>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRight" title="User Guide Index"/></a>

<h3><a href="../index.html#commands">Command</a>: alphafold</h3>
<p>
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold</a>
is an artificial intelligence method for predicting protein structures
that has been highly successful in
<a href="https://en.wikipedia.org/wiki/AlphaFold#Competitions"
target="_blank">recent tests</a>.
The method is described in:
</p>
<blockquote>
<a href="https://www.nature.com/articles/s41586-021-03819-2" target="_blank">
Highly accurate protein structure prediction with AlphaFold.</a>
Jumper J, Evans R, Pritzel A, <i>et al.</i>
<i>Nature.</i> 2021 Aug;596(7873):583-589.
<br>
<a name="alphafold-multimer"></a>
<a href="https://www.biorxiv.org/content/10.1101/2021.10.04.463034"
target="_blank">Protein complex prediction with AlphaFold-Multimer.</a>
Evans R, O'Neill M, Pritzel A, <i>et al.</i> bioRxiv 2021.
</blockquote>
<p>
The <b>alphafold</b> command:
</p>
<ul>
<li><a href="#fetch">finds and retrieves existing models</a>
from the <a href="https://alphafold.ebi.ac.uk/"
target="_blank">AlphaFold Database</a>:
</p>
<blockquote>
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8728224/" target="_blank">
AlphaFold Protein Structure Database: massively expanding the structural coverage of protein-sequence space with high-accuracy models.</a>
Varadi M, Anyango S, Deshpande M, <i>et al.</i>
<i>Nucleic Acids Res.</i> 2022 Jan 7;50(D1):D439-D444.
</blockquote>
<p>
The database contains models for protein sequences
in <a href="https://www.uniprot.org" target="_blank">UniProt</a>
(single chains only, not complexes); 
see the <a href="#version"><b>version</b></a> option.
</p>
<li><a href="#predict">runs new AlphaFold predictions</a>
on <a href="https://colab.research.google.com/signup" target="_blank">Google
Colab</a> using <b>ColabFold</b>, an open-source, optimized version of
<a href="https://en.wikipedia.org/wiki/AlphaFold#AlphaFold_2,_2020"
target="_blank">AlphaFold 2</a>:
<blockquote>
<a href="https://www.nature.com/articles/s41592-022-01488-1"
target="_blank">ColabFold: making protein folding accessible to all.</a>
Mirdita M, Sch&uuml;tze K, Moriwaki Y, Heo L, Ovchinnikov S, Steinegger M.
<i>Nat Methods.</i> 2022 Jun;19(6):679-682.
</blockquote>
<li><a href="#pae">plots residue-residue alignment errors</a>
for AlphaFold structures and 
<a href="#contacts">shows them with colored pseudobonds</a>
</ul>
<p>
AlphaFold-predicted structures vary in confidence levels
(see <a href="#colorConfidence">coloring</a>) 
and should be interpreted with caution.
The <b>alphafold</b> command is also implemented as the tools
<a href="../tools/alphafold.html"><b>AlphaFold</b></a>
and <a href="../tools/alphafold.html#pae"><b>AlphaFold Error Plot</b></a>.
Several ChimeraX
<a href="https://www.rbvi.ucsf.edu/chimerax/docs/presentations.html"
target="_blank">presentations</a> and
<a href="https://www.rbvi.ucsf.edu/chimerax/docs/videos/"
target="_blank">videos</a> show modeling with AlphaFold and related analyses.
See also:
<a href="esmfold.html"><b>esmfold</b></a>,
<a href="blastprotein.html"><b>blastprotein</b></a>,
<a href="modeller.html"><b>modeller</b></a>,
<a href="swapaa.html"><b>swapaa</b></a>
</p>

<a name="fetch"></a>
<a name="match"></a>
<a name="search"></a>
<p class="nav">
[<a href="#top">back to top: alphafold</a>]
</p>
<h3>Getting Models from the AlphaFold Database</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>alphafold fetch</b> &nbsp;<i>uniprot-id</i>&nbsp;
<span class="nowrap">[&nbsp;<a href="#alignTo"><b>alignTo</b></a>&nbsp;&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>&nbsp;[&nbsp;<a href="#trim"><b>trim</b></a>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]]</span>
[&nbsp;<a href="#colorConfidence"><b>colorConfidence</b></a>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<a href="#ignoreCache"><b>ignoreCache</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#pae"><b>pae</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#version"><b>version</b></a>&nbsp;&nbsp;1&nbsp;|&nbsp;2&nbsp;|&nbsp;3&nbsp;|&nbsp;<b>4</b>&nbsp;]
<br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>alphafold match</b> &nbsp;<i>sequence</i>&nbsp;
[&nbsp;<a href="#search-option"><b>search</b></a>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<a href="#trim"><b>trim</b></a>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<a href="#colorConfidence"><b>colorConfidence</b></a>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<a href="#ignoreCache"><b>ignoreCache</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#pae"><b>pae</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
<br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>alphafold search</b> &nbsp;<i>sequence</i>&nbsp;
[&nbsp;<b>matrix</b>&nbsp;&nbsp;<i>similarity-matrix</i>&nbsp;]
[&nbsp;<b>cutoff</b>&nbsp;&nbsp;<i>evalue</i>&nbsp;]
[&nbsp;<b>maxSequences</b>&nbsp;&nbsp;<i>M</i>&nbsp;]
[&nbsp;<a href="#version"><b>version</b></a>&nbsp;&nbsp;1&nbsp;|&nbsp;2&nbsp;|&nbsp;3&nbsp;|&nbsp;<b>4</b>&nbsp;]
</blockquote>
<ul>
<li>The <b>alphafold fetch</b> command retrieves the model (if available)
for a specific UniProt name or accession number. It is equivalent to using the
<b><a href="open.html#fetch">open</a></b> command to fetch
<b>from alphafold</b>. Examples:
<blockquote>
<b>alphafold fetch p29474</b>
<br>&ndash; OR &ndash; (equivalent)<br>
<b><a href="open.html#fetch">open</a> p29474 from alphafold</b>
</blockquote>
<li>The <b>alphafold match</b> command retrieves models for sequences
the same as or similar to those of experimentally determined protein structures
already open in ChimeraX, or other sequences independent of structure.
Giving the model number of an atomic structure already open in ChimeraX 
specifies all of its protein chains. Examples with <i>sequence</i> given as a
<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>:
<blockquote>
<b>alphafold match #1</b>
<!-- example structure 3p5b -->
<br>
<b>alphafold match #3/B,D trim false</b>
</blockquote>
Alternatively, the <i>sequence</i> can be given as any of the following:
  <ul>
  <li>the <a href="seqspec.html"><i>sequence-spec</i></a> of a sequence
  in the <a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>,
in the form:
&nbsp;<span class="nowrap"><i>alignment-ID</i>:<i>sequence-ID</i></span>&nbsp;
(<a href="seqspec.html">details...</a>)
  <li>a UniProt name or accession number
  <li>plain text pasted directly into the command line
  </ul>
<p>
For a specified structure chain, a model is obtained for its exact 
UniProt entry if available, otherwise the single top hit identified by
<a href="https://www.rbvi.ucsf.edu/chimerax/data/kmer-aug2022/kmer_search.html"
target="_blank">K-mer search</a> of the
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>
(<a href="#search-option">details...</a>).
For each model with a corresponding structure chain
from the <b>alphafold match</b> command or the
<a href="#alignTo"><b>alignTo</b></a> option of <b>alphafold fetch</b>:
</p>
<a name="structmatch"></a>
<ol>
<li>the chain ID of the predicted structure is made the same as the
corresponding chain of the existing model
<li>the predicted structure is superimposed onto the existing chain using
<a href="matchmaker.html"><b>matchmaker</b></a>,
and the following are reported in a table
in the <a href="../tools/log.html"><b>Log</b></a>:
  <ul>
  <li><b>Chain</b> &ndash; chain ID
  <li><a href="https://www.uniprot.org/help/entry_name" 
  target="_blank"><b>UniProt Name</b></a>
  and <a href="https://www.uniprot.org/help/accession_numbers"
  target="_blank"><b>UniProt Id</b></a> (accession number)
  <li><b>RMSD</b> &ndash; C&alpha; root-mean-square deviation between
  the predicted and experimental structures, over all residues of the latter
  <li><b>Length</b> &ndash; number of residues in the predicted structure
  <li><b>Seen</b> &ndash; number of residues with atomic coordinates 
  in the experimental structure
  <li><b>% Id</b> &ndash; percent identity in the sequence alignment generated 
  by <a href="matchmaker.html"><b>matchmaker</b></a> for superposition;
  the number of positions with identical residues
  divided by the length of the shorter sequence
  </ul>
<li>the following <a href="../attributes.html">attributes</a>
 are assigned to the residues of the predicted structure:
  <ul>
  <li><b>c_alpha_distance</b>
&ndash; C&alpha; distance between corresponding positions of the predicted
and existing chains after their superposition (step 2 above)
  <li><b>missing_structure</b>
&ndash; positions missing from the coordinates of the existing chain
  <li><b>same_sequence</b>
&ndash; positions with the same residue type as the existing chain
  </ul>
These <a href="../attributes.html">attributes</a>
can be used for <a href="../tools/alphafold.html#coloring">coloring</a> 
and other purposes.
</ol>
<br>
<li>The <b>alphafold search</b> command uses 
a <a href="../tools/blastprotein.html#notes">BLAST</a> web service
hosted by the <a href="https://www.cgl.ucsf.edu/Overview/index.html" 
target="_blank">UCSF RBVI</a> to search the 
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>.
It differs from <b>alphafold match</b> in that it uses BLAST instead
of fast (but low-sensitivity)
<a href="https://www.rbvi.ucsf.edu/chimerax/data/kmer-aug2022/kmer_search.html"
target="_blank">K-mer searching</a>,
accepts only a single chain or sequence as input,
and returns a list of hits for the user to inspect, 
rather than fetching the single top hit per chain automatically.
The query <i>sequence</i> can be given as any of the following:
  <ul>
  <li>a <a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
  corresponding to a single chain in an atomic structure open in ChimeraX
  <li>the <a href="seqspec.html"><i>sequence-spec</i></a> of a sequence
  in the <a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>
  <li>a <a href="https://www.uniprot.org/help/entry_name"
  target="_blank">UniProt name</a> or
  <a href="https://www.uniprot.org/help/accession_numbers"
  target="_blank">accession number</a>
  <li>plain text pasted directly into the command line
  </ul>
<p>
The <b>matrix</b> option indicates which amino acid <i>similarity-matrix</i>
to use for scoring the hits (uppercase or lowercase can be used): BLOSUM45,
BLOSUM50, <b>BLOSUM62</b> (default), BLOSUM80, BLOSUM90, PAM30, PAM70, PAM250,
or IDENTITY.
The <b>cutoff</b> <i>evalue</i> is the maximum or least significant
expectation value needed to qualify as a hit (default <b>1e-3</b>).
Results can also be limited with the <b>maxSequences</b> option
(default <b>100</b>); this is the maximum number of unique sequences to return.
</p><p>
When results are returned, the hits are listed in a
<a href="../tools/blastprotein.html#results"><b>Blast Protein</b></a> window.
Double-clicking a hit uses
<a href="#fetch"><b>alphafold fetch</b></a> to retrieve the model,
or multiple chosen hits can be retrieved at once by using the results panel
<a href="../window.html#context-menu">context menu</a>
or <b>Load Structures</b> button
(<a href="../tools/blastprotein.html#results">details...</a>).
</p>
</ul>

<a name="options"></a>
<p class="nav">
[<a href="#top">back to top: alphafold</a>]
</p>
<h4>Options</h4>

<blockquote>
<a name="alignTo"></a>
<b>alignTo</b>&nbsp;&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
<br>
Superimpose the predicted structure from 
<a href="#fetch"><b>alphafold fetch</b></a> onto
a single chain in an already-open structure, and make its chain ID the same
as that chain's. See also the <a href="#trim"><b>trim</b></a> option. 
</blockquote>
<blockquote>
<a name="colorConfidence"></a>
<b>colorConfidence</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
Whether to <a href="color.html#byattribute">color</a>
the predicted structures by the
<a href="https://alphafold.ebi.ac.uk/faq#faq-5" target="_blank">pLDDT
confidence measure</a> in the B-factor field (default <b>true</b>):
   <ul>
     <li>100
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=blue width="10px" title="blue"></td></tr>
</table>
to 90
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=cornflowerblue width="10px" title="cornflower blue"></td></tr>
</table>
&ndash; high accuracy expected
     <li>90
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=cornflowerblue width="10px" title="cornflower blue"></td></tr>
</table>
to 70
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=yellow width="10px" title="yellow"></td></tr>
</table>
&ndash; backbone expected to be modeled well
     <li>70
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=yellow width="10px" title="yellow"></td></tr>
</table>
to 50
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=orange width="10px" title="orange"></td></tr>
</table>
&ndash; low confidence, caution
     <li>50
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=orange width="10px" title="orange"></td></tr>
</table>
to 0
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=red width="10px" title="red"></td></tr>
</table>
&ndash; should not be interpreted, may be disordered
     </ul>
<p>
...in other words, using
</p>
<blockquote><b>
<a href="color.html#bfactor">color bfactor</a>
<a href="color.html#palette-options">palette</a> alphafold
<!-- 0,red:50,orange:70,yellow:90,cornflowerblue:100,blue -->
<!-- 100,blue:90,cornflowerblue:70,yellow:50,orange:0,red -->
</b></blockquote>
<p>
The <a href="../tools/colorkey.html"><b>Color Key</b></a> graphical interface
or a command can be used to draw a corresponding color key, for example:
</p>
  <blockquote>
  <a href="key.html"><b>key</b></a> <b>red:low orange: yellow: cornflowerblue: blue:high</b>
  &nbsp;[<a href="key.html#options"><i>other-key-options</i></a>]
<!--
<br>
  <b>key</b> <b>blue:100 cornflowerblue:90 yellow:70 orange:50 red:0</b>
  &nbsp;<b>numericLabelSpacing equal</b>
  &nbsp;[<a href="key.html#options"><i>other-key-options</i></a>]
-->
  </blockquote>
</blockquote>
<blockquote>
<a name="directory"></a>
<b>directory</b>&nbsp;&nbsp;<i>results-folder</i>
<br>
For <a href="#predict"><b>alphafold predict</b></a>, 
specify a <i>results-folder</i> pathname (name and location)
or the word <a href="usageconventions.html#browse"><b>browse</b></a>
to specify it interactively in a file browser window.
The directory or folder does not need to exist already, as it will be
created by the command. The pathname can include [N] to indicate 
substitution with the smallest positive integer that makes a new directory
(default <b>~/Downloads/ChimeraX/AlphaFold/prediction_[N]</b>).
If the specified pathname does not include [N]
but a directory of that name and location already exists
and contains a results.zip file, _[N] will be appended automatically 
to avoid overwriting the existing directory.
</blockquote>
<blockquote>
<a name="ignoreCache"></a>
<b>ignoreCache</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
The fetched models are stored locally in 
<b>~/Downloads/ChimeraX/AlphaFold/</b>,
where ~ indicates a user's home directory.
If a file specified for opening is not found in this local cache or 
<b>ignoreCache</b> is set to <b>true</b>, the file will be fetched and cached.
</blockquote>
<blockquote>
<a name="search-option"></a>
<b>search</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
When fetching models with <a href="#match"><b>alphafold match</b></a>,
whether to search the database for the most similar sequence
if the <a href="uniprot.org/help/accession_numbers"
target="_blank">UniProt accession number</a>
for a chain is not provided in the experimental structure's input file,
or is provided but not found in the 
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>
(<b>true</b>, default). A fast but low-sensitivity (requiring high % identity)
<a href="https://www.rbvi.ucsf.edu/chimerax/data/kmer-aug2022/kmer_search.html"
target="_blank">K-mer search</a> of the the database is performed.
The closest sequence match for which a model is available will be retrieved.
With <b>search false</b>, only the experimental structure's input file will
be used as a potential source of UniProt accession numbers.
When present, these are given in 
<a href="https://www.wwpdb.org/documentation/file-format-content/format33/sect3.html" target="_blank">DBREF records</a>
in PDB format and in 
<a href="https://mmcif.wwpdb.org/dictionaries/mmcif_std.dic/Categories/struct_ref.html" target="_blank">struct_ref</a> and
<a href="https://mmcif.wwpdb.org/dictionaries/mmcif_std.dic/Categories/struct_ref_seq.html" target="_blank">struct_ref_seq</a>
tables in mmCIF.
</blockquote>
<blockquote>
<a name="minimize"></a>
<b>minimize</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
This option allows skipping energy-minimization of the result from
<a href="#predict"><b>alphafold predict</b></a>, for faster job completion
and/or to avoid failures that may occur during minimization.
</blockquote>
<blockquote>
<a name="templates"></a>
<b>templates</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
Whether a calculation run with <a href="#predict"><b>alphafold predict</b></a>
should use known structures from the PDB as templates.
AlphaFold can use up to four structures as templates;
when this option is <b>true</b>, ColabFold will search the PDB
sequences for similarity to the target and report in the Colab log
which entries (if any) are used as templates.
</blockquote>
<blockquote>
<a name="trim"></a>
<b>trim</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
Whether to trim a predicted protein structure to the same residue range
as the corresponding experimental structure given with the 
<a href="#match"><b>alphafold match</b></a> command or the <b>alignTo</b>
option of <a href="#fetch"><b>alphafold fetch</b></a>.
With <b>trim true</b> (default):
<ul>
<li>Predictions with UniProt identifier determined by 
<a href="#match"><b>alphafold match</b></a>
from the experimental structure's input file will be
trimmed to the same residue ranges as used in the experiment.
These ranges are given in 
<a href="https://www.wwpdb.org/documentation/file-format-content/format33/sect3.html" target="_blank">DBREF records</a>
in PDB format and in 
<a href="https://mmcif.wwpdb.org/dictionaries/mmcif_std.dic/Categories/struct_ref.html" target="_blank">struct_ref</a> and
<a href="https://mmcif.wwpdb.org/dictionaries/mmcif_std.dic/Categories/struct_ref_seq.html" target="_blank">struct_ref_seq</a>
tables in mmCIF.
<li>Predictions retrieved with <a href="#fetch"><b>alphafold fetch</b></a> or
found by <a href="#match"><b>alphafold match</b></a> searching for similar 
sequences in the <a href="https://alphafold.ebi.ac.uk/" 
target="_blank">AlphaFold Database</a> will be trimmed to start and end 
with the first and last aligned positions in the sequence alignment 
calculated by <a href="matchmaker.html"><b>matchmaker</b></a>
as part of the superposition step.
</ul>
<p>
Using <b>trim false</b> indicates retaining the full-length models
for the <a href="https://www.uniprot.org/" target="_blank">UniProt</a>
sequences, which could be longer. 
</p>
</blockquote>
<blockquote>
<a name="version"></a>
<b>version</b>&nbsp;&nbsp;1&nbsp;|&nbsp;2&nbsp;|&nbsp;3&nbsp;|&nbsp;<b>4</b>
<br>
The <a href="https://alphafold.ebi.ac.uk/"
target="_blank">AlphaFold Database</a> 
contains single-chain models for protein sequences
in <a href="https://www.uniprot.org" target="_blank">UniProt</a>.
This option specifies which version of the database to use with
<a href="#fetch"><b>alphafold fetch</b></a>,
<a href="#pae"><b>alphafold pae</b></a>, or
<a href="#search"><b>alphafold search</b></a>
(as well as <a href="blastprotein.html"><b>blastprotein</b></a> with 
<b>database alphafold</b>):
  <ul>
  <li><b>version 1</b> (Jul 2021): ~360,000 sequences,
reference proteomes of 21 species including <i>Homo sapiens</i>;
used by Chimera 1.3, which does not have a <b>version</b> option
  <li><b>version 2</b> (Dec 2021 and Jan 2022 releases combined):
~1 million sequences, v1 + most of the manually curated entries (SwissProt) +
sequences relevant to neglected tropical disease or antimicrobial resistance;
default in ChimeraX 1.4
  <li><b>version 3</b> (Jul 2022): &gt;200 million sequences
  <li><b>version 4</b> (Nov 2022): bugfix of version 3, updating the 
coordinates of ~4% of the entries; default in ChimeraX 1.5 and later
</ul>
<p>
The <a href="#match"><b>alphafold match</b></a> command always uses version 4
and does not have this option.
</p>
</blockquote>

<a name="predict"></a>
<p class="nav">
[<a href="#top">back to top: alphafold</a>]
</p>
<h3>Running an AlphaFold Prediction</h3>
<p>
The <b>alphafold predict</b> command runs a calculation on
<a href="https://colab.research.google.com/signup"
target="_blank">Google Colab</a> using <b>ColabFold</b>,
an open-source, optimized version of
<a href="https://en.wikipedia.org/wiki/AlphaFold#AlphaFold_2,_2020"
target="_blank">AlphaFold 2</a>. Users should cite:
<blockquote>
<a href="https://www.nature.com/articles/s41592-022-01488-1"
target="_blank">ColabFold: making protein folding accessible to all.</a>
Mirdita M, Sch&uuml;tze K, Moriwaki Y, Heo L, Ovchinnikov S, Steinegger M.
<i>Nat Methods.</i> 2022 Jun;19(6):679-682.
</blockquote>
<p>
For <b>monomer</b> prediction:
</p>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>alphafold predict</b> &nbsp;<i>sequence</i>
[&nbsp;<a href="#minimize"><b>minimize</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#templates"><b>templates</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#directory"><b>directory</b></a>&nbsp;&nbsp;<i>results-folder</i>&nbsp;]
</blockquote>
<p>
<a name="sequence-spec"></a>
The protein <i>sequence</i> to predict can be given as any of the following:
</p>
<ol>
<li>a <a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
corresponding to a single chain in an atomic structure open in ChimeraX
<li>the <a href="seqspec.html"><i>sequence-spec</i></a> of a sequence
in the <a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>,
in the form:
&nbsp;<span class="nowrap"><i>alignment-ID</i>:<i>sequence-ID</i></span>&nbsp;
(<a href="seqspec.html">details...</a>)
<li>a UniProt name or accession number
<li>plain text pasted directly into the command line
</ol>
<p>
These methods specify the entire sequence only, although a truncated version
could be pasted.
</p><p>
For <b>multimer</b> (protein complex) prediction:
</p>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>alphafold predict</b> 
&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
[&nbsp;<a href="#minimize"><b>minimize</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#templates"><b>templates</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#directory"><b>directory</b></a>&nbsp;&nbsp;<i>results-folder</i>&nbsp;]
<br>&ndash; or &ndash;</br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>alphafold predict</b> &nbsp;<i>sequence1</i>,<i>sequence2</i>[,<i>sequence3</i>...,<i>sequenceN</i>]
[&nbsp;<a href="#minimize"><b>minimize</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#templates"><b>templates</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#directory"><b>directory</b></a>&nbsp;&nbsp;<i>results-folder</i>&nbsp;]
</blockquote>
<p>
The sequences of two or more protein chains can be specified
either collectively as a <a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
(for atomic-structure chains already open in ChimeraX), or
individually within a comma-separated list using any combination of
specifier types #2-4 listed <a href="#sequence-spec">above</a>.
The comma-separated list should not contain any spaces.
If the same protein chain occurs multiple times in the complex, its sequence
should be repeated that number of times. For example, to predict a
homodimer, the same sequence (or its specifier) would need to be given twice.
Prediction may only be feasible for smaller complexes
(<a href="#caveats">details...</a>).
</p><p>
A warning will appear saying that this Colab notebook is from github
(was not authored by Google), with a button to click to run anyway.
Users will need to have a Google account and to sign into it via
a browser. Once that is done, the sign-in may be remembered depending
on the user's browser settings; it is not kept in the
<a href="../preferences.html">ChimeraX preferences</a>.
See the <a href="https://www.youtube.com/watch?v=gIbCAcMDM7E"
target="_blank">example video</a> for an explanation of the images/plots
from <b>ColabFold</b> that appear in the Colab log
and where to find downloaded files.
</p><p>
The model will be opened automatically and
<a href="#colorConfidence">colored by confidence value</a>.
The model for a sequence that was specified by structure chain will
be superimposed on that chain and assigned structure-comparison attributes
for further analysis (<a href="#structmatch">details...</a>).
</p>

<a name="caveats"></a>
<h4>Caveats</h4>
<ul>
<li>The process includes
installing various software packages on a virtual machine, 
searching sequence databases, generating a multiple sequence alignment,
predicting atomic coordinates, and <a href="#options">optionally</a>,
energy-minimizing the best structure. In addition,
predicting a multimer (complex) structure may take longer than predicting
the structure of a monomer with the same total number of residues.
<b>The free version of Colab limits jobs</b> to 12 hours 
and may terminate them at shorter times at Google's
discretion (see the <a href="https://colab.research.google.com/signup"
target="_blank">FAQ</a>).
Those who want to run longer and/or more frequent calculations
may need to sign up for one of the paid
<a href="https://colab.research.google.com/signup"
target="_blank">Colab plans</a>.
<br><br>
<li><b>Each chain must contain at least 16 residues.</b>
Shorter sequences are not accepted because they cannot be used to generate
a reliable multiple sequence alignment.
<br><br>
<li><b>Total sequence length cannot be very large</b>.
AlphaFold runs out of graphics memory for long sequences (~1200 amino acids
on old Google Colab GPUs with 16 GB memory).
Multimer predictions face the same limit on the total number of residues,
so <b>only smaller complexes can be predicted</b>.
As mentioned above, paid <a href="https://colab.research.google.com/signup"
target="_blank">Colab plans</a> provide more computational resources
than the free plan. Structures with up to 3000 amino acids can be predicted
using an Nvidia A100 GPU on Google Colab, costing about $1.50 for a
2000-residue prediction (May 2023); this 
<a href="https://youtu.be/H-pDs9rZtkw" target="_blank">video</a> explains how.
</ul>

<a name="pae"></a>
<p class="nav">
[<a href="#top">back to top: alphafold</a>]
</p>
<h3>AlphaFold Predicted Aligned Error (PAE)</h3>
<p>
Besides the per-residue
<a href="https://alphafold.ebi.ac.uk/faq#faq-5" target="_blank">pLDDT
confidence measure</a>, AlphaFold gives
for each pair of residues (X,Y) the expected position error
at residue X if the predicted and true structures were aligned on residue Y.
These residue-residue &ldquo;predicted aligned error&rdquo; or PAE values can 
be shown as a 2D plot using the command <b>alphafold pae</b> (details below),
the <a href="../tools/alphafold.html#pae"><b>AlphaFold Error Plot</b></a>
tool, or <a href="#fetch"><b>alphafold fetch</b></a> or 
<a href="#fetch"><b>alphafold match</b></a> with the option <b>pae true</b>.
See also the
<a href="https://www.rbvi.ucsf.edu/chimerax/data/pae-apr2022/pae.html"
target="_blank">AlphaFold Error Estimates example</a> and
<a href="https://www.youtube.com/watch?v=oxblwn0_PMM" target="_blank">video</a>.
</p>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>alphafold pae</b>
[&nbsp;<a href="atomspec.html#hierarchy"><i>model-spec</i></a>&nbsp;]
(&nbsp;<b>uniprotId</b></a>&nbsp;&nbsp;<i>uniprot</i>&nbsp;|&nbsp;<b>file</b>&nbsp;&nbsp;<i>filename</i>&nbsp;)
[&nbsp;<a href="color.html#palette-options"><b>palette</b></a>&nbsp;&nbsp;<i>palette</i>&nbsp;]
[&nbsp;<a href="color.html#palette-options"><b>range</b></a>&nbsp;&nbsp;<i>low,high</i>&nbsp;|&nbsp;full&nbsp;]
[&nbsp;<b>plot</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<b>colorDomains</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>minSize</b>&nbsp;&nbsp;<i>M</i>&nbsp;]
[&nbsp;<b>connectMaxPae</b>&nbsp;&nbsp;<i>N</i>&nbsp;]
[&nbsp;<b>cluster</b>&nbsp;&nbsp;<i>resolution</i>&nbsp;]
[&nbsp;<b>dividerLines</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<a href="#version"><b>version</b></a>&nbsp;&nbsp;1&nbsp;|&nbsp;2&nbsp;|&nbsp;3&nbsp;|&nbsp;<b>4</b>&nbsp;]
</blockquote>
<p>
With <b>alphafold pae</b>, the matrix of PAE values can be either:
<ul>
<li>fetched from the <a href="https://alphafold.ebi.ac.uk/"
target="_blank">AlphaFold Database</a> with the <b>uniprotId</b> option,
where <i>uniprot</i> is the the UniProt name or accession number of an entry 
in the database
<br>&ndash; or &ndash;
<li>a json or pkl file from AlphaFold specified with the <b>file</b> option.
The <i>filename</i> is generally a pathname to a local file, 
either absolute or relative to the current working directory as reported by 
<a href="pwd.html"><b>pwd</b></a>. Substituting the word 
<a href="usageconventions.html#browse"><b>browse</b></a>
for <i>filename</i> brings up a file browser window for choosing the
name and location interactively.
</ul>
<a name="association"></a>
<p>
The corresponding AlphaFold structure (already open) can be given
as a <a href="atomspec.html#hierarchy"><i>model-spec</i></a>
to associate it with the plot. This association allows
coloring by domain as described below, and for selections
on the plot to highlight the corresponding parts of the structure.
</p><p>
By default, the PAE plot is drawn when domain coloring is not done 
(<b>plot</b> is default <b>true</b> when <b>colorDomains</b> is <b>false</b>)
and <i>vice versa</i>.
</p><p>
<a name="colorDomains"></a>
Setting <b>colorDomains</b> to <b>true</b> clusters the residues into
coherent domains (sets of residues with relatively low pairwise PAE values)
and uses randomly chosen colors to distinguish these domains in the structure.
The residues are assigned an integer domain identifier (starting with 1) as an
<a href="../attributes.html">attribute</a> named <b>pae_domain</b>
that can be used to <a href="atomspec.html#attributes">specify</a> 
them in commands (for example, to recolor or select specific domains).
Residues not grouped into any domain are assigned a <b>pae_domain</b> 
value of <b>None</b>. The clustering uses the
NetworkX greedy_modularity_communities algorithm with parameters:
</p>
<ul>
<li><b>minSize</b> (default <b>10</b> residues)
&ndash; minimum number of residues allowed in a domain
<li><b>connectMaxPae</b> (default <b>5.0</b> &Aring;)
&ndash; the maximum PAE value allowed between residues for them to be 
clustered into the same domain. Larger values give larger domains
and generally increase the time to compute the clustering, 
which is ~5 seconds for 1000 residues when the default of 5.0 is used.
<li><b>cluster</b> (default <b>0.5</b>, typical range 0.5&ndash;5.0) 
&ndash; graph resolution; larger values give smaller domains
</ul>
<p>
The <b>dividerLines</b> option (default <b>true</b>) indicates whether,
for multimer predictions, to draw lines on the plot demarcating the
end of one chain and the start of another. The lines may obscure a few
chain-terminal residues in the plot, and <b>dividerLines false</b>
can be used if this is problematic.
</p><p>
The default <a href="color.html#palette-options"><b>palette</b></a> 
for coloring the PAE plot is <b>pae</b>, with colors assigned to values
as follows:
</p>
<table border cellpadding="0" cellspacing="0"
style="display:inline-block; vertical-align:top"><tr>
<td bgcolor="blue" width="30px" title="blue"
align="center"><font size="-1" color="white">0</font></td>
<td bgcolor="cornflowerblue" width="30px" title="cornflowerblue"
align="center"><font size="-1" color="white">5</font></td>
<td bgcolor="yellow" width="30px" title="yellow"
align="center"><font size="-1">10</font></td>
<td bgcolor="orange" width="30px" title="orange"
align="center"><font size="-1">15</font></td>
<td bgcolor="gray" width="30px" title="gray"
align="center"><font size="-1" color="white">20</font></td>
<td bgcolor="lightgray" width="30px" title="lightgray"
align="center"><font size="-1">25</font></td>
<td bgcolor="white" width="30px" title="white"
align="center"><font size="-1">30</font></td>
</tr></table>
<p>
Another <a href="color.html#palette-options"><b>palette</b></a> 
with value range suitable for PAE plots is <b>paegreen</b>:
</p>
<table border cellpadding="0" cellspacing="0"
style="display:inline-block; vertical-align:top"><tr>
<td bgcolor="#1e3546661e35" width="30px" title="#1e3546661e35"
align="center"><font size="-1" color="white">0</font></td>
<td bgcolor="#245a922c245a" width="30px" title="#245a922c245a"
align="center"><font size="-1" color="white">5</font></td>
<td bgcolor="#374cb168374c" width="30px" title="#374cb168374c"
align="center"><font size="-1" color="white">10</font></td>
<td bgcolor="#5687c9ba5687" width="30px" title="#5687c9ba5687"
align="center"><font size="-1">15</font></td>
<td bgcolor="#828fddf3828f" width="30px" title="#828fddf3828f"
align="center"><font size="-1">20</font></td>
<td bgcolor="#bae1efdebae1" width="30px" title="#bae1efdebae1"
align="center"><font size="-1">25</font></td>
<td bgcolor=white width="30px" title="white"
align="center"><font size="-1">30</font></td>
</tr></table>
<p>
Although these palettes include value-color pairs, it may be helpful
to give a value <a href="color.html#range"><b>range</b></a>
if a colors-only palette is used instead (default <b>0,30</b>).
A range can also be used to override the values in a value-color palette,
instead spacing the colors evenly across the specified range.
</p><p>
The plot window has a 
<a href="../tools/alphafold.html#context">context menu</a>
with options for recoloring the plot and 
<a href="#association">associated</a> structure,
hiding/showing the divider lines, and saving the plot as an image,
as well as buttons for recoloring the structure: 
</p>
<ul>
<li><b>Color PAE Domains</b> applies coloring by PAE cluster as
described <a href="#colorDomains">above</a>.
<li><b>Color pLDDT</b> returns the structure to
the default <a href="#colorConfidence">confidence coloring</a>.
</ul>
<p>
The <a href="../tools/colorkey.html"><b>Color Key</b></a> graphical interface
or a command can be used to draw (in the main graphics window)
a color key for the PAE plot. For example, to make a color key that
matches the <b>pae</b> or <b>paegreen</b> scheme, respectively:
</p>
  <blockquote>
  <a href="key.html"><b>key</b></a>&nbsp;<b>pae&nbsp;:0&nbsp;:&nbsp;:&nbsp;:15&nbsp;:&nbsp;:&nbsp;:30</b>
  &nbsp;<a href="key.html#options"><b>showTool true</b></a>
<br>
  <a href="key.html"><b>key</b></a>&nbsp;<b>paegreen&nbsp;:0&nbsp;:&nbsp;:&nbsp;:15&nbsp;:&nbsp;:&nbsp;:30</b>
  &nbsp;<a href="key.html#options"><b>showTool true</b></a>
  </blockquote>
<p>
A title for the color key (<i>e.g.</i>, &ldquo;<b>Predicted Aligned
Error (&Aring;)</b>&rdquo;) would need to be created separately with
<a href="2dlabels.html"><b>2dlabels</b></a>.
</p>

<a name="contacts"></a>
<p class="nav">
[<a href="#top">back to top: alphafold</a>]
</p>
<h3>Pseudobonds Colored by PAE</h3>
<p>
Residue-residue <a href="#pae">PAE values</a> can also be shown with colored 
<a href="../pseudobonds.html">pseudobonds</a> in the predicted structure:
</p>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>alphafold contacts</b>
&nbsp;<a href="atomspec.html"><i>res-spec1</i></a>&nbsp;
[&nbsp;<b>toResidues</b>&nbsp;&nbsp;<a href="atomspec.html"><i>res-spec2</i></a>&nbsp;
[&nbsp;<b>flip</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>distance</b>&nbsp;&nbsp;<i>d</i>&nbsp;]
[&nbsp;<b>maxPae</b>&nbsp;&nbsp;<i>max-error</i>&nbsp;]
[&nbsp;<a href="color.html#palette-options"><b>palette</b></a>&nbsp;&nbsp;<i>palette</i>&nbsp;]
[&nbsp;<a href="color.html#palette-options"><b>range</b></a>&nbsp;&nbsp;<i>low,high</i>&nbsp;|&nbsp;full&nbsp;]
[&nbsp;<b>radius</b>&nbsp;&nbsp;<i>r</i>&nbsp;]
[&nbsp;<b>dashes</b>&nbsp;&nbsp;<i>N</i>&nbsp;]
[&nbsp;<b>name</b>&nbsp;&nbsp;<i>model-name</i>&nbsp;]
[&nbsp;<b>replace</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<b>outputFile</b>&nbsp;&nbsp;<i>pae-file</i>&nbsp;]
</blockquote>
<p>
See the
<a href="https://www.rbvi.ucsf.edu/chimerax/data/alphafold-contacts-jun2022/afcontacts.html" target="_blank">AlphaFold Contacts example</a>
and <a href="https://www.youtube.com/watch?v=TMcjEecFHaI" 
target="_blank">video</a>.
See also:
<a href="size.html"><b>size</b></a>,
<a href="style.html"><b>style</b></a>,
<a href="clashes.html"><b>contacts</b></a>,
<a href="crosslinks.html"><b>crosslinks</b></a>,
<a href="rename.html"><b>rename</b></a>
</p><p>
A <a href="#pae">PAE plot</a> 
containing the specified residues must already be shown.
The PAE matrix is not symmetrical. The first specification
<a href="atomspec.html"><i>res-spec1</i></a> gives the aligned residues,
whereas <b>toResidues</b> <a href="atomspec.html"><i>res-spec2</i></a> 
gives the residues whose error values are reported, except that using
<b>flip true</b> swaps the meaning of 
<a href="atomspec.html"><i>res-spec1</i></a> and
<a href="atomspec.html"><i>res-spec2</i></a>.
If one set of residues is higher-confidence (lower in 
<a href="https://alphafold.ebi.ac.uk/faq#faq-5" target="_blank">pLDDT</a>)
than the other, it is usually best to specify them as the aligned residues
so that the coloring will show the error values of the lower-confidence set.
</p><p>
Omitting the <b>toResidues</b> option defines 
<a href="atomspec.html"><i>res-spec2</i></a>
as all residues covered by the PAE plot except for those in
<a href="atomspec.html"><i>res-spec1</i></a>; however, if <b>toResidues</b> 
is omitted and <a href="atomspec.html"><i>res-spec1</i></a>
includes all residues in the plot, 
<a href="atomspec.html"><i>res-spec2</i></a>
will also be defined as all residues in the plot.
</p><p>
The <b>distance</b> option limits the number of 
<a href="../pseudobonds.html">pseudobonds</a> by 
only drawing them between pairs of residues with any interresidue distance
&le; <i>d</i> &Aring; (default <b>3.0</b>). These pseudobonds are drawn between
&alpha;-carbons regardless of which atoms were within the distance cutoff.
The <b>maxPae</b> option can be used to further limit the pseudobonds
to only those representing PAE values &le; <i>max-error</i>
(no default value; if the option is omitted, there is no restriction by PAE).
</p><p>
The default <a href="color.html#palette-options"><b>palette</b></a> for
coloring the pseudobonds by <a href="#pae">PAE value</a> is <b>paecontacts</b>,
with colors assigned to values as follows:
</p>
<table border cellpadding="0" cellspacing="0"
style="display:inline-block; vertical-align:top"><tr>
<td bgcolor="blue" width="30px" title="blue"
align="center"><font size="-1" color="white">0</font></td>
<td bgcolor="cornflowerblue" width="30px" title="cornflowerblue"
align="center"><font size="-1" color="white">5</font></td>
<td bgcolor="yellow" width="30px" title="yellow"
align="center"><font size="-1">10</font></td>
<td bgcolor="orange" width="30px" title="orange"
align="center"><font size="-1">15</font></td>
<td bgcolor="red" width="30px" title="red"
align="center"><font size="-1">20</font></td>
</tr></table>
<p>
Although this palette includes value-color pairs, it may be helpful
to give a value <a href="color.html#range"><b>range</b></a>
if a colors-only palette is used instead (default <b>0,30</b>).
A range can also be used to override the values in a value-color palette,
instead spacing the colors evenly across the specified range.
</p><p>
The <a href="../pseudobonds.html">pseudobond</a> stick <b>radius</b>
(default <b>0.2</b> &Aring;) and number of <b>dashes</b> (default <b>1</b>,
meaning a solid stick) can also be specified. 
</p><p>
The <b>name</b> option allows specifying the pseudobond <i>model-name</i> 
(default <b>PAE Contacts</b>).  If a model by that name already exists, any
pre-existing pseudobonds will be removed from that model and replaced by the
new ones (<b>replace true</b>, default) unless <b>replace false</b> is used.
</p><p>
The <b>outputFile</b> option allows saving a list
of the residue pairs (those meeting the distance criterion) and
their PAE values to a plain text file. 
The <i>pae-file</i> argument is the output file pathname,
enclosed in quotation marks if it includes spaces,
or the word <a href="usageconventions.html#browse"><b>browse</b></a>
to specify it interactively in a file browser window.
</p><p>
<a name="examples">Examples</a>:
<blockquote><b>
alphafold contacts #1 
<br>
alphafold contacts /A to /B distance 8
<br>
alphafold contacts sel palette blue:red range 1,5
</b></blockquote>
</p>
The following would select all pseudobonds and label them with the names and
numbers of the residues that they connect:
<blockquote><b>
<a href="select.html">sel</a> pbonds
<br>
<a href="label.html">label</a>&nbsp;sel&nbsp;pseudobonds&nbsp;text&nbsp;"{0.atoms[0].residue.name}&nbsp;{0.atoms[0].residue.number}&nbsp;to&nbsp;{0.atoms[1].residue.name}&nbsp;{0.atoms[1].residue.number}"
</b></blockquote>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics / 
July 2023</address>
</body></html>
