<html>

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Command: blastprotein</title>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRight" title="User Guide Index"/></a>

<h3><a href="../index.html#commands">Command</a>: blastprotein</h3>
<h3 class="usage"><a href="usageconventions.html">Usage</a>:
<br><b>blastprotein</b> &nbsp;<i>sequence</i>&nbsp;
[&nbsp;<b>database</b>&nbsp;&nbsp;<i>sequence-database</i>&nbsp;]
[&nbsp;<a href="alphafold.html#version"><b>version</b></a>&nbsp;&nbsp;1&nbsp;|&nbsp;2&nbsp;|&nbsp;<b>3</b>&nbsp;]
[&nbsp;<b>matrix</b>&nbsp;&nbsp;<i>similarity-matrix</i>&nbsp;]
[&nbsp;<b>cutoff</b>&nbsp;&nbsp;<i>evalue</i>&nbsp;]
[&nbsp;<b>maxSeqs</b>&nbsp;&nbsp;<i>M</i>&nbsp;]
[&nbsp;<b>log</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>name</b>&nbsp;&nbsp;<i>N</i>&nbsp;]
</h3>
<p>
The <b>blastprotein</b> command runs a protein sequence similarity search
using a <a href="../tools/blastprotein.html#notes">BLAST</a> web service 
hosted by the
<a href="https://www.cgl.ucsf.edu/Overview/index.html" target="_blank">UCSF
Resource for Biocomputing, Visualization, and Informatics (RBVI)</a>.
It is the command implementation of the
<a href="../tools/blastprotein.html"><b>Blast Protein</b></a> tool.
One use is to search with a target sequence of unknown structure
to find templates for <a href="modeller.html">comparative modeling</a>.
See also: <a href="alphafold.html#search"><b>alphafold search</b></a>,
<a href="esmfold.html#search"><b>esmfold search</b></a>
</p><p>
The query <i>sequence</i> can be given as any of the following:
</p>
  <ul>
  <li>a <a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
  corresponding to a single chain in an atomic structure open in ChimeraX
  <li>the <a href="seqspec.html"><i>sequence-spec</i></a> of a sequence
  in the <a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>
  <li>a <a href="https://www.uniprot.org/help/entry_name"
  target="_blank">UniProt name</a> or 
  <a href="https://www.uniprot.org/help/accession_numbers"
  target="_blank">accession number</a>
  <li>plain text pasted directly into the command line
  </ul>
<p>
The protein <i>sequence-database</i> to search can be:
</p>
<ul>
<li><b>pdb</b> (default) 
&ndash; experimentally determined structures in the
<a href="https://www.rcsb.org" target="_blank">Protein Data Bank (PDB)</a>
<li><b>nr</b>
&ndash; NCBI &ldquo;non-redundant&rdquo; database containing
<a href="https://www.ncbi.nlm.nih.gov/genbank/" target="_blank">GenBank</a>
CDS translations + <a href="https://www.rcsb.org" target="_blank">PDB</a>
+ <a href="https://www.uniprot.org/" target="_blank">SwissProt</a> +
<a href="https://proteininformationresource.org/"  target="_blank">PIR</a>
+ <a href="https://www.prf.or.jp/index-e.html" target="_blank">PRF</a>
excluding environmental samples from whole-genome sequencing; this database
is much larger than <b>pdb</b> alone and takes much longer to search
<li><b>alphafold</b>
&ndash; artificial-intelligence-predicted structures in the
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>
(<a href="alphafold.html">more...</a>)
  <ul>
  <li>the <b>version</b> option applies only to the <b>alphafold</b> database
(<a href="alphafold.html#version">details...</a>)
  </ul>
<li><b>esmfold</b> &ndash; artificial-intelligence-predicted structures in the
<a href="https://esmatlas.com/about#about_this_website" target="_blank">ESM
Metagenomic Atlas</a> (<a href="esmfold.html">more...</a>)
<li><b>uniref100</b> &ndash;
<a href="https://www.uniprot.org/help/uniref"
target="_blank">UniProt Reference Cluster</a> at 100% identity
(identical sequences and subfragments collapsed into a single entry)
   <li><b>uniref90</b> &ndash; based on <b>uniref100</b>, but omitting
sequences shorter than 11 residues and clustering at 90% identity
   <li><b>uniref50</b> &ndash; based on <b>uniref100</b>, but omitting
sequences shorter than 11 residues and clustering at 50% identity
</ul>
<p>
The <b>matrix</b> option indicates which amino acid <i>similarity-matrix</i> 
to use for alignment scoring (uppercase or lowercase can be used):
</p>
<ul>
  <li><b>BLOSUM45</b>
  <li><b>BLOSUM50</b>
  <li><b>BLOSUM62</b> (default)
  <li><b>BLOSUM80</b>
  <li><b>BLOSUM90</b>
  <li><b>PAM30</b>
  <li><b>PAM70</b>
  <li><b>PAM250</b>
  <li><b>IDENTITY</b>
</ul>
<p>
The <b>cutoff</b> <i>evalue</i> is the maximum or least significant
<a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=FAQ#expect" target="_blank">E-value</a>
needed to qualify as a hit (default <b>1e-3</b>).
Results can also be limited with the <b>maxSeqs</b> option
(default <b>100</b>); this is the
maximum number of unique sequences to return; more hits than this number
may be obtained because multiple structures or other sequence-database entries
may have the same sequence.
</p><p>
When the search completes, results are shown in a
<a href="../tools/blastprotein.html#results">separate window</a>.
Many other types of information can be shown and used to sort the hits:
alignment scores, structure resolution, ligand residue names, <i>etc.</i>
(<a href="../tools/blastprotein.html#results">details...</a>)
</p><p>
Double-clicking a row with an associated structure (AlphaFold or PDB)
<a href="../commands/open.html#fetch">fetches</a> the structure,
and if a structure chain was used as the query,
automatically superimposes the hit and query structures with
<a href="../commands/matchmaker.html"><b>matchmaker</b></a>.
AlphaFold-predicted structures are
<a href="alphafold.html#colorConfidence">colored by confidence 0-100</a>.
ESMFold-predicted structures are
<a href="esmfold.html#colorConfidence">colored by confidence 0-1</a>.
<!-- CHECK THIS 
Double-clicking rows for sequence-only hits shows the
corresponding web pages in the <a href="https://www.ncbi.nlm.nih.gov/protein"
target="_blank">NCBI Protein</a> database.
-->
One or more hits can be chosen (highlighted) in the list and
the panel's <a href="../window.html#context-menu">context menu</a>
used to fetch and superimpose all of the corresponding structures, or to
show their <a href="../tools/blastprotein.html#notes">multiple 
sequence alignment</a> with the query.
</p><p>
The <b>log</b> option indicates whether to also list the results in the
<a href="../tools/log.html"><b>Log</b></a> (default <b>false</b>).
</p><p>
The <b>name</b> option allows supplying a name for a specific set of
<a href="../tools/blastprotein.html"><b>Blast Protein</b></a> results,
which may be useful when several sets of results are shown at the same time. 
The name appears in the title bar of the results panel.
</p>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics / 
April 2023</address>
</body></html>
