<html>

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Command: blastprotein</title>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRight" title="User Guide Index"/></a>

<h3><a href="../index.html#commands">Command</a>: blastprotein</h3>
<h3 class="usage"><a href="usageconventions.html">Usage</a>:
<br><b>blastprotein</b> 
(&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>&nbsp;|&nbsp;<a href="seqspec.html"><i>sequence-spec</i></a>&nbsp;)
[&nbsp;<b>database</b>&nbsp;&nbsp;<b>pdb</b>&nbsp;|&nbsp;nr&nbsp;]
[&nbsp;<b>matrix</b>&nbsp;&nbsp;<i>similarity-matrix</i>&nbsp;]
[&nbsp;<b>cutoff</b>&nbsp;&nbsp;<i>evalue</i>&nbsp;]
[&nbsp;<b>maxSeqs</b>&nbsp;&nbsp;<i>M</i>&nbsp;]
[&nbsp;<b>log</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>toolId</b>&nbsp;&nbsp;<i>N</i>&nbsp;]
</h3>
<p>
The <b>blastprotein</b> command runs a protein sequence similarity search
using a <a href="../tools/blastprotein.html#notes">BLAST</a> web service 
hosted by the
<a href="http://www.cgl.ucsf.edu/Overview/index.html" target="_blank">UCSF
Resource for Biocomputing, Visualization, and Informatics (RBVI)</a>.
It is the command implementation of the
<a href="../tools/blastprotein.html"><b>Blast Protein</b></a> tool.
One use is to search with a target sequence of unknown structure
to find templates for <a href="modeller.html">comparative modeling</a>.
</p><p>
The query sequence can be taken from a protein structure or a sequence 
alignment open in ChimeraX: 
<ul>
<li>&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
<br>&ndash; OR &ndash;
<li>&nbsp;<a href="seqspec.html"><i>sequence-spec</i></a>
</ul>
<p>
The protein sequence <b>database</b> to search can be:
</p>
<ul>
<li><b>pdb</b> (default) 
&ndash; experimentally determined structures in the
<a href="http://www.rcsb.org" target="_blank">Protein Data Bank (PDB)</a>
<li><b>nr</b>
&ndash; NCBI &ldquo;non-redundant&rdquo; database containing
<a href="https://www.ncbi.nlm.nih.gov/genbank/" target="_blank">GenBank</a>
CDS translations + <a href="http://www.rcsb.org" target="_blank">PDB</a>
+ <a href="https://www.uniprot.org/" target="_blank">SwissProt</a> +
<a href="https://proteininformationresource.org/"  target="_blank">PIR</a>
+ <a href="https://www.prf.or.jp/index-e.html" target="_blank">PRF</a>
excluding environmental samples from whole-genome sequencing; this database
is much larger than <b>pdb</b> alone and takes much longer to search
</ul>
<p>
The <b>matrix</b> option indicates which amino acid <i>similarity-matrix</i> 
to use for alignment scoring (uppercase or lowercase can be used):
</p>
<ul>
  <li><b>BLOSUM45</b>
  <li><b>BLOSUM50</b>
  <li><b>BLOSUM62</b> (default)
  <li><b>BLOSUM80</b>
  <li><b>BLOSUM90</b>
  <li><b>PAM30</b>
  <li><b>PAM70</b>
  <li><b>PAM250</b>
  <li><b>IDENTITY</b>
</ul>
<p>
The <b>cutoff</b> <i>evalue</i> is the maximum or least significant
expectation value needed to qualify as a hit (default <b>1e-3</b>).
Results can also be limited with the <b>maxSeqs</b> option
(default <b>100</b>); this is the
maximum number of unique sequences to return; more hits than this number
may be obtained because multiple structures or other sequence-database entries
may have the same sequence.
</p><p>
When results are returned, the hits are listed in a
<a href="../tools/blastprotein.html"><b>Blast Protein</b></a> window.
The names of hits from the <b>pdb</b> database (a subset of <b>nr</b>),
are links that can be clicked to 
<a href="../commands/open.html#fetch">fetch</a> the structure,
and if a structure chain was used as the query,
to automatically superimpose the hit and query structures with
<a href="../commands/matchmaker.html"><b>matchmaker</b></a>.
The names of other hits are GI numbers that can be clicked to show the
corresponding web pages in the <a href="https://www.ncbi.nlm.nih.gov/protein"
target="_blank">NCBI Protein</a> database.
Checkboxes allow designating one or more hit sequences to show as a
<a href="../tools/blastprotein.html#notes">multiple alignment</a>
with the query.
Many other types of information can be shown and used to sort the hits:
alignment scores, structure resolution, ligand residue names, <i>etc.</i>
(<a href="../tools/blastprotein.html#results">details...</a>)
</p><p>
The <b>log</b> option indicates whether to also list the results in the
<a href="../tools/log.html"><b>Log</b></a> (default <b>false</b>).
</p><p>
The <b>toolId</b> option allows supplying an integer identifier for 
a specific instance of the 
<a href="../tools/blastprotein.html"><b>Blast Protein</b></a> tool,
allowing multiple panels for different searches to be shown simultaneously. 
If only one set of results is shown at a time, this option is not needed.
</p>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics / 
May 2020</address>
</body></html>
