<html>

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Command: morph</title>
<style>@media (prefers-color-scheme: dark) { :root { color-scheme: dark; } }</style>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRighticon" title="User Guide Index"/></a>

<h3><a href="../index.html#commands">Command</a>: morph</h3>

<h3 class="usage"><a href="usageconventions.html">Usage</a>:
<br><b>morph</b>
&nbsp;<a href="atomspec.html#hierarchy"><i>model-spec</i></a>&nbsp;
[&nbsp;<b>wrap</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>frames</b>&nbsp;&nbsp;<i>N</i>&nbsp;]
[&nbsp;<b>same</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>hideModels</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<b>play</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<b>slider</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
&nbsp;<a href="#morph-parameters"><i>morph-parameters</i></a></h3>
<p>
The <b>morph</b> command creates a 
<a href="../trajectories.html">trajectory</a> that morphs between two or
more atomic models. The trajectory can be played back
with the <a href="coordset.html"><b>coordset</b></a> command
or the <a href="#slider">slider</a> graphical interface.
See also: 
<a href="volume.html#morph"><b>volume morph</b></a>,
<a href="../movies.html">making movies</a>, 
the <a href="https://www.rbvi.ucsf.edu/chimerax/features.html#morphing"
target="_blank">ChimeraX morphing highlight</a>, movie tutorials:
[<a href="https://www.rbvi.ucsf.edu/chimerax/tutorials.html#movies-atomic"
target="_blank">2019</a>]
[<a href="https://www.rbvi.ucsf.edu/chimerax/tutorials.html#movies-volume"
target="_blank">2017</a>]
</p><p>
The method is based on that used by the
<a href="http://www2.molmovdb.org/wiki/info/index.php/Morph_Server"
target="_blank">Yale Morph Server</a>:
</p>
<blockquote>
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC102811/" target="_blank">
The morph server: a standardized system for analyzing and visualizing 
macromolecular motions in a database framework.</a>
Krebs WG, Gerstein M.
<i>Nucleic Acids Res.</i> 2000 Apr 15;<b>28</b>(8):1665-75.
</blockquote>
<p>
Before morphing, the atomic models should be superimposed
(such as with <a href="matchmaker.html"><b>matchmaker</b></a>).
In the <b>morph</b> command, the models should be specified in the order
desired for morphing.
Although the same model number cannot be given more than once, 
using <b>wrap true</b> indicates the morph should also go from the
last specified model to the first. For a more complicated trajectory
that visits the same conformation more than once, another approach is to open
a given structure multiple times to give copies with different model numbers.
Examples:
</p>
<blockquote><b>
morph #3,1,2 frames 40
<br>
morph #1.1-20 same t
<br>
morph #2,5 core 0.1 frames 75 wrap t
</b></blockquote>
<p>
Only the <a href="#pairing">atoms in common</a>
among the input structures will be included in the resulting trajectory.
The models can have different numbers of residues or different
sequences (homologs or mutants can be compared), but currently
they must contain equal numbers of chains
(see <a href="#limitations">limitations</a>).
</p><p>
<a name="stage"></a>
Each sequential pair of input structures serves as the
starting and ending points of one <b><i>stage</i></b> of the
trajectory, and a morph trajectory can have multiple stages.
Within each stage, intermediate sets of coordinates are generated:
<ol>
<li><a href="#hinges">Hinges are identified</a> to divide the set of
<a href="#pairing">atoms in common</a> into different groups.
<li>For each group of atoms, intermediates are generated by rigid-body
rotation and translation: each intermediate represents an appropriate
fraction of (a) the total path between the starting and ending centroid positions,
and (b) the total rotation to best fit the starting and ending structures
when their centroids are superimposed. Centroids are not mass-weighted.
Details of the movement path depend on the <a href="#method"><b>method</b></a>.
<li>The atoms within each group are moved in a non-rigid way,
according to the <a href="#cartesian"><b>cartesian</b></a> option
(<a href="#details">details...</a>).
</ol>
<p>
The endpoint and intermediate coordinate sets are opened as a new
<a href="../trajectories.html">trajectory</a> model. 
The <b>frames</b> option (default <i>N</i> = <b>50</b>) indicates generating 
<i>N</i>&ndash;1 intermediate coordinate sets per <a href="#stage">stage</a>,
giving a total number of frames in the trajectory
= 1 + <i>N</i>(number of stages).
</p><p>
Unless <a href="#hinges">colored to show hinge locations</a>,
a single-chain morph trajectory will exhibit its
<a href="colortables.html#autocolors">unique model color</a>, whereas
a multichain morph trajectory will retain the coloring scheme of the
first model used in its creation.
</p><p>
<a name="same"></a>
The <b>same</b> option indicates that the input structures 
have the same sequences, atom names, residue numbering, and chain IDs, 
as is typically the case for members of an NMR ensemble 
(default <b>false</b>, atoms paired as described 
<a href="#pairing">below</a>). Setting this option to <b>true</b> may 
make the calculation slightly faster, but otherwise there is little harm
in staying with the default if there is any doubt as to 
whether numbering/naming are the same.
</p><p>
The <b>hideModels</b> option (default <b>true</b>) hides the input models,
and the <b>play</b> option 
(default <b>true</b>) plays through the resulting trajectory a single time. 
Regardless of the <b>play</b> option, the
<a href="coordset.html"><b>coordset</b></a> command 
and/or the <a href="#slider">slider</a> can be used to replay the trajectory.
</p><p>
<a name="slider"></a>
The <b>slider</b> option (default <b>true</b>) shows a graphical interface
for trajectory playback.
The slider can be dragged or a frame number entered directly.
The interface also includes a play/pause button, a
&nbsp;<img style="height:16px; margin: 1px -1px; vertical-align:top"
src="../tools/ui-icons/snail.png">&nbsp; value to increase for slower playback,
and a button for recording a movie
(<img class="iconbar" src="../tools/ui-icons/record.png">).
Sequential integers are added to the movie filename (movie1.mp4, movie2.mp4, ...)
so that repeated recordings will not overwrite the previous ones, and the save
location can be set with the <a href="snapshot.html"><b>snapshot</b></a> command.
The movie will start at the current slider position, so
to include the whole morph, place the slider at the far left
before clicking the record button.
</p><p>
The <b>Loop Playback</b> setting in the slider
<a href="../window.html#context-menu">context menu</a> controls whether
interactive playback should continue until explicitly paused (initially on).
The <b>Bounce Playback</b> setting in the slider
<a href="../window.html#context-menu">context menu</a> is only available
when <b>Loop Playback</b> is on, and controls whether looping wraps from
end to beginning so that playback is only in the forward direction
(initial setting, <b>Bounce Playback</b> off) or alternates between forward
and backward (<b>Bounce Playback</b> on). These loop/bounce settings apply only
to interactive viewing, not recording a movie with the button mentioned above.
See also:
<a href="coordset.html#slider"><b>coordset slider</b></a>
</p>

<a name="morph-parameters"></a>
<p class="nav">
[<a href="#top">back to top: morph</a>]
</p>
<h3>Morphing Parameters</h3>
<blockquote>
<a name="method"></a>
<b>method</b>&nbsp;(&nbsp;linear&nbsp;|&nbsp;<b>corkscrew</b>&nbsp;)
<br>
How to calculate the rigid-body transformations
within a <a href="#stage">stage</a>:
  <ul>
  <li><b>linear</b>
&ndash; rotate the group of atoms around its centroid and translate it along 
the straight line between the initial and final positions of the centroid
  <li><b>corkscrew</b> (default)
&ndash; move the group of atoms along a circular arc in the plane perpendicular 
to the rotation axis, and linearly along the rotation axis. 
When moving along the circular arc, the group rotates at the same rate 
as the tangent to the arc. The combination of circular and linear motion 
makes the centroid traverse a helical path between its initial and final
positions. In general, if the rotation angle is small, the helical path 
will be nearly straight (a small segment on a cylinder of large diameter).
In the case of the largest possible angle, 180&deg;, the helical path 
turns by half a circle, and the helical axis passes through
a point halfway between the initial and final centroids.
  </ul>
</blockquote>
<blockquote>
<a name="cartesian"></a>
<b>cartesian</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
Whether coordinates within a group of atoms
should be linearly interpolated in Cartesian (x,y,z coordinate) space.
Linear interpolation in Cartesian space simply moves each atom 
in a straight line between its starting and ending positions.
Otherwise, internal coordinates will be used for the
interpolation where possible (<a href="#details">details...</a>).
Using internal coordinates is slower
but produces less distortion. A trajectory made with Cartesian forcing
may be acceptable if few atomic details will be shown (for example,
if only ribbons will be displayed).
</blockquote>
<blockquote>
<b>rate</b>&nbsp;&nbsp;<i>function</i>
<br>
How conformational changes within a <a href="#stage">stage</a>
are distributed across the trajectory, where the <i>function</i> can be:
  <ul>
  <li><b>linear</b> (default)
  &ndash; coordinate changes will be distributed approximately evenly
  <li><b>ramp down</b>
  &ndash; coordinates will change most rapidly near the starting conformation
  <li><b>ramp up</b>
  &ndash; coordinates will change most rapidly near the ending conformation
  <li><b>sinusoidal</b> &ndash; coordinates will change most rapidly halfway
  between the starting and ending conformations
  </ul>
</blockquote>
<blockquote>
<a name="hinges"></a>
<b>coreFraction</b>&nbsp;&nbsp;<i>f</i>
<br>
<b>minHingeSpacing</b>&nbsp;&nbsp;<i>M</i>
<br>
<b>colorSegments</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
<b>colorCore</b>&nbsp;&nbsp;<a href="colornames.html"><i>color</i></a>
<br><br style="line-height:5px">
Hinges are identified using a reimplementation of the method described in
<a href="https://www.ncbi.nlm.nih.gov/pubmed/10734184" target="_blank">Krebs
and Gerstein</a>, <i>Nucleic Acids Res</i> <b>28</b>:1665 (2000).
The <b>coreFraction</b> <i>f</i> (default <b>0.5</b>) is the
fraction of the residues in a chain
to group in the first step of hinge detection.
The best-matching fraction <i>f</i> of residues (best-fitting between
the starting and ending structures) comprises one group
and the remaining (1&ndash;<i>f</i>) a second group. Going through the
residues in order, hinges are added between consecutive residues where
there is a change from one group to the other, as long as a hinge is
at least <b>minHingeSpacing</b> <i>M</i> residues 
away from the previous hinge (does not create a
segment shorter than <i>M</i> residues, default <b>6</b>).
The <b>colorSegments</b> option
shows the resulting segments in different colors (chosen automatically),
whereas <b>colorCore</b> uses the specified
<a href="colornames.html"><i>color</i></a> for residues in the initial core.
Although segments are recomputed for each <a href="#stage">stage</a> 
of a multistage morph, segment coloring only shows the first set.
</blockquote>

<a name="pairing"></a>
<p class="nav">
[<a href="#top">back to top: morph</a>]
</p>
<h3>Atoms in Common</h3>
<p>
Coordinate sets are generated by
interpolating between starting and ending structures.
Interpolation requires a pairing of atoms in the starting structure
with atoms in the ending structure.
Only atoms common to <b><i>all</i></b> of the input structures
are included in the morph trajectory.
</p><p>
The structures must contain equal numbers of biopolymer chains
(see <a href="#limitations">limitations</a>).
Unless <a href="#same"><b>same</b></a> is <b>true</b>, atoms are paired
as follows:
</p>
<blockquote>
<p>
If the sets of chain IDs are identical (for example, each structure contains
chains A and B), the IDs will be used to indicate pairing; if they differ
(for example, one structure contains chains A and B, whereas another contains
chains A and D), the chains will be paired by their order of occurrence
in the input files.
Residues are paired by aligning the chain sequences.
The sequence alignment is performed using the
<a href="matchmaker.html"><b>matchmaker</b></a> defaults
(Needleman-Wunsch algorithm, BLOSUM-62 matrix, secondary structure reassignment
with <a href="dssp.html"><b>dssp</b></a>,
30% secondary structure scoring, <i>etc.</i>),
except that the Nucleic matrix is used for nucleic acids.
Only the sequence alignment stage of
<a href="matchmaker.html"><b>matchmaker</b></a> is
performed, not the superposition of structures.
</p><p>
HET residues are paired if they are covalently bonded to the same (equivalent)
atoms in already paired residues, including other HET paired residues.
<!--
A <a href="../pseudobonds.html">pseudobond</a> such as a 
<a href="distance.html">distance measurement</a>
between the biopolymer chain and a noncovalently bound HET residue
can be added to each structure to force including the noncovalently bound
residue in the morph.
The <a href="../pseudobonds.html">pseudobond</a> must connect the same
(paired) atoms in the two structures.
-->
</p><p>
Once residues are paired, atoms in common within those residues are paired.
In paired residues of the same type, atom pairing is straightforward.
In paired residues of different types, only atoms with the same names
are paired, and only a single connected fragment is kept per residue.
For example (disregarding hydrogens),
phenylalanine and tyrosine have in common all atoms of phenylalanine.
</p>
</blockquote>
<p>
To summarize, HET residues such as ligands and ions are only included
if at least one of the following applies:
</p>
<ul>
<li>the entire atomic models have identical sets of atoms with the 
same atom names, residue numbers, and chain IDs
<li><a href="#same"><b>same</b></a> is <b>true</b> and the HET residues
in question are present in all of the input structures with the same 
atom names, residue names, residue numbers, and chain IDs 
(even if other atoms are different)
<li>the HET residues are bonded to the paired chains either directly or
indirectly through other paired HET residues
</ul>

<a name="limitations"></a>
<p class="nav">
[<a href="#top">back to top: morph</a>]
</p>
<h3>Limitations</h3>
<p>
<b>Secondary structure is not automatically re-evaluated during playback</b>.
Protein <a href="cartoon.html">cartoon</a> (ribbon) display
depends on which residues are in helices and strands,
but these secondary structure assignments are not recomputed automatically
as the coordinates change.  The <b>computeSs</b> option of
<a href="coordset.html"><b>coordset</b></a> or an
<a href="coordset.html#ssnote">alternative approach</a> should be used to
update the assignments.
</p>
<p>
<b>Structures with different numbers of chains are not handled</b>.
Currently, a morph trajectory can only be generated from input structures
with equal numbers of biopolymer chains.
Extra chains in the input models should be
<a href="delete.html">deleted</a> beforehand
or <a href="split.html">split</a> into separate models not used in morphing.
</p>
<p>
<b>Sequences should be easy to align</b>.
The sequences of the structures must be
aligned to determine the <a href="#pairing">atoms in common</a>
for interpolation.  When the sequences are dissimilar,
parts of the sequence alignment may be wrong, leading to a jumbled
and unattractive morph trajectory.
</p>

<a name="details"></a>
<p class="nav">
[<a href="#top">back to top: morph</a>]
</p>
<h3>Interpolating Cartesian or Internal Coordinates</h3>
<p>
Linear interpolation in Cartesian (x,y,z coordinate) space simply moves 
each atom in a straight line between its starting and ending positions.
Even with <a href="#cartesian"><b>cartesian false</b></a>, which  means
to interpolate dihedral angles where possible, many atoms will still
be linearly interpolated in Cartesian space. These atoms include:
</p>
<ul>
<li>for any residues bonded to other residues, every atom that is
participating in an interresidue bond
<li>for residues not bonded to other residues, the first atom of the residue
<li>three atoms in every residue that contains at least three atoms; only 
with the positions of these three set by interpolation in Cartesian space can
the positions of any remaining atoms in the residue be set by interpolating
dihedral angles. For example, in an amino acid in the middle of a peptide chain,
backbone atoms N, CA, and C are first interpolated in Cartesian space, and then
the remaining atoms in the residue are positioned by interpolating 
intraresidue dihedral angles, moving outward so that
three previously positioned atoms are always used to position the next atom.
</ul>
<p>
With <a href="#cartesian"><b>cartesian true</b></a>, the positions of
all of the atoms will be interpolated in Cartesian space.
</p>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics /
September 2024</address>
</body></html>
