<html>

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Command: open</title>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRight" title="User Guide Index"/></a>

<h3><a href="../index.html#commands">Command</a>: open</h3>

<h3><a href="usageconventions.html">Usage</a>: <br><b>open formats</b>
</h3>
<h3><a href="usageconventions.html">Usage</a>:
<br><b>open</b> &nbsp;<i>filename</i>&nbsp;
[&nbsp;format&nbsp;&nbsp;<a href="#file-table"><i>format-name</i></a>&nbsp;]
[&nbsp;coordsets&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
[&nbsp;vseries&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<b>alignment</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
&nbsp;<a href="#options"><i>general-options</i></a>&nbsp;
&nbsp;<a href="#trajectory-options"><i>trajectory-options</i></a>
</h3>
<h3><a href="usageconventions.html">Usage</a>:
<br><b>open</b> &nbsp;<i>URL</i>&nbsp;
[&nbsp;format&nbsp;&nbsp;<a href="#file-table"><i>format-name</i></a>&nbsp;]
[&nbsp;coordsets&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
[&nbsp;vseries&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<b>alignment</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
&nbsp;<a href="#options"><i>general-options</i></a>&nbsp;
&nbsp;<a href="#trajectory-options"><i>trajectory-options</i></a>
</h3>
<h3><a href="usageconventions.html">Usage</a>:
<br><b>open</b> 
[&nbsp;<a href="#fetch-table"><i>prefix</i></a>:]<i>identifier</i>&nbsp;
[&nbsp;fromDatabase&nbsp;&nbsp;<a href="#fetch-table"><i>source</i></a>&nbsp;]
[&nbsp;format&nbsp;&nbsp;<a href="#fetch-table"><i>format-name</i></a>&nbsp;]
[&nbsp;coordsets&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
[&nbsp;ignoreCache&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
&nbsp;<a href="#options"><i>general-options</i></a>
</h3>
<p>
The command <b>open</b> reads data from:
<ul>
<li><a href="#local">local files</a>
<li><a href="#url">URLs</a>
<li><a href="#fetch">online databases</a>
</ul>
Input files may contain data to be displayed or commands or code to be executed.
Local files can also be browsed/opened with <b>File... Open</b> in the menu,
in which case the proper filename <a href="#local">suffix</a> is required.
See also: 
<a href="close.html"><b>close</b></a>,
<a href="save.html"><b>save</b></a>,
<a href="help.html"><b>help</b></a>,
<a href="sym.html"><b>sym</b></a>,
<a href="info.html"><b>info</b></a>,
<a href="../tools/filehistory.html"><b>File History</b></a>,
<b>Rapid Access</b> <img class="icon" src="../rapidoff.png"></a>,
<a href="colortables.html#autocolors">initial colors</a>,
<a href="../autostyle.html">automatic styling</a>
</p><p>
The <b>open formats</b> command lists the possible formats and databases
in the <a href="../tools/log.html"><b>Log</b></a>. 
</p><p>
Examples:
</p>
<blockquote>
<b>open ~/Desktop/maps/myfile.mrc name refined3</b>
<br>
<b>open 2gbp</b>
<br>
<b>open 1plx format pdb coord T</b>
<br>
<b>open emdb:3087</b>
<br>
<b>open 3237 from emdb name "gamma-secretase complex"</b>
<br>
<b>open http://www.rbvi.ucsf.edu/chimerax/features.html</b>
</blockquote>

<a name="local"></a>
<p class="nav">
[<a href="#top">back to top</a>]
</p>
<h3>Local Files</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>open</b> &nbsp;<i>filename</i>&nbsp;
[&nbsp;<b>format</b>&nbsp;&nbsp;<a href="#file-table"><i>format-name</i></a>&nbsp;]
[&nbsp;<b>coordsets</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>vseries</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<b>alignment</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
&nbsp;<a href="#options"><i>general-options</i></a>&nbsp;
&nbsp;<a href="#trajectory-options"><i>trajectory-options</i></a>
</blockquote>
<p>
File types can be categorized as:
</p>
<ul>
<li><a href="#atomic">Atomic Structures and Related Data</a>
<li><a href="#trajectory">Trajectory Coordinates</a>
<li><a href="#volume">Volume Data and Related</a>
<li><a href="#object">Other 3D Objects</a>
<li><a href="#sequence">Sequences</a>
<li><a href="#composite">Composite</a>
<li><a href="#miscellaneous">Miscellaneous</a>
</ul>
<p>
The <i>filename</i> is generally a pathname to a local file, 
either absolute or relative to the current working directory as
reported by <a href="pwd.html"><b>pwd</b></a>.
The working directory can be changed with <a href="cd.html"><b>cd</b></a>.
A pathname may start with &ldquo;<b>~</b>&rdquo; to indicate the user's 
home directory. One or more files can be specified at a time by including 
<b>*</b> (wild card), <b>?</b> (single-character wild card), 
and/or square-bracketed character ranges in the <i>filename</i>
(Unix-style pathname expansion or <b><i>globbing</i></b>).
</p><p>
The file format can be indicated with either the <i>filename</i> suffix or
the <b>format</b> option (where <i>format-name</i> can be truncated).
Files that are gzipped, as indicated by their additional .gz suffix, 
can also be read.
</p><p>
A multi-model file in PDB format can be opened as a
<a href="../trajectories.html">trajectory</a>
with <b>coordsets true</b> (default <b>false</b>).
A <a href="#trajectory">trajectory coordinate</a> file is
automatically opened as <a href="../trajectories.html">trajectory</a>,
but to provide topology information, the <b>model</b> option must be used 
to specify an existing model with the same set of atoms
(for example: <b>open myfile.trr model #3</b>).
The coordinates of the existing model will be replaced with the new,
unless appending is indicated with <b>replace false</b>.
</p><p>
When multiple maps with the same grid dimensions are opened 
in a single command (multiple files specified with wild cards or globbing, 
or a single multi-map file), the <b>vseries</b> option can be used to enforce 
loading them as a <a href="../trajectories.html#vseries">volume series</a>
for playback with 
<a href="vseries.html"><b>vseries</b></a> or the associated slider
(if <b>true</b>), or as independent models (if <b>false</b>).
If the option is not used, 3D TIFF images and
<a href="save.html#chimap">Chimera map</a> files containing five or more
same-sized maps are automatically interpreted as series, if possible.
</p><p>
When a multi-sequence file is opened, the <b>alignment</b> option 
indicates whether it should be interpreted as an alignment 
(default <b>true</b>, all sequences in one <b>Multalign Viewer</b> window).
Otherwise, each sequence will be opened in a separate window. 
If the sequences (including any gap characters) are of differing lengths,
the option must be set to <b>false</b>. 
</p>
<center>
<a name="file-table"><font size="+1"><b>File Types</b></font></a>
</center>
<table border cellpadding="4" cellspacing="0">
<tr>
<td align="center" style="background-color:#cccccc"><b>format</b></td>
<td align="center" style="background-color:#cccccc"><b><i>format-name</i></b></td>
<td align="center" style="background-color:#cccccc"><b>suffix</b></td>
<td align="center" style="background-color:#cccccc"><b>description</b></td>
<tr><tr>
<td colspan=4 align="center" style="background-color:#eaeaea">
<a name="atomic"></a>
<b>Atomic Structures and Related Data</b></td>
</tr><tr>
<td align="center"><a href="http://www.rbvi.ucsf.edu/chimera/docs/UsersGuide/tutorials/framepdbintro.html" 
target="_blank">Protein Data Bank (PDB)</a></td>
<td align="center"><b>pdb</b></td>
<td align="center">.pdb<br>.pdb1<br>.ent<br>.pqr</td>
<td align="center">atomic coordinates and associated annotations</td>
</tr><tr>
<td align="center"><a href="http://mmcif.wwpdb.org/docs/faqs/pdbx-mmcif-faq-general.html"
target="_blank">PDBx/mmCIF</a></td>
<td align="center"><b>mmcif</b></td>
<td align="center">.cif<br>.mmcif</td>
<td align="center">atomic coordinates and associated annotations
<br>(see <a href="sym.html#assembly">assemblies from mmCIF</a>)</td>
</tr><tr>
<td align="center"><a href="../pseudobonds.html">Pseudobond</a> file</td>
<td align="center"><b>pseudobonds</b></td>
<td align="center">.pb</td>
<td align="center">
plain text file with each line <a href="atomspec.html">specifying</a>
a pair of atoms (already open in ChimeraX)
to connect with <a href="../pseudobonds.html">pseudobonds</a></td>
</tr><tr>
<td colspan=4 align="center" style="background-color:#eaeaea">
<a name="trajectory"></a>
<b>Trajectory Coordinates</b></td>
</tr><tr>
<td align="center">DCD coordinates</a> (binary)</td>
<td align="center"><b>dcd</b></td>
<td align="center">.dcd</td>
<td align="center"><a href="../trajectories.html">trajectory</a> coordinates;
<br>existing model provides topology
(<a href="#trajectory-options">details...</a>)</td>
</tr><tr>
<td align="center"><a href="http://manual.gromacs.org/online/xtc.html"
target="_blank">Gromacs compressed coordinates</a></td>
<td align="center"><b>xtc</b></td>
<td align="center">.xtc</td>
<td align="center"><a href="../trajectories.html">trajectory</a> coordinates;
<br>existing model provides topology
(<a href="#trajectory-options">details...</a>)</td>
</tr><tr>
<td align="center"><a href="http://manual.gromacs.org/online/trr.html"
target="_blank">Gromacs full-precision coordinates</a> (binary)</td>
<td align="center"><b>trr</b></td>
<td align="center">.trr</td>
<td align="center"><a href="../trajectories.html">trajectory</a> coordinates;
<br>existing model provides topology
(<a href="#trajectory-options">details...</a>)</td>
<tr><tr>
<td colspan=4 align="center" style="background-color:#eaeaea">
<a name="volume"></a>
<b>Volume Data and Related</b> (associated tool: 
<a href="../tools/volumeviewer.html"><b>Volume Viewer</b></a>)</td>
</tr><tr>
<td align="center"><a href="https://www.fei.com/software/amira-for-life-sciences/"
target="_blank">Amira</a> mesh</td>
<td align="center"><b>amira</b></td>
<td align="center">.am</td>
<td align="center">3D data</td>
</tr><tr>
<td align="center">
<a href="http://www.poissonboltzmann.org/docs/file-format-info/"
target="_blank">Adaptive Poisson-Boltzmann Solver (APBS) potential</a>
<br>(see the
<a href="http://www.rbvi.ucsf.edu/chimera/docs/ContributedSoftware/apbs/apbs.html"
target="_blank"><b>APBS</b></a> interface in 
<a href="http://www.rbvi.ucsf.edu/chimera/"
target="_blank"><b>UCSF Chimera</b></a>)</td>
<td align="center"><b>apbs</b></td>
<td align="center">.dx</td>
<td align="center">electrostatic potential map</td>
</tr><tr>
<td align="center">BRIX density map</td>
<td align="center"><b>dsn6</b></td>
<td align="center">.brix</td>
<td align="center">density map</td>
</tr><tr>
<td align="center">
<a href="http://www.ccp4.ac.uk/html/maplib.html" target="_blank">CCP4</a>
density map</td>
<td align="center"><b>ccp4</b></td>
<td align="center">.ccp4<br>.map</td>
<td align="center">density map</td>
</tr><tr>
<td align="center">Chimera/ChimeraX map
<br>(based on the hierarchical data format
<a href="http://www.hdfgroup.org/HDF5/" target="_blank">HDF5</a>)</td>
<td align="center"><b>cmap</b></td>
<td align="center">.cmp<br>.cmap</td>
<td align="center">density map</td>
</tr><tr>
<td align="center">Directional FSC (Fourier shell correlation)</td>
<td align="center"><b>dres</b></td>
<td align="center">.dres</td>
<td align="center">directional resolution to show on a colored globe
<br>(plain text file with four whitespace-separated columns of numbers,
an XYZ vector and a resolution or other value
to colormap onto the globe)</td>
</tr><tr>
<td align="center">DSN6 density map</td>
<td align="center"><b>dsn6</b></td>
<!-- dsn6 format name is shared, program figures out which variant -->
<td align="center">.omap</td>
<td align="center">density map</td>
</tr><tr>
<td align="center">Image stack
<br>(grayscale image series in multiple files or a single multipage file,
in formats supported by
<a href="http://www.pythonware.com/products/pil/" target="_blank">PIL</a>)</td>
<td align="center"><b>images</b></td>
<td align="center">.tif, .tiff<br>.png<br>.pgm</td>
<td align="center">density map</td>
</tr><tr>
<td align="center">
<a href="http://www.bitplane.com/imaris" target="_blank">Imaris</a> map</td>
<td align="center"><b>ims</b></td>
<td align="center">.ims</td>
<td align="center">3D-5D image<br>(could be time series and/or multichannel)</td>
</tr><tr>
<td align="center">
<a href="http://www2.mrc-lmb.cam.ac.uk/research/locally-developed-software/image-processing-software/#image"
target="_blank">MRC</a> density map</td>
<td align="center"><b>mrc</b></td>
<td align="center">.mrc</td>
<td align="center">density map</td>
<tr><tr>
<td align="center"><a href="http://huanglab.ucsf.edu/STORM.html"
target="_blank">STochastic Optical Reconstruction Microscopy (STORM)</a></td>
<td align="center"><b>storm</b></td>
<td align="center">.storm</td>
<td align="center">super-resolution optical microscopy data</td>
<tr><tr>
<td colspan=4 align="center" style="background-color:#eaeaea">
<a name="object"></a>
<b>Other 3D Objects</b></td>
</tr><tr>
<td align="center"><a href="http://www.khronos.org/collada/"
target="_blank">COLLADA</a></td>
<td align="center"><b>collada</b></td>
<td align="center">.dae</td>
<td align="center">graphical objects<br>(geometry nodes define surfaces)</td>
</tr><tr>
<td align="center"><a href="http://en.wikipedia.org/wiki/STL_%28file_format%29#Binary_STL"
target="_blank">STereoLithography (binary)</a></td>
<td align="center"><b>stl</b></td>
<td align="center">.stl</td>
<td align="center">triangle-based format (binary)
native to stereolithography CAD software
<tr><tr>
<td align="center"><a href="http://vtk.org/" 
target="_blank">Visualization Toolkit (VTK) PolyData</a></td>
<td align="center"><b>vtk</b></td>
<td align="center">.vtk</td>
<td align="center">lines, polygons</td>
<tr><tr>
<td colspan=4 align="center" style="background-color:#eaeaea">
<a name="sequence"></a>
<b>Sequences</b> (associated tool: <b>Multalign Viewer</b>)</td>
<tr><tr>
<td align="center">FASTA, aligned FASTA</td>
<td align="center"><b>fasta</b></td>
<td align="center">.fasta, .fa<br>.afasta, .afa</td>
<td align="center">individual sequence(s) or sequence alignment</td>
<tr><tr>
<td align="center">Multiple Sequence Format (MSF)</a></td>
<td align="center"><b>msf</b></td>
<td align="center">.msf</td>
<td align="center">sequence alignment</td>
<tr><tr>
<td align="center">PIR, aligned PIR</a></td>
<td align="center"><b>pir</b></td>
<td align="center">.pir<br>.ali</td>
<td align="center">individual sequence(s) or sequence alignment</td>
<tr><tr>
<td colspan=4 align="center" style="background-color:#eaeaea">
<a name="composite"></a>
<b>Composite</b></td>
</tr><tr>
<td align="center"><a href="save.html">ChimeraX session file</a></td>
<td align="center"><b>ses</b></td>
<td align="center">.cxs</td>
<td align="center">ChimeraX session</td>
</tr><tr>
<td align="center"><a href="https://pdb-dev.rcsb.rutgers.edu/about/"
target="_blank">Integrative Hybrid Model (IHM)</a>
</td>
<td align="center"><b>ihm</b></td>
<td align="center">.ihm</td>
<td align="center">integrative hybrid model, potentially including
atomic and coarse-grained representations, crosslinks, <i>etc.</i>
(examples can be downloaded from the
<a href="https://pdb-dev.rcsb.rutgers.edu/" target="_blank">archive</a>)</td>
<tr><tr>
<td colspan=4 align="center" style="background-color:#eaeaea">
<a name="miscellaneous"></a>
<b>Miscellaneous</b></td>
</tr><tr>
<td align="center"><a href="usageconventions.html#cxc-files">ChimeraX
command file</a></td>
<td align="center"><b>cmd</b></td>
<td align="center">.cxc</td>
<td align="center">
<a href="../index.html#commands">ChimeraX commands</a></td>
</tr><tr>
<td align="center">Python</td>
<td align="center"><b>py</b></td>
<td align="center">.py<br>.pyc<br>.pyo</td>
<td align="center">Python code</td>
</tr><tr>
<td align="center">HTML, XHTML</td>
<td align="center"></td>
<td align="center">.html, .htm<br>.xhtml,&nbsp;.xht,&nbsp;.xml</td>
<td align="center">
hypertext to render in the 
<a href="../window.html#browser">ChimeraX browser</a>
</tr></table>

<a name="url"></a>
<p class="nav">
[<a href="#top">back to top</a>]
</p>
<h3>URLs</h3>

<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>open</b> &nbsp;<i>URL</i>&nbsp;
[&nbsp;<b>format</b>&nbsp;&nbsp;<a href="#file-table"><i>format-name</i></a>&nbsp;]
[&nbsp;<b>coordsets</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>vseries</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<b>alignment</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
&nbsp;<a href="#options"><i>general-options</i></a>&nbsp;
&nbsp;<a href="#trajectory-options"><i>trajectory-options</i></a>
</blockquote>
<p>
Uniform Resource Locators (URLs) starting with
<b>http:</b>, <b>https:</b>, and <b>ftp:</b> are handled.
File types and command options are as described for 
<a href="#local">local files</a>.
The file type can be indicated by suffix, MIME type, or the <b>format</b> option.
</p><p>
Web-fetched data other than those handled by the
<a href="../window.html#browser">ChimeraX browser</a>
are stored locally as described <a href="#fetch">below</a>.
</p>

<a name="fetch"></a>
<p class="nav">
[<a href="#top">back to top</a>]
</p>
<h3>Fetch from Online Databases</h3>

<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>: <b>open</b> 
[&nbsp;<a href="#fetch-table"><i>prefix</i></a>:]<i>identifier</i>&nbsp;
[&nbsp;<b>fromDatabase</b>&nbsp;&nbsp;<a href="#fetch-table"><i>source</i></a>&nbsp;]
[&nbsp;<b>format</b>&nbsp;&nbsp;<a href="#fetch-table"><i>format-name</i></a>&nbsp;]
[&nbsp;<b>coordsets</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>ignoreCache</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
&nbsp;<a href="#options"><i>general-options</i></a>
</blockquote>
<p>
An entry can be fetched from a database if its <i>identifier</i> is supplied.
The source database can be indicated with either a <i>prefix</i>
before the identifier or the <b>fromDatabase</b> option.
The default database for a 4-character identifier without prefix is the
<td align="center"><a href="http://www.rcsb.org/"
target="_blank">RCSB Protein Data Bank</a>; the default format
for that database is <b>mmcif</b>, although <b>pdb</b> or <b>mmtf</b>
can be specified with the <b>format</b> option.
</p><p>
Opening a structure in mmCIF format automatically shows a clickable table of
<a href="sym.html#assembly">assembly information</a>
in the <a href="../tools/log.html"><b>Log</b></a>.
</p><p>
A multi-model file in PDB format can be opened as a
<a href="../trajectories.html">trajectory</a>
with <b>coordsets true</b> (default <b>false</b>).
</p><p>
Web-fetched data are stored locally in <b>~/Downloads/ChimeraX/</b> 
(where ~ indicates a user's home directory), with subdirectories
<b>PDB</b> for atomic coordinates from the Protein Data Bank,
<b>EMDB</b> for density maps from the
<td align="center"><a href="http://www.ebi.ac.uk/pdbe/emdb/"
target="_blank">Electron Microscopy Data Bank</a>, <i>etc.</i>
For caching, no distinction is made among the different Protein Data Bank sites.
If a data file specified for opening is not found in this local cache or 
<b>ignoreCache</b> is set to <b>true</b>, the file will be fetched and cached. 
<!--
ChimeraX preferences, command history, file history, and package installation 
files are stored in a standard platform-dependent location
(for example, <b>~/Library/Caches/ChimeraX</b> on a Mac;
see <a href="http://pypi.python.org/pypi/appdirs" 
target="_blank">python appdirs</a>). 
-->
</p>
<center>
<a name="fetch-table"><font size="+1"><b>Fetch Sources</b></font></a>
</center>
<table border cellpadding="4" cellspacing="0"><tr>
<td align="center" style="background-color:#cccccc"><b>database</b></td>
<td align="center" style="background-color:#cccccc"><b><i>prefix</i></b></td>
<td align="center" style="background-color:#cccccc"><b><i>source</i></b></td>
<td align="center" style="background-color:#cccccc"><b>description</b></td>
<td align="center" style="background-color:#cccccc"><b><i>format-name</i></b></td>
</tr><tr>
<td align="center"><a href="http://www.rcsb.org/"
target="_blank">RCSB Protein Data Bank</a></td>
<td align="center"><b>pdb:</b></td>
<td align="center"><b>pdb</b></td>
<td align="center">atomic coordinates and associated annotations</td>
<td align="center"><b>mmcif</b> (default) or <b>pdb</b>
<br>(see <a href="http://www.rcsb.org/pdb/static.do?p=file_formats/index.jsp"
target="_blank">wwPDB formats</a>)
<br>or <a href="http://mmtf.rcsb.org/" target="_blank"><b>mmtf</b></a></td>
</tr><tr>
<td align="center"><a href="http://www.ebi.ac.uk/pdbe/"
target="_blank">Protein Data Bank <br>in Europe (PDBe)</a></td>
<td align="center"><b>pdbe:</b></td>
<td align="center"><b>pdbe</b></td>
<td align="center">atomic coordinates and associated annotations</td>
<td align="center"><b>mmcif</b> (default) or  <b>pdb</b></td>
</td>
</tr><tr>
<td align="center"><a href="http://www.ebi.ac.uk/pdbe/"
target="_blank">PDBe</a></td>
<td align="center"><b>pdbe_updated:</b></td>
<td align="center"><b>pdbe_updated</b></td>
<td align="center">atomic coordinates and associated annotations,
&nbsp;&nbsp;updated to include small-molecule connectivity 
and binding-site information as described in 
<a href="http://nar.oxfordjournals.org/content/44/D1/D385.long"
target="_blank">Velankar <i>et al.</i></a>,
<i>Nucleic Acids Res</i> <b>44</b>:D385 (2016)</td>
<td align="center"><b>mmcif</b>
</td>
</tr><tr>
<td align="center"><a href="http://pdbj.org/" target="_blank">Protein 
Data Bank Japan (PDBj)</a></td>
<td align="center"><b>pdbj:</b></td>
<td align="center"><b>pdbj</b></td>
<td align="center">atomic coordinates and associated annotations</td>
<td align="center"><b>mmcif</b></a>
</td>
</tr><tr>
<td align="center"><a href="http://www.ebi.ac.uk/pdbe/eds/"
target="_blank">PDBe Electron Density Server (EDS)</a>
</td>
<td align="center"><b>eds:</b></td>
<td align="center"><b>eds</b></td>
<td align="center">(2Fo-Fc) electron density map for a PDB entry;
not available for all PDB entries</td>
<td align="center"><b>ccp4</b></td>
</tr><tr>
<td align="center"><a href="http://www.ebi.ac.uk/pdbe/eds/"
target="_blank">PDBe EDS</a>
</td>
<td align="center"><b>edsdiff:</b></td>
<td align="center"><b>edsdiff</b></td>
<td align="center">(Fo-Fc) electron density difference map for a PDB entry;
not available for all PDB entries</td>
<td align="center"><b>ccp4</b></td>
</tr><tr>
<td align="center"><a href="http://www.ebi.ac.uk/pdbe/emdb/"
target="_blank">Electron Microscopy Data Bank</a>
<br>
&nbsp;&nbsp;(<a href="ftp://ftp.wwpdb.org/pub/emdb/README"
target="_blank">ftp</a> from wwpdb.org to .edu/.gov hosts,
Chinese mirror site to .cn hosts, otherwise from ebi.ac.uk)&nbsp;
</td>
<td align="center"><b>emdb:</b></td>
<td align="center"><b>emdb</b></td>
<td align="center">density map</td>
<td align="center"><b>ccp4</b></td>
</tr><tr>
<td align="center"><a href="http://www.cellpack.org"
target="_blank">cellPACK</a></td>
<td align="center"><b>cellpack:</b></td>
<td align="center"><b>cellpack</b></td>
<td align="center">hierarchical model;
see <a href="http://www.cellpack.org/use/" target="_blank">cellPACK
Chimera instructions</a> for available identifiers</td>
<td align="center"><b>cellpack</b></td>
</tr></table>

<a name="options"></a>
<p class="nav">
[<a href="#top">back to top</a>]
</p>
<h3>General Options</h3>

<blockquote>
<b>name</b>&nbsp;&nbsp;<i>model-name</i>
<br>
The name of the model (shown in the <b>Model Panel</b>) will be generated 
automatically from the input <i>filename</i> or database <i>identifier</i>
unless another name is given with the <b>name</b> option.
If <i>model-name</i> contains spaces, it should be enclosed in quotation marks.
Model number is also assigned automatically, starting with the lowest available
number. Model name and number can be changed with the
<a href="rename.html"><b>rename</b></a> command.
</blockquote>
<blockquote>
<b>autoStyle</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
Setting <b>autoStyle</b> to <b>false</b> bypasses 
<a href="../autostyle.html">automatic styling</a>, <i>i.e.</i>,
automatically adjusting the appearance of a newly opened atomic model
based on its contents.
Bypassing this calculation shows all atoms as white spheres, and may be 
useful in rare cases where generating the rule-based representation is slow
(such as drawing cartoons for very large numbers of models).
</blockquote>
<blockquote>
<b>autoAssociate</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false
<br>
Whether structure-chain sequences should be compared to sequences in 
<b>Multalign Viewer</b> and automatically associated if the number of
mismatches is no more than 1/10 the number of residues in the structure chain.
Unlike the other options of <b>open</b>, this sets a state that persists 
within the ChimeraX session unless explicitly changed. 
The initial setting is <b>true</b>.
<p>
For automatic association, the sequence and structure files
can be opened in either order.
Gaps in the structure sequence relative to the sequence in the alignment file 
can only occur where residues are missing from the structure (for example, 
a flexible loop with insufficient density to determine coordinates).
<!-- check the following -->
Associations are reported in the <a href="../window.html">status line</a>
and the <a href="../tools/log.html"><b>Log</b></a>.
A structure (even if it has multiple chains) cannot be associated with more
than one sequence, but a single sequence can be associated with
more than one structure.
If more than one sequence matches a given structure chain,
the single best-matching sequence is associated. 
</p>
</blockquote>

<a name="trajectory-options"></a>
<p class="nav">
[<a href="#top">back to top</a>]
</p>
<h3>Trajectory Options</h3>
<p>
A <a href="#trajectory">trajectory coordinate</a> file
must be opened as the same model number as an existing atomic model 
or trajectory, specified with the <b>model #<i>N</i></b> option. 
The existing model is needed to provide topology information,
and should contain the same set of atoms as the new trajectory.
The coordinates of the existing model will be replaced with the new,
unless <b>replace false</b> is used to indicate appending the new frames instead.
</p>
<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics / 
June 2017</address>
</body></html>
