<html lang="en">

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Command: open</title>
<style>@media (prefers-color-scheme: dark) { :root { color-scheme: dark; } }</style>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRighticon" title="User Guide Index"/></a>

<h3><a href="../index.html#commands">Command</a>: open</h3>

<h3 class="usage"><a href="usageconventions.html">Usage</a>:
<br><b>open</b> (&nbsp;<i>filename</i>&nbsp;|&nbsp;<i>URL</i>&nbsp;|&nbsp;[&nbsp;<a href="#fetch-table"><i>prefix</i></a>:]<i>identifier</i>&nbsp;)
[&nbsp;<a href="#maxModels"><b>maxModels</b></a>&nbsp;&nbsp;<i>M</i>&nbsp;]
[&nbsp;<a href="#coordsets"><b>coordsets</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
&nbsp;<a href="#options"><i>options</i></a>&nbsp;
</h3>
<h3 class="usage"><a href="usageconventions.html">Usage</a>: <br><b>open formats</b>
</h3>
<p>
The command <b>open</b> reads data in several
<a href="#formats">input formats</a> from:
<p>
<ul>
<li><a href="#local">local files</a>
<li><a href="#url">URLs</a>
<li><a href="#fetch">online sources</a>;
see also <a href="../fetch.html"><b>Fetch by ID</b></a>
</ul>
<p>
Input files may contain data to be displayed or commands or code to be executed.
</p><p>
For some types of files,
<a href="../modelinfo.html">associated information</a> is automatically
shown in the <a href="../tools/log.html"><b>Log</b></a> when the file is read.
See also: 
<a href="runscript.html"><b>runscript</b></a>,
<a href="alphafold.html"><b>alphafold</b></a>,
<a href="close.html"><b>close</b></a>,
<a href="save.html"><b>save</b></a>,
<a href="help.html"><b>help</b></a>,
<a href="sym.html"><b>sym</b></a>,
<a href="info.html"><b>info</b></a>,
<a href="../tools/filehistory.html"><b>File History</b></a>,
<b>Rapid Access</b>
<img class="iconbar" src="../tools/ui-icons/lightning_night.png">
<img class="iconbar" src="../tools/ui-icons/lightning_day.png">,
<a href="colortables.html#autocolors">initial colors</a>,
<a href="../autostyle.html">automatic styling</a>,
<a href="../../devel/bundles/mmcif/src/mmcif_guidelines.html"
target="_blank">ChimeraX Fast mmCIF Guidelines</a>
</p><p>
Examples:
</p>
<blockquote>
<b>open ~/Desktop/maps/myfile.mrc name refined3</b>
<br>
<b>open 2gbp</b>
<br>
<b>open 1plx format pdb coord T</b>
<br>
<b>open 1iee 2blx from redo</b>
<br>
<b>open emdb:3087</b>
<br>
<b>open 3237 from emdb name "gamma-secretase complex"</b>
<br>
<b>open https://www.rbvi.ucsf.edu/chimerax/features.html</b>
</blockquote>

<a name="formats"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>Input Formats</h3>
<p>
The command <b>open formats</b> command lists the possible formats and 
<a href="#fetch">databases</a> in the 
<a href="../tools/log.html"><b>Log</b></a>. File types can be categorized as:
</p>
<ul>
<li><a href="#atomic">Atomic Structures and Related Data</a>
<li><a href="#trajectory">Trajectory Coordinates</a>
<li><a href="#trajtop">Trajectory Topology</a>
<li><a href="#volume">Volume Data and Related</a>
<li><a href="#segmentation">Segmentations</a>
<li><a href="#object">Other 3D Objects</a>
<li><a href="#sequence">Sequences</a>
<li><a href="#composite">Composite</a>
<li><a href="#miscellaneous">Miscellaneous</a>
</ul>
<center>
<a name="file-table"><font size="+1"><b>File Types</b></font></a>
</center>
<table border cellpadding="4" cellspacing="0">
<tr>
<td align="center" style="color:black; background-color:#cccccc"><b>format</b></td>
<td align="center" style="color:black; background-color:#cccccc"><span class="nowrap"><b><i>format-name</i></b></span></td>
<td align="center" style="color:black; background-color:#cccccc"><b>suffix</b></td>
<td align="center" style="color:black; background-color:#cccccc"><b>description</b></td>
</tr><tr>
<td colspan=4 align="center" class="shaded">
<a name="atomic"></a>
<a href="#formats" class="nounder"><b>&larr;</b></a>
<b>Atomic Structures and Related Data</b></td>
</tr><tr>
<td align="center"><a href="../formats/pdbintro.html">PDB
(Protein Data Bank)</a></td>
<td align="center"><b>pdb</b></td>
<td align="center">.pdb, .pdb1<br>.ent, .pqr
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">atomic coordinates and associated annotations;
<br>if &gt;99,999 atoms, <b>amber</b> or <b>h36</b>
<a href="save.html#pdb">serial numbering</a> must be used;
if <a href="../formats/pdbintro.html#PQR">PQR format</a>,
atomic <a href="../radii.html">radius</a> is set;
see also <a href="altlocs.html">alternate locations</a></td>
</tr><tr>
<td align="center"><a href="https://mmcif.wwpdb.org/docs/faqs/pdbx-mmcif-faq-general.html"
target="_blank">PDBx/mmCIF</a></td>
<td align="center"><b>mmcif</b></td>
<td align="center">.cif<br>.mmcif</td>
<td align="center">atomic coordinates and associated annotations
<br>(see <a href="sym.html#assembly">assemblies from mmCIF</a>,
<a href="hbonds.html#related">H-bonds from mmCIF</a>,
<br><a href="../../devel/bundles/mmcif/src/mmcif_guidelines.html"
target="_blank">ChimeraX Fast mmCIF Guidelines</a>,
and <a href="altlocs.html">alternate locations</a>;
<br>see also the <a href="#fetchEmdbMap"><b>fetchEmdbMap</b></a> option)</td>
</tr><tr>
<td align="center" class="text"><a href="https://mmtf.rcsb.org/"
target="_blank">MMTF (Macromolecular Transmission Format)</a></td>
<td align="center"><b>mmtf</b></td>
<td align="center">.mmtf
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">atomic coordinates and associated annotations</td>
</tr><tr>
<td align="center">
<a href="http://ww1.iucr.org/iucr-top/cif/standard/cifstd1.html" 
target="_blank">Crystallographic Information File (CIF)</td>
<td align="center"><b>corecif<br>smallcif</b></td>
<td align="center">.cif</td>
<td align="center">atomic coordinates and crystallographic information</td>
</tr><tr>
<td align="center">SwissDock</td>
<td align="center"><b>swissdock</b></td>
<td align="center">.pdb</td>
<td align="center" class="text">atomic coordinates of docked ligands
and/or target receptor in the modified PDB format written by
<a href="http://www.swissdock.ch/" target="_blank">SwissDock</a>
<br>
(see <a href="../tools/viewdock.html"><b>ViewDock</b></a>
for analysis of docking results from
<a href="http://www.swissdock.ch/" target="_blank">SwissDock</a>;
<br>see also the <a href="#showTool"><b>showTool</b></a> option)
</td>
</tr><tr>
<td align="center">Sybyl Mol2</td>
<td align="center"><b>mol2</b></td>
<td align="center">.mol2</td>
<td align="center" class="text">atomic coordinates
<br>
(see <a href="../tools/viewdock.html"><b>ViewDock</b></a>
for analysis of docking results from <a href="http://dock.compbio.ucsf.edu/" 
target="_blank">UCSF&nbsp;DOCK</a>&nbsp;v4-6, 
<a href="http://dock.compbio.ucsf.edu/DOCK3.7/" 
target="_blank">DOCK&nbsp;v3.7</a>, or
<a href="https://www.ccdc.cam.ac.uk/solutions/csd-discovery/components/gold/"
target="_blank">GOLD</a>;
<br>see also the <a href="#showTool"><b>showTool</b></a> option)
</td>
</tr><tr>
<td align="center"><a href="http://autodock.scripps.edu/faqs-help/faq/what-is-the-format-of-a-pdbqt-file"
target="_blank">PDBQT</a></td>
<td align="center"><b>pdbqt</b></td>
<td align="center">.pdbqt</td>
<td align="center" class="text">atomic coordinates
<br>
(see <a href="../tools/viewdock.html"><b>ViewDock</b></a>
for analysis of docking results from
<a href="http://vina.scripps.edu/" target="_blank">AutoDock Vina</a>;
<br>see also the <a href="#showTool"><b>showTool</b></a> option)
</td>
</tr><tr>
<td align="center"><a href="https://gromos.net/"
target="_blank">Gromos87</a></td>
<td align="center"><b>gro</td>
<td align="center">.gro</td>
<td align="center">atomic coordinates<br>(not read as a trajectory)</td>
</tr><tr>
<td align="center">
<a href="https://en.wikipedia.org/wiki/Chemical_table_file"
target="_blank">MDL Molfile/SDF</a></td>
<td align="center"><b>sdf</b></td>
<td align="center">.mol, .sdf</td>
<td align="center">atomic coordinates</td>
</tr><tr>
<td align="center">Schr&ouml;dinger Maestro</td>
<td align="center"><b>mae</b></td>
<td align="center">.mae<br>.maegz</td>
<td align="center" class="text">atomic coordinates
<br>
(see <a href="../tools/viewdock.html"><b>ViewDock</b></a>
for analysis of docking results from 
<a href="https://www.schrodinger.com/glide" target="_blank">Glide</a>;
<br>see also the <a href="#showTool"><b>showTool</b></a> option)
</td>
</tr><tr>
<td align="center">
<a href="http://viperdb.scripps.edu/" target="_blank">VIPERdb
(Virus Particle Explorer database)</a></td>
<td align="center"><b>viperdb</b></td>
<td align="center">.vdb</td>
<td align="center">PDB-format asymmetric unit in the VIPERdb
coordinate system; capsid built automatically using 
<a href="sym.html#symmetry"><b>sym</b></a>
with <b>i222</b> symmetry</td>
</tr><tr>
<td align="center">
<a href="https://zdock.umassmed.edu/" target="_blank">ZDOCK</a>
quasi-PDB format</td>
<td align="center"><b>zdock</b>
<br>(required)</td>
<td align="center">.pdb
<br>(insufficient to specify format)</td>
<td align="center">
atomic coordinates
<br>(use <b>format zdock</b> to disambiguate from standard PDB)
</td>
</tr><tr>
<td align="center"><a href="../pseudobonds.html#pbfiles">Pseudobond
file</a></td>
<td align="center" class="text"><b>pseudobonds</b></td>
<td align="center">.pb
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">
list of atom pairs
to connect with <a href="../pseudobonds.html">pseudobonds</a>,
but see also <a href="pbond.html"><b>pbond</b></a>
</td>
</tr><tr>
<td align="center">
<a href="https://link.springer.com/article/10.1007/s10858-018-0220-3"
target="_blank">PDB NMR-STAR</a></td>
<td align="center"><b>nmrstar</b></td>
<td align="center">.str
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center" class="text">nuclear magnetic resonance (NMR) 
distance constraints; will be displayed on atomic model(s) specified
with the <a href="#nmrstar"><b>structures</b></a> option</td>
</tr><tr>
<td align="center">
<a href="#pae">AlphaFold PAE</a></td>
<td align="center"><b>pae</b></td>
<td align="center">.json
<td align="center" class="text">
<a href="https://alphafold.ebi.ac.uk" target="_blank">AlphaFold</a>
predicted aligned error (PAE) matrix, to be shown as an 
<a href="../tools/alphafold.html#pae">interactive plot</a> 
tied to the corresponding AlphaFold model, which should be specified
with the <a href="#pae"><b>structure</b></a> option</td>
</tr><tr>
<td align="center">
<a href="../formats/sms.html">ChimeraX similar structures file</a></td>
<td align="center"><b>sms</b></td>
<td align="center">.sms
<td align="center" class="text">
ChimeraX-specific format containing the results of searching 
structure databases for similar proteins
(<a href="../formats/sms.html">details...</a>)</td>
</tr><tr>
<td colspan=4 align="center" class="shaded">
<a name="trajectory"></a>
<a href="#formats" class="nounder"><b>&larr;</b></a>
<b>Trajectory Coordinates</b>
<br>(also requires trajectory topology, either from being opened at the same
time as a <a href="#trajtop">PSF</a> file, or from an already-open atomic model
given with the <a href="#structureModel"><b>structureModel</b></a> option;
see also <a href="#traj-options">trajectory options</a>)</td>
</tr><tr>
<td align="center"><a href="https://ambermd.org/netcdf/nctraj.xhtml" 
target="_blank">Amber NetCDF</a> (binary)</td>
<td align="center"><b>amber</b></td>
<td align="center">.nc</td>
<td align="center">atomic <a href="../trajectories.html">trajectory</a>
coordinates</td>
</tr><tr>
<td align="center">DCD coordinates (binary)</td>
<td align="center"><b>dcd</b></td>
<td align="center">.dcd</td>
<td align="center">atomic <a href="../trajectories.html">trajectory</a>
coordinates
<br>(DCD support courtesy of
<a href="https://www.ks.uiuc.edu/~jim/mdtools/index.html"
target="_blank">MDTools</a>)</td>
</tr><tr>
<td align="center"><a href="http://manual.gromacs.org/documentation/5.1/user-guide/file-formats.html#xtc"
target="_blank">Gromacs compressed coordinates</a></td>
<td align="center"><b>xtc</b></td>
<td align="center">.xtc</td>
<td align="center">atomic <a href="../trajectories.html">trajectory</a>
coordinates</td>
</tr><tr>
<td align="center"><a href="http://manual.gromacs.org/documentation/5.1/user-guide/file-formats.html#trr"
target="_blank">Gromacs full-precision coordinates</a> (binary)</td>
<td align="center"><b>trr</b></td>
<td align="center">.trr</td>
<td align="center">atomic <a href="../trajectories.html">trajectory</a>
coordinates</td>
</tr><tr>
<td align="center"><a href="https://docs.lammps.org/dump.html"
target="_blank">LAMMPS dump</a></td>
<td align="center"><b>dump</b></td>
<td align="center">.dump</td>
<td align="center">atomic <a href="../trajectories.html">trajectory</a>
coordinates
<br>(LAMMPS support courtesy of Mitch Murphy)</td>
</tr><tr>
<td colspan=4 align="center" class="shaded">
<a name="trajtop"></a>
<a href="#formats" class="nounder"><b>&larr;</b></a>
<b>Trajectory Topology</b>
<br>(also requires a <a href="#trajectory">trajectory coordinate file</a>
given with the <a href="#coords"><b>coords</b></a> option;
see also <a href="#traj-options">trajectory options</a>)</td>
</tr><tr>
<td align="center">PSF (protein structure file;
<a href="https://www.ks.uiuc.edu/Training/Tutorials/namd/namd-tutorial-unix-html/node23.html"
target="_blank">X-PLOR-style topology</a>)</td>
<td align="center"><b>psf</b></td>
<td align="center">.psf</td>
<td align="center">atomic <a href="../trajectories.html">trajectory</a>
topology 
<br>(PSF support courtesy of
<a href="https://www.ks.uiuc.edu/~jim/mdtools/index.html"
target="_blank">MDTools</a>)</td>
<!--
</tr><tr>
<td colspan=4 align="center" class="shaded">
<a name="docking"></a>
<a href="#formats" class="nounder"><b>&larr;</b></a>
<b>Docking Results</b> (associated tool: 
<a href="../tools/viewdock.html"><b>ViewDock</b></a>)</td>
</tr><tr>
<td align="center">Dock 4, 5 or 6 Mol2</td>
<td align="center"><b>mol2</b></td>
<td align="center">.mol2</td>
<td align="center">Mol2 format from 
<a href="https://dock.compbio.ucsf.edu/">UCSF DOCK</a> versions 4, 5, 6</td>
-->
</tr><tr>
<td colspan=4 align="center" class="shaded">
<a name="volume"></a>
<a href="#formats" class="nounder"><b>&larr;</b></a>
<b>Volume Data and Related</b> (associated tool: 
<a href="../tools/volumeviewer.html"><b>Volume Viewer</b></a>;
see also the <a href="#difference"><b>difference</b></a> option)</td>
</tr><tr>
<td align="center">
<a href="https://www.fei.com/software/amira-for-life-sciences/"
target="_blank">Amira</a> mesh</td>
<td align="center"><b>amira</b></td>
<td align="center">.am</td>
<td align="center">3D lattice volume data, float, unsigned 16-bit, uncompressed only</td>
</tr><tr>
<td align="center">APBS potential</td>
<td align="center"><b>apbs</b></td>
<td align="center">.dx</td>
<td align="center">
<a href="https://apbs.readthedocs.io/en/latest/formats/opendx.html"
target="_blank">electrostatic potential map</a> from 
<a href="https://www.poissonboltzmann.org/"
target="_blank">Adaptive Poisson-Boltzmann Solver</a>
<br>(see the
<a href="https://www.rbvi.ucsf.edu/chimera/docs/ContributedSoftware/apbs/apbs.html"
target="_blank"><b>APBS</b></a> tool in 
<a href="https://www.rbvi.ucsf.edu/chimera/"
target="_blank"><b>UCSF Chimera</b></a>)</td>
</tr><tr>
<td align="center">BRIX map</td>
<td align="center"><b>brix</b></td>
<td align="center">.brix</td>
<td align="center">electron density map</td>
</tr><tr>
<td align="center">
<a href="https://en.wikipedia.org/wiki/CCP4_(file_format)" 
target="_blank">CCP4</a> map</td>
<td align="center"><b>ccp4</b></td>
<td align="center">.ccp4, .map</td>
<td align="center">electron density map</td>
</tr><tr>
<td align="center"><a href="save.html#chimap">Chimera map</a></td>
<td align="center"><b>cmap</b></td>
<td align="center">.cmp, .cmap</td>
<td align="center">volume data from Chimera/ChimeraX
<br>(based on the hierarchical data format
<a href="https://www.hdfgroup.org/HDF5/" target="_blank">HDF5</a>)</td>
</tr><tr>
<td align="center">CNS or XPLOR map</td>
<td align="center"><b>xplor</b></td>
<td align="center">.cns, .xplor</td>
<td align="center">unformatted ASCII density map</td>
</tr><tr>
<td align="center">DelPhi or GRASP potential</td>
<td align="center"><b>delphi</b></td>
<td align="center">.phi</td>
<td align="center">electrostatic potential map from
<a href="http://compbio.clemson.edu/delphi" target="_blank">DelPhi</a>
(the academic version) or 
<a href="https://honig.c2b2.columbia.edu/grasp/" target="_blank">GRASP</a></td>
</tr><tr>
<td align="center">
<a href="https://docs.openmicroscopy.org/bio-formats/5.7.1/formats/deltavision.html"
target="_blank">DeltaVision</a> map</td>
<td align="center"><b>dv</b></td>
<td align="center">.dv</td>
<td align="center">light microscopy data</td>
</tr><tr>
<td align="center"><a href="../dicom/dicom.html">DICOM</a> map</td>
<td align="center"><b>dicom</b></td>
<td align="center">.dcm</td>
<td align="center">medical image stack; 
see the <a href="#verbose"><b>verbose</b></a> option</td>
</tr><tr>
<td align="center">Directional FSC (Fourier shell correlation)</td>
<td align="center"><b>dres</b></td>
<td align="center">.dres
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">directional resolution to show on a colored globe
<br>(plain text file with four whitespace-separated columns of numbers,
an XYZ vector and a resolution or other value
to colormap onto the globe)</td>
</tr><tr>
<td align="center">DOCK scoring grid</td>
<td align="center"><b>dock</b></td>
<td align="center">.bmp<br>.cnt<br>.nrg</td>
<td align="center">
<a href="https://dock.compbio.ucsf.edu"
target="_blank">UCSF DOCK</a> (versions 4, 5, 6)
bump, contact, and energy scoring grids<br>
(suffixes not interchangeable; <i>gridname</i>.bmp required for reading
<i>gridname</i>.cnt and/or <i>gridname</i>.nrg)</td>
</tr><tr>
<td align="center">DSN6 map</td>
<td align="center"><b>dsn6</b></td>
<td align="center">.omap</td>
<td align="center">electron density map</td>
</tr><tr>
<td align="center">EMAN HDF map</td>
<td align="center"><b>emanhdf</b></td>
<td align="center">.hdf, .h5, .hdf5</td>
<td align="center">electron density map
<br>(based on the hierarchical data format
<a href="https://www.hdfgroup.org/HDF5/" target="_blank">HDF5</a>;
if the EMAN HDF file reader fails, a <a href="#generic-hdf">generic HDF
file reader</a> is tried)</td>
</tr><tr>
<td align="center">Gaussian cube grid</td>
<td align="center"><b>cube</b></td>
<td align="center">.cube, .cub</td>
<td align="center">orbitals, electron densities, other</td>
</tr><tr>
<td align="center"><a href="https://research.csc.fi/-/gopenmol_download"
target="_blank">gOpenMol</a> grid</td>
<td align="center"><b>gopenmol</b></td>
<td align="center">.plt</td>
<td align="center">orbitals, electron densities, other</td>
</tr><tr>
<td align="center">ImageJ or
<a href="https://www-legacy.openmicroscopy.org/site/products/ome-tiff"
target="_blank">OME</a> TIFF</td>
<td align="center"><b></b></td>
<td align="center">.tif, .tiff</td>
<td align="center">microscopy data
<br>(multiple channels and/or time points specified in file header)</td>
</tr><tr>
<td align="center">Image stack</td>
<td align="center"><b>images</b></td>
<td align="center">.tif, .tiff<br>.png<br>.pgm</td>
<td align="center"><a name="image-stacks">image stack</a>
<br>(grayscale or RGB image series in multiple files or a single multipage file,
in formats supported by
<a href="https://en.wikipedia.org/wiki/Python_Imaging_Library" 
target="_blank">PIL</a>);
see <a href="#image-wildcards">image stack wildcards</a> and
<a href="#photo">single image</a>
</td>
</tr><tr>
<td align="center">
<a href="https://www.imagescience.de/formats.html"
target="_blank">IMAGIC map</a></td>
<td align="center"><b>imagic</b></td>
<td align="center">.hed<br>.img</td>
<td align="center">density map
<br>(the format consists of two files, <i>filename</i>.hed with
header information and <i>filename</i>.img with the map values; 
both are required, but either can be specified in the <b>open</b> command;
IMAGIC support
courtesy of Ralf Schmidt and Michael Schatz, Image Science Software)</td>
</tr><tr>
<td align="center">
<a href="https://www.bitplane.com/imaris" target="_blank">Imaris</a> map</td>
<td align="center"><b>ims</b></td>
<td align="center">.ims</td>
<td align="center">3D-5D image
<br>(could be time series and/or multichannel)</td>
</tr><tr>
<td align="center">
<a href="https://bio3d.colorado.edu/imod/doc/mrc_format.txt"
target="_blank">IMOD</a> map</td>
<td align="center"><b>imodmap</b></td>
<td align="center">.rec</td>
<td align="center">electron density map
<br>(MRC map with signed 8-bit mode interpreted as unsigned)</td>
</tr><tr>
<td align="center">
<a href="https://brettbode.github.io/wxmacmolplt/MacMolPlt_Manual.html"
target="_blank">MacMolPlt</a> grid</td>
<td align="center"><b>macmolplt</b></td>
<td align="center">.mmp</td>
<td align="center">
<a href="https://brettbode.github.io/wxmacmolplt/Manual_pages/MacMolPlt_Files.html#SurfExport"
target="_blank">3D surfaces</a></td>
</tr><tr>
<td align="center">
<a href="https://www2.mrc-lmb.cam.ac.uk/research/locally-developed-software/image-processing-software/#image"
target="_blank">MRC</a> map</td>
<td align="center"><b>mrc</b></td>
<td align="center">.mrc</td>
<td align="center">electron density map</td>
</tr><tr>
<td align="center">
<a href="http://www.sb.fsu.edu/~xray/Manuals/CCP4/mtzformat.html"
target="_blank">MTZ</a> crystallographic reflection file
<br><br>*requires installing the
<a href="https://cxtoolshed.rbvi.ucsf.edu/apps/chimeraxclipper"
target="_blank"><b>Clipper</b> bundle</a> from the
<a href="https://cxtoolshed.rbvi.ucsf.edu/"
target="_blank">ChimeraX Toolshed</a>
and specifying a corresponding atomic model
(<a href="#clipper">details...</a>)
</td>
<td align="center"><b>mtz</b></td>
<td align="center">.mtz</td>
<td align="center" class="text">reciprocal space map
</td>
</tr><tr>
<td align="center">NetCDF map</td>
<td align="center"><b>netcdfmap</b><br>(required)</td>
<td align="center">.nc<br>(insufficient to specify format)</td>
<td align="center">electron density map
<br>(use <b>format netcdfmap</b> to disambiguate from 
<a href="#trajectory">Amber NetCDF trajectories</a>)</td>
</tr><tr>
<td align="center"><a href="https://nifti.nimh.nih.gov/"
target="_blank">Neuroimaging Informatics Technology Initiative (NIfTI)</a></td>
<td align="center"><b>nifti</b></td>
<td align="center">.nii
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">medical image</td>
</tr><tr>
<td align="center"><a href="https://teem.sourceforge.net/nrrd/"
target="_blank">Nearly Raw Raster Data (NRRD)</a></td>
<td align="center"><b>nrrd</b></td>
<td align="center">.nrrd</td>
<td align="center">medical image</td>
</tr><tr>
<td align="center">Priism data</td>
<td align="center"><b>priism</b></td>
<td align="center">.xyzw<br>.xyzt</td>
<td align="center">3D light or EM data
<br>(.xyzt is time series)</td>
</tr><tr>
<td align="center">PROFEC grid</td>
<td align="center"><b>profec</b></td>
<td align="center">.profec</td>
<td align="center">interaction free energy grid from PROFEC
(in <a href="https://ambermd.org" target="_blank"><b>Amber</b></a>
versions 6 and 7)</td>
</tr><tr>
<td align="center">Purdue image format</td>
<td align="center"><b>pif</b></td>
<td align="center">.pif</td>
<td align="center">electron density map</td>
<!-- via Signal Viewer bundle in Toolshed
</tr><tr>
<td align="center">
Signal 3D time series</td>
<td align="center"><b>signal</b></td>
<td align="center">.npz</td>
<td align="center">3D fluorescence time series</td>
-->
</tr><tr>
<td align="center"><a href="https://situs.biomachina.org/index.html"
target="_blank">Situs</a> map</td>
<td align="center"><b>situs</b></td>
<td align="center">.situs, .sit</td>
<td align="center">electron density map</td>
</tr><tr>
<td align="center">
<a href="https://spider.wadsworth.org/spider_doc/spider/docs/spider.html"
target="_blank">SPIDER</a> map</td>
<td align="center"><b>spider</b></td>
<td align="center">.spi, .vol</td>
<td align="center">electron density map</td>
</tr><tr>
<td align="center">
<a href="http://huanglab.ucsf.edu/STORM.html"
target="_blank">STORM (STochastic Optical Reconstruction Microscopy)</a>
</td>
<td align="center"><b>storm</b></td>
<td align="center">.storm
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">super-resolution microscopy data
(<a href="../formats/storm.html">details...</a>)</td>
</tr><tr>
<td align="center"><a href="https://www.biochem.mpg.de/6348566/tom_e"
target="_blank">TOM toolbox</a> map</td>
<td align="center"><b>tom_em</b></td>
<td align="center">.em</td>
<td align="center">electron density map</td>
</tr><tr>
<td align="center">UHBD potential (binary)</td>
<td align="center"><b>uhbd</b></td>
<td align="center">.grd</td>
<td align="center">electrostatic potential map from
University of Houston Brownian Dynamics</td>
</tr><tr>
<td colspan=4 align="center" class="shaded">
<a name="segmentation"></a>
<a href="#formats" class="nounder"><b>&larr;</b></a>
<b>Segmentations</b></td>
</tr><tr>
<td align="center">
<a href="../formats/sff.html">EMDB segmentation</a></td>
<td align="center"><b>sff</b></td>
<td align="center">.sff (XML)<br>.hff (HDF5)</td>
<td align="center">EM segmentation</td>
</tr><tr>
<td align="center">
<a href="../formats/imod.html">IMOD segmentation</a></td>
<td align="center"><b>imod</b></td>
<td align="center">.imod<br>.mod</td>
<td align="center">EM segmentation meshes and contours</td>
</tr><tr>
<td align="center">
<a href="../tools/segment.html">Segger segmentation</a></td>
<td align="center"><b>segger</b></td>
<td align="center">.seg</td>
<td align="center">segmentation regions</td>
</tr><tr>
<td colspan=4 align="center" class="shaded">
<a name="object"></a>
<a href="#formats" class="nounder"><b>&larr;</b></a>
<b>Other 3D Objects</b></td>
</tr><tr>
<td align="center">
<a href="../formats/bild.html">Chimera/ChimeraX BILD</a>
</td>
<td align="center"><b>bild</b></td>
<td align="center">.bild, .bld
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">graphical objects</td>
</tr><tr>
<td align="center">
<a href="../markers.html#markerfiles">Chimera/ChimeraX markers</a>
</td>
<td align="center"><b>markers</b></td>
<td align="center">.cmm</td>
<td align="center"><a href="../markers.html">markers and links</a>
<br>(analogous to atoms and bonds)</td>
</tr><tr>
<td align="center"><a href="https://www.khronos.org/collada/"
target="_blank">COLLADA</a></td>
<td align="center"><b>collada</b></td>
<td align="center">.dae
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">graphical objects
<br>(TriangleSets only)</td>
</tr><tr>
<td align="center"><a href="https://www.khronos.org/gltf"
target="_blank">GL Transmission Format</a></td>
<td align="center"><b>gltf</b></td>
<td align="center">.glb
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">graphical objects</td>
</tr><tr>
<td align="center"><a href="https://mole.upol.cz/" 
target="_blank">MOLEonline</a> channels</td>
<td align="center"><b>mole</b></td>
<td align="center">.json</td>
<td align="center" class="text">channels in JSON format from
<a href="https://mole.upol.cz/" target="_blank">MOLEonline</a>;
shown as <a href="../markers.html">marker models</a> with
the specified <a href="#transparency">transparency</a>
and a <a href="mseries.html#slider">slider interface</a>
for flipping through them</td>
</tr><tr>
<td align="center">Neuron trace
<br>(<a href="http://www.neuronland.org/NLMorphologyConverter/MorphologyFormats/SWC/Spec.html" target="_blank">SWC</a>, see also the
<a href="http://neuromorpho.org/myfaq.jsp#QS3"
target="_blank">NeuroMorpho FAQ</a>)</td>
<td align="center"><b>swc</b></td>
<td align="center">.swc</td>
<td align="center">neuron reconstruction, such as from
<a href="http://neuromorpho.org/" target="_blank">NeuroMorpho.Org</a></td>
</tr><tr>
<td align="center"><a href="https://en.wikipedia.org/wiki/STL_%28file_format%29#Binary_STL"
target="_blank">STereoLithography</a></td>
<td align="center"><b>stl</b></td>
<td align="center">.stl
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">triangle-based format
native to stereolithography CAD software</td>
</tr><tr>
<td align="center">
<a href="https://vtk.org/" target="_blank">VTK (Visualization ToolKit)</a>
PolyData</td>
<td align="center"><b>vtk</b></td>
<td align="center">.vtk
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">lines, polygons
[documentation
<a href="https://vtk.org/VTK/img/file-formats.pdf" target="_blank">PDF</a>]
<br>(<a href="../formats/vtk.html">details</a> on ChimeraX VTK export)</td>
</tr><tr>
<td align="center">
<a href="https://www.fileformat.info/format/wavefrontobj/egff.htm"
target="_blank">Wavefront OBJ</a></td>
<td align="center"><b>obj</b></td>
<td align="center">.obj
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">graphical objects
<br>(triangles only: vertices, vertex normals, faces; texture colors)</td>
</tr><tr>
<td colspan=4 align="center" class="shaded">
<a name="sequence"></a>
<a href="#formats" class="nounder"><b>&larr;</b></a>
<b>Sequences</b> (associated tool: 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>
or <a href="../tools/profilegrid.html"><b>Profile Grid</b></a>,
see the <a href="#viewer"><b>viewer</b></a> option;
see also <a href="#ident"><b>ident</b></a>)</td>
</tr><tr>
<td align="center">Clustal ALN</td>
<td align="center"><b>aln</b></td>
<td align="center" class="text">.aln, .clustal<br>.clustalw,&nbsp;.clustalx
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">sequence alignment</td>
</tr><tr>
<td align="center">FASTA, aligned FASTA</td>
<td align="center"><b>fasta</b></td>
<td align="center">.fasta, .fa<br>.afasta, .afa</a>
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">individual sequence(s) or sequence alignment</td>
</tr><tr>
<td align="center">GCG MSF (Multiple Sequence Format)</td>
<td align="center"><b>msf</b></td>
<td align="center">.msf
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">sequence alignment</td>
</tr><tr>
<td align="center">GCG RSF (Rich Sequence Format)</td>
<td align="center"><b>rsf</b></td>
<td align="center">.rsf
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">sequence alignment</td>
</tr><tr>
<td align="center">
<a href="https://nar.oxfordjournals.org/content/25/1/226.full"
target="_blank">HSSP</a></td>
<td align="center"><b>hssp</b></td>
<td align="center">.hssp
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">sequence alignment
<br>(other information not read; see
<a href="https://swift.cmbi.ru.nl/gv/hssp/" target="_blank">HSSP database</a>)
</td>
</tr><tr>
<td align="center" class="text">PIR (NBRF), aligned PIR</td>
<td align="center"><b>pir</b></td>
<td align="center">.pir, .ali
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">individual sequence(s) or sequence alignment</td>
</tr><tr>
<td align="center">Selex (Pfam)</td>
<td align="center"><b>pfam</b></td>
<td align="center">.selex, .slx<br>.pfam
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">sequence alignment</td>
</tr><tr>
<td align="center">
<a href="https://en.wikipedia.org/wiki/Stockholm_format"
target="_blank">Stockholm</a></td>
<td align="center"><b>stockholm</b></td>
<td align="center">.sth, .sto
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">sequence alignment</td>
</tr><tr>
<td colspan=4 align="center" class="shaded">
<a name="composite"></a>
<a href="#formats" class="nounder"><b>&larr;</b></a>
<b>Composite</b></td>
</tr><tr>
<td align="center"><a href="save.html#session">ChimeraX session file</a></td>
<td align="center"><b>session</b></td>
<td align="center">.cxs</td>
<td align="center">ChimeraX session; see the <b>open</b> command options
<a href="#combine"><b>combine</b></a> and
<a href="#resizeWindow"><b>resizeWindow</b></a>; see also the
<a href="../preferences.html#window"><b>Window</b> preferences</a> and
settings in the <a href="../tools/log.html"><b>Log</b> context menu</a>
</td>
</tr><tr>
<td align="center"><a href="https://github.com/ihmwg/IHM-dictionary"
target="_blank">IHM (Integrative Hybrid Model)</a>
</td>
<td align="center"><b>ihm</b></td>
<td align="center">.ihm<br>.cif</td>
<td align="center">integrative hybrid model,
examples available from <a href="https://pdb-ihm.org/"
target="_blank">PDB-IHM</a> (see <a href="#fetch">database fetch</a>);
may create <a href="../attributes.html#residue">residue attributes</a>
</td>
</tr><tr>
<td colspan=4 align="center" class="shaded">
<a name="miscellaneous"></a>
<a href="#formats" class="nounder"><b>&larr;</b></a>
<b>Miscellaneous</b></td>
</tr><tr>
<td align="center"><a href="usageconventions.html#cxc-files">ChimeraX
command file</a></td>
<td align="center"><b>cmd</b></td>
<td align="center">.cxc</td>
<td align="center">
<a href="../index.html#commands">ChimeraX commands</a></td>
</tr><tr>
<td align="center"><a href="../formats/defattr.html">attribute
assignment file</a></td>
<td align="center"><b>defattr</b></td>
<td align="center">.defattr</td>
<td align="center">
<a href="../attributes.html">attribute</a> assignments
<br>(<a href="#defattr">details...</a>)</td>
</tr><tr>
<td align="center"><a href="../formats/mutationcsv.html">mutation
scores file</a></td>
<td align="center"><b>ms</b></td>
<td align="center">.csv</td>
<td align="center">
assay scores from deep mutational scanning
(<a href="../formats/mutationcsv.html">details...</a>);
corresponding structure chains can be specified with the
<a href="#chains"><b>chains</b></a> option</td>
</tr><tr>
<td align="center">photo (single image)</td>
<td align="center"><b>photo</b></td>
<td align="center">.jpg
<br>(other image types .tif, .png, <i>etc.</i>
accepted if specified as <b>format photo</b>;
<a href="#photo">details...</a>)
</td>
<td align="center">2D image;
<br>see also <a href="#image-stacks">image stack</a></td> 
</tr><tr>
<td align="center">Python</td>
<td align="center"><b>py</b></td>
<td align="center">.py, .pyc, .pyo
<br><a href="#compressed">+&nbsp;compressed</td>
<td align="center">Python code
<br>(see also <a href="runscript.html"><b>runscript</b></a>)</td>
</tr><tr>
<td align="center">HTML, XHTML</td>
<td align="center"><b>html</b></td>
<td align="center">.html, .htm<br>.xhtml,&nbsp;.xht,&nbsp;.xml</td>
<td align="center">
hypertext to render in the 
<a href="../tools/helpviewer.html">ChimeraX browser</a></td>
</tr><tr>
<td align="center"><a href="../formats/positions.html">ChimeraX 
model positions</a></td>
<td align="center"><b>positions</b></td>
<td align="center">.positions</td>
<td align="center">
ChimeraX model names and positions to restore</td>
</tr></table>

<a name="local"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>Local Files</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>open</b> &nbsp;<i>filename</i>&nbsp;
[&nbsp;<i>filename2</i>...&nbsp;]
[&nbsp;<b>format</b>&nbsp;&nbsp;<a href="#file-table"><i>format-name</i></a>&nbsp;]
[&nbsp;<a href="#maxModels"><b>maxModels</b></a>&nbsp;&nbsp;<i>M</i>&nbsp;]
[&nbsp;<a href="#coordsets"><b>coordsets</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#vseries"><b>vseries</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<a href="#resizeWindow"><b>resizeWindow</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
&nbsp;<a href="#options"><i>options</i></a>&nbsp;
</blockquote>
<p>
The <i>filename</i> is generally a pathname to a local file, 
either absolute or relative to the current working directory as
reported by <a href="pwd.html"><b>pwd</b></a>.
The working directory can be changed with <a href="cd.html"><b>cd</b></a>.
A pathname may start with &ldquo;<b>~</b>&rdquo; to indicate the user's 
home directory.
One or more files can be specified with a single filename
by including <b>*</b> (wild card), <b>?</b> (single-character wild card), 
and/or square-bracketed character ranges
(Unix-style pathname expansion or <b><i>globbing</i></b>).
For <a href="#image-stacks">image stacks</a> only, 
<a href="#image-wildcards">additional wild cards</a> may be used.
</p><p>
Multiple filenames separated by spaces can also be given,
which is equivalent to using the <b>open</b> command 
multiple times with the same options for each filename.
</p><p>
The file format can be indicated with either the <i>filename</i> suffix or
the <b>format</b> option
(where <a href="#file-table"><i>format-name</i></a> can be truncated).
Certain formats can be read when compressed (.gz, .bz2, or .xz, see
<a href="#compressed">details</a>).
</p><p>
Giving <a href="usageconventions.html#browse"><b>browse</b></a> as the 
<i>filename</i> brings up a dialog for browsing the filesystem interactively.
</p><p>
Additional ways to open local files:
</p>
<ul>
<li>menu <a href="../menu.html#file"><b>File... Open</b></a>, in which case 
the proper filename <a href="#file-table">suffix</a> is required
<li><a href="../tools/toolbar.html#home"><b>Toolbar</b></a> icon
<a href="../tools/toolbar.html#home"><img class="icon"
border=1 src="../tools/toolbar-icons/open-in-app.png"></a>
<li>drag-and-drop onto the ChimeraX icon or a running instance of ChimeraX
</ul>
<p>
On a Mac, ChimeraX-associated file types can also be opened by 
double-clicking them in a Finder window or using &ldquo;open&rdquo; 
in a Terminal window. 
</p>

<a name="options"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>Options</h3>
<blockquote>
<b>alignment</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
Whether to interpret a multi-sequence file as an alignment 
(default, all sequences in one 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window).
Otherwise, each sequence will be opened in a separate window. 
If the sequences (including any gap characters) are of differing lengths,
the option must be set to <b>false</b>. 
</blockquote>
<blockquote>
<a name="alignTo"></a>
<b>alignTo</b>&nbsp;&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>&nbsp;[&nbsp;<b>trim</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
<br>
When <a href="#fetch">fetching</a>
an <a href="alphafold.html#fetch">AlphaFold prediction</a>
or <a href="esmfold.html#fetch">ESMFold prediction</a>, 
superimpose the predicted structure onto the specified chain using 
<a href="matchmaker.html"><b>matchmaker</b></a> defaults, and make the
chain ID of the prediction the same as the specified chain's.
If the <b>trim</b> option is <b>true</b> (default),
the predicted structure will be trimmed to start and end with the first 
and last aligned positions in the sequence alignment calculated by
<a href="matchmaker.html"><b>matchmaker</b></a>
as part of the superposition step.
</blockquote>
<blockquote>
<a name="associate"></a>
<b>associate</b>&nbsp;&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
<br>
When <a href="#fetch">fetching</a> a sequence from
<a href="https://www.uniprot.org" target="_blank">UniProt</a>,
<a href="../tools/sequenceviewer.html#association">associate</a> it
with the specified structure chain(s) only. Normally,
<a href="../tools/sequenceviewer.html#association">association</a> 
will occur automatically; however, this option handles cases in which
the desired chain does not associated automatically due to a high number of 
sequence mismatches, truncations, large insertions, <i>etc.</i>,
and/or one wishes to exclude some of the associations that would otherwise
occur automatically.
After a sequence is opened, its associations can also be adjusted with
the <a href="../tools/sequenceviewer.html#context"><b>Sequence Viewer</b>
context menu</a> or the
<a href="sequence.html#associate"><b>sequence associate</b></a> command.
</blockquote>
<blockquote>
<a name="atomic"></a>
<b>atomic</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
Whether to interpret coordinates read from mmCIF or PDB format as 
an atomic model (<b>true</b>, default).
Setting the option to <b>false</b> is appropriate for data such as SAXS models 
that are simply collections of points 
(possibly with connecting &ldquo;bonds&rdquo;)
rather than realistically spaced atoms;
it classifies the models as Structure rather than atomicStructure, 
bypassing chemical perception steps like identifying
<a href="../atomtypes.html">atom types</a>.
</blockquote>
<blockquote>
<b>autoAssociate</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false
<br>
Whether structure-chain sequences should be compared to sequences in 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>
and automatically associated if the number of
mismatches is no more than 1/10 the number of residues in the structure chain.
Unlike the other options of <b>open</b>, this sets a state that persists 
within the ChimeraX session unless explicitly changed. 
The initial setting is <b>true</b>.
<p>
For automatic association, the sequence and structure files
can be opened in either order.
Gaps in the structure sequence relative to the sequence in the alignment file 
can only occur where residues are missing from the structure (for example, 
a flexible loop with insufficient density to determine coordinates).
Associations are reported in the <a href="../window.html">status line</a>
and the <a href="../tools/log.html"><b>Log</b></a>.
A structure (even if it has multiple chains) cannot be associated with more
than one sequence, but a single sequence can be associated with
more than one structure.
If more than one sequence matches a given structure chain,
the single best-matching sequence is associated. 
</p>
</blockquote>
<blockquote>
<b>autoStyle</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
Setting <b>autoStyle</b> to <b>false</b> bypasses 
<a href="../autostyle.html">automatic styling</a>, <i>i.e.</i>,
automatically adjusting the appearance of a newly opened atomic model
based on its contents.
Bypassing this calculation shows all atoms as white spheres, and may be 
useful in rare cases where generating the rule-based representation is slow
(such as drawing cartoons for very large numbers of models) or 
when chemical perception is inappropriate (such as for a SAXS bead model). 
See also: <a href="../markers.html">markers and links</a>
</blockquote>
<blockquote>
<a name="center"></a>
<b>center</b>&nbsp;&nbsp;<a href="pointspec.html#point"><i>point-spec</i></a>
<br>
Center the new model at the specified point, which can be given as
<i>x,y,z</i> (three values separated by commas only),
an <a href="atomspec.html"><i>atom-spec</i></a>, or any of the other
standard <a href="pointspec.html#vector">point specifications</a>.
</blockquote>
<blockquote>
<a name="chains"></a>
<b>chains</b>&nbsp;&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>&nbsp;[&nbsp;<b>allowMismatches</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
<br>
When opening a <a href="../formats/mutationcsv.html">mutation scores</a> 
(.csv) file, designate one or more structure chains 
to be associated with the data for analysis with the
<a href="mutationscores.html"><b>mutationscores</b></a> command.
Otherwise, the mutation scores will be associated automatically with each 
open structure chain that has exactly the same sequence and residue 
numbering (as in the mutation data) the first time some analysis action 
is performed, such as interacting with a mutation plot.
The <b>chains</b> option allows explicitly specifying which structure
chain(s) to use for this purpose, and if a chain does not have exactly 
the same sequence as the mutation data, <b>allowMismatches true</b>
is needed to force the association.
</p><p>
An association can also be made later with the command
<a href="mutationscores.html#structure"><b>mutationscores structure</b></a>.
If the chain residue numbering is different than in the mutation scores file,
it is necessary to use
<a href="mutationscores.html#structure"><b>mutationscores structure</b></a>
with the <a href="mutationscores.html#structure"><b>alignSequences</b></a> 
option to make the association
(it cannot be done with the <b>open</b> command).
</blockquote>
<blockquote>
<a name="channel"></a>
<b>channel</b>&nbsp;&nbsp;<i>N</i>
<br>
Assign channel number <i>N</i> to map; typically used for light microscopy data.
When multiple maps are opened in a single command
(multiple files specified with wild cards or globbing), 
channel numbers are automatically discerned from the filenames, 
if possible: for example, a file will be identified as channel 0
if its name contains either of the following:
<ul class="none">
<li>_ch0_
<li>_0ch_
</ul>
(and 0 could be any whole number).
<!-- OME TIFF header of one file can specify names of all files in the dataset, 
which ChimeraX will open automatically; maybe rare, currently left undescribed. -->
Multichannel data opened from a single file or from multiple files
in a single command (as described above) are automatically grouped,
in that changing the display style, step size, or plane for one channel
in <a href="../tools/volumeviewer.html"><b>Volume Viewer</b></a>
changes it for all of the channels in the group
(<a href="../tools/volumeviewer.html#multichannel">details...</a>).
The <b>channel</b> option allows manually assigning a channel to a map
so that it can be saved along with other data in a
single (multichannel) <a href="save.html#chimap">Chimera map</a> file.
See also: <a href="#image-wildcards">image stack wildcards</a>,
<a href="volume.html#channels"><b>volume channels</b></a>
</blockquote>
<blockquote>
<a name="childModels"></a>
<b>childModels</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
When opening a <a href="../formats/positions.html">ChimeraX positions file</a>,
whether to explicitly apply matrices for submodels rather than simply
allowing them to follow their parent models.
</blockquote>
<blockquote>
<a name="colorConfidence"></a>
<b>colorConfidence</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
When <a href="#fetch">fetching</a>
an <a href="alphafold.html#fetch">AlphaFold prediction</a>
or <a href="esmfold.html#fetch">ESMFold prediction</a>, 
whether to <a href="color.html#byattribute">color</a> it by the
<a href="https://alphafold.ebi.ac.uk/faq#faq-12" target="_blank">pLDDT
confidence measure</a> in the B-factor field (default <b>true</b>):
     <ul>
     <li>90
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=cornflowerblue width="10px" title="cornflower blue"></td></tr>
</table>
to 100
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=blue width="10px" title="blue"></td></tr>
</table>
&ndash; high accuracy expected
     <li>70
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=yellow width="10px" title="yellow"></td></tr>
</table>
to 90
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=cornflowerblue width="10px" title="cornflower blue"></td></tr>
</table>
&ndash; expected to be modeled well (a generally good backbone prediction)
     <li>50
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=orange width="10px" title="orange"></td></tr>
</table>
to 70
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=yellow width="10px" title="yellow"></td></tr>
</table>
&ndash; low confidence, treat with caution
     <li>0
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=red width="10px" title="red"></td></tr>
</table>
to 50
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=orange width="10px" title="orange"></td></tr>
</table>
&ndash; should not be interpreted, may be disordered
     </ul>
...in other words, using
<a href="color.html#bfactor"><b>color bfactor</b></a>
<b><a href="color.html#palette-options">palette</a>
0,red:50,orange:70,yellow:90,cornflowerblue:100,blue</b>
<!-- 100,blue:90,cornflowerblue:70,yellow:50,orange:0,red</b> -->
<br>
The <a href="../tools/colorkey.html"><b>Color Key</b></a> graphical interface
or <a href="key.html"><b>key</b></a> command
can be used to draw a corresponding color key, for example:
</p>
  <blockquote>
  <b>key</b> <b>red:low orange: yellow: cornflower blue: blue:high</b>
  &nbsp;[<a href="key.html#options"><i>other-key-options</i></a>]
<!--
<br>
  <b>key</b> <b>blue:100 cornflowerblue:90 yellow:70 orange:50 red:0</b>
  &nbsp;<b>numericLabelSpacing equal</b>
  &nbsp;[<a href="key.html#options"><i>other-key-options</i></a>]
-->
  </blockquote>
</blockquote>
<blockquote>
<a name="combine"></a>
<b>combine</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
When opening a <a href="save.html#session">session</a> file,
whether to retain the current session (existing models and other data)
or to close the current session first (default, <b>combine false</b>).
The model numbers of the incoming session will be changed as needed
to avoid duplicate model numbers.
See also the <a href="#resizeWindow"><b>resizeWindow</b></a> option.
</blockquote>
<blockquote>
<b>combineSymAtoms</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
Nonstandard PDB or mmCIF files may contain duplicate copies of atoms
at symmetry positions; this option indicates whether to merge
multiple singleton atoms of the same element at the same 3D position 
into one (default <b>true</b>) to prevent errors in subsequent
calculations such as <a href="addh.html">adding hydrogens</a>.
</blockquote>
<blockquote>
<a name="contours"></a>
<b>contours</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
Whether to read the contours from an
<a href="../formats/imod.html">IMOD segmentation file</a>.
</blockquote>
<blockquote>
<a name="debug"></a>
<b>debug</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
Whether to send verbose information to the 
<a href="../tools/log.html"><b>Log</b></a> when opening a
<a href="../formats/sff.html">EMDB segmentation file</a>.
</blockquote>
<blockquote>
<a name="difference"></a>
<b>difference</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false
<br>
When opening a <a href="#volume">density map</a>, whether to display it
as a difference map, <i>i.e.</i>, showing positive and negative contour surfaces
and turning off <a href="volume.html#capFaces">capping at box faces</a>
(default <b>false</b>, except <b>true</b> when <a href="#fetch">fetching</a>
from <b>edsdiff</b>).
</blockquote>
<blockquote>
<b>ensembles</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
Whether to fetch ensembles associated with an IHM dataset 
(default <b>false</b>, as these can be quite large).
</blockquote>
<blockquote>
<a name="fetchEmdbMap"></a>
<b>fetchEmdbMap</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
When opening or <a href="#fetch">fetching</a> an mmCIF file,
whether to additionally fetch the associated EMDB map, if any
(default <b>false</b>). See also the <a href="#fits"><b>fits</b></a> option.
</blockquote>
<blockquote>
<a name="fits"></a>
<b>fits</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
When <a href="#fetch">fetching</a> a map from the EMDB, whether to 
additionally fetch associated atomic structures, if any, from the PDB.
Alternatively, specifying the <a href="#fetch-table">data source</a>
as <b>emdb_fits</b> (rather than <b>emdb</b>) will automatically
fetch both types of data without the use of this option.
See also the <a href="#fetchEmdbMap"><b>fetchEmdbMap</b></a> option.
</blockquote>
<blockquote>
<a name="forEachFile"></a>
<b>forEachFile</b>&nbsp;&nbsp;<i>input-files</i>
<br>
The <b>forEachFile</b> option allows processing multiple data files
iteratively with the same
<a href="usageconventions.html#cxc-files">ChimeraX command file</a>
without having to open the command file over and over.
The data files over which to iterate are specified in a single pathname 
using wild cards or globbing, and would typically contain
<a href="#volume">maps</a> or <a href="#atomic">atomic structures</a>.
For example:
<blockquote>
<b>open myscript.cxc foreach ~/data/*.cif</b>
</blockquote>
...would open the first data file, run the script, open the next data file,
run the script, <i>etc.</i>
Any existing models are closed before each data file is opened.
For saving input-specific output files, 
&ldquo;$file&rdquo; can be used within the command file 
to indicate the base name of each data file.
For example, the following command would save 1a0m.png for input file 1a0m.pdb,
2gbp.png for input file 2gbp.pdb, <i>etc.</i>:
<blockquote>
<b>save $file.png</b>
</blockquote>
Output files will be placed in the same directory as the data files
unless the file-saving command gives a different pathname.
See also: 
<a href="alias.html"><b>alias</b></a>,
<a href="runscript.html"><b>runscript</b></a>
</blockquote>
<blockquote>
<a name="height"></a>
<b>height</b>&nbsp;&nbsp;<i>h</i>
<br>
Height in physical units of distance (&Aring;) for the
<a href="#photo">photo</a> (single image) being opened.
</blockquote>
<blockquote>
<a name="id"></a>
<b>id</b>&nbsp;&nbsp;<i>model-id</i>
<br>
A model identifier (model number, as shown in the 
<a href="../tools/modelpanel.html"><b>Model Panel</b></a>) 
will be assigned automatically, starting with the lowest available number,
unless another identifier is given with the <b>id</b> option.
If the same command opens multiple models or the <i>model-id</i>
is already &ldquo;taken,&rdquo; submodel numbers will be used.
Similarly, the <a href="#name"><b>name</b></a> option can be used to specify
the model name. The name and/or number of an existing (already open) model
can be changed with the <a href="rename.html"><b>rename</b></a> command.
</blockquote>
<blockquote>
<a name="ident"></a>
<b>ident</b>&nbsp;&nbsp;<i>alignment-ID</i>
<br>
When opening <a href="#sequence">sequence data</a>, assign the specified text
as the sequence alignment identifier. If this option is not used,
an identifier will be generated automatically. This identifier is used to 
<a href="seqspec.html">specify</a> the sequence alignment in other commands
and appears in the title bar of the alignment viewer window,
if shown. See also the <a href="#viewer"><b>viewer</b></a> option.
</blockquote>
<blockquote>
<a name="log"></a>
<b>log</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
When opening a 
<a href="usageconventions.html#cxc-files">ChimeraX command file</a>,
whether to echo each command within the file to the 
<a href="../tools/log.html"><b>Log</b></a>
as it is executed (default <b>true</b>).
If <b>false</b>, although the commands themselves will not be logged,
any informational outputs they generate will still appear there.
</blockquote>
<blockquote>
<a name="logInfo"></a>
<b>logInfo</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
Whether to show available <a href="../modelinfo.html">model information</a>
(other than structure title) in the <a href="../tools/log.html"><b>Log</b></a>.
See also: <a href="log.html#metadata"><b>log metadata</b></a>
</blockquote>
<blockquote>
<a name="matchNames"></a>
<b>matchNames</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
When opening a <a href="../formats/positions.html">ChimeraX positions file</a>,
whether to match models by name instead of simply applying the
matrices in order of appearance in the file.
</blockquote>
<blockquote>
<a name="maxModels"></a>
<b>maxModels</b>&nbsp;&nbsp;<i>M</i>
<br>
Maximum number of models to keep when opening a multi-model file 
in mmCIF or PDB format as multiple models in ChimeraX
(<b><a href="#coordsets">coordsets</a> false</b>).
</blockquote>
<blockquote>
<a name="maxSurfaces"></a>
<b>maxSurfaces</b>&nbsp;&nbsp;<i>N</i>
<br>
Maximum number of surface models to create when opening an
<a href="../formats/sff.html">EMDB segmentation file</a>.
</blockquote>
<blockquote>
<a name="meshes"></a>
<b>meshes</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
Whether to read the meshes from an
<a href="../formats/imod.html">IMOD segmentation file</a>.
</blockquote>
<blockquote>
<a name="models"></a>
<b>models</b>&nbsp;&nbsp;<a href="atomspec.html#hierarchy"><i>model-spec</i></a>
<br>
Limit <a href="#defattr">attribute assignment</a> to the specified models.
</blockquote>
<blockquote>
<a name="name"></a>
<b>name</b>&nbsp;&nbsp;<i>model-name</i>
<br>
The name of a model 
(as shown in the <a href="../tools/modelpanel.html"><b>Model Panel</b></a>) 
will be generated 
automatically from the input <i>filename</i> or database <i>identifier</i>
unless another name is given with the <b>name</b> option.
If <i>model-name</i> contains spaces, it should be enclosed in quotation marks.
Similarly, the <a href="#id"><b>id</b></a> option can be used to specify
the model identifier (model number).
The name and/or number of an existing (already open) model
can be changed with the <a href="rename.html"><b>rename</b></a> command.
</blockquote>
<blockquote>
<b>newTab</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
Whether to open an HTML file in a new tab of the
<a href="../tools/helpviewer.html">ChimeraX browser</a>
(default <b>false</b>, replace any current page).
</blockquote>
<blockquote>
<a name="pixelSize"></a>
<b>pixelSize</b>&nbsp;&nbsp;<i>s</i>
<br>
Pixel size in physical units of distance (&Aring;) for the
<a href="#photo">photo</a> (single image) being opened.
</blockquote>
<blockquote>
<a name="resizeWindow"></a>
<b>resizeWindow</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false
<br>
When opening a <a href="save.html#session">session</a>,
whether to restore its window size along with its contents; 
overrides the corresponding
<a href="../preferences.html#window"><b>Window</b> preference</a>.
See also the <a href="#combine"><b>combine</b></a> option.
</blockquote>
<blockquote>
<a name="segidChains"></a>
<b>segidChains</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
Whether to assign segment IDs in PDB format as the chain IDs in ChimeraX.
</blockquote>
<blockquote>
<a name="transferMethod"></a>
<b>transferMethod</b>&nbsp;&nbsp;https&nbsp;|&nbsp;ftp
<br>
Method to use when <a href="#fetch">fetching</a> a map
from a specific EMDB mirror site.
</blockquote>
<blockquote>
<a name="showTool"></a>
<b>showTool</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
Whether to show the 
<a href="../tools/viewdock.html"><b>ViewDock</b></a> tool
when opening a file of molecular docking results
(certain files in the <a href="#atomic">formats</a> <b>swissdock</b>,
<b>pdbqt</b>, <b>mae</b>, and <b>mol2</b>; default <b>showTool true</b>).
If the file is not found to contain molecular docking results, however,
<a href="../tools/viewdock.html"><b>ViewDock</b></a> will not be shown.
Alternatively, one or more files of docking results can be opened with
<b>showTool false</b>, and subsequently (in a separate step) the 
<a href="../tools/viewdock.html"><b>ViewDock</b></a> tool
started from the <b>Tools</b> menu.
</blockquote>
<blockquote>
<a name="transparency"></a>
<b>transparency</b>&nbsp;&nbsp;<i>percent</i>
<br>
Assign the specified <i>percent</i> transparency to
<a href="https://mole.upol.cz/" target="_blank">MOLEonline</a> 
channels read from <a href="#object">JSON format</a>
(default <b>0</b>, not transparent).
See also the <a href="transparency.html"><b>transparency</b></a> command.
</blockquote>
<blockquote>
<a name="viewer"></a>
<b>viewer</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;|&nbsp;<i>viewer-type</i>
<br>
Whether to show the 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>
or a <a href="../tools/profilegrid.html"><b>Profile Grid</b></a>
when opening <a href="#sequence">sequence data</a>.
See also the <a href="#ident"><b>ident</b></a> option.
True (default) means to show the default viewer 
given the number of sequences in the alignment (according to the
<a href="../preferences.html#sequences"><b>Sequences</b> preferences</a>),
whereas false means not to display the alignment in either viewer.
The viewer type can also be specified directly as <i>viewer-type</i>
regardless of alignment size, with possible values:
<ul>
<li><b>seqview</b> (or <b>view</b> or <b>sv</b>)
&ndash; <a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>
<li><b>profile</b> (or <b>grid</b> or <b>pg</b>)
&ndash; <a href="../tools/profilegrid.html"><b>Profile Grid</b></a>
</ul>
</blockquote>
<blockquote>
<a name="vseries"></a>
<b>vseries</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false
<br>
When multiple maps with the same grid dimensions are opened 
in a single command (multiple files specified with wild cards or globbing, 
or a single multi-map file), this option can be used to enforce 
loading them as a <a href="../trajectories.html#vseries">volume series</a>
for playback with 
<a href="vseries.html"><b>vseries</b></a> or the associated slider
(if <b>true</b>), or as independent models (if <b>false</b>).
If the option is not used, 3D TIFF images and
<a href="save.html#chimap">Chimera map</a> files containing five or more
same-sized maps are automatically interpreted as series, if possible.
When multiple maps are opened in a single command (multiple files specified with
wild cards or globbing), time ordering is automatically discerned from the
filenames, if possible: for example, a file will be identified as time point 12 
if its name contains either of the following:
<ul class="none">
<li>_stack0012_
<li>_0012stack_
</ul>
(and 0012 could be any integer).
</blockquote>
<blockquote>
<a name="verbose"></a>
<b>verbose</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
Whether to show detailed debugging information for 
<a href="../dicom/dicom.html">DICOM</a> input.
If <b>format dicom</b> is specified explicitly,
all of the DICOM files in a directory and its subdirectories 
can be opened with a single <b>open</b> command, for example:
</p>
<blockquote>
<b>open "~/Desktop/2-CT 5.0 H30s-55580" format dicom</b>
</blockquote>
<p>
Each stack that is found will be opened as a separate model.
Alternatively, the multiple files of a stack can be specified 
with filename globbing as described <a href="#local">above</a>.
The option to specify a directory rather than filenames
is only available for this format. See also:
<a href="../dicom-quickref.html">ChimeraX DICOM Reference</a>
</p>
</blockquote>
<blockquote>
<a name="width"></a>
<b>width</b>&nbsp;&nbsp;<i>w</i>
<br>
Width in physical units of distance (&Aring;) for the
<a href="#photo">photo</a> (single image) being opened.
</blockquote>

<a name="traj-options"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>Trajectory Options</h3>
<p>
Only one frame (coordinate set) of a 
<a href="../trajectories.html">trajectory</a> can be viewed at a time.
Different frames can be accessed or played back sequentially with the
<a href="coordset.html"><b>coordset</b></a> command or the 
<a href="#slider">slider interface</a>.
</p><p>
Only multi-model mmCIF or PDB files provide all of the trajectory information 
in a single file. The <a href="#coordsets"><b>coordsets</b></a> option 
indicates reading the file as a trajectory instead of as multiple models,
for example:  
<ul class="none">
<li><b>open 1plx.cif coordsets true</b>
</ul>
<p>
Otherwise, the trajectory topology and coordinates come 
from two different files, either:
</p>
<ul>
<li>a <a href="#trajectory">trajectory coordinate file</a>
with topology provided by an already-open atomic model specified with the
<a href="#structureModel"><b>structureModel</b></a> option, for example:
<br><b>open myfile.trr structureModel #1</b>
<li>a PSF <a href="#trajtop">trajectory topology file</a> 
with the corresponding <a href="#trajectory">trajectory coordinate file</a>
specified with the <a href="#coords"><b>coords</b></a> option, for example:
<br><b>open myfile.psf coords myfile.dcd</b>
</ul>
<p>
Option details:
</p>
<blockquote>
<a name="coordsets"></a>
<a name="missingCoordsets"></a>
<b>coordsets</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
[&nbsp;<b>missingCoordsets</b>&nbsp;&nbsp;fill&nbsp;|&nbsp;ignore&nbsp;|&nbsp;<b>renumber</b>&nbsp;]
<br>
Whether to open a multi-model file in mmCIF or PDB format as a
<a href="../trajectories.html">trajectory</a>.
By default (<b>coordsets false</b>), 
the multiple models in the file will be opened not as a
<a href="../trajectories.html">trajectory</a> but as
separate submodels that can be shown/hidden/specified independently 
(for example, using #1.1, #1.2, #1.3...) 
or collectively (for example, using #1).
<p>
The <b>missingCoordsets</b> sub-option indicates how
to treat gaps in MODEL numbering when reading a PDB multi-model file
as a <a href="../trajectories.html">trajectory</a>
with <b>coordsets true</b>.
In PDB format, MODEL records are typically numbered consecutively.
The numbers from these records are used as the frame numbers of the 
coordinate sets in the trajectory.
When gaps in the numbering are detected, options are to:
</p>
<ul>
<li><b>fill</b>
&ndash; generate duplicate coordinate sets (additional trajectory frames) 
to fill in the gaps
<li><b>ignore</b> 
&ndash; use the MODEL numbers without filling gaps
<li><b>renumber</b> (default)
&ndash; renumber the coordinate sets consecutively
</ul>
<p>
For example, if the first coordinate set MODEL 1 is immediately
followed in the file by another coordinate set MODEL 6,
<b>fill</b> would create four additional copies of the first 
coordinate set as frames 2-5,
<b>ignore</b> would give trajectory frame 1 immediately followed by
trajectory frame 6, and
<b>renumber</b> would renumber the MODEL 6 coordinate set as trajectory frame 2.
</blockquote>
<blockquote>
<a name="structureModel"></a>
<b>structureModel</b>&nbsp;&nbsp;<a href="atomspec.html#hierarchy"><i>model-spec</i></a>
<br>
To provide the topology information needed to open a 
<a href="#trajectory">trajectory coordinate file</a> as a 
<a href="../trajectories.html">trajectory</a>, the <b>structureModel</b>
option is used to specify an existing model with the same set of atoms
(for example: <b>open myfile.trr structureModel #1</b>).
This model can itself be a <a href="../trajectories.html">trajectory</a>. 
If the <b>structureModel</b> option is omitted and only one atomic model
is present, that model will be used; if more than one is present,
a dialog will appear for specifying which to use.
If the model is not a trajectory (only has one set of coordinates),
the coordinates will be replaced with those in the
incoming coordinate file unless <a href="#replace"><b>replace false</b></a>
is used to indicate appending the new frames instead.
If the model is a trajectory (contains multiple frames), the incoming
frames will be appended at the end of the existing ones unless 
<a href="#replace"><b>replace true</b></a> is used to indicate replacing
them instead.
The <b>structureModel</b> option is also needed to open local MTZ files
(<a href="#clipper">details...</a>).
</blockquote>
<blockquote>
<a name="replace"></a>
<b>replace</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false
<br>
Whether to replace the coordinates 
of the existing <a href="#structureModel">structure model</a>
(which could be a <a href="../trajectories.html">trajectory</a> 
with multiple frames) with those from the incoming
<a href="#trajectory">trajectory coordinate file</a>.
The default behavior (option not specified) is to replace the coordinate
set if the existing model is not a trajectory, <i>i.e.</i>, has only one 
coordinate set, and to append the new frames if the model is a trajectory
and thus already contains multiple frames.
Giving the option as <b>true</b> indicates replacing all one or more existing
frames, and <b>false</b> indicates appending the new frames to the existing
ones regardless of how many frames the model already has.
</blockquote>
<blockquote>
<a name="coords"></a>
<b>coords</b> &nbsp;<i>trajectory-coordinate-file</i>&nbsp;
<br>
As an alternative to opening
a <a href="#trajectory">trajectory coordinate file</a> and supplying the
topology with the <a href="#structureModel"><b>structureModel</b></a> option,
one can open a PSF <a href="#trajtop">trajectory topology file</a> 
and use the <b>coords</b> option to supply
the corresponding <a href="#trajectory">trajectory coordinate file</a>.
Any of the <a href="#trajectory">trajectory coordinate formats</a> may be used.
The <i>trajectory-coordinate-file</i> is specified by pathname, or by the word
<a href="usageconventions.html#browse"><b>browse</b></a>
to specify it interactively in a file browser window.
Omitting this option also raises a file browser window
for specifying the file interactively.
</blockquote>
<blockquote>
<a name="start"></a>
<a name="end"></a>
<a name="step"></a>
<b>start</b>&nbsp;&nbsp;<i>start</i></a>
<br>
<b>end</b>&nbsp;&nbsp;<i>end</i></a>
<br>
<b>step</b>&nbsp;&nbsp;<i>step</i></a>
<br>
The <b>start</b>, <b>end</b> and <b>step</b> options allow loading a subset of 
the coordinate sets from a <a href="#trajectory">trajectory coordinate file</a>,
starting with frame number <i>start</i> and proceeding with steps of 
<i>step</i> but not exceeding the specified <i>end</i>. For example,
<b>start 2 end 25 step 5</b> would read in coordinate sets 2 
(the second one in the file), 7 (the seventh), 12, 17, and 22. 
These options take integer values only, and they do not apply to 
reading trajectories from multi-model mmCIF or PDB files.
</blockquote>
<blockquote>
<a name="slider"></a>
<b>slider</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false
<br>
Whether to show a graphical interface for 
<a href="../trajectories.html">trajectory</a> playback (default <b>true</b>).
The slider can be dragged or a frame number entered directly.
The interface also includes a play/pause button, a
&nbsp;<img style="height:16px; margin: 1px -1px; vertical-align:top"
src="../tools/ui-icons/snail.png">&nbsp; value to increase for slower playback,
and a button for recording a movie
(<img class="iconbar" src="../tools/ui-icons/record.png">).
The movie will start at the current slider position, so
to include the whole trajectory, place the slider at the far left
before clicking the record button.
See also: <a href="coordset.html#slider"><b>coordset slider</b></a>
</blockquote>

<a name="nmrstar"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>NMR-STAR Options</h3>
<p>
<a href="https://link.springer.com/article/10.1007/s10858-018-0220-3"
target="_blank">NMR-STAR</a> files contain distance constraints
that will be shown as <a href="../pseudobonds.html">pseudobonds</a>
between atoms. These files can be <a href="#fetch">fetched</a> from the
PDB for certain structures determined by nuclear magnetic resonance (NMR),
or obtained directly from the PDB website using the Download option
&ldquo;Combined NMR data (NMR-STAR) (.gz).&rdquo;
</p>
<blockquote>
<b>structures</b>&nbsp;&nbsp;<a href="atomspec.html#hierarchy"><i>model-spec</i></a>
<br>
Specifies the atomic model(s) on which to draw
<a href="../pseudobonds.html">pseudobonds</a> representing constraints
read from an NMR-STAR file. If this option is not given, every currently
open atomic model that contains all of the atoms associated with 
constraints in the NMR-STAR file will be used.
If an atomic model that does not contain all of the constraint-associated
atoms is specified, a warning will be issued, but constraint pseudobonds 
will still be generated for the atom pairs that are present.
</blockquote>
<blockquote>
<b>type</b>&nbsp;&nbsp;<i>type-name</i>
<br>
A single NMR-STAR may contain multiple types of constraints, where each
type has a name such as &ldquo;NOE&rdquo; or &ldquo;hydrogen bond.&rdquo;
By default, each type will be opened as a separate
<a href="../pseudobonds.html">pseudobond group</a> with the type included
in the model name.
The <b>type</b> option can be used to limit the input to a specific type.
If not specified, all types of constraints are read and each type is placed
in a separate <a href="../pseudobonds.html">pseudobond group</a>.
</blockquote>
<p>
Examples:
</p>
<blockquote>
<b>open 8r4w</b>
<br>
<b>open 8r4w from pdb_nmr structures #1 type NOE</b>
<br>
<br><b>open ~/Downloads/8r4w_nmr-data.str structures #1</b>
</blockquote>
<p>
Each restraint has a maximum distance. If the distance in a structure is
within the maximum, it is classified satisfied 
(<a href="atomspec.html#builtin">command-line specifier</a> for the
corresponding <a href="../pseudobonds.html">pseudobonds</a> is
<b>nmr-satisfied</b>), otherwise it is long 
(<a href="atomspec.html#builtin">command-line specifier</a> <b>nmr-long</b>).
Examples of using these in commands:
<blockquote>
<b>hide satisfied</b>
<br>
<b>show satisfied</b>
<br>
<b>hide long</b>
</blockquote>
<p>
NMR-STAR constraints and their maxima are saved in
<a href="save.html#session">sessions</a>.
</p>

<a name="pae"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>AlphaFold PAE Options</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>open</b> &nbsp;<i>filename</i>&nbsp; 
[&nbsp;<b><a href="#atomic">format</a>&nbsp;pae</b>&nbsp;]
[&nbsp;<b>structure</b>&nbsp;&nbsp;<a href="atomspec.html#hierarchy"><i>model-spec</i></a>&nbsp;]
[&nbsp;<b>plot</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<b>colorDomains</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>minSize</b>&nbsp;&nbsp;<i>M</i>&nbsp;]
[&nbsp;<b>connectMaxPae</b>&nbsp;&nbsp;<i>N</i>&nbsp;]
[&nbsp;<b>cluster</b>&nbsp;&nbsp;<i>resolution</i>&nbsp;]
[&nbsp;<b>dividerLines</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
<br><br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>open</b> &nbsp;&nbsp;<i>uniprot-ID</i>&nbsp; <b>from 
<a href="#fetch-table">alphafold_pae</a></b>
[&nbsp;<b>structure</b>&nbsp;&nbsp;<a href="atomspec.html#hierarchy"><i>model-spec</i></a>&nbsp;]
[&nbsp;<b>plot</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<b>colorDomains</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>minSize</b>&nbsp;&nbsp;<i>M</i>&nbsp;]
[&nbsp;<b>connectMaxPae</b>&nbsp;&nbsp;<i>N</i>&nbsp;]
[&nbsp;<b>cluster</b>&nbsp;&nbsp;<i>resolution</i>&nbsp;]
[&nbsp;<b>dividerLines</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
</blockquote>
<p>
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold</a>
is an artificial intelligence method for predicting the atomic structures
of biomolecules and their complexes. Besides the per-residue
<a href="https://alphafold.ebi.ac.uk/faq#faq-12" target="_blank">pLDDT
confidence measure</a>, AlphaFold gives
for each pair of <i>structural entities</i> (X,Y) the expected position error
at entity X if the predicted and true structures were aligned on Y.
Structural entities include standard biopolymer residues as well as the
individual atoms of other types of residues: ligands, ions, glycans,
and post-translationally modified residues.
Only AlphaFold 3 (not earlier versions) generates
predictions that include these other types of residues.
The &ldquo;predicted aligned error&rdquo; or PAE values can
be read and shown as a <a href="../tools/alphafold.html#pae">2D plot</a> 
using <b>open</b> (as described here),
<a href="alphafold.html#pae"><b>alphafold pae</b></a>,
the <a href="../tools/alphafold.html#pae"><b>AlphaFold Error Plot</b></a>
tool, or <a href="#fetch"><b>alphafold fetch</b></a> or
<a href="#fetch"><b>alphafold match</b></a> with the option <b>pae true</b>.
See also the
<a href="https://www.rbvi.ucsf.edu/chimerax/data/pae-apr2022/pae.html"
target="_blank">AlphaFold Error Estimates example</a> and
<a href="https://www.youtube.com/watch?v=oxblwn0_PMM" target="_blank">video</a>.
</p><p>
AlphaFold writes the PAE values to a JSON (.json) file, typically named
similarly to the corresponding model file (.cif).
Giving <a href="usageconventions.html#browse"><b>browse</b></a> as the 
<i>filename</i> brings up a dialog for browsing the local
filesystem interactively to find this JSON file. Alternatively,
for entries in the <a href="https://alphafold.ebi.ac.uk/"
target="_blank">AlphaFold Database</a>, the PAE data can be fetched directly
by <i>uniprot-id</i> (<a href="https://www.uniprot.org" 
target="_blank">UniProt</a> name or accession number). 
Either way, the following option may be required:
</p>
<blockquote>
<b>structure</b>&nbsp;&nbsp;<a href="atomspec.html#hierarchy"><i>model-spec</i></a>
<br>
Specifies the corresponding AlphaFold model structure. 
If this option is not given, ChimeraX will look for an
open atomic model opened from the same directory as the PAE file,
and if none is found, will raise a dialog for specifying which model to use.
</blockquote>
<p>
Setting <b>colorDomains</b> to <b>true</b> clusters the entities into
coherent domains (sets with relatively low pairwise PAE values)
and uses randomly chosen colors to distinguish these domains in the structure.
The entities are assigned an integer domain identifier (starting with 1) as an
<a href="../attributes.html">attribute</a> named <b>pae_domain</b>
that can be used to <a href="atomspec.html#attributes">specify</a>
them in commands (for example, to recolor or select specific domains).
Entities not grouped into any domain are assigned a <b>pae_domain</b>
value of <b>None</b>. The clustering uses the
NetworkX greedy_modularity_communities algorithm with parameters:
</p>
<ul>
<li><b>minSize</b> (default <b>10</b>)
&ndash; minimum number of entities allowed in a domain
<li><b>connectMaxPae</b> (default <b>5.0</b> &Aring;)
&ndash; the maximum PAE value allowed between entities for them to be
clustered into the same domain. Larger values give larger domains
and generally increase the time to compute the clustering,
which is ~5 seconds for 1000 residues when the default of 5.0 is used.
<li><b>cluster</b> (default <b>0.5</b>, typical range 0.5&ndash;5.0)
&ndash; graph resolution; larger values give smaller domains
</ul>
<p>
The <b>dividerLines</b> option (default <b>true</b>) indicates whether,
for multimer predictions, to draw lines on the plot demarcating the
end of one chain and the start of another. The lines may obscure a few
chain-terminal residues in the plot, and <b>dividerLines false</b>
can be used if this is problematic.
For predictions that include nonstandard residues and/or covalent
modifications, divider lines also segregate the entire set of such
entities from the biopolymer chain(s).
</p>

<a name="photo"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>Photo (Single Image)</h3>
<p>
A single image file can be opened as a flat model. Although flat,
the model exists in 3D and can be rotated and translated like other models.
Several image types (filename suffixes .png, .tif, <i>etc.</i>) 
are accepted, but except for .jpg, require specifying 
<a href="#miscellaneous"><b>format photo</b></a> 
in the <a href="#local"><b>open</b></a> command. Any of the following 
options can be used to specify the size of the image within the scene:
</p>
<blockquote>
<a href="#width"><b>width</b></a>&nbsp;&nbsp;<i>w</i>
<br>
<a href="#height"><b>height</b></a>&nbsp;&nbsp;<i>h</i>
<br>
<a href="#pixelSize"><b>pixelSize</b></a>&nbsp;&nbsp;<i>s</i>
</blockquote>
<p>
Width, height, and pixel size values are specified in physical units of 
distance (&Aring;). If width or height is given,
the other is set according to the aspect ratio of the image.
If both are given and their ratio differs
from that of the image, the image will be stretched.
Width and height values override any pixel size specification.
If no width, height, or pixel size is given,
a pixel size of <b>1.0</b> will be used.
Examples:
<blockquote>
<b>open ~/Desktop/headshot.png format photo width 5</b>
<br>
<b>open figure1.jpg height 30</b>
</blockquote>
</p>

<a name="image-wildcards"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>Image Stack Wildcards</h3>
<p>
For <a href="#image-stacks">image stacks</a> only, special wildcards
{t}, {c}, {z} can be used in the <a href="#top"><b>open</b></a>
command file pathname to specify the stack collectively.
These wildcards indicate the parts of the filename with
time, channel, and Z-indices, respectively. For example,
to open a stack where each file is a 3D volume
(each file contains all Z planes for one channel at one timepoint):
<blockquote>
open "/Volumes/data/micro/cell8_3 stars/cell8_{c}ch_{t}stack_*msec.tif"
</blockquote>
...and the file names look like:
<blockquote>
cell8_0ch_0000stack_000000msec.tif
<br>cell8_0ch_0001stack_001383msec.tif
<br>cell8_0ch_0002stack_002766msec.tif
<br>cell8_0ch_0003stack_004149msec.tif
<br>cell8_1ch_0000stack_000000msec.tif
<br>cell8_1ch_0001stack_001383msec.tif
<br>cell8_1ch_0002stack_002766msec.tif
<br>cell8_1ch_0003stack_004149msec.tif
<br>[...]
</blockquote>
The following example is for files each containing a single Z plane:
<blockquote>
open ~/Desktop/latest/r12c09f01p{z}-ch{c}sk{t}fk1fl1.tiff
</blockquote>
...where the file names look like:
<blockquote>
r12c09f01p01-ch1sk01fk1fl1.tiff
<br>r12c09f01p01-ch1sk02fk1fl1.tiff
<br>r12c09f01p01-ch1sk03fk1fl1.tiff
<br>r12c09f01p01-ch1sk04fk1fl1.tiff
<br>[...]
</blockquote>
See also the <a href="#channel"><b>channel</b></a>
option of the <a href="#top"><b>open</b></a> command.
</p>

<a name="compressed"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>Compressed Files</h3>
<p>
Certain <a href="#formats">file types</a> 
can be read when compressed, as indicated with
a suffix for the type of compression (.gz, .bz2, or .xz following the 
usual suffix for the type of file). 
For example, myfile.pdb.gz is a gzipped PDB file.
</p><p>
The file types that can be read by ChimeraX in compressed form
(as well as uncompressed) are
ALN, BILD, Collada, compiled Python code, directional FSC,
FASTA, glTF, HSSP, MMTF, MSF, NIfTI, PDB, Pfam, PIR, pseudobonds, Python code,
RSF, STL, Stockholm, STORM, VTK PolyData, and Wavefront OBJ.
</p>

<a name="defattr"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>Attribute Assignments</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>open</b> &nbsp;<i>attr-file</i>&nbsp;
[&nbsp;<a href="#miscellaneous"><b>format&nbsp;&nbsp;defattr</b></a>&nbsp;]
[&nbsp;<b>models</b>&nbsp;&nbsp;<a href="atomspec.html"><i>atom-spec</i></a>&nbsp;]
[&nbsp;<b>log</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
</blockquote>
<p>
The <b>open</b> command can read and assign
<a href="../attributes.html">attribute</a> values
from an <a href="../formats/defattr.html">attribute assignment file</a>.
It is possible to define new attributes or
to (re)assign values for existing attributes
of atoms, residues, and atomic models.
The <i>attr-file</i> is specified by its pathname, or by the word
<a href="usageconventions.html#browse"><b>browse</b></a>
to specify it interactively in a file browser window.
If the <a href="#models"><b>models</b></a> option is used, assignment
will consider only the models containing the specified atoms.
Of course, whether attribute values are actually assigned to those models
or their parts depends on the contents of the
<a href="../formats/defattr.html">attribute assignment file</a>.
The <b>log</b> option indicates whether to report details of the
assignments to the <a href="../tools/log.html"><b>Log</b></a>.
</p><p>
The related command <a href="setattr.html"><b>setattr</b></a>
assigns attribute values directly, but is less practical for large-scale
assignment since only a single value can be given per command.
See also:
<a href="info.html"><b>info</b></a>,
<a href="../hydrophobicity.html">amino acid hydrophobicity</a>
</p>

<a name="generic-hdf"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>Generic HDF File Reader</h3>
<p>
If <a href="#volume">EMAN HDF format</a> is expected, 
but that file reader fails, a generic HDF file reader is tried instead. 
This generic reader does not look for any metadata; it simply looks for
3-dimensional arrays and tries to open them as <a href="../tools/volumeviewer.html">volume models</a> with grid spacing of 1. The following 
<a href="#local"><b>open</b></a> command option can be used to
limit which 3D arrays are read:
</p>
<blockquote>
<b>arrayName</b>&nbsp;&nbsp;<i>name</i>
<br>
In <a href="https://www.hdfgroup.org/HDF5/" target="_blank">HDF5</a> format,
each array has a directory path within the file and 
a <i>name</i> (the last part of the pathname). This option indicates reading
only arrays with the specified <i>name</i>. Example:
 <blockquote>
 <b>open Cam_long_00003.h5 arrayName Data</b>
 </blockquote>
</blockquote>

<a name="url"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>URLs</h3>

<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>open</b> &nbsp;<i>URL</i>&nbsp;
[&nbsp;<b>dataFormat</b>&nbsp;&nbsp;<a href="#file-table"><i>format-name</i></a>&nbsp;]
[&nbsp;<a href="#maxModels"><b>maxModels</b></a>&nbsp;&nbsp;<i>M</i>&nbsp;]
[&nbsp;<a href="#coordsets"><b>coordsets</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#vseries"><b>vseries</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<a href="#resizeWindow"><b>resizeWindow</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;]
&nbsp;<a href="#options"><i>options</i></a>&nbsp;
</blockquote>
<p>
Uniform Resource Locators (URLs) starting with
<b>http:</b>, <b>https:</b>, and <b>ftp:</b> are handled.
File types and command options are the same as described 
<a href="#local">above</a> for local files.
The file type can be indicated by <a href="#file-table">suffix</a>
or the <b>dataFormat</b> option 
(where <a href="#file-table"><i>format-name</i></a> can be truncated).
</p><p>
Web-fetched data other than those handled by the
<a href="../tools/helpviewer.html">ChimeraX browser</a>
are stored locally as described <a href="#ignoreCache">below</a>.
</p>

<a name="fetch"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>Fetch from Online Sources</h3>

<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>: <b>open</b> 
[&nbsp;<a href="#fetch-table"><i>source1</i></a>:]<i>identifier</i>&nbsp;
[[&nbsp;<a href="#fetch-table"><i>source2</i></a>:]<i>identifier2</i>...&nbsp;]
[&nbsp;<b>fromDatabase</b>&nbsp;&nbsp;<a href="#fetch-table"><i>source</i></a>&nbsp;]
[&nbsp;<b>format</b>&nbsp;&nbsp;<a href="#fetch-table"><i>format-name</i></a>&nbsp;]
[&nbsp;<a href="#maxModels"><b>maxModels</b></a>&nbsp;&nbsp;<i>M</i>&nbsp;]
[&nbsp;<a href="#coordsets"><b>coordsets</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#ignoreCache"><b>ignoreCache</b></a>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<a href="#maxAssemblies"><b>maxAssemblies</b></a>&nbsp;&nbsp;<i>N</i>&nbsp;]
[&nbsp;<a href="#resName"><b>resName</b></a>&nbsp;&nbsp;<i>name</i>&nbsp;]
&nbsp;<a href="#options"><i>options</i></a>&nbsp;
</blockquote>
<p>
Atomic structures, density maps, and protein sequences can be fetched directly
into ChimeraX from online sources. A specific compound or entry is specified 
by an <i>identifier</i>, and its source (database or web service)
can be indicated with either a <i>prefix</i>
before the identifier or the <b>fromDatabase</b> option.
The default database for a 4-character identifier without prefix is the
<a href="https://www.rcsb.org/" target="_blank">RCSB Protein Data Bank</a>;
the default format
for that database is <b>mmcif</b>, although <b>pdb</b> or <b>mmtf</b>
can be specified with the <b>format</b> option.
Specifying multiple identifiers is equivalent to using the <b>open</b> command
multiple times with the same options for each identifier.
As an alternative to the command, <a href="../fetch.html"><b>Fetch by ID</b></a>
provides a graphical interface to many common fetch types.
</p><p>
See also: 
<a href="../tools/downloaddicom.html"><b>Download DICOM</b></a>,
<a href="../preferences.html#webaccess"><b>Web Access</b> preferences</a>
</p><p>
<a name="ignoreCache"></a>
Web-fetched data are stored locally in <b>~/Downloads/ChimeraX/</b> 
(where ~ indicates a user's home directory), with subdirectories
<b>PDB</b>, <b>EMDB</b>, <i>etc.</i>
For caching, no distinction is made among the different Protein Data Bank sites.
If a data file specified for opening is not found in this local cache or 
<b>ignoreCache</b> is set to <b>true</b>, the file will be fetched and cached.
</p><p>
<a name="maxAssemblies"></a>
Multiple biological assemblies are defined for many entries in the PDB.
By default, fetching from <b>rcsb_bio</b> or <b>pdbe_bio</b>
will return every assembly available for that entry in that database.
The <b>maxAssemblies</b> option limits fetching to the
first <i>N</i> assemblies.
</p><p>
<a name="resName"></a>
The <b>resName</b> option applies only to <a href="#pubchem">PubChem3D</a>
and <a href="#smiles">SMILES</a> fetches;
it assigns a residue name to the compound (default UNL).
For compatibility with other programs and PDB format,
three-character residue names consisting of uppercase letters 
and possibly numbers are recommended. 
</p><p>
The <b>colorConfidence</b> option (default <b>true</b>) applies only to 
<b>alphafold</b> or <b>esmfold</b>
fetch and specifies whether to color the predicted structures
by a confidence value (<a href="#colorConfidence">details...</a>).
The <b>alignTo</b> and <b>trim</b> options also apply only to 
<b>alphafold</b> or <b>esmfold</b> fetch (<a href="#alignTo">details...</a>).
See also: <a href="alphafold.html"><b>alphafold</b></a>,
<a href="esmfold.html"><b>esmfold</b></a>
</p>

<center>
<a name="fetch-table"><font size="+1"><b>Fetch Sources</b></font></a>
</center>
<table border cellpadding="4" cellspacing="0"><tr>
<td align="center" style="color:black; background-color:#cccccc"><b>database
or web service</b></td>
<td align="center" style="color:black; background-color:#cccccc"><b><i>source</i></b></td>
<td align="center" style="color:black; background-color:#cccccc"><b>description</b></td>
<td align="center" style="color:black; background-color:#cccccc"><span class="nowrap"><b><i>format-name</i></b></span></td>
</tr><tr>
<td align="center" rowspan="2"><a href="https://www.rcsb.org/"
target="_blank">RCSB Protein Data Bank</a></td>
<td align="center"><b>pdb</b></td>
<td align="center">atomic coordinates and associated annotations</td>
<td align="center"><b>mmcif</b> (default) or <b>pdb</b>
<br>or <a href="https://mmtf.rcsb.org/" target="_blank"><b>mmtf</b></a></td>
</tr><tr>
<td align="center" class="text"><b>rcsb_bio</b></td>
<td align="center" class="text">atomic coordinates for biological assembly
or assemblies, see <a href="#maxAssemblies"><b>maxAssemblies</b></a></td>
<td align="center">(no choice, <b>mmcif</b> or <b>pdb</b> depending on
the specific entry)</td>
</tr><tr>
<td align="center" rowspan="3"><a href="https://www.ebi.ac.uk/pdbe/"
target="_blank">Protein Data Bank <br>in Europe (PDBe)</a></td>
<td align="center"><b>pdbe</b></td>
<td align="center">atomic coordinates and associated annotations</td>
<td align="center"><b>mmcif</b> (default) or <b>pdb</b></td>
</tr><tr>
<td align="center" class="text"><b>pdbe_bio</b></td>
<td align="center" class="text">atomic coordinates for biological assembly
or assemblies, see <a href="#maxAssemblies"><b>maxAssemblies</b></a></td>
<td align="center"><b>mmcif</b></td>
</tr><tr>
<td align="center" class="text"><b>pdbe_updated</b></td>
<td align="center" class="text">atomic coordinates and associated annotations,
&nbsp;updated to include small-molecule connectivity 
and binding-site information as described in 
<a href="https://nar.oxfordjournals.org/content/44/D1/D385.long"
target="_blank">Velankar <i>et al.</i></a>,
<i>Nucleic Acids Res</i> <b>44</b>:D385 (2016)</td>
<td align="center"><b>mmcif</b>
</td>
</tr><tr>
<td align="center"><a href="https://pdbj.org/" target="_blank">Protein 
Data Bank Japan (PDBj)</a></td>
<td align="center"><b>pdbj</b></td>
<td align="center">atomic coordinates and associated annotations</td>
<td align="center"><b>mmcif</b></td>
</tr><tr>
<td align="center"><a href="https://www.wwpdb.org/data/ccd"
target="_blank">PDB Chemical Component Dictionary (CCD)</a>
</td>
<td align="center"><b>ccd</b></td>
<td align="center">idealized structure of PDB chemical component 
specified by CCD ID (residue name)</td>
<td align="center"><b>ccd</b></td>
</tr><tr>
<td align="center">
<a href="https://pdb-redo.eu/" target="_blank">PDB-REDO</a>
</td>
<td align="center"><b>redo</b></td>
<td align="center" class="text">
atomic coordinates or diffraction data (if MTZ)
specified by PDB identifier</td>
<td align="center"><b>mmcif</b> (default) or <b>pdb</b> or <b>mtz</b>*
<br><br>*<b>mtz</b> requires installing the
<a href="https://cxtoolshed.rbvi.ucsf.edu/apps/chimeraxclipper"
target="_blank"><b>Clipper</b> bundle</a> from the
<a href="https://cxtoolshed.rbvi.ucsf.edu/"
target="_blank">ChimeraX Toolshed</a>
and specifying the corresponding atomic model
(<a href="#clipper">details...</a>)
</td>
</tr><tr>
<td align="center">
<a href="https://link.springer.com/article/10.1007/s10858-018-0220-3"
target="_blank">PDB NMR-STAR</a>
</td>
<td align="center"><b>pdb_nmr</b></td>
<td align="center" class="text">nuclear magnetic resonance (NMR) 
distance constraints; will be displayed on atomic model(s) specified
with the <a href="#nmrstar"><b>structures</b></a> option</td>
<td align="center"><b>nmrstar</b></td>
</tr><tr>
<td align="center" rowspan="2">
<a href="https://www.ebi.ac.uk/pdbe/" target="_blank">PDBe</a>
Electron Density Server (EDS)
</td>
<td align="center"><b>eds</b></td>
<td align="center">(2Fo-Fc) electron density map for a PDB entry;
<br>not available for all PDB entries</td>
<td align="center" rowspan="2"><b>ccp4</b></td>
</tr><tr>
<td align="center"><b>edsdiff</b></td>
<td align="center">(Fo-Fc) electron density difference map for a PDB entry;
<br>not available for all PDB entries</td>
</tr><tr>
<td align="center" class="text"><a href="https://www.ebi.ac.uk/emdb/"
target="_blank">Electron Microscopy Data Bank (EMDB)</a></td>
<td align="center"><b>emdb</b></td>
<td align="center">electron density map
<br>
(from wwpdb.org to .edu/.gov hosts, Chinese mirror site to .cn hosts, 
Japanese mirror to .jp hosts, otherwise ebi.ac.uk; 
Japanese site uses https, others
<a href="ftp://ftp.wwpdb.org/pub/emdb/README" target="_blank">ftp</a>)
<br>
(for associated atomic structures, if any,
see command option <a href="#fits"><b>fits</b></a> or
data source <b>emdb_fits</b>, below)
</td>
<td align="center" rowspan="5"><b>ccp4</b></td>
</tr><tr>
<td align="center" class="text"><a href="https://www.ebi.ac.uk/emdb/"
target="_blank">EMDB</a> ebi.ac.uk</td>
<td align="center"><b>emdb_europe</b></td>
<td align="center">electron density map
<br>
(command option <a href="#transferMethod"><b>transferMethod</b></a>
can be <b>https</b> or <b>ftp</b>)
</td>
</tr><tr>
<td align="center" class="text"><a href="https://www.ebi.ac.uk/emdb/"
target="_blank">EMDB</a> Chinese mirror site</td>
<td align="center"><b>emdb_china</b></td>
<td align="center">electron density map
<br>
(<a href="#transferMethod"><b>transferMethod</b></a> <b>ftp</b> only)
</td>
</tr><tr>
<td align="center" class="text"><a href="https://www.ebi.ac.uk/emdb/"
target="_blank">EMDB</a> Japanese mirror site</td>
<td align="center"><b>emdb_japan</b></td>
<td align="center">electron density map
<br>
(command option <a href="#transferMethod"><b>transferMethod</b></a>
can be <b>https</b> or <b>ftp</b>)
</td>
</tr><tr>
<td align="center" class="text"><a href="https://www.ebi.ac.uk/emdb/"
target="_blank">EMDB</a> wwpdb.org</td>
<td align="center"><b>emdb_us</b></td>
<td align="center">electron density map
<br>
(command option <a href="#transferMethod"><b>transferMethod</b></a>
can be <b>https</b> or <b>ftp</b>)
</td>
</tr><tr>
<td align="center" class="text">
<a href="https://www.ebi.ac.uk/emdb/" target="_blank">EMDB</a> and PDB</td>
<td align="center"><b>emdb_fits</b></td>
<td align="center">electron density map and associated atomic models, if any
</td>
<td align="center"><b>ccp4</b> and <b>mmcif</b></td>
</tr><tr>
<td align="center">
<a href="https://pdb-ihm.org/" target="_blank">PDB-IHM</a></td>
<td align="center"><b>pdbihm</b></td>
<td align="center" class="text">integrative hybrid model;
<br>
to be used with the current 4-character IDs, for example:
<br><b>open pdbihm:8zzc</b> &nbsp;&ndash;or&ndash;&nbsp;
<b>open 8zzc from pdbihm</b>
</td>
<td align="center"><b>ihm</b></td>
</tr><tr>
<td align="center">
<a href="https://pdb-dev.wwpdb.org/" target="_blank">PDB-Dev</a></td>
<td align="center"><b>pdbdev</b></td>
<td align="center" class="text">integrative hybrid model;
<br>
to be used with the old-style PDB-Dev identifiers, 
where leading zeroes can be omitted; for example, 
accession code PDBDEV_00000001 can be opened with command:
<br><b>open pdbdev:1</b>
</td>
<td align="center"><b>ihm</b></td>
</tr><tr>
<td align="center"><a href="https://alphafold.ebi.ac.uk/"
target="_blank">AlphaFold</a></td>
<td align="center"><b>alphafold</b></td>
<td align="center">theoretical protein structure modeled by
<a href="https://www.nature.com/articles/s41586-021-03819-2"
target="_blank">AlphaFold</a> and specified by
<a href="https://www.uniprot.org/" target="_blank">UniProt</a>
name or accession number (<a href="alphafold.html#fetch">details...</a>);
any of the <a href="alphafold.html#fetch"><b>alphafold fetch</b></a>
options can be used</td>
<td align="center"><b>mmcif</b></td>
</tr><tr>
<td align="center"><a href="https://alphafold.ebi.ac.uk/"
target="_blank">AlphaFold</a> PAE</td>
<td align="center"><b>alphafold_pae</b></td>
<td align="center"><a href="#pae">AlphaFold predicted aligned error</a> (PAE) 
specified by <a href="https://www.uniprot.org/" target="_blank">UniProt</a>
name or accession number (<a href="alphafold.html#fetch">details...</a>)
and associated with the corresponding AlphaFold model using the 
<a href="#pae"><b>structure</b></a> option</td>
<td align="center"><b>pae</b></td>
</tr><tr>
<td align="center"><a href="https://modelarchive.org/"
target="_blank">ModelArchive</a></td>
<td align="center"><b>modelarchive</b></td>
<td align="center">theoretical structure specified by 
<a href="https://modelarchive.org/" target="_blank">ModelArchive</a> identifier;
use option <b>pae true</b> to also plot its pairwise scores, if any,
with <a href="../tools/alphafold.html#pae"><b>AlphaFold Error Plot</b></a>;
see also <a href="modelcif.html#pae"><b>modelcif pae</b></a>
<br>(<a href="https://modelarchive.org/" target="_blank">ModelArchive</a>
file-reading support courtesy of Gerardo Tauriello)</td>
<td align="center"><b>mmcif</b></td>
</tr><tr>
<td align="center">
<a href="https://esmatlas.com/about#about_this_website"
target="_blank">ESM Metagenomic Atlas</a></td>
<td align="center"><b>esmfold</b></td>
<td align="center">theoretical protein structure modeled by
<a href="https://pubmed.ncbi.nlm.nih.gov/36927031/" target="_blank">ESMFold</a>
(Evolutionary Scale Modeling) and specified by its MGnify identifier
(<a href="esmfold.html#fetch">details...</a>)</td>
<td align="center"><b>pdb</b></td>
</tr><tr>
<td align="center"><a href="http://www.crystallography.net/cod/"
target="_blank">Crystallography Open Database</a></td>
<td align="center"><b>cod</b></td>
<td align="center">crystal structures of small molecules and 
inorganic compounds</td>
<td align="center"><b>corecif</b></td>
</tr><tr>
<td align="center"><a href="http://www.crystallography.net/pcod/"
target="_blank">Predicted Crystallography Open Database</a></td>
<td align="center"><b>pcod</b></td>
<td align="center">predicted structures of inorganic compounds</td>
<td align="center"><b>corecif</b></td>
</tr><tr>
<td align="center"><a href="http://www.cellpack.org"
target="_blank">cellPACK</a></td>
<td align="center"><b>cellpack</b></td>
<td align="center">hierarchical model;
<br>
<!--
only the model with identifier <b>HIV-1_0.1.6</b> is available currently
-->
see <a href="http://www.cellpack.org/use/" target="_blank">cellPACK
Chimera instructions</a> for available identifiers
</td>
<td align="center"><b>cellpack</b></td>
</tr><tr>
<td align="center"><a href="https://pubchemdocs.ncbi.nlm.nih.gov/pubchem3d"
target="_blank">PubChem3D</a></td>
<td align="center"><b>pubchem</b></td>
<td align="center"><a name="pubchem"></a>
modeled atomic coordinates specified by
<a href="https://pubchem.ncbi.nlm.nih.gov/" target="_blank">PubChem</a>
compound identifier (CID); 
<br>available for most but not all entries in PubChem Compound
<br>(partial charges are assigned as the atom 
<a href="../attributes.html">attribute</a> named <b>charge</b>)
</td>
<td align="center"><b>sdf</b></td>
</tr><tr>
<td align="center">IUPAC &rarr; SMILES &rarr; 3D translator</td>
<td align="center"><b>iupac</b></td>
<td align="center"><a name="iupac"></a>
modeled atomic coordinates specified by
IUPAC name, first translated to
a <a href="https://www.daylight.com/dayhtml/doc/theory/theory.smiles.html"
target="_blank">SMILES</a> string using the
<a href="https://opsin.ch.cam.ac.uk/index.html" target="_blank">OPSIN
web service</a> provided by the <a href="https://www-cmi.ch.cam.ac.uk/"
target="_blank">Centre for Molecular Informatics</a>
at the University of Cambridge, followed by the process
described <a href="#smiles">below</a> for SMILES; see also
<a href="../tools/buildstructure.html"><b>Build Structure</b></a>
</td>
<td align="center"><b>iupac</b></td>
</tr><tr>
<td align="center">SMILES &rarr; 3D translator</td>
<td align="center"><b>smiles</b></td>
<td align="center"><a name="smiles"></a>
modeled atomic coordinates specified by
<a href="https://www.daylight.com/dayhtml/doc/theory/theory.smiles.html"
target="_blank">SMILES</a> string, generated using the
<a href="https://cactus.nci.nih.gov/translate/" target="_blank">SMILES
translator</a> provided by the
<a href="https://cactus.nci.nih.gov/" target="_blank">National Cancer Institute
CADD group</a>; see also
<a href="../tools/buildstructure.html"><b>Build Structure</b></a>
</td>
<td align="center"><b>smiles</b></td>
</tr><tr>
<td align="center"><a href="https://www.uniprot.org" 
target="_blank">UniProt</a></td>
<td align="center"><b>uniprot</b></td>
<td align="center">protein sequence
<br>
(command option <a href="#associate"><b>associate</b></a>)
<br>
(associated tool: 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>;
see <a href="../tools/sequenceviewer.html#uniprot">UniProt
Sequence Features</a>)</td>
<td align="center"><b>uniprot</b></td>
</tr><tr>
<td align="center">
<a href="https://www.science.org/doi/10.1126/science.adg7492"
target="_blank">AlphaMissense</a></td>
<td align="center"><b>alpha_missense</b></td>
<td align="center">
<a href="https://www.science.org/doi/10.1126/science.adg7492"
target="_blank">AlphaMissense</a>
predicted pathogenicity scores for missense mutations,
ranging from 0 (benign) to 1 (pathogenic) and assigned score name <b>amiss</b>
for analysis with <a href="mutationscores.html"><b>mutationscores</b></a>
<br>(command option <a href="#chains"><b>chains</b></a>)
</td>
<td align="center"><b>amiss</b></td>
</tr><tr>
<td align="center">
<a href="https://www.uniprot.org/help/disease_phenotypes_variants_section"
target="_blank">UniProt Variants</a></td>
<td align="center"><b>uniprot_variants</b></td>
<td align="center">
<a href="http://genetics.bwh.harvard.edu/pph2/" target="_blank">PolyPhen-2</a>
and <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC168916/" 
target="_blank">SIFT</a> predicted pathogenicity values for clinically
observed human protein variants, assigned score names
<b>PolyPhen</b> and <b>SIFT</b>, respectively,
for analysis with <a href="mutationscores.html"><b>mutationscores</b></a>
<br>(command option <a href="#chains"><b>chains</b></a>)
</td>
<td align="center"><b>variants</b></td>
</tr></table>

<a name="clipper"></a>
<p class="nav">
[<a href="#top">back to top: open</a>]
</p>
<h3>Open/Save Options from Clipper</h3>
<p>
The <a href="https://cxtoolshed.rbvi.ucsf.edu/apps/chimeraxclipper"
target="_blank"><b>Clipper</b> bundle</a> from the
<a href="https://cxtoolshed.rbvi.ucsf.edu/"
target="_blank">ChimeraX Toolshed</a> allows
<a href="#top">opening</a> and <a href="save.html">saving</a>
crystallographic reflection files in MTZ format.
See also: 
<a href="https://tristanic.github.io/chimerax-clipper/commands/clipper.html"
target="_blank"><b>clipper</b> commands</a>,
<a href="#fetch-table">PDB-REDO fetch</a>
</p><p>
<a href="#top">Opening</a>
an MTZ file requires specifying an already-open atomic structure
with the <a href="#structureModel"><b>structureModel</b></a> option,
for example, if that atomic structure is open as model #1:
</p>
<blockquote>
<b>open myfile.mtz structureModel #1</b>
</blockquote>
<p>
If the <a href="#structureModel"><b>structureModel</b></a> option
is not given in the command, however, a dialog will appear to allow specifying 
which atomic structure should be used.
</p><p>
Structure factors can also be <a href="#fetch">fetched</a> along
with the corresponding atomic structure from the
<a href="https://www.rcsb.org/" target="_blank">RCSB Protein Data Bank</a>
by specifying <b>structureFactors true</b>, for example:
<blockquote>
<b>open 6cmn structureFactors t</b>
</blockquote>
Not all PDB entries have structure factors available.
In either case (opening an MTZ file or fetching structure factors),
an <b>overSampling</b> <i>value</i> can be given to set the
fineness of the real-space grid relative to the data resolution;
values in the 1.5-4.0 range are generally recommended, with higher
values giving a finer grid.
</p><p>
<a href="save.html">Saving</a>:
Only previously fetched structure factors or those from
a previously opened MTZ file can be saved as MTZ,
with additional <a href="save.html"><b>save</b></a> command options:
</p>
<blockquote>
<b>preserveInput</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
Whether to pass through the original, unmodified structure factors 
as well as those actually used by Clipper (which may have been modified).
</blockquote>
<blockquote>
<b>saveMapCoeffs</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>
<br>
Whether to save the current amplitudes and phases for all maps 
generated by Clipper. If <b>false</b>, just save the experimental data.
</blockquote>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics / 
November 2025</address>
</body></html>
