<html>

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Command: rna</title>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRight" title="User Guide Index"/></a>

<h3><a href="../index.html#commands">Command</a>: rna</h3>

<h3 class="usage"><a href="usageconventions.html">Usage</a>:
<br><b>rna</b> &nbsp;<i>operation</i> &nbsp;<i>arguments</i></h3>
<p>
The <b>rna</b> command builds rough but potentially large-scale
models of single-stranded RNA, given base-pairing information.
See also: <a href="build.html"><b>build</b></a>
</p><p>
The <i>operation</i> can be:
<ul>
<li><a href="#model"><b>model</b></a>
&ndash; create an atomic model of single-stranded RNA 
from base-pairing information or a previously calculated path
<li><a href="#path"><b>path</b></a>
&ndash; create a path of <a href="../markers.html">markers</a> 
representing stems and loops from base-pairing information
</ul>
<p>

<a name="model"></a>
<a href="#top" class="nounder">&bull;</a>
<b>rna model</b></a> &nbsp;<i>sequence</i>&nbsp;
<b>path</b>&nbsp;&nbsp;<a href="atomspec.html#hierarchy"><i>marker-spec</i></a>
[&nbsp;<b>startSequence</b>&nbsp;&nbsp;<i>i</i>&nbsp;]
[&nbsp;<b>name</b>&nbsp;&nbsp;<i>n</i>&nbsp;]
<br>&nbsp;&nbsp;&ndash; or &ndash;<br>
<a href="#top" class="nounder">&bull;</a>
<b>rna model</b></a> &nbsp;<i>sequence</i>&nbsp;
<b>pairs</b>&nbsp;&nbsp;<i>pair-list</i>
[&nbsp;<b>startSequence</b>&nbsp;&nbsp;<i>i</i>&nbsp;]
[&nbsp;<b>name</b>&nbsp;&nbsp;<i>n</i>&nbsp;]
[&nbsp;<b>stemColor</b>&nbsp;&nbsp;<a href="colornames.html"><i>color-spec</i></a>&nbsp;]
[&nbsp;<b>loopColor</b>&nbsp;&nbsp;<a href="colornames.html"><i>color-spec</i></a>&nbsp;]
[&nbsp;<b>pColor</b>&nbsp;&nbsp;<a href="colornames.html"><i>color-spec</i></a>&nbsp;]
[&nbsp;<b>length</b>&nbsp;&nbsp;<i>l</i>&nbsp;]
[&nbsp;<b>pattern</b>&nbsp;&nbsp;<b>circle</b>&nbsp;|&nbsp;helix&nbsp;&nbsp;|&nbsp;line&nbsp;|&nbsp;sphere&nbsp;]
&nbsp;<a href="#pattern-options"><i>pattern-options</i></a>&nbsp;
[&nbsp;<b>loopPattern</b>&nbsp;&nbsp;<b>helix</b>&nbsp;&nbsp;|&nbsp;horseshoe&nbsp;]
&nbsp;<a href="#loop-pattern-options"><i>loop-pattern-options</i></a>&nbsp;
[&nbsp;<b>branchTilt</b>&nbsp;&nbsp;<i>angle<sub>b</b></i>&nbsp;]
[&nbsp;<b>loopTwist</b>&nbsp;&nbsp;<i>angle<sub>l</b></i>&nbsp;]
<blockquote>
<p>
Create an atomic model of single-stranded RNA.
The <i>sequence</i> can be a string of upper-case letters (such as AGCTU)
or the pathname of a FASTA file containing the desired sequence,
as in the example <a href="#examples">below</a>.
The <b>startSequence</b> option
(default <b>1</b>) indicates where to start within the supplied sequence.
The <b>name</b> option gives the name of the resulting molecule model
(default <b>RNA</b>).
</p><p>
Either a <b>path</b> or <b>pairs</b> must be supplied.
</p>
<ul>
<li>The <b>path</b> to be followed by the RNA is
<a href="atomspec.html#hierarchy">specified</a> as a series of
<a href="../markers.html">markers</a> previously generated 
with <a href="#path"><b>rna path</b></a>,
which also assigns the nucleotide orientations required to create the model.
<li>The <b>pairs</b> <i>pair-list</i> specification is the same as
for <a href="#path"><b>rna path</b></a>.
Base-paired and non-base-paired nucleotides in the
resulting structure will be colored according to the
<b>stemColor</b> and <b>loopColor</b> values
(defaults <b>yellow</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="yellow" width="10px" title="yellow"></td></tr></table>
and <b>cornflower blue</b> 
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="cornflowerblue" width="10px" 
title="cornflowerblue"></td></tr></table>),
respectively, and phosphorus atoms according to <b>pColor</b>
(default <b>orange</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="orange" width="10px" title="orange"></td></tr></table>).
The remaining options
are the same as described for <a href="#path"><b>rna path</b></a>.
</ul>
<p>
<b>Algorithm</b>: The path of the RNA is either precomputed with
<a href="#path"><b>rna path</b></a>
or computed implicitly using the same algorithm as that command.
The algorithm generates a series of markers and assigns an orientation
to each marker (<a href="#path-alg">details...</a>).
Atomic model templates are placed in the computed orientations
on top of the markers.
In stem regions, the orientations are chosen to form base-pairing
hydrogen bonds.
The atomic templates are in the file rna_templates_6pj6.cif
within the ChimeraX rna_layout bundle.
</p>
</blockquote>

<a name="path"></a>
<a href="#top" class="nounder">&bull;</a>
<b>rna path</b> &nbsp;<i>pair-list</i>&nbsp;
[&nbsp;<b>length</b>&nbsp;&nbsp;<i>l</i>&nbsp;]
[&nbsp;<b>markerRadius</b>&nbsp;&nbsp;<i>rad<sub>m</sub></i>&nbsp;]
[&nbsp;<b>stemColor</b>&nbsp;&nbsp;<a href="colornames.html"><i>color-spec</i></a>&nbsp;]
[&nbsp;<b>loopColor</b>&nbsp;&nbsp;<a href="colornames.html"><i>color-spec</i></a>&nbsp;]
[&nbsp;<b>name</b>&nbsp;&nbsp;<i>n</i>&nbsp;]
[&nbsp;<b>pattern</b>&nbsp;&nbsp;<b>circle</b>&nbsp;|&nbsp;helix&nbsp;&nbsp;|&nbsp;line&nbsp;|&nbsp;sphere&nbsp;]
&nbsp;<a href="#pattern-options"><i>pattern-options</i></a>&nbsp;
[&nbsp;<b>loopPattern</b>&nbsp;&nbsp;<b>helix</b>&nbsp;&nbsp;|&nbsp;horseshoe&nbsp;]
&nbsp;<a href="#loop-pattern-options"><i>loop-pattern-options</i></a>&nbsp;
[&nbsp;<b>branchTilt</b>&nbsp;&nbsp;<i>angle<sub>b</b></i>&nbsp;]
[&nbsp;<b>loopTwist</b>&nbsp;&nbsp;<i>angle<sub>l</b></i>&nbsp;]
<blockquote>
Create a path of <a href="../markers.html">markers</a>, 
one marker per nucleotide, to represent an RNA molecule with specified 
base-pairing interactions.
The <i>pair-list</i> can be either a comma-separated list of
numbers or the pathname of a text file containing three columns,
one line per stem, as in the example <a href="#examples">below</a>.
</p><p>
Consecutive markers and base-paired markers are connected by links.
Stretches of base-paired nucleotides are <b><i>stems</i></b>, whereas
nucleotides that are not base-paired are said to belong to <b><i>loops</i></b>.
Stems are described with triples of integers. For example,
1,50,10 indicates pairing nucleotide 1 with nucleotide 50 at the start of a
10-bp stem, such that 2 and 49, 3 and 48, ... 10 and 41 are also paired.
Multiple stems can be specified with additional triples (<i>e.g.</i>,
1,50,10,60,70,2 describes two stems), and any number of stems can be given.
A limitation is that residues within a stem range cannot be
paired with residues beyond that range; for example, if a stem starts
at 1,50, no other stem can pair residues in the range 1-50 with residues
outside that range.
</p><p>
Residue numbers in the resulting marker set start at 1.
The last residue number is the highest number specified for a stem,
unless a higher number is given with the <b>length</b> option.
The <b>markerRadius</b> option specifies marker radius 
(default <b>2.0</b> &Aring;).
Paired and unpaired markers are colored according to the
<b>stemColor</b> and <b>loopColor</b> values
(defaults <b>yellow</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="yellow" width="10px" title="yellow"></td></tr></table>
and <b>cornflower blue</b> 
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="cornflowerblue" width="10px" 
title="cornflowerblue"></td></tr></table>), respectively,
and the <b>name</b> option gives the name
of the resulting marker model (default <b>RNA path</b>).
<p>
<a name="pattern-options"></a>
The <b>pattern</b> option controls the overall RNA layout:
</p>
<ul>
<li><b>circle</b> (default)
<li><b>helix</b> &ndash; with associated <i>pattern-options</i>:
  <ul class="none">
  <li>[&nbsp;<b>helixRadius</b>&nbsp;&nbsp;<i>rad<sub>h</sub></i>&nbsp;]
(default <b>300.0</b> &Aring;) 
  <li>[&nbsp;<b>helixRise</b>&nbsp;&nbsp;<i>rise</i>&nbsp;]
(default <b>50.0</b> &Aring;) &ndash; rise per helical turn
  </ul>
<li><b>line</b> &ndash; linear
<li><b>sphere</b> &ndash; a spiral on a sphere
with associated <i>pattern-options</i>:
  <ul class="none">
  <li>[&nbsp;<b>sphereTurnSpacing</b>&nbsp;&nbsp;<i>spacing</i>&nbsp;]
(default <b>60.0</b> &Aring;) &ndash; distance on the sphere
between adjacent turns of the spiral
  <li>[&nbsp;<b>sphereRadius</b>&nbsp;&nbsp;<i>rad<sub>s</sub></i>&nbsp;]
(default calculated from the sequence length and turn spacing)
  <li>[&nbsp;<b>sphereTurns</b>&nbsp;&nbsp;<i>turns</sub></i>&nbsp;]
(default calculated from the sequence length and turn spacing)
&ndash; number of times the RNA wraps around the sphere from pole to pole;
does not have to be an integer
<p>
Specifying a sphere radius and number of turns that do not accommodate the
full sequence will give an error, but values that leave extra room on the
sphere are accepted.
</p>
  </ul>
</ul>
<p>
Stem and loop orientations are produced by random rotations, where each angle
of rotation is obtained by multiplying the <b>branchTilt</b> 
<i>angle<sub>b</sub></i> (default <b>0</b>&deg;) 
by a random number uniformly distributed between &ndash;1 and 1.
</p><p>
<a name="loop-pattern-options"></a>
The <b>loopPattern</b> option controls the layout of 
non-base-paired segments of the sequence:
</p>
<ul>
<li><b>helix</b> (default) &ndash; with associated <i>loop-pattern-options</i>:
  <ul class="none">
  <li>[&nbsp;<b>helixLoopSize</b>&nbsp;&nbsp;<i>N<sub>loop</sub></i>&nbsp;]
(default <b>8</b>) 
&ndash; minimum number of nucleotides per helical turn: in a given loop,
this number or 2x or 3x the value may be used (the largest possible) 
if the loop contains at least that many residues.  
For example, if the parameter is 8, there will be 24 nucleotides per turn
in loops of 24 or more residues, 16 per turn in loops of 16-23 residues, 
and 8 per turn in loops of 15 residues or fewer.
  <li>[&nbsp;<b>helixLoopRise</b>&nbsp;&nbsp;<i>rise<sub>loop</i></i>&nbsp;]
(default <b>20.0</b> &Aring;) &ndash; rise per helical turn
  </ul>
<li><b>horseshoe</b> &ndash; U-shaped, with associated <i>loop-pattern-options</i>
  <ul class="none">
  <li>[&nbsp;<b>horseshoeSideSize</b>&nbsp;&nbsp;<i>N<sub>side</sub></i>&nbsp;]
(default <b>10</b>) 
&ndash; number of nucleotides in each side of the &ldquo;U&rdquo;
  <li>[&nbsp;<b>horseshoeCurveSize</b>&nbsp;&nbsp;<i>N<sub>curve</sub></i>&nbsp;]
(default <b>8</b>) 
&ndash; number of nucleotides in the curve of the &ldquo;U&rdquo;
  <li>[&nbsp;<b>horseshoeSpacing</b>&nbsp;&nbsp;<i>N<sub>spacing</sub></i>&nbsp;]
(default <b>1</b>) 
&ndash; number of nucleotides between horseshoes, <i>i.e.</i>
when the stretch of unpaired residues is longer than the number of
nucleotides allocated to a single horseshoe
  </ul>
</ul>
<p>
Orientations of individual nucleotides within a loop are produced by random
rotations about the P-P line, where each angle of rotation is obtained by 
multiplying the 
<b>loopTwist</b> <i>angle<sub>l</sub></i> (default <b>0</b>&deg;)
by a random number uniformly distributed between &ndash;1 and 1.
</p><p>
<a name="path-alg">
<b>Algorithm</b></a>:
The markers are P backbone atom positions and are spaced so that backbone
bond lengths are reasonable.
Stem regions are generated as twisted double helices rotating
31.5&deg; per base pair, with pairs 2.55 &Aring; apart along the
helix axis.
The overall layout of the series of loops and stems is given by the
<b>pattern</b> option. Loops and stems radiate directly outward from
the pattern, with additional variations in angle according to the 
<b>branchTilt</b> parameter.
The <b>loopPattern</b> option specifies loop layout,
with the twist of each nucleotide in the loop varying according to the
<b>loopTwist</b> parameter.
</p>
</blockquote>

<a name="examples"></a>
<h4>Example</h4>

<blockquote><b>
rna path pairings.txt length 100
<br>
rna model hiv-pNL4-3.fasta #1 start 455
</b></blockquote>
<p>
These example commands generate a model for the first
100 nucleotides of HIV RNA, with secondary structure as described in:
</p>
<blockquote>
<a href="https://www.ncbi.nlm.nih.gov/pubmed/19661910" target="_blank">
Architecture and secondary structure of an entire HIV-1 RNA genome.</a>
Watts JM, Dang KK, Gorelick RJ, Leonard CW, Bess JW Jr, Swanstrom R,
Burch CL, Weeks KM.
<i>Nature.</i> 2009 Aug 6;460(7256):711-6.
</blockquote>
<p>
...using a base-pairing file named pairings.txt containing:
<p>
<pre>
1       57      3
5       54      11
17      43      5
25      38      4
58      104     8
67      94      3
70      90      4
</pre>
<p>
...and a sequence file named hiv-pNL4-3.fasta containing:
</p>
<pre>
>gi|296556482|gb|AF324493.2| HIV-1 vector pNL4-3, 1-9709
TGGAAGGGCTAATTTGGTCCCAAAAAAGACAAGAGATCCTTGATCTGTGGATCTACCACACACAAGGCTA
CTTCCCTGATTGGCAGAACTACACACCAGGGCCAGGGATCAGATATCCACTGACCTTTGGATGGTGCTTC
AAGTTAGTACCAGTTGAACCAGAGCAAGTAGAAGAGGCCAATGAAGGAGAGAACAACAGCTTGTTACACC
CTATGAGCCAGCATGGGATGGAGGACCCGGAGGGAGAAGTATTAGTGTGGAAGTTTGACAGCCTCCTAGC
ATTTCGTCACATGGCCCGAGAGCTGCATCCGGAGTACTACAAAGACTGCTGACATCGAGCTTTCTACAAG
GGACTTTCCGCTGGGGACTTTCCAGGGAGGTGTGGCCTGGGCGGGACTGGGGAGTGGCGAGCCCTCAGAT
GCTACATATAAGCAGCTGCTTTTTGCCTGTACTGGGTCTCTCTGGTTAGACCAGATCTGAGCCTGGGAGC
TCTCTGGCTAACTAGGGAACCCACTGCTTAAGCCTCAATAAAGCTTGCCTTGAGTGCTCAAAGTAGTGTG
</pre>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics / 
June 2020</address>
</body></html>
