<html>

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Command: sequence</title>
<style>@media (prefers-color-scheme: dark) { :root { color-scheme: dark; } }</style>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRighticon" title="User Guide Index"/></a>

<h3><a href="../index.html#commands">Command</a>: sequence</h3>
<p>
The <b>sequence</b> subcommands perform actions related to the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>
and associated structures:
</p>
<ul>
<li><a href="#align"><b>sequence align</b></a>
&ndash; calculate a new sequence alignment using a web service
<li><a href="#associate"><b>sequence associate</b></a> and
<a href="#dissociate"><b>sequence dissociate</b></a>
&ndash; control sequence-structure associations
<li><a href="#chain"><b>sequence chain</b></a>
&ndash; show the sequence of a structure chain (start the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>)
<li><a href="#expandsel"><b>sequence expandsel</b></a>
&ndash; expand the current <a href="../selection.html">selection</a>
to all residues
<a href="../tools/sequenceviewer.html#association">associated</a>
with the same column(s) of the sequence alignment
<li><a href="#header"><b>sequence header</b></a>
&ndash; show, hide, save alignment headers
<li><a href="#identity"><b>sequence identity</b></a>
&ndash; calculate pairwise percent identities
<li><a href="#refreshAttrs"><b>sequence refreshAttrs</b></a>
&ndash; for structure chains 
<a href="../tools/sequenceviewer.html#association">associated</a>
with more than one sequence alignment,
assign <a href="../attributes.html">attributes</a> 
consistent with a specific alignment
<li><a href="#refseq"><b>sequence refseq</b></a>
&ndash; set which sequence to use as the reference for numbering
<li><a href="#update"><b>sequence update</b></a>
&ndash; replace the sequence data of a structure chain with its
associated sequence in the 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>
</ul>
<p>
The <a href="../menu.html#tools"><b>Tools</b> menu</a> 
can also be used to show the sequence of a structure chain in the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>.
All of the other actions can be accessed from the
<a href="../tools/sequenceviewer.html#context"><b>Sequence Viewer</b>
context menu</a>.
</p>

<a name="chain"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Show Sequence from Structure</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;chain</b>
&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
[&nbsp;<b>viewer</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
</blockquote>
<p>
The command <b>sequence chain</b> shows the sequence of the specified 
biopolymer chain in the 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>,
although the graphical interface can be suppressed
(for example, to run a script that uses the sequence data but not its display)
with <b>viewer false</b>.
Only one structure chain should be specified per command.
See also: 
<a href="../tools/moldisplay.html"><b>Molecule Display</b></a> icon
<a href="../tools/moldisplay.html" title="Molecule Display Toolbar...">
<img class="iconbar" src="../tools/shortcut-icons/sequence.png"></a>
</p><p>
Other ways to start the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>:
Independent of structure, sequence alignments and individual sequences
can also be <a href="open.html#local">opened</a>
from <a href="open.html#sequence">files</a>
or <a href="open.html#fetch">fetched</a>
from <a href="https://www.uniprot.org/" target="_blank">UniProt</a>.
Other tools or commands may generate new sequence alignments
(<i>e.g.</i>, <a href="../tools/blastprotein.html#results"><b>Blast Protein</b>
results</a>, <a href="../tools/matchmaker.html"><b>Matchmaker</b></a>,
<a href="#align">sequence realignment</a>).
</p>

<a name="associate"></a>
<a name="disassociate"></a>
<a name="dissociate"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Sequence-Structure Association</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;associate</b>
&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
<span class="nowrap">[&nbsp;<i>alignment-ID</i>:<i>sequence-ID</i>&nbsp;]</span>
<br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence</b> 
(&nbsp;<b>dissociate</b>&nbsp;|&nbsp;<b>disassociate</b>&nbsp;)
&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
&nbsp;<i>alignment-ID</i>
</blockquote>
<p>
<a href="../tools/sequenceviewer.html#association">Sequence-structure
association</a> 
(such as for synchronized <a href="../selection.html">selection</a>)
occurs automatically, but the commands 
<b>sequence associate</b> and <b>sequence dissociate</b> 
(same as <b>sequence disassociate</b>)
allow more precise control, for example, of which structure chains are used 
for <a href="../tools/sequenceviewer.html#headers">header</a> calculations, 
or forcing or removing associations regardless of 
whether the number of mismatches would be tolerated by the 
<a href="../tools/sequenceviewer.html#association">automatic procedure</a>.
</p><p>
The command <b>sequence associate</b> associates
one or more structure chains with a sequence.
The target sequence for association is specified by alignment ID,
as reported in the title bar of the 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window,
and the name or index number of the target sequence in the alignment,
in the form: 
&nbsp;<span class="nowrap"><i>alignment-ID</i>:<i>sequence-ID</i></span>&nbsp;
(<a href="seqspec.html">details...</a>).
</p><p>
Alternatively, the <i>sequence-ID</i> can be omitted 
to associate each specified structure chain with the
the best-matching sequence in the alignment.
The <i>alignment-ID</i> can be omitted if only one alignment is open, or
if the <i>sequence-ID</i> is also omitted; in the latter case,
each specified structure chain will be associated with the 
best-matching sequence in each open alignment.
If either or both are omitted, the colon (:) should also be omitted
except in rare cases to disambiguate an alignment and sequence 
that have the same name.
</p><p>
For <b>sequence dissociate</b>, only the alignment needs to be specified, 
not an individual sequence, because a structure chain can only be 
associated with one sequence per alignment.
</p>

<a name="expandsel"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Expanding a Selection to Alignment Columns</h3>
<blockquote>
<a href="usageconventions.html">Usage</a>:
<b>sequence &nbsp;expandsel</b> [&nbsp;<i>alignment-ID</i>&nbsp;]
</blockquote>
<p>
Expand the current <a href="../selection.html">selection</a> to all residues
<a href="../tools/sequenceviewer.html#association">associated</a>
with the same column(s) of the specified sequence alignment, or if none
is specified, all sequence alignments.
The <i>alignment-ID</i> is shown in the title bar of the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window.
</p>

<a name="header"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Sequence Viewer Headers and Numbering</h3>
<blockquote>
<a href="usageconventions.html">Usage</a>:
<b>sequence &nbsp;header</b> [&nbsp;<i>alignment-ID</i>&nbsp;]
&nbsp;<i>header-name</i>
(&nbsp;<b>show</b>&nbsp;|&nbsp;<b>hide</b>&nbsp;|&nbsp;<b>save</b>&nbsp;&nbsp;<i>filename</i>&nbsp;)
</blockquote>
<p>
The command <b>sequence header</b> shows, hides, or saves a
<a href="../tools/sequenceviewer.html#headers">sequence header</a> to a file.
(It can also be used to change the sequence
<a href="../tools/sequenceviewer.html#settings-headers"><b>Headers</b>
preferences</a>, but command details are omitted here because normally
the <a href="../tools/sequenceviewer.html#settings"><b>Settings</b></a>
dialog will be used instead.)
</p><p>
The <i>header-name</i> can be <b>consensus</b>, <b>conservation</b>,
or <b>rmsd</b>, although there will only be an effect when that header
is available (for example, an RMSD header is only available for alignments
associated with at least two structures).
Headers are saved to a simple text format that lists the alignment positions 
and values. The <i>filename</i> can be given as a pathname or the word 
<a href="usageconventions.html#browse"><b>browse</b></a>
to bring up a file browser window for choosing the
name and location interactively. 
If multiple alignments are open but an <i>alignment-ID</i> is not specified, 
showing/hiding affects all applicable alignments. 
However, saving only works for a single header at a time, so an
<i>alignment-ID</i> must be given when more than one alignment is open.
</p>
<a name="refseq"></a>
<blockquote>
<a href="usageconventions.html">Usage</a>:
<b>sequence &nbsp;refseq</b>
<span class="nowrap">[&nbsp;<i>alignment-ID</i>:<i>sequence-ID</i>&nbsp;]</span>
</blockquote>
<p>
The overall numbering of an alignment is shown across the top at the
first position and every tenth position thereafter. By default, there
is no reference sequence, and the alignment is simply numbered 1, 11, 21, 
<i>etc.</i>, regardless of whether the columns contain gaps.
However, with <b>sequence refseq</b>, any of the individual sequences 
can be specified as the reference for numbering instead.
The sequence to use is specified by its alignment ID, shown in the title bar
of the <a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window,
plus the sequence's name or index number in the alignment, in the form:
&nbsp;<span class="nowrap"><i>alignment-ID</i>:<i>sequence-ID</i></span>&nbsp;
(<a href="seqspec.html">details...</a>).
</p><p>
When the reference sequence has a gap character at a position
where a number is to be shown (first column, eleventh column, <i>etc.</i>),
how that position falls relative to the reference is shown in parentheses.
For example, &ldquo;(<1)&rdquo; would be shown for alignment positions
before the reference's first residue, and &ldquo;(43/44)&rdquo;
for positions that fall between residues 43 and 44 of the reference.
</p><p>
Omitting the sequence identifier returns to the default numbering:
</p>
<blockquote>
<a href="usageconventions.html">Usage</a>:
<b>sequence &nbsp;refseq</b>
<span class="nowrap">[&nbsp;<i>alignment-ID</i>&nbsp;]</span>
</blockquote>
<p>
If more than one 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window
is present, the <i>alignment-ID</i> is required.
The <i>alignment-ID</i> is shown in the title bar of the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window.
</p>

<a name="identity"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Calculate Percent Identities</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;identity</b> &nbsp;<i>alignment-ID</i>&nbsp;
[&nbsp;<b>denominator</b>&nbsp;&nbsp;<b>shorter</b>&nbsp;|&nbsp;longer&nbsp;|&nbsp;nongap&nbsp;]
<br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;identity</b> &nbsp;<i>alignment-ID</i>&nbsp;
<span class="nowrap">&nbsp;<i>alignment-ID</i>:<i>sequence-ID</i>&nbsp;</span>
[&nbsp;<b>denominator</b>&nbsp;&nbsp;<b>shorter</b>&nbsp;|&nbsp;longer&nbsp;|&nbsp;nongap&nbsp;]
<br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;identity</b> 
<span class="nowrap">&nbsp;<i>alignment-ID</i>:<i>sequence-ID</i>&nbsp;</span>
<span class="nowrap">&nbsp;<i>alignment-ID</i>:<i>sequence-ID</i>&nbsp;</span>
[&nbsp;<b>denominator</b>&nbsp;&nbsp;<b>shorter</b>&nbsp;|&nbsp;longer&nbsp;|&nbsp;nongap&nbsp;]
</blockquote>
<p>
The <b>sequence identity</b> command calculates the pairwise percent identity
between sequences of the same length (including gaps, as shown in
the <a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window).
The calculation is always pairwise, but can be performed for all-by-all pairs
within a single alignment, or all-by-one, or between two specific sequences.
An entire alignment is specified by its ID, shown in the title bar of the 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window,
and an individual sequence by the alignment ID plus the sequence's
name or index number in the alignment, in the form: 
&nbsp;<span class="nowrap"><i>alignment-ID</i>:<i>sequence-ID</i></span>&nbsp;
(<a href="seqspec.html">details...</a>).
</p><p>
Results are listed in the <a href="../tools/log.html"><b>Log</b></a>. 
For each pair, the number of columns with identical residues is given 
as a percentage of the specified <b>denominator</b>:
</p>
<ul>
<li><b>shorter</b> (default) 
&ndash; the number of residues in the shorter of the two sequences
<li><b>longer</b>
&ndash; the number of residues in the longer of the two sequences
<li><b>nongap</b>
&ndash; the number of columns where neither sequence has a gap
</ul> 

<a name="refreshAttrs"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Refresh Attributes</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;refreshAttrs</b> &nbsp;<i>alignment-ID</i>&nbsp;
</blockquote>
<p>
For structure chains
<a href="../tools/sequenceviewer.html#association">associated</a>
with more than one sequence alignment, make sure that the
current <a href="../tools/sequenceviewer.html#headers">attribute values</a>
are consistent with a specific alignment.
The <i>alignment-ID</i> is shown in the title bar of the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window.
</p>

<a name="update"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Update Structure Metadata with Associated Sequence</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;update</b> 
&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
[&nbsp;<b>alignment</b>&nbsp;&nbsp;<i>alignment-ID</i>&nbsp;]
</blockquote>
<p>
Standard PDB files have SEQRES records containing the sequence of
each biopolymer chain in the structure, and similarly, standard mmCIF files
include this information in specific tables. In many structures, some of
the residues in the sequence are missing from the atomic coordinates
because their positions could not be resolved from the experimental density
(due to disorder, <i>etc.</i>).
However, atomic coordinate files output from various modeling programs often
lack sequence information, so that the sequence is only inferred from 
the coordinates. The <b>sequence update</b> command is useful for
adding full sequence information to such structures.
For example, the full sequence could be opened from a FASTA file or by 
<a href="open.html#fetch">fetching</a> it from
<a href="https://www.uniprot.org" target="_blank">UniProt</a>,
associated with the chain, and then used to update the chain's sequence
information. This command only applies to the sequences associated with
the specified chains, limited to a specific <b>alignment</b> as needed.
The <i>alignment-ID</i> is shown in the title bar of the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window.
The sequence must cover all of the residues in the structure chain.
</p>

<a name="align"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Align Sequences using Clustal Omega or MUSCLE</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;align</b> &nbsp;<i>alignment-ID</i>
[&nbsp;<b>replace</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>program</b>&nbsp;&nbsp;<b>clustalOmega</b>&nbsp;|&nbsp;muscle&nbsp;]
<br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;align</b> 
&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
[&nbsp;<b>program</b>&nbsp;&nbsp;<b>clustalOmega</b>&nbsp;|&nbsp;muscle&nbsp;]
<br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;align</b> 
&nbsp;<i>sequence1</i>,<i>sequence2</i>[,<i>sequence3</i>...,<i>sequenceN</i>]
[&nbsp;<b>program</b>&nbsp;&nbsp;<b>clustalOmega</b>&nbsp;|&nbsp;muscle&nbsp;]
</blockquote>
<p>
The <b>sequence align</b> command calculates a new alignment of the
specified protein sequences using a web service hosted by the
<a href="https://www.rbvi.ucsf.edu/Overview/index.html" target="_blank">UCSF
RBVI</a>. The result is opened in a new
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window,
except that <b>replace true</b> (default <b>false</b>)
can be used to specify overwriting an 
existing alignment when all of its sequences are being realigned.
</p><p>
The sequences to align can be specified collectively by:
</p>
<ul>
<li>alignment ID, as shown in the title bar of an existing
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window;
<li>a <a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
(for atomic-structure protein chains already open in ChimeraX)
</ul>
...or individually as a comma-separated list (without spaces) 
of any combination of:
<ul>
<li>plain text of the entire amino acid sequence pasted directly 
into the command line
<li><a href="https://www.uniprot.org" target="_blank">UniProt</a>
name or accession number, for example:
<blockquote>
<b>sequence align ldlr_rat,ldlr_mouse,ldlr_human</b>
</blockquote>
<li>the <a href="seqspec.html"><i>sequence-spec</i></a> of a sequence
in the <a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>,
in the form: 
&nbsp;<span class="nowrap"><i>alignment-ID</i>:<i>sequence-ID</i></span>&nbsp;
(<a href="seqspec.html">details...</a>). Example:
<blockquote>
<b>open myfile.msf</b>
<br>
<b>sequence align 1,2,3,-1</b>
<br><i>&ndash; OR (if multiple sequence windows are open) &ndash;</i><br>
<span class="nowrap"><b>sequence align myfile.msf:1,myfile.msf:2,myfile.msf:3,myfile.msf:-1</b></span>
</blockquote>
</ul>
<p>
The <b>program</b> can be either of two choices:
</p>
<ul>
<li><a name="clustal"></a> 
<b>clustalOmega</b> (default, same as <b>clustal</b> or <b>omega</b>) 
&ndash; <a href="http://www.clustal.org/omega/" 
target="_blank">Clustal Omega</a> v1.1.0 with parameters: 
</p>
  <ul>
  <li>Number of guide-tree/HMM iterations: <b>1</b>
  <li>Full distance matrix during initial alignment: <b>true</b>
  <li>Full distance matrix during alignment iteration: <b>true</b>
  </ul>
<p>
See the <a href="http://www.clustal.org/omega/README"
target="_blank">README file</a> at the Clustal Omega website for details.
Users should cite:
</p>
<blockquote>
<a href="https://www.ncbi.nlm.nih.gov/pubmed/21988835" target="_blank">
Fast, scalable generation of high-quality protein multiple sequence alignments
using Clustal Omega.</a>
Sievers F, Wilm A, Dineen D, Gibson TJ, Karplus K, Li W, Lopez R, McWilliam H,
Remmert M, S&ouml;ding J, Thompson JD, Higgins DG.
<i>Mol Syst Biol.</i> 2011 Oct 11;7:539.
</blockquote>
<li><a name="muscle"></a>
<b>muscle</b> &ndash; <a href="https://drive5.com/muscle/" 
target="_blank">MUSCLE</a> v3.8.31 with parameters:
</p>
  <ul>
  <li>Maximum number of iterations: <b>16</b>
  <li>Maximum time to iterate: <b>unlimited</b>
  <li>Find diagonals (faster execution if sequences are similar,
  with possible decrease in accuracy): <b>false</b>
  </ul>
<p>
See the <a href="https://www.drive5.com/muscle/manual/options.html"
target="_blank">command-line reference</a> at the MUSCLE website
and <a href="https://www.biomedcentral.com/1471-2105/5/113"
target="_blank"><i>BMC Bioinformatics</i> <b>5</b>:113 (2004)</a> for details.
Users should cite:
</p>
<blockquote>
<a href="https://www.ncbi.nlm.nih.gov/pubmed/15034147" target="_blank">
MUSCLE: multiple sequence alignment with high accuracy and high throughput.</a>
Edgar RC.
<i>Nucleic Acids Res.</i> 2004 Mar 19;32(5):1792-7.
</blockquote>
</ul>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics / 
October 2024</address>
</body></html>
