<html lang="en"><!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
--><head><meta charset="utf-8">
<link rel="stylesheet" type="text/css" href="../userdocs.css">
<title>Command: sequence</title>
<style>@media (prefers-color-scheme: dark) { :root { color-scheme: dark; } }</style>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon" class="clRighticon" title="User Guide Index"></a>

<h3><a href="../index.html#commands">Command</a>: sequence</h3>
<p>
The <b>sequence</b> subcommands perform actions related to the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>
and associated structures:
</p>
<ul>
<li><a href="#align"><b>sequence align</b></a>
&ndash; calculate a new sequence alignment using a web service
</li><li><a href="#associate"><b>sequence associate</b></a> and
<a href="#dissociate"><b>sequence dissociate</b></a>
&ndash; control sequence-structure associations
</li><li><a href="#chain"><b>sequence chain</b></a>
&ndash; show the sequence of a structure chain (start the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>)
</li><li><a href="#expandsel"><b>sequence expandsel</b></a>
&ndash; expand the current <a href="../selection.html">selection</a>
to all residues
<a href="../tools/sequenceviewer.html#association">associated</a>
with the same column(s) of the sequence alignment
</li><li><a href="#header"><b>sequence header</b></a>
&ndash; show, hide, save alignment headers
</li><li><a href="#identity"><b>sequence identity</b></a>
&ndash; calculate pairwise percent identities
</li><li><a href="#match"><b>sequence match</b></a>
&ndash; superimpose structures based on the alignment of their 
<a href="../tools/sequenceviewer.html#association">associated</a> sequences
</li><li><a href="#refreshAttrs"><b>sequence refreshAttrs</b></a>
&ndash; for structure chains 
<a href="../tools/sequenceviewer.html#association">associated</a>
with more than one sequence alignment,
assign <a href="../attributes.html">attributes</a> 
consistent with a specific alignment
</li><li><a href="#refseq"><b>sequence refseq</b></a>
&ndash; set which sequence to use as the reference for numbering
</li><li><a href="#rename"><b>sequence rename</b></a>
&ndash; change the name of a sequence
</li><li><a href="#search"><b>sequence search</b></a>
&ndash; fast MMSeqs2 search for structures similar in sequence 
to a structure chain
</li><li><a href="#update"><b>sequence update</b></a>
&ndash; replace the sequence data of a structure chain with its
associated sequence in the 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>
</li></ul>
<p>
The <a href="../menu.html#tools"><b>Tools</b> menu</a> 
can also be used to show the sequence of a structure chain in the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>.
All of the other actions can be accessed from the
<a href="../tools/sequenceviewer.html#context"><b>Sequence Viewer</b>
context menu</a>.
</p>

<a name="chain"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Show Sequence from Structure</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;chain</b>
&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
[&nbsp;<b>viewer</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
</blockquote>
<p>
The command <b>sequence chain</b> shows the sequence of the specified 
biopolymer chain in the 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>,
although the graphical interface can be suppressed
(for example, to run a script that uses the sequence data but not its display)
with <b>viewer false</b>.
Only one structure chain should be specified per command.
See also: 
<a href="../tools/moldisplay.html"><b>Molecule Display</b></a> icon
<a href="../tools/moldisplay.html" title="Molecule Display Toolbar...">
<img class="iconbar" src="../tools/shortcut-icons/sequence.png" alt="computer generated image"></a>
</p><p>
Other ways to start the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>:
Independent of structure, sequence alignments and individual sequences
can also be <a href="open.html#local">opened</a>
from <a href="open.html#sequence">files</a>
or <a href="open.html#fetch">fetched</a>
from <a href="https://www.uniprot.org/" target="_blank">UniProt</a>.
Other tools or commands may generate new sequence alignments
(<i>e.g.</i>, <a href="../tools/blastprotein.html#results"><b>Blast Protein</b>
results</a>, <a href="../tools/matchmaker.html"><b>Matchmaker</b></a>,
<a href="#align">sequence realignment</a>).
</p>

<a name="associate"></a>
<a name="disassociate"></a>
<a name="dissociate"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Sequence-Structure Association</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;associate</b>
&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
<span class="nowrap">[&nbsp;<i>alignment-ID</i>:<i>sequence-ID</i>&nbsp;]</span>
<br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence</b> 
(&nbsp;<b>dissociate</b>&nbsp;|&nbsp;<b>disassociate</b>&nbsp;)
&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
&nbsp;<i>alignment-ID</i>
</blockquote>
<p>
<a href="../tools/sequenceviewer.html#association">Sequence-structure
association</a> 
(such as for synchronized <a href="../selection.html">selection</a>)
occurs automatically, but the commands 
<b>sequence associate</b> and <b>sequence dissociate</b> 
(same as <b>sequence disassociate</b>)
allow more precise control, for example, of which structure chains are used 
for <a href="../tools/sequenceviewer.html#headers">header</a> calculations, 
or forcing or removing associations regardless of 
whether the number of mismatches would be tolerated by the 
<a href="../tools/sequenceviewer.html#association">automatic procedure</a>.
</p><p>
The command <b>sequence associate</b> associates
one or more structure chains with a sequence.
The target sequence for association is specified by alignment ID,
as reported in the title bar of the 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window,
and the name or index number of the target sequence in the alignment,
in the form: 
&nbsp;<span class="nowrap"><i>alignment-ID</i>:<i>sequence-ID</i></span>&nbsp;
(<a href="seqspec.html">details...</a>).
</p><p>
Alternatively, the <i>sequence-ID</i> can be omitted 
to associate each specified structure chain with the
the best-matching sequence in the alignment.
The <i>alignment-ID</i> can be omitted if only one alignment is open, or
if the <i>sequence-ID</i> is also omitted; in the latter case,
each specified structure chain will be associated with the 
best-matching sequence in each open alignment.
If either or both are omitted, the colon (:) should also be omitted
except in rare cases to disambiguate an alignment and sequence 
that have the same name.
</p><p>
For <b>sequence dissociate</b>, only the alignment needs to be specified, 
not an individual sequence, because a structure chain can only be 
associated with one sequence per alignment.
</p>

<a name="match"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Structure Superposition</h3>
<blockquote>
<a href="usageconventions.html">Usage</a>:
<b>sequence &nbsp;match</b> [&nbsp;<i>alignment-ID</i>&nbsp;]
&nbsp;<a href="atomspec.html#hierarchy"><i>matchchain(s)</i></a>&nbsp; <b>to</b>
&nbsp;<a href="atomspec.html#hierarchy"><i>refchain</i></a>&nbsp;
[&nbsp;<b>cutoffDistance</b>&nbsp;&nbsp;<i>cutoff</i>&nbsp;|&nbsp;none&nbsp;]
[&nbsp;<b>conservation</b>&nbsp;&nbsp;<i>percent</i>&nbsp;]
[&nbsp;<b>columns</b>&nbsp;&nbsp;<i>list-of-positions</i>&nbsp;]
</blockquote>
<p>
Superimpose the specified structures based on the alignment of their 
<a href="../tools/sequenceviewer.html#association">associated</a> sequences.
The <i>alignment-ID</i> is shown in the title bar of the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window.
</p><p>
By default (<b>cutoffDistance</b> option not used), 
structures are fit iteratively with a <i>cutoff</i> of
<b>2.0</b> &Aring; for omitting farther-apart pairs from the fit.
Specifying the <i>cutoff</i> as <b>none</b> turns iteration off.
When iteration is performed,
the sequence alignment is not changed, but residue pairs in the
alignment can be pruned from the &ldquo;match list&rdquo; used to
superimpose the structures. 
Fitting uses <a href="../tools/sequenceviewer.html#onepoint">one atom</a>
per residue. In each cycle of fitting,
pairs of atoms are removed from the match list and the remaining
pairs are fitted until no matched pair is
more than <i>cutoff</i> apart (default <b>2.0</b> &Aring;).
The atom pairs removed are either the 10% farthest apart of all pairs
or the 50% farthest apart of all pairs exceeding the cutoff, whichever
is the lesser number of pairs. The result of iteration is that the
best-matching &ldquo;core&rdquo; regions are maximally superimposed,
and conformationally dissimilar regions such as flexible
loops are not included in the final fit, even though they may be aligned
in the sequence alignment.
</p><p>
The <b>conservation</b> option specifies omitting columns with less than
<i>percent</i> identity from the initial fit.
</p><p>
The <b>columns</b> option specifies omitting columns other than those
listed from the initial fit.
The list must consist of positive integers separated by commas only;
hyphenated ranges cannot be used.
</p><p>
The number of atom pairs fitted and the resulting RMSD are reported in the
<a href="../tools/log.html"><b>Log</b></a> and the status line at the bottom
of the <a href="../window.html">ChimeraX window</a>. See also:
<a href="matchmaker.html"><b>matchmaker</b></a>,
<a href="align.html"><b>align</b></a>
</p>

<a name="expandsel"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Expanding a Selection to Alignment Columns</h3>
<blockquote>
<a href="usageconventions.html">Usage</a>:
<b>sequence &nbsp;expandsel</b> [&nbsp;<i>alignment-ID</i>&nbsp;]
</blockquote>
<p>
Expand the current <a href="../selection.html">selection</a> to all residues
<a href="../tools/sequenceviewer.html#association">associated</a>
with the same column(s) of the specified sequence alignment, or if none
is specified, all sequence alignments.
The <i>alignment-ID</i> is shown in the title bar of the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window.
</p>

<a name="rename"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Rename a Sequence</h3>
<blockquote>
<a href="usageconventions.html">Usage</a>:
<b>sequence &nbsp;rename</b>
<span class="nowrap">[&nbsp;<i>alignment-ID</i>:<i>sequence-ID</i>&nbsp;]</span>
&nbsp;<i>new-name</i>
</blockquote>
<p>
Rename the specified sequence.
The sequence is specified by its alignment ID, shown in the title bar
of the <a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window,
plus the sequence's name or index number in the alignment, in the form:
&nbsp;<span class="nowrap"><i>alignment-ID</i>:<i>sequence-ID</i></span>&nbsp;
(<a href="seqspec.html">details...</a>).
The <i>new-name</i> cannot include a colon (:) and should be enclosed in 
quotation marks if it includes any spaces.
</p>

<a name="header"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Sequence Viewer Headers and Numbering</h3>
<blockquote>
<a href="usageconventions.html">Usage</a>:
<b>sequence &nbsp;header</b> [&nbsp;<i>alignment-ID</i>&nbsp;]
&nbsp;<i>header-name</i>
(&nbsp;<b>show</b>&nbsp;|&nbsp;<b>hide</b>&nbsp;|&nbsp;<b>save</b>&nbsp;&nbsp;<i>filename</i>&nbsp;)
</blockquote>
<p>
The command <b>sequence header</b> shows, hides, or saves a
<a href="../tools/sequenceviewer.html#headers">sequence header</a> to a file.
(It can also be used to change the sequence
<a href="../tools/sequenceviewer.html#settings-headers"><b>Headers</b>
preferences</a>, but command details are omitted here because normally
the <a href="../tools/sequenceviewer.html#settings"><b>Settings</b></a>
dialog will be used instead.)
</p><p>
The <i>header-name</i> can be <b>consensus</b>, <b>conservation</b>,
or <b>rmsd</b>, although there will only be an effect when that header
is available (for example, an RMSD header is only available for alignments
associated with at least two structures).
Headers are saved to a simple text format that lists the alignment positions 
and values. The <i>filename</i> can be given as a pathname or the word 
<a href="usageconventions.html#browse"><b>browse</b></a>
to bring up a file browser window for choosing the
name and location interactively. 
If multiple alignments are open but an <i>alignment-ID</i> is not specified, 
showing/hiding affects all applicable alignments. 
However, saving only works for a single header at a time, so an
<i>alignment-ID</i> must be given when more than one alignment is open.
</p>
<a name="refseq"></a>
<blockquote>
<a href="usageconventions.html">Usage</a>:
<b>sequence &nbsp;refseq</b>
<span class="nowrap">[&nbsp;<i>alignment-ID</i>:<i>sequence-ID</i>&nbsp;]</span>
</blockquote>
<p>
The overall numbering of an alignment is shown across the top at the
first position and every tenth position thereafter. By default, there
is no reference sequence, and the alignment is simply numbered 1, 11, 21, 
<i>etc.</i>, regardless of whether the columns contain gaps.
However, with <b>sequence refseq</b>, any of the individual sequences 
can be specified as the reference for numbering instead.
The sequence to use is specified by its alignment ID, shown in the title bar
of the <a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window,
plus the sequence's name or index number in the alignment, in the form:
&nbsp;<span class="nowrap"><i>alignment-ID</i>:<i>sequence-ID</i></span>&nbsp;
(<a href="seqspec.html">details...</a>).
</p><p>
When the reference sequence has a gap character at a position
where a number is to be shown (first column, eleventh column, <i>etc.</i>),
how that position falls relative to the reference is shown in parentheses.
For example, &ldquo;(&lt;1)&rdquo; would be shown for alignment positions
before the reference's first residue, and &ldquo;(43/44)&rdquo;
for positions that fall between residues 43 and 44 of the reference.
</p><p>
Omitting the sequence identifier returns to the default numbering:
</p>
<blockquote>
<a href="usageconventions.html">Usage</a>:
<b>sequence &nbsp;refseq</b>
<span class="nowrap">[&nbsp;<i>alignment-ID</i>&nbsp;]</span>
</blockquote>
<p>
If more than one 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window
is present, the <i>alignment-ID</i> is required.
The <i>alignment-ID</i> is shown in the title bar of the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window.
</p>

<a name="identity"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Calculate Percent Identities</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;identity</b> &nbsp;<i>alignment-ID</i>&nbsp;
[&nbsp;<b>denominator</b>&nbsp;&nbsp;<b>shorter</b>&nbsp;|&nbsp;longer&nbsp;|&nbsp;nongap&nbsp;]
<br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;identity</b> &nbsp;<i>alignment-ID</i>&nbsp;
<span class="nowrap">&nbsp;<i>alignment-ID</i>:<i>sequence-ID</i>&nbsp;</span>
[&nbsp;<b>denominator</b>&nbsp;&nbsp;<b>shorter</b>&nbsp;|&nbsp;longer&nbsp;|&nbsp;nongap&nbsp;]
<br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;identity</b> 
<span class="nowrap">&nbsp;<i>alignment-ID</i>:<i>sequence-ID</i>&nbsp;</span>
<span class="nowrap">&nbsp;<i>alignment-ID</i>:<i>sequence-ID</i>&nbsp;</span>
[&nbsp;<b>denominator</b>&nbsp;&nbsp;<b>shorter</b>&nbsp;|&nbsp;longer&nbsp;|&nbsp;nongap&nbsp;]
</blockquote>
<p>
The <b>sequence identity</b> command calculates the pairwise percent identity
between sequences of the same length (including gaps, as shown in
the <a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window).
The calculation is always pairwise, but can be performed for all-by-all pairs
within a single alignment, or all-by-one, or between two specific sequences.
An entire alignment is specified by its ID, shown in the title bar of the 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window,
and an individual sequence by the alignment ID plus the sequence's
name or index number in the alignment, in the form: 
&nbsp;<span class="nowrap"><i>alignment-ID</i>:<i>sequence-ID</i></span>&nbsp;
(<a href="seqspec.html">details...</a>).
</p><p>
Results are listed in the <a href="../tools/log.html"><b>Log</b></a>. 
For each pair, the number of columns with identical residues is given 
as a percentage of the specified <b>denominator</b>:
</p>
<ul>
<li><b>shorter</b> (default) 
&ndash; the number of residues in the shorter of the two sequences
</li><li><b>longer</b>
&ndash; the number of residues in the longer of the two sequences
</li><li><b>nongap</b>
&ndash; the number of columns where neither sequence has a gap
</li></ul> 

<a name="refreshAttrs"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Refresh Attributes</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;refreshAttrs</b> &nbsp;<i>alignment-ID</i>&nbsp;
</blockquote>
<p>
For structure chains
<a href="../tools/sequenceviewer.html#association">associated</a>
with more than one sequence alignment, make sure that the
current <a href="../tools/sequenceviewer.html#headers">attribute values</a>
are consistent with a specific alignment.
The <i>alignment-ID</i> is shown in the title bar of the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window.
</p>

<a name="search"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Fast MMSeqs2 Search for Similar Structures </h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;search</b> 
&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>&nbsp;
[&nbsp;<b>database</b>&nbsp;&nbsp;<b>pdb</b>&nbsp;|&nbsp;afdb&nbsp;]
[&nbsp;<b>showTable</b>&nbsp;&nbsp;<b>true</b>&nbsp;|&nbsp;false&nbsp;]
[&nbsp;<b>evalueCutoff</b>&nbsp;&nbsp;<i>max-evalue</i>&nbsp;]
[&nbsp;<b>identityCutoff</b>&nbsp;&nbsp;<i>min-pid</i>&nbsp;]
[&nbsp;<b>maxHits</b>&nbsp;&nbsp;<i>N</i>&nbsp;]
<span class="nowrap">[&nbsp;<b>trim</b>&nbsp;&nbsp;true&nbsp;|&nbsp;false&nbsp;|&nbsp;chains,sequence,ligands&nbsp;]</span>
[&nbsp;<b>alignmentCutoffDistance</b>&nbsp;&nbsp;<i>dist</i>&nbsp;]
[&nbsp;<b>saveDirectory</b>&nbsp;&nbsp;<i>results-folder</i>&nbsp;]
</blockquote>
<p>
The <b>sequence search</b> command performs a fast MMseqs2 search of the
<a href="https://www.rcsb.org" target="_blank">PDB</a> or
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>
with a protein structure chain as the query, using the
<a href="https://search.rcsb.org" target="_blank">RCSB web service</a>.
The <a href="../tools/foldseek.html"><b>Foldseek</b>
(<b>Similar Structures</b>)</a> tool provides a graphical interface
to running this command. See also:
<a href="similarstructures.html"><b>similarstructures</b></a>,
<a href="foldseek.html"><b>foldseek</b></a>,
<a href="blastprotein.html"><b>blastprotein</b></a>
</p>
MMseqs2 (<a href="https://github.com/soedinglab/MMseqs2" target="_blank">Many-against-Many sequence searching</a>) is described in:
<blockquote>
<a href="https://www.ncbi.nlm.nih.gov/pubmed/29035372" target="_blank">
MMseqs2 enables sensitive protein sequence searching for the analysis
of massive data sets.</a>
Steinegger M, S&ouml;ding J.
<i>Nat Biotechnol.</i> 2017 Nov;35(11):1026-1028.
<br><br>
<a href="https://www.ncbi.nlm.nih.gov/pubmed/30615063" target="_blank">
MMseqs2 desktop and local web server app for fast, interactive sequence
searches.</a>
Mirdita M, Steinegger M, S&ouml;ding J.
<i>Bioinformatics.</i> 2019 Aug 15;35(16):2856-2858.
</blockquote>
<p>
The <b>showTable</b> option (default <b>true</b>)
indicates whether to show the results in the
<a href="../tools/foldseek.html#results"><b>Similar Structures</b></a> tool,
which facilitates exploring large sets of protein structures
by efficiently showing them in 3D as backbone traces and in
2D as sequence alignment schematics or scatter plots based on conformation.
These analyses can also be performed with the
<a href="similarstructures.html"><b>similarstructures</b></a> command
by referring to a set of results by name.
The name assigned to the results (<b>mm1</b> or <b>mm2</b>)
is reported in the <a href="../tools/log.html"><b>Log</b></a>
when the search is run.
</p><p>
Choices of <b>database</b> to search:
</p><ul>
<li><b>pdb</b> (default)
&ndash; <a href="https://www.rcsb.org" target="_blank">PDB</a>
</li><li><b>afdb</b>
&ndash; <a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>
</li></ul>
<p>
The hits can be limited with the following options:
</p>
<ul>
<li><b>evalueCutoff</b>&nbsp;&nbsp;<i>max-evalue</i>
&ndash; maximum E-value of a hit to be retained (default <b>1e-3</b>)
</li><li><b>identityCutoff</b>&nbsp;&nbsp;<i>min-pid</i>
&ndash; minimum percent sequence identity with the query
for a hit to be retained (default <b>0</b>)
</li><li><b>maxHits</b>&nbsp;&nbsp;<i>N</i>
&ndash; maximum number of hits retained (default <b>1000</b>)
</li></ul>
<p>
The <b>trim</b> and <b>alignmentCutoffDistance</b> options do not act
immediately. Instead, they specify how to process the hit structures
if later opened from the
<a href="../tools/foldseek.html#results"><b>Similar Structures</b></a>
tool or with the command
<a href="../commands/similarstructures.html#open"><b>similarstructures
open</b></a>. The structures are fetched from the respective database
(<a href="https://www.rcsb.org/" target="_blank">Protein Data Bank</a>
or <a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>) and processed as follows:
</p><ul>
<li>The <b>trim</b> option indicates deleting all of the following
(if true), or none of them (if false), or a comma-separated list of:
  <ul>
  <li><b>chains</b> &ndash; for PDB entries, chains other than the hit chain
  </li><li><b>sequence</b> &ndash; N- and C-terminal segments of the hit chain
that were not included in the sequence alignment returned by the search method
  </li><li><b>ligands</b> &ndash; ligands, solvent, and ions &gt; 3 &Aring;
from the hit chain
  </li></ul>
The <b>trim</b> default is the current setting in the
<a href="../tools/foldseek.html#options"><b>Similar Structures</b> options</a>,
otherwise (if the tool is not open) <b>true</b>.
<br><br>
</li><li>The hit chain is superimposed onto the query chain by
least-squares fitting the &alpha;-carbons of the paired residues
and iteratively pruning far-apart pairs as described for the
<a href="align.html"><b>align</b></a> command.
The <b>alignmentCutoffDistance</b> <i>d</i> is the alignment pruning distance 
so that only &alpha;-carbon pairs within the specified distance
are used in the final fit (default as per the
<a href="../tools/foldseek.html#options"><b>Similar Structures</b> options</a>,
otherwise <b>2.0</b> &Aring;)
</li></ul>
<p>
Search results are saved in a
<a href="../formats/sms.html">similar structures file</a>
(suffix .sms, a JSON file format specific to ChimeraX)
with filename based on the query name and the database searched.
The file will be listed in the
<a href="../tools/filehistory.html"><b>File History</b></a> for easy access,
and simply opening it loads the set of results into the
<a href="../tools/foldseek.html#results"><b>Similar Structures</b></a>
interface.
The <b>saveDirectory</b> option allows specifying the save location,
either directly or as the word
<a href="usageconventions.html#browse"><b>browse</b></a>
to specify it interactively in a file browser window
(default location <b>~/Downloads/ChimeraX/MMseqs2/</b>).
</p>

<a name="update"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Update Structure Metadata with Associated Sequence</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;update</b> 
&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
[&nbsp;<b>alignment</b>&nbsp;&nbsp;<i>alignment-ID</i>&nbsp;]
</blockquote>
<p>
Standard PDB files have SEQRES records containing the sequence of
each biopolymer chain in the structure, and similarly, standard mmCIF files
include this information in specific tables. In many structures, some of
the residues in the sequence are missing from the atomic coordinates
because their positions could not be resolved from the experimental density
(due to disorder, <i>etc.</i>).
However, atomic coordinate files output from various modeling programs often
lack sequence information, so that the sequence is only inferred from 
the coordinates. The <b>sequence update</b> command is useful for
adding full sequence information to such structures.
For example, the full sequence could be opened from a FASTA file or by 
<a href="open.html#fetch">fetching</a> it from
<a href="https://www.uniprot.org" target="_blank">UniProt</a>,
associated with the chain, and then used to update the chain's sequence
information. This command only applies to the sequences associated with
the specified chains, limited to a specific <b>alignment</b> as needed.
The <i>alignment-ID</i> is shown in the title bar of the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window.
The sequence must cover all of the residues in the structure chain.
</p>

<a name="align"></a>
<p class="nav">
[<a href="#top">back to top: sequence</a>]
</p>
<h3>Align Sequences using Clustal Omega or MUSCLE</h3>
<blockquote>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;align</b> &nbsp;<i>alignment-ID</i>
[&nbsp;<b>replace</b>&nbsp;&nbsp;true&nbsp;|&nbsp;<b>false</b>&nbsp;]
[&nbsp;<b>program</b>&nbsp;&nbsp;<b>clustalOmega</b>&nbsp;|&nbsp;muscle&nbsp;]
<br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;align</b> 
&nbsp;<a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
[&nbsp;<b>program</b>&nbsp;&nbsp;<b>clustalOmega</b>&nbsp;|&nbsp;muscle&nbsp;]
<br>
<a href="usageconventions.html"><b>Usage</b></a>:
<b>sequence &nbsp;align</b> 
&nbsp;<i>sequence1</i>,<i>sequence2</i>[,<i>sequence3</i>...,<i>sequenceN</i>]
[&nbsp;<b>program</b>&nbsp;&nbsp;<b>clustalOmega</b>&nbsp;|&nbsp;muscle&nbsp;]
</blockquote>
<p>
The <b>sequence align</b> command calculates a new alignment of the
specified protein sequences using a web service hosted by the
<a href="https://www.rbvi.ucsf.edu/Overview/index.html" target="_blank">UCSF
RBVI</a>. The result is opened in a new
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window,
except that <b>replace true</b> (default <b>false</b>)
can be used to specify overwriting an 
existing alignment when all of its sequences are being realigned.
</p><p>
The sequences to align can be specified collectively by:
</p>
<ul>
<li>alignment ID, as shown in the title bar of an existing
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window;
</li><li>a <a href="atomspec.html#hierarchy"><i>chain-spec</i></a>
(for atomic-structure protein chains already open in ChimeraX)
</li></ul>
...or individually as a comma-separated list (without spaces) 
of any combination of:
<ul>
<li>plain text of the entire amino acid sequence pasted directly 
into the command line
</li><li><a href="https://www.uniprot.org" target="_blank">UniProt</a>
name or accession number, for example:
<blockquote>
<b>sequence align ldlr_rat,ldlr_mouse,ldlr_human</b>
</blockquote>
</li><li>the <a href="seqspec.html"><i>sequence-spec</i></a> of a sequence
in the <a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>,
in the form: 
&nbsp;<span class="nowrap"><i>alignment-ID</i>:<i>sequence-ID</i></span>&nbsp;
(<a href="seqspec.html">details...</a>). Example:
<blockquote>
<b>open myfile.msf</b>
<br>
<b>sequence align 1,2,3,-1</b>
<br><i>&ndash; OR (if multiple sequence windows are open) &ndash;</i><br>
<span class="nowrap"><b>sequence align myfile.msf:1,myfile.msf:2,myfile.msf:3,myfile.msf:-1</b></span>
</blockquote>
</li></ul>
<p>
The <b>program</b> can be either of two choices:
</p>
<ul>
<li><a name="clustal"></a> 
<b>clustalOmega</b> (default, same as <b>clustal</b> or <b>omega</b>) 
&ndash; <a href="http://www.clustal.org/omega/" target="_blank">Clustal Omega</a> v1.1.0 with parameters: 
<p></p>
  <ul>
  <li>Number of guide-tree/HMM iterations: <b>1</b>
  </li><li>Full distance matrix during initial alignment: <b>true</b>
  </li><li>Full distance matrix during alignment iteration: <b>true</b>
  </li></ul>
<p>
See the <a href="http://www.clustal.org/omega/README" target="_blank">README file</a> at the Clustal Omega website for details.
Users should cite:
</p>
<blockquote>
<a href="https://www.ncbi.nlm.nih.gov/pubmed/21988835" target="_blank">
Fast, scalable generation of high-quality protein multiple sequence alignments
using Clustal Omega.</a>
Sievers F, Wilm A, Dineen D, Gibson TJ, Karplus K, Li W, Lopez R, McWilliam H,
Remmert M, S&ouml;ding J, Thompson JD, Higgins DG.
<i>Mol Syst Biol.</i> 2011 Oct 11;7:539.
</blockquote>
</li><li><a name="muscle"></a> <b>muscle</b> 
&ndash; the <a href="https://drive5.com/muscle5/" target="_blank">MUSCLE5</a>
command <a href="https://drive5.com/muscle5/manual/cmd_align.html">align</a> 
with its default parameters
<p>
Users should cite:
</p>
<blockquote>
<a href="https://www.ncbi.nlm.nih.gov/pubmed/36379955" target="_blank">
Muscle5: High-accuracy alignment ensembles enable unbiased assessments of sequence homology and phylogeny.</a> Edgar RC.
<i>Nat Commun.</i> 2022 Nov 15;13(1):6968.
</blockquote>
</li></ul>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics / 
March 2025</address>

</body></html>