*** OpenMM_CUDA_12-8.2.0.dist-info/METADATA.orig	Tue Jul 15 09:31:42 2025
--- OpenMM_CUDA_12-8.2.0.dist-info/METADATA	Tue Jul 15 09:31:52 2025
*************** Version: 8.2.0
*** 4,12 ****
  Summary: CUDA platform for OpenMM
  Home-page: https://openmm.org
  Author: Peter Eastman
- Requires-Dist: nvidia-cuda-runtime-cu12
- Requires-Dist: nvidia-cuda-nvcc-cu12
- Requires-Dist: nvidia-cuda-nvrtc-cu12
- Requires-Dist: nvidia-cuda-cupti-cu12
- Requires-Dist: nvidia-cufft-cu12
  
--- 4,7 ----
