*** OpenMM_CUDA_12-8.2.0.dist-info/METADATA.orig	2025-07-10 13:41:23
--- OpenMM_CUDA_12-8.2.0.dist-info/METADATA	2025-07-10 13:41:28
*************** Author: Peter Eastman
*** 4,12 ****
  Summary: CUDA platform for OpenMM
  Home-page: https://openmm.org
  Author: Peter Eastman
- Requires-Dist: nvidia-cuda-runtime-cu12
- Requires-Dist: nvidia-cuda-nvcc-cu12
- Requires-Dist: nvidia-cuda-nvrtc-cu12
- Requires-Dist: nvidia-cuda-cupti-cu12
- Requires-Dist: nvidia-cufft-cu12
- 
--- 4,6 ----
