*** openvr/__init__.py.orig	Fri Oct 26 10:48:58 2018
--- openvr/__init__.py	Fri Oct 26 10:58:38 2018
***************
*** 1947,1962 ****
          ("seatedZeroPoseReset", VREvent_SeatedZeroPoseReset_t),
      ]
  
! 
! class VREvent_t(Structure):
!     "An event posted by the server to all running applications"
! 
!     _fields_ = [
!         ("eventType", c_uint32),
!         ("trackedDeviceIndex", TrackedDeviceIndex_t),
!         ("eventAgeSeconds", c_float),
!         ("data", VREvent_Data_t),
!     ]
  
  
  class VROverlayIntersectionMaskPrimitive_Data_t(Union):
--- 1947,1973 ----
          ("seatedZeroPoseReset", VREvent_SeatedZeroPoseReset_t),
      ]
  
! # TODO: Fix the generator of this file (translate/translate.pl) as
! # explained in https://github.com/cmbruns/pyopenvr/pull/42#issuecomment-346866490
! if sizeof(c_void_p) != 4 and platform.system() == 'Linux':
!     class VREvent_t(Structure):
!         "An event posted by the server to all running applications"
!         _pack_ = 4
!         _fields_ = [
!             ("eventType", c_uint32),
!             ("trackedDeviceIndex", TrackedDeviceIndex_t),
!             ("eventAgeSeconds", c_float),
!             ("data", VREvent_Data_t),
!         ]
! else:
!     class VREvent_t(Structure):
!         "An event posted by the server to all running applications"
!         _fields_ = [
!             ("eventType", c_uint32),
!             ("trackedDeviceIndex", TrackedDeviceIndex_t),
!             ("eventAgeSeconds", c_float),
!             ("data", VREvent_Data_t),
!         ]
  
  
  class VROverlayIntersectionMaskPrimitive_Data_t(Union):
