*** OpenMM_CUDA_12-8.4.0.dist-info/METADATA.orig	Thu Jan  8 14:22:36 2026
--- OpenMM_CUDA_12-8.4.0.dist-info/METADATA	Thu Jan  8 15:49:01 2026
***************
*** 4,14 ****
  Summary: CUDA platform for OpenMM
  Home-page: https://openmm.org
  Author: Peter Eastman
- Requires-Dist: nvidia-cuda-runtime-cu12
- Requires-Dist: nvidia-cuda-nvcc-cu12
- Requires-Dist: nvidia-cuda-nvrtc-cu12
- Requires-Dist: nvidia-cuda-cupti-cu12
- Requires-Dist: nvidia-cufft-cu12
  Dynamic: author
  Dynamic: home-page
  Dynamic: requires-dist
--- 4,9 ----
