*** PIL/TiffImagePlugin.py.orig	2019-03-11 02:36:51.000000000 -0700
--- PIL/TiffImagePlugin.py	2019-03-12 14:05:56.000000000 -0700
***************
*** 1042,1054 ****
              # reset python3 buffered io handle in case fp
              # was passed to libtiff, invalidating the buffer
              self.fp.tell()
!             self.fp.seek(self.__next)
              self._frame_pos.append(self.__next)
-             if DEBUG:
-                 print("Loading tags, location: %s" % self.fp.tell())
-             self.tag_v2.load(self.fp)
-             self.__next = self.tag_v2.next
-             self.__frame += 1
          self.fp.seek(self._frame_pos[frame])
          self.tag_v2.load(self.fp)
          self.__next = self.tag_v2.next
--- 1042,1059 ----
              # reset python3 buffered io handle in case fp
              # was passed to libtiff, invalidating the buffer
              self.fp.tell()
!             last_frame = len(self._frame_pos) - 1
!             if self.__frame != last_frame:
!                 # Current frame is not the last frame in self._frame_pos list.
!                 # Read last frame in self._frame_pos list to get next frame position
!                 # to add to list.
!                 self.fp.seek(self._frame_pos[last_frame])
!                 if DEBUG:
!                     print("Loading tags, location: %s" % self.fp.tell())
!                 self.tag_v2.load(self.fp)
!                 self.__frame = last_frame
!                 self.__next = self.tag_v2.next
              self._frame_pos.append(self.__next)
          self.fp.seek(self._frame_pos[frame])
          self.tag_v2.load(self.fp)
          self.__next = self.tag_v2.next
