diff --git a/prereqs/PyQt/Makefile b/prereqs/PyQt/Makefile
index 183b38a01..7b8981d50 100644
--- a/prereqs/PyQt/Makefile
+++ b/prereqs/PyQt/Makefile
@@ -12,11 +12,12 @@ endif
 # PYQT_LICENSE can be "commercial" for commercial version
 PYQT_LICENSE = gpl
 
-VERSION = 5.15
+VERSION = 6.2
 PYQT_MINOR_VERSION = 2
-QT_MINOR_VERSION = 2
+QT_MINOR_VERSION = 3
+PYQTWE_MINOR_VERSION = 1
 
-MAKE_ARGS = -f Makefile.$(METHOD) VERSION=$(VERSION) PYQT_MINOR_VERSION=$(PYQT_MINOR_VERSION) QT_MINOR_VERSION=$(QT_MINOR_VERSION) PYQT_LICENSE=$(PYQT_LICENSE)
+MAKE_ARGS = -f Makefile.$(METHOD) VERSION=$(VERSION) PYQT_MINOR_VERSION=$(PYQT_MINOR_VERSION) QT_MINOR_VERSION=$(QT_MINOR_VERSION) PYQTWE_MINOR_VERSION=$(PYQTWE_MINOR_VERSION) PYQT_LICENSE=$(PYQT_LICENSE)
 
 all:
 
diff --git a/prereqs/PyQt/Makefile.fromsrc b/prereqs/PyQt/Makefile.fromsrc
index b6dcb96cf..92919d061 100644
--- a/prereqs/PyQt/Makefile.fromsrc
+++ b/prereqs/PyQt/Makefile.fromsrc
@@ -9,12 +9,12 @@ SUBDIRS = qt sip
 include $(TOP)/mk/subdir.make
 
 PYQT_LICENSE = commercial
-REMOTE_DIR = plato.cgl.ucsf.edu:/usr/local/src/PyQt5
+REMOTE_DIR = plato.cgl.ucsf.edu:/usr/local/src/PyQt6
 
 # VERSION is overriden by Makefile
-VERSION = 5.13
-QT_MINOR_VERSION = 2
-PYQT_MINOR_VERSION = 1
+VERSION = 6.2
+QT_MINOR_VERSION = 3
+PYQT_MINOR_VERSION = 2
 PYQTWE_MINOR_VERSION = 1
 QT_VERSION = $(VERSION).$(QT_MINOR_VERSION)
 ifeq (,$(PYQT_MINOR_VERSION))
@@ -28,8 +28,8 @@ else
 PYQTWE_VERSION = $(VERSION).$(PYQTWE_MINOR_VERSION)
 endif
 
-PYQT_DISTRIBUTION = PyQt5_$(PYQT_LICENSE)-$(PYQT_VERSION).tar.gz
-PYQT_SOURCE = $(tmpdir)/PyQt5_$(PYQT_LICENSE)-$(PYQT_VERSION)
+PYQT_DISTRIBUTION = PyQt6_$(PYQT_LICENSE)-$(PYQT_VERSION).tar.gz
+PYQT_SOURCE = $(tmpdir)/PyQt6_$(PYQT_LICENSE)-$(PYQT_VERSION)
 PYQTWE_DISTRIBUTION = PyQtWebEngine_$(PYQT_LICENSE)-$(PYQTWE_VERSION).tar.gz
 PYQTWE_SOURCE = $(tmpdir)/PyQtWebEngine_$(PYQT_LICENSE)-$(PYQTWE_VERSION)
 
@@ -56,7 +56,7 @@ install: $(PYQT_SOURCE) $(PYQTWE_SOURCE)
 	cd $(PYQTWE_SOURCE) && $(PYTHON_EXE) configure.py $(CONFIG_ARGS) && $(MAKE) && $(MAKE) install
 
 app-install:
-	$(RSYNC) $(PYSITEDIR)/PyQt5 $(APP_PYSITEDIR)
+	$(RSYNC) $(PYSITEDIR)/PyQt6 $(APP_PYSITEDIR)
 	# TODO: web engine
 
 distribution: $(PYQT_DISTRIBUTION) $(PYQTWE_DISTRIBUTION)
diff --git a/prereqs/PyQt/Makefile.wheel b/prereqs/PyQt/Makefile.wheel
index 144bedbba..a72820c63 100644
--- a/prereqs/PyQt/Makefile.wheel
+++ b/prereqs/PyQt/Makefile.wheel
@@ -5,6 +5,8 @@ include ../pips/Makefile.pip
 
 ifeq ($(OS),Windows)
 export PATH := $(bindir)/Scripts:$(PATH)
+else ifeq ($(OS),Darwin)
+export PATH := $(PYTHON_FRAMEWORK)/bin:$(PATH)
 else
 export PATH := $(bindir):$(PATH)
 endif
@@ -12,13 +14,13 @@ endif
 # PYQT_LICENSE can be "commercial" for commercial version
 PYQT_LICENSE = gpl
 REMOTE_HOST = plato.cgl.ucsf.edu:
-REMOTE_DIR = $(REMOTE_HOST)/usr/local/src/PyQt5
+REMOTE_DIR = $(REMOTE_HOST)/usr/local/src/PyQt6
 
 # VERSION is overriden by Makefile
-VERSION = 5.15
-QT_MINOR_VERSION = 2
+VERSION = 6.2
+QT_MINOR_VERSION = 3
 PYQT_MINOR_VERSION = 2
-PYQTWE_MINOR_VERSION = 2
+PYQTWE_MINOR_VERSION = 1
 QT_VERSION = $(VERSION).$(QT_MINOR_VERSION)
 ifeq (,$(PYQT_MINOR_VERSION))
 PYQT_VERSION = $(VERSION)
@@ -33,47 +35,45 @@ endif
 
 # Versions of Riverbank Computing Limited packages/programs from when
 # unlicensed wheels where built and presumably matches the gpl wheel in PyPI.
-SIP_VERSION = 12.8.1
-BUILDER_VERSION = 1.5.0
+SIP_VERSION = 13.2.1
+BUILDER_VERSION = 1.12.2
 
-QT_WHEEL = PyQt5_$(PYQT_LICENSE)-$(PYQT_VERSION)-$(BUILD_TAGS).whl
-WB_WHEEL = PyQtWebEngine_$(PYQT_LICENSE)-$(PYQTWE_VERSION)-$(WE_BUILD_TAGS).whl
-
-PYTHON_TAGS = cp35.cp36.cp37.cp38.cp39
+PYTHON_TAGS = cp36
+WHEEL_ABI = abi3
 
 ifeq ($(OS),Darwin)
 BIN_PLATFORM = macos
-WHEEL_ABI = abi3
-WHEEL_PLATFORM = macosx_10_13_intel
-WE_WHEEL_PLATFORM = macosx_10_13_intel
-QTDIR = $(HOME)/Qt/$(QT_VERSION)/clang_64
+WHEEL_PLATFORM = macosx_10_14_universal2
+WE_WHEEL_PLATFORM = $(WHEEL_PLATFORM)
+QTDIR = $(HOME)/Qt/$(QT_VERSION)/macos
 else ifeq ($(OS),Linux)
 BIN_PLATFORM = linux
-WHEEL_ABI = abi3
-WHEEL_PLATFORM = manylinux2014_x86_64
+WHEEL_PLATFORM = manylinux1_x86_64
 WE_WHEEL_PLATFORM = $(WHEEL_PLATFORM)
 #QTDIR = $(HOME)/Qt$(QT_VERSION)/$(QT_VERSION)/gcc_64
 QTDIR = $(HOME)/Qt/$(QT_VERSION)/gcc_64
 else ifeq ($(OS),Windows)
 BIN_PLATFORM = win.exe
-WHEEL_ABI = none
 WHEEL_PLATFORM = win_amd64
 WE_WHEEL_PLATFORM = $(WHEEL_PLATFORM)
 QTDIR = c:\\Qt\\$(QT_VERSION)\\msvc2019_64
 endif
-BUILD_TAGS = $(QT_VERSION)-$(PYTHON_TAGS)-$(WHEEL_ABI)-$(WHEEL_PLATFORM)
-WE_BUILD_TAGS = $(QT_VERSION)-$(PYTHON_TAGS)-$(WHEEL_ABI)-$(WE_WHEEL_PLATFORM)
+BUILD_TAGS = $(PYTHON_TAGS)-$(WHEEL_ABI)-$(WHEEL_PLATFORM)
+WE_BUILD_TAGS = $(PYTHON_TAGS)-$(WHEEL_ABI)-$(WE_WHEEL_PLATFORM)
+
+QT_WHEEL = PyQt6_$(PYQT_LICENSE)-$(PYQT_VERSION)-$(QT_VERSION)-$(BUILD_TAGS).whl
+WB_WHEEL = PyQt6_WebEngine_$(PYQT_LICENSE)-$(PYQTWE_VERSION)-$(QT_VERSION)-$(WE_BUILD_TAGS).whl
 
 BUNDLE_ARGS = --qt-dir $(QTDIR) --ignore-missing --no-openssl
 ifdef WIN32
 BUNDLE_ARGS += --no-msvc-runtime
 endif
 
-SIP_WHEEL := PyQt5-sip==$(SIP_VERSION)
+SIP_WHEEL := PyQt6-sip==$(SIP_VERSION)
 
 ifneq ($(PYQT_LICENSE),commercial)
-QT_WHEEL := PyQt5==$(PYQT_VERSION)
-WB_WHEEL := PyQtWebEngine==$(PYQT_VERSION)
+QT_WHEEL := PyQt6==$(PYQT_VERSION)
+WB_WHEEL := PyQt6_WebEngine==$(PYQTWE_VERSION)
 endif
 
 
@@ -94,12 +94,12 @@ endif
 
 $(WB_WHEEL):
 ifeq ($(PYQT_LICENSE),commercial)
-	rsync -av $(REMOTE_DIR)/$(PYQT_VERSION)/$(WB_WHEEL) .
+	rsync -av $(REMOTE_DIR)/$(PYQTWE_VERSION)/$(WB_WHEEL) .
 endif
 
 REMOTE_SOURCE =  \
-		$(REMOTE_DIR)/$(PYQT_VERSION)/PyQt5_commercial-$(PYQT_VERSION)-cp\*-$(WHEEL_PLATFORM).whl \
-		$(REMOTE_DIR)/$(PYQT_VERSION)/PyQtWebEngine_commercial-$(PYQTWE_VERSION)-cp\*-$(WE_WHEEL_PLATFORM).whl \
+		$(REMOTE_DIR)/$(PYQT_VERSION)/PyQt6_commercial-$(PYQT_VERSION)-$(BUILD_TAGS).whl \
+		$(REMOTE_DIR)/$(PYQTWE_VERSION)/PyQt6_WebEngine_commercial-$(PYQTWE_VERSION)-$(WE_BUILD_TAGS).whl \
 		$(REMOTE_DIR)/pyqt-commercial.sip
 
 pyqt-commercial.sip:
@@ -127,14 +127,21 @@ ifeq ($(OS),Linux)
 	cd $(QTDIR)/qml && find . -name \*.so | xargs $(bindir)/chrpath -C
 	$(bindir)/chrpath -C $(QTDIR)/libexec/QtWebEngineProcess
 endif
-	pyqt-bundle $(BUNDLE_ARGS) $(BUNDLE_ARGS) PyQt5_commercial-$(PYQT_VERSION)-$(PYTHON_TAGS)-*.whl
-	pyqt-bundle $(BUNDLE_ARGS) $(BUNDLE_ARGS) PyQtWebEngine_commercial-$(PYQTWE_VERSION)-$(PYTHON_TAGS)-*.whl
+	pyqt-bundle $(BUNDLE_ARGS) PyQt6_commercial-$(PYQT_VERSION)-$(BUILD_TAGS).whl
+	pyqt-bundle $(BUNDLE_ARGS) PyQt6_WebEngine_commercial-$(PYQTWE_VERSION)-$(WE_BUILD_TAGS).whl
+
+upload-commercial: wheels
+	rsync -av $(QT_WHEEL) $(REMOTE_DIR)/$(PYQT_VERSION)/.
+	rsync -av $(WB_WHEEL) $(REMOTE_DIR)/$(PYQTWE_VERSION)/.
 
-upload: wheels
-	rsync -av $(QT_WHEEL) $(WB_WHEEL) $(REMOTE_DIR)/$(PYQT_VERSION)/.
+upload:
+	$(MAKE) -f Makefile.wheel PYQT_LICENSE=commercial upload-commercial
 
 clean:
-	rm -rf $(QT_WHEEL) $(WB_WHEEL) pip.log
+	rm -f pip.log
+ifeq ($(PYQT_LICENSE),commercial)
+	rm -f $(QT_WHEEL) $(WB_WHEEL)
+endif
 
 distclean: clean
-	rm -f *commercial* pyqtlicense-*
+	rm -rf *commercial*
