--- setup.py.save	2014-03-15 08:27:06.000000000 -0700
+++ setup.py	2014-05-08 12:49:03.317574451 -0700
@@ -191,7 +191,7 @@
             _add_directory(library_dirs, os.path.join(
                 "/usr/lib", "python%s" % sys.version[:3], "config"))
 
-        elif sys.platform == "darwin":
+        elif sys.platform == "darwin_SKIP":
             # attempt to make sure we pick freetype2 over other versions
             _add_directory(include_dirs, "/sw/include/freetype2")
             _add_directory(include_dirs, "/sw/lib/freetype2/include")
@@ -446,7 +446,7 @@
             exts.append(Extension(
                 "PIL._webp", ["_webp.c"], libraries=libs, define_macros=defs))
 
-        if sys.platform == "darwin":
+        if sys.platform == "darwin_SKIP":
             # locate Tcl/Tk frameworks
             frameworks = []
             framework_roots = [
