--- Tk/tkImaging.c.orig	2014-08-13 09:32:20.000000000 -0700
+++ Tk/tkImaging.c	2014-10-24 19:43:51.478254024 -0700
@@ -57,14 +57,16 @@
 static Imaging
 ImagingFind(const char* name)
 {
-    Py_ssize_t id;
+    Imaging i;
 
-    /* FIXME: use CObject instead? */
-    id = atol(name);
-    if (!id)
-	return NULL;
-
-    return (Imaging) id;
+    PyObject *o = PyLong_FromString(name, NULL, 10);
+    if (o == NULL) {
+	    PyErr_Clear();
+	    return NULL;
+    }
+    i = (Imaging) PyLong_AsVoidPtr(o);
+    Py_DECREF(o);
+    return i;
 }
 
 
