diff -ur client/cxservices/api/default_api.py client.new/cxservices/api/default_api.py
--- client/cxservices/api/default_api.py	2020-06-28 19:09:35.000000000 -0700
+++ client.new/cxservices/api/default_api.py	2020-06-28 19:01:07.000000000 -0700
@@ -129,9 +129,9 @@
             auth_settings=auth_settings,
             async_req=params.get('async_req'),
             _return_http_data_only=params.get('_return_http_data_only'),
-            _preload_content=params.get('_preload_content', True),
+            _preload_content=params.get('_preload_content', False),
             _request_timeout=params.get('_request_timeout'),
-            collection_formats=collection_formats)
+            collection_formats=collection_formats).data
 
     def file_post(self, body, job_id, file_name, **kwargs):  # noqa: E501
         """Upload job file to server  # noqa: E501
@@ -335,9 +335,9 @@
             auth_settings=auth_settings,
             async_req=params.get('async_req'),
             _return_http_data_only=params.get('_return_http_data_only'),
-            _preload_content=params.get('_preload_content', True),
+            _preload_content=params.get('_preload_content', False),
             _request_timeout=params.get('_request_timeout'),
-            collection_formats=collection_formats)
+            collection_formats=collection_formats).data
 
     def files_list(self, job_id, **kwargs):  # noqa: E501
         """Return list of job files on server  # noqa: E501
diff -ur client/cxservices/rest.py client.new/cxservices/rest.py
--- client/cxservices/rest.py	2020-06-28 19:09:36.000000000 -0700
+++ client.new/cxservices/rest.py	2020-06-28 19:01:07.000000000 -0700
@@ -187,7 +187,7 @@
                 # Pass a `string` parameter directly in the body to support
                 # other content types than Json when `body` argument is
                 # provided in serialized form
-                elif isinstance(body, str):
+                elif isinstance(body, str) or isinstance(body, bytes):
                     request_body = body
                     r = self.pool_manager.request(
                         method, url,
diff -ur client/setup.py client.new/setup.py
--- client/setup.py	2020-06-28 19:09:36.000000000 -0700
+++ client.new/setup.py	2020-06-28 19:18:52.000000000 -0700
@@ -31,7 +31,7 @@
     url="",
     keywords=["Swagger", "RBVI ChimeraX Web Services"],
     install_requires=REQUIRES,
-    packages=find_packages(),
+    packages=find_packages(exclude=["test"]),
     include_package_data=True,
     long_description="""\
     REST API for RBVI web services supporting ChimeraX tools  # noqa: E501
