<html>
  <title>ChimeraX Boltz</title>
  <style>
    li { margin: 5px 0; }
  </style>
  
  <body>
    <h1>Making Boltz Structure Predictions in ChimeraX</h1>
    <p>
      Tom Goddard<br>
      April 23, 2025
    </p>

    <p>
      ChimeraX can run the <a href="https://github.com/jwohlwend/boltz/">Boltz-1</a> structure prediction method
      to compute atomic structures of proteins and nucleic acids, including modified residues, ligands, ions and
      solvent on your laptop or desktop computer.
      See the <a href="https://youtu.be/377V9A_0ECc">demonstration video</a>.
      The method is inspired by AlphaFold 3 and described in the following
      <a href="https://www.biorxiv.org/content/10.1101/2024.11.19.624167v2">preprint</a>
      and is fully open source with a very permissive MIT license.
    </p>

    <pre>
      Boltz-1: Democratizing Biomolecular Interaction Modeling
      Jeremy Wohlwend, Gabriele Corso, Saro Passaro, Mateo Reveiz, Ken Leidal,
        Wojtek Swiderski, Tally Portnoi, Itamar Chinn, Jacob Silterra, Tommi Jaakkola, Regina Barzilay
      bioRxiv 2024.11.19.624167; doi: https://doi.org/10.1101/2024.11.19.624167</pre>

    <p>
      A ChimeraX graphical user interface (menu Tools / Structure Prediction / Boltz) and ChimeraX command (boltz)
      are provide to make predictions.
      The <a href="https://www.rbvi.ucsf.edu/chimerax/data/boltz-apr2025/boltz_help.html">latest version of this documentation</a> is available online.
    </p>
    
    <h2>Boltz installation</h2>

    <img src="images/boltz_gui.png" width=400 align=right>

    <p>
      When you first start the Boltz tool within ChimeraX (menu Tools / Structure Prediction / Boltz) it will
      show a button <b>Install Boltz</b>.  Boltz is a large software package, taking about 1 Gbyte of disk space
      and uses the Torch machine learning package.  It also requires the neural network weights (3.3 Gbytes)
      and PDB Chemical Component Dictionary (CCD, 0.3 Gbytes) to make predictions.  Downloading and installing all of these
      files will take 10 minutes or more depending on network speed and install the software in your home
      directory in ~/boltz and the network weights and CCD in ~/.boltz.
    </p>

    <p>
      The ChimeraX <b>boltz install</b> command can also be used to do this one time installation.
    </p>
    
    <h2>Predicting a structure with Boltz</h2>

    <table align=right>
      <tr><td><img src="images/boltz_assembly.png" width=400>
	<td width=100 align=center>
	  Menu of molecular components<br>
	  <img src="images/boltz_menu.png" width=150>
    </table>

    <p>
      To predict a structure made up of proteins, nucleic acids and small molecules you first specify
      all the molecular components.   Choose entries from the <b>Add</b> menu and press the <b>Add</b> button
      to add them to your assembly specification in the table below.  You can specify component molecules
      in several ways.
    </p>

    <ul>
      <li>Protein and nucleic acids can be specified by choosing chains open models.
      <li>Sequences of one letter codes for proteins (20 amino acids) or DNA or RNA can be pasted in.
      <li>UniProt database identifiers can be give for proteins.
      <li>Ligands, ions and solvent can be specified by 3-letter or 5-letter chemical component dictinary codes (e.g. ATP or HEM).
      <li>Small molecules can be specified by SMILES strings.
    </ul>

    <p>
      Components can be added multiple times to have more instances of that molecule in the assembly.
      Press the <b>Predict</b> button after the assembly is completed by adding each component to start the prediction.
    </p>

    <table align=right>
      <tr><td width=250 align=center><img src="images/hemoglobin.png" width=250>
	  <br>Hemoglobin dimer prediction (blue) aligned to experimental structure PDB 1hho
	  <tr><td align=center><img src="images/hemoglobin_pae.png" width=200>
	      <br>Predicted aligned error
    </table>
    

    <h2>Results</h2>

    <p>
      The results are put in a new folder on your desktop named by default <b>boltz_</b><i>assembly_name</i>
      where the prediction name can be specified at the top of the ChimeraX Boltz panel.  Using the <b>Options</b>
      described below you can change where the result files are placed.
    </p>
    
    <p>
      Predictions for small assemblies, for example 500 residues and ligand atoms, take one to several minutes
      depending on the computer (e.g. Nvidia GPU vs CPU only).  The run in the background (a separate process)
      so ChimeraX can continue to be used while the calculation runs.  The predicted structure will be opened
      in ChimeraX when the calculation completes.  If the assembly specification involved proteins or nucleic
      acids specified using chains of open models, the predicted structure will be aligned (using matchmaker)
      to the open model for the first such component.
    </p>
    
    <h2>Coloring</h2>
    <p>
      The prediction will be colored using the standard AlphaFold pLDDT type of coloring where blue indicates
      high confidence, yellow and red moderate to low confidence.
    </p>

    <h2>Predicted aligned error</h2>
    <p>
      Finer grained estimates of prediction confidence can be displayed by pressing the <b>Error Plot</b> button
      to show the <a href="https://www.rbvi.ucsf.edu/chimerax/data/pae-apr2022/pae.html">predicted aligned error</a>.
    </p>

    <a name="runtimes">
    <h2>Run times and size limits</h2>
    </a>
    
    <p>
      Here are run times for several desktop and laptop computers for various size molecular assemblies from the Protein Databank predicting a single structure.
    </p>

    <ul>
      <li>MacBook Pro - M1 Max, 32 core GPU, 32 GB
      <li>Mac Studio - M2 Ultra, 60 core GPU, 64 GB, descartes.cgl.ucsf.edu
      <li>Linux - Nvidia RTX 4090 24 GB, cpu i9-13900K (24 cores) 64 GB (DDR5 5200MHz), minsky.cgl.ucsf.edu
      <li>Windows 11 - Nvidia RTX 3070 8 GB PCIe 4.0, cpu i7-12700K (12 cores) 64 GB DDR5 4000MHz, vizvault.cgl.ucsf.edu
      <li>Windows 10 - Nvidia RTX 4070 12 GB PCIe 3.0, cpu i5-6700K (4 cores) 64 GB DDR4 2133MHz, vive.cgl.ucsf.edu
    </ul>

    <p>Prediction times in minutes</p>
    <table border=1 cellpadding=4 cellspacing=0 bgcolor="#f0f0f0">
      <tr><th>PDB code<th>Tokens<th>MacBook Pro<th>Mac Studio<th>Linux<br>Nvidia 4090<th>Windows<br>i7 CPU<th>Windows<br>Nvidia 3070<th>Windows<br>Nvidia 4070<th>residues and atoms and prediction error
      <tr><td align=center>8rf4<td align=center>129<td align=center>0.9<td align=center>0.8<td align=center>0.5<td align=center>2.2<td align=center>1<td align=center>1.6<td>118 amino acids, 11 ligand atoms, 0.8A RMSD 108 residues
      <tr><td align=center>1hho<td align=center>382<td align=center>2.1<td align=center>1.4<td align=center>0.8<td align=center>12<td align=center>5<td align=center>4<td>287 amino acids, 95 ligand atoms, 0.4A RMSD 146 residues
      <tr><td align=center>9moj<td align=center>660<td align=center>3.9<td align=center>2.5<td align=center>1.2<td align=center>29<td align=center>7<td align=center>3.2<td>660 amino acids, heterotetramer, 0.8A RMSD 132 residues
      <tr><td align=center>9h1k<td align=center>671<td align=center>4.6<td align=center>2.7<td align=center>1.5<td align=center>32<td align=center>17<td align=center>27<td>560 amino acids, 59 rna bases, 52 ligand atoms, 0.7A RMSD protein, RNA wrong
      <tr><td align=center>9b3h<td align=center>911<td align=center>17<td align=center>4.7<td align=center>2.7<td align=center>96<td align=center>49<td align=center>75<td>911 amino acids, heterodimer, 1A RMSD 509 residues
      <tr><td align=center>9fz5<td align=center>1025<td align=center>15<td align=center>6.6<td align=center>cpu 78<br>gpu out of memory<td align=center><td align=center><td align=center><td>1025 amino acids, heterotrimer, 3.7A RMSD 740 residues
      <tr><td align=center>9mcw<td align=center>1154<td align=center>13<td align=center>6.8<td align=center>4.4<td align=center><td align=center><td align=center><td>1154 rna bases, homodimer, wrong dimer and monomer conformations
      <tr><td align=center>8sao<td align=center>1371<td align=center>out of memory<td align=center>23<td align=center>cpu 151<br>gpu out of memory<td align=center><td align=center><td align=center><td>1274 amino acids, 97 ligand atoms, 3.4A RMSD 1151 residues
    </table>

    <h3>Performance notes</h3>
    <p>
      <b>Mac GPU acceleration.</b>
      The reported Mac performance is for Mac M1/M2/M3/M4 series GPUs.  Boltz uses machine learning package torch which
      has GPU acceleration called Metal Performance Shaders (MPS) on these Mac M series GPUs which have speed only 2-3x slower than an Nvidia 4090 but with the advantage that the Mac can handle larger molecular systems using the entire computer memory (e.g. 32 or 64 GB).  Older <a href="https://discuss.pytorch.org/t/pytorch-support-for-intel-gpus-on-mac/151996">Mac Intel machines do not have GPU acceleration in Torch</a> and run at speeds similar to Windows Intel CPU-only predictions.
    </p>
    
    <p>
      <b>Windows Nvidia GPU performance.</b>
      The above table shows a significant slow-down in predictions beyond about 600 residues on Windows with Nvidia 3070 (8 GB) and 4070 (12 GB) graphics.  This is probably because the GPU memory is insufficient for larger structures and the machine learning toolkit falls back to a mix of CPU and GPU calculation.  Notice that the 4070 GPU took more time than the 3070 GPU for large structures probably because the CPU on the 4070 machine (i5-6700K) is significantly slower than the CPU on the 3070 machine (i7-12900K).
    </p>

    <p>
      <b>Linux Nvidia GPU out of memory.</b>
      On Linux Nvidia 4090 with 24 GB of GPU memory the maximum prediction appears to be about 1000 residues plus ligand atoms before an "out of memory" error occurs.  This contrasts with Windows where Torch appears to fallback to using CPU and not run out of GPU memory.
    </p>

    <p>
      <b>Optimizing for larger structures.</b>
      It looks like structure predictions that are twice as large will be possible by using bfloat16 instead of float32 for boltz weights and activations, and this will also give faster predictions.  I tested this on Nvidia 4090 on Linux and it was able to predict 9fz5 (1025 tokens) in 3 minutes and 8sa0 (1371 tokens) in 5 minutes.  Details in ChimeraX ticket <a href="https://www.rbvi.ucsf.edu/trac/ChimeraX/ticket/17555">#17555</a>.  I hope to enable this option after more testing.
    </p>

    <h2>Options</h2>

    <img src="images/boltz_options.png" width=450 align=right>

    <p>
      Pressing the <b>Options</b> button shows additional settings for boltz predictions.
    </p>

    <ul>
      <li><b>Results directory</b>: You can change where the results are placed.
	The path contains "[name]" which is replaced with the prediction name.
      <li><b>Number of predictions</b>: The default number of predicted structures is 1, but you can request more to get an ensemble of structures that usually have small variations.
      <li><b>Compute device</b>: The computation can run on Nvidia GPUs with CUDA or Mac M series GPUs and complete faster than CPU-only calculations.  The default setting tries to use the fastest available device.
      <li><b>Multiple sequence alignment cache</b>: Boltz uses deep multiple sequence alignments for proteins
	that are computed using the Colabfold server.  To avoid recomputing the alignments when you run predictions
	with the same proteins but different ligands ChimeraX caches those alignments in ~/Downloads/ChimeraX/BoltzMSA
	and the cached alignments will be used if available as long as this option enables using the cache.
      <li><b>Boltz install location</b>: Boltz is installed in a virtual Python environment contained in this folder.
    </ul>

    <p>
      Additional advanced options are avaiable by using the ChimeraX boltz command.
    </p>
    
    <h2>ChimeraX boltz command</h2>

    <p>
      The ChimeraX Boltz graphical interface runs a prediction by running the ChimeraX <b>boltz</b> command.
      That command is recorded in the ChimeraX Log panel, and looking at that command can help you understand
      the command options.
    </p>

    <pre>
      <b>boltz predict</b> [<i>sequences</i>] [<b>protein</b> <i>sequences</i>] [<b>dna</b> <i>sequences</i>] [<b>rna</b> <i>sequences</i>]
         [<b>ligands</b> <i>residue-spec</i>] [<b>excludeLigands</b> <i>ccd-codes</i>]
         [<b>ligandCcd</b> <i>ccd-codes</i>] [<b>ligandSmiles</b> <i>smiles-string</i>]
         [<b>name</b> <i>prediction-name</i>] [<b>resultsDirectory</b> <i>directory</i>]
         [<b>device</b> default|cpu|gpu] [<b>samples</b> <i>n</i>] [<b>recycles</b> <i>n</i>] [<b>seed</b> <i>n</i>]
         [<b>useMsaCache</b> <b>true</b>|false] [<b>open</b> <b>true</b>|false] [<b>installLocation</b> <i>directory</i>] [<b>wait</b> true|<b>false</b>]</pre>

    <p>
      Options descriptions
    </p>

     <ul>
       <li><i>sequences</i> - Sequences can be specified using chain ids, UniProt identifiers, explicit strings of amino acid 1-letter codes, or ChimeraX sequence viewer ids.
       <li><b>protein</b> <i>sequences</i> - Like the previous sequences option only this explicitly excludes non-protein sequences in specifiers for open models.  For example "protein #1" would not include any DNA/RNA chains of model #1.  This option can be used more than once.
       <li><b>dna</b> <i>sequences</i> - Uses only DNA sequences from open model specifiers and treats explicit 1-letter code sequences as DNA.  UniProt ids cannot be used.  This option can be used more than once.
       <li><b>rna</b> <i>sequences</i> - Uses only RNA sequences from open model specifiers and treats explicit 1-letter code sequences as RNA.  UniProt ids cannot be used.  This option can be used more than once.
       <li><b>ligands</b> <i>residue-spec</i> - Specify ligands using residue specifiers for open models.
       <li><b>excludeLigands</b> <i>ccd-codes</i> - Exclude these CCD codes when interpreting the ligands option.  By default it excludes ccd code "HOH", ie water.
       <li><b>ligandCcd</b> <i>ccd-codes</i> - Comma-separated list of 3 or 5-letter CCD codes.  This option can be used more than once.
       <li><b>ligandSmiles</b> <i>smiles-string</i> - Comma-separated list of SMILES strings.  This option can be used more than once.
       <li><b>name</b> <i>prediction-name</i> - Name of prediction used in the results folder name.
       <li><b>resultsDirectory</b> <i>directory</i> - Path to the results directory that will be created.  "[name]" in the path will be replaced by the prediction name.  If the folder already exists an numeric suffix _1, _2, _3... will be appended.
       <li><b>device</b> <b>default</b>|cpu|gpu - Whether to run the computation on GPU or CPU.  The default setting chooses based on GPU availability and torch support for the GPU.
       <li><b>samples</b> <i>n</i> - Number of predictions.  Default is 1.  This is what Boltz calls "diffusion samples".  Creating additional structures takes much less time than creating the first structure.
       <li><b>recycles</b> <i>n</i> - Number of passes through the neural net that derives spatial information from which structures will be computed.  Defaults to 3.  Better structures make result from 10 but runtime will be increased.
       <li><b>seed</b> <i>n</i> - Random numbers seed (integer) to initialize calculation.  Runs with different seeds will give different results.
       <li><b>useMsaCache</b> <b>true</b>|false - Whether to use protein deep sequences alignments from the ChimeraX Boltz MSA cache in ~/Downloads/ChimeraX/BoltzMSA.  Because the alignments for different proteins in an assembly are paired to match ones from the same organisms, using the cache requires that an assembly have the exact same set of proteins.  It cannot use alignments computed for individual proteins from multiple different runs.
       <li><b>open</b> <b>true</b>|false - Whether to open the predicted structures when the prediction finishes.
	 The structures will be aligned to an already open model if that open model was used (the first used) in
	 specifying the assembly.
       <li><b>installLocation</b> <i>directory</i> - Where Boltz is installed.
       <li><b>wait</b> true|<b>false</b> - Whether ChimeraX should wait frozen while the prediction is computed or return immediately and allow ChimeraX use during the computation.
    </ul>

     <h3>Installation command</h3>
     
     <pre>
       <b>boltz install</b> <i>directory</i> [<b>downloadModelWeightsAndCcd</b> <b>true</b> | false]</pre>

     <p>
       The boltz install command creates a Python virtual environment to install
       <a href="https://pypi.org/project/boltz/">Boltz from PyPi</a>.  The specified directory
       will be created or if it already exists must be empty.  It then downloads
       the Boltz network parameters and Chemical Component Dictionary to ~/.boltz.  Finally it makes
       an index of the atom counts for each CCD code so that the ChimeraX Boltz interface can
       report the total number of tokens (residues plus ligand atoms) in an assembly in order
       to judge whether the computer has enough memory to make the requested prediction.
     </p>

    <p>
      The install process executes these commands to make the virtual environment and install Boltz.
      It uses the ChimeraX Python executable to create the virtual environment.  Boltz will no longer
      work if ChimeraX is moved or deleted and will need to be reinstalled in that case.
    </p>

    <pre>
      python -m venv <i>directory</i>
      <i>directory</i>/bin/python -m pip install boltz
    </pre>
    
    <p>
    </p>
    
    <h2>Limitations</h2>

    <ul>
      <li><b>Structure size</b>.  Boltz uses a lot of memory and the amount of available memory
	limits the size of structures that can be predicted.  For a computer with 32 Gbytes the size limit
	is roughly 1000 residues plus ligand atoms (called "tokens").  Consumer Nvidia GPUs with 8 or 12 GB of
	memory (e.g. RTX 3070) may only be capable of much smaller predictions.  Consumer Nvidia GPUs with
	24 GB (RTX 3090 and RTX 4090) appear able to predict 1000 tokens.  This is perhaps the most important
	shortcoming of Boltz compared to AlphaFold 3 which handles memory more efficiently and is able to
	predict 5000 tokens with 80GB of GPU memory, about twice the size that Boltz can predict.
	A drawback of AlphaFold 3 is that it requires Linux and an Nvidia GPU in addition to various
	licensing restrictions.  We hope in the future Boltz will optimize memory use to predict larger
	structures.
      <li><b>Run time</b>.  The computation time increase quadratically with the number of tokens.  So
	a prediction with 3 times the number of residue and ligand atoms will take approximately
	9 times longer to run.
      <li><b>Nvidia GPU support on Windows</b>.  Currently the torch 2.6.0 from PyPi used with the standard
	Boltz installation does not include Nvidia GPU support on Windows.  If you wish to use the Nvidia GPU
	with CUDA version 12.6 on a Windows machine run the following commands in a Windows Command Prompt.
	For other CUDA version refer to the
	<a href="https://pytorch.org/get-started/locally/">Torch installation page</a>
	for the correct install command.
	<pre>
	  > cd C:\Users\username\boltz\Scripts
	  > pip.exe uninstall torch
	  > pip.exe install torch --index-url https://download.pytorch.org/whl/cu126</pre>
      <li><b>No covalently linked ligands</b>.  Although Boltz can predict covalently linked ligands I have
	not yet included that capability in the ChimeraX interface or command.
      <li><b>No assigning chain identifiers</b>.  It can be helpful to assign chain identifiers (A,B,C...) to
	the different molecular components to match existing structures.  Boltz is capable of this but the
	ChimeraX user interface does not currently allow it.
      <li><b>MSA sequence alignments</b>. Boltz uses the Colabfold MSA server (https://api.colabfold.com)
	for computing deep sequence alignments.  This requires internet connectivity and is subject to
	outages it that server (located in Korea currently) is down.
    </ul>
    
  </body>
  
</html>
