--- src/plugins/bearer/corewlan/corewlan.pro.orig	2012-11-23 02:09:55.000000000 -0800
+++ src/plugins/bearer/corewlan/corewlan.pro	2013-03-13 16:11:20.000000000 -0700
@@ -5,7 +5,5 @@
 LIBS += -framework Foundation -framework SystemConfiguration
 
 contains(QT_CONFIG, corewlan) {
-    isEmpty(QMAKE_MAC_SDK)|contains(QMAKE_MAC_SDK, "/Developer/SDKs/MacOSX10\.[67]\.sdk") {
          LIBS += -framework CoreWLAN -framework Security
-    }
 }
