--- Modules/pyexpat.c.orig	2008-05-23 11:58:04.000000000 -0700
+++ Modules/pyexpat.c	2008-05-23 11:56:18.000000000 -0700
@@ -2,7 +2,7 @@
 #include <ctype.h>
 
 #include "frameobject.h"
-#include "expat.h"
+#include "expat/expat.h"
 
 #include "pyexpat.h"
 
--- Modules/_elementtree.c.orig	2008-07-10 13:38:23.000000000 -0700
+++ Modules/_elementtree.c	2008-07-10 13:38:33.000000000 -0700
@@ -1826,7 +1826,7 @@
 
 #if defined(USE_EXPAT)
 
-#include "expat.h"
+#include "expat/expat.h"
 
 #if defined(USE_PYEXPAT_CAPI)
 #include "pyexpat.h"
