--- Makefile.pre.in.orig	2014-03-16 19:31:30.000000000 -0700
+++ Makefile.pre.in	2014-04-24 17:16:15.000000000 -0700
@@ -1162,7 +1162,8 @@
 		venv venv/scripts venv/scripts/posix \
 		curses pydoc_data $(MACHDEPS)
 libinstall:	build_all $(srcdir)/Lib/$(PLATDIR) $(srcdir)/Modules/xxmodule.c
-	@for i in $(SCRIPTDIR) $(LIBDEST); \
+	if test "$(BINLIBDEST)" != "$(LIBDEST)"; then $(LN) -s $(DESTDIR)$(BINLIBDEST) $(DESTDIR)$(LIBDEST); fi
+	@for i in $(SCRIPTDIR) $(BINLIBDEST); \
 	do \
 		if test ! -d $(DESTDIR)$$i; then \
 			echo "Creating directory $$i"; \
