*** numpy/distutils/system_info.py~	2006-10-02 21:35:22.000000000 -0700
--- numpy/distutils/system_info.py	2007-05-15 16:13:20.000000000 -0700
***************
*** 152,157 ****
--- 152,159 ----
      default_include_dirs.append(os.path.join(sys.prefix, 'include'))
      default_src_dirs.append(os.path.join(sys.prefix, 'src'))
  
+ default_lib_dirs = [] # Avoid system dependencies for Chimera build.
+ 
  default_lib_dirs = filter(os.path.isdir, default_lib_dirs)
  default_include_dirs = filter(os.path.isdir, default_include_dirs)
  default_src_dirs = filter(os.path.isdir, default_src_dirs)
