--- lib/python3.7/site-packages/simtk/openmm/app/forcefield.py	2019-07-03 17:34:24.494184453 +0100
+++ lib/python3.7/site-packages/simtk/openmm/app/forcefield_new.py	2019-07-03 17:35:29.492779395 +0100
@@ -183,7 +183,7 @@
         self._templateGenerators = []
         self.loadFile(files)
 
-    def loadFile(self, files):
+    def loadFile(self, files, prefix=''):
         """Load an XML file and add the definitions from it to this ForceField.
 
         Parameters
@@ -194,6 +194,8 @@
             directory, a path relative to this module's data subdirectory (for
             built in force fields), or an open file-like object with a read()
             method from which the forcefield XML data can be loaded.
+        prefix : string
+            An optional string to prepend to each residue name in the files
         """
 
         if isinstance(files, tuple):
@@ -259,7 +261,7 @@
         for tree in trees:
             if tree.getroot().find('Residues') is not None:
                 for residue in tree.getroot().find('Residues').findall('Residue'):
-                    resName = residue.attrib['name']
+                    resName = prefix+residue.attrib['name']
                     template = ForceField._TemplateData(resName)
                     if 'override' in residue.attrib:
                         template.overrideLevel = int(residue.attrib['override'])
