--- python/setup.py.orig	2016-09-22 12:50:01.343416334 -0700
+++ python/setup.py	2016-09-22 12:49:38.190719810 -0700
@@ -200,6 +200,8 @@
             # that have GCC installed from MacPorts or homebrew *and* Anaconda
             os.environ['CC'] = 'clang'
             os.environ['CXX'] = 'clang++'
+    if platform.system() == 'Linux':
+        extra_compile_args += ['-D_GLIBCXX_USE_CXX11_ABI=0']
 
 
     library_dirs=[openmm_lib_path]
