*** python/setup.py~	Wed May 25 21:48:10 2016
--- python/setup.py	Wed Sep 21 13:28:34 2016
***************
*** 212,219 ****
                      "libraries": libraries,
                      "extra_compile_args": extra_compile_args,
                      "extra_link_args": extra_link_args}
!     if platform.system() != "Windows":
!         extensionArgs["runtime_library_dirs"] = library_dirs
      setupKeywords["ext_modules"] = [Extension(**extensionArgs)]
  
      outputString = ''
--- 212,219 ----
                      "libraries": libraries,
                      "extra_compile_args": extra_compile_args,
                      "extra_link_args": extra_link_args}
! #    if platform.system() != "Windows":
! #        extensionArgs["runtime_library_dirs"] = library_dirs
      setupKeywords["ext_modules"] = [Extension(**extensionArgs)]
  
      outputString = ''
