--- filelock.py.orig	2016-10-24 15:42:44.707016200 -0700
+++ filelock.py	2016-10-24 15:43:41.856728000 -0700
@@ -318,17 +318,14 @@
     def _acquire(self):
         open_mode = os.O_RDWR | os.O_CREAT | os.O_TRUNC
 
+        fd = os.open(self._lock_file, open_mode)
+
         try:
-            fd = os.open(self._lock_file, open_mode)
-        except OSError:
-            pass
+            msvcrt.locking(fd, msvcrt.LK_NBLCK, 1)
+        except (IOError, OSError):
+            os.close(fd)
         else:
-            try:
-                msvcrt.locking(fd, msvcrt.LK_NBLCK, 1)
-            except (IOError, OSError):
-                os.close(fd)
-            else:
-                self._lock_file_fd = fd
+            self._lock_file_fd = fd
         return None
 
     def _release(self):
