--- Modules/FindQt4.cmake.orig	2013-03-19 16:35:51.000000000 -0700
+++ Modules/FindQt4.cmake	2013-03-19 16:35:55.000000000 -0700
@@ -670,7 +670,7 @@
       # Qt/Mac frameworks has two include dirs.
       # One is the framework include for which CMake will add a -F flag
       # and the other is an include dir for non-framework Qt modules
-      set(QT_INCLUDE_DIR ${QT_HEADERS_DIR} ${QT_QTCORE_LIBRARY_RELEASE} )
+      #set(QT_INCLUDE_DIR ${QT_HEADERS_DIR} ${QT_QTCORE_LIBRARY_RELEASE} )
     else()
       set(QT_INCLUDE_DIR ${QT_HEADERS_DIR})
     endif()
