--- OpenGL/GL/VERSION/GL_2_0.py.orig	2013-07-26 15:52:12.275008993 -0700
+++ OpenGL/GL/VERSION/GL_2_0.py	2013-07-26 16:06:36.013958360 -0700
@@ -20,6 +20,8 @@
 from OpenGL.raw.GL.ARB.shader_objects import GL_OBJECT_LINK_STATUS_ARB as GL_OBJECT_LINK_STATUS
 from OpenGL.raw.GL.ARB.shader_objects import GL_OBJECT_ACTIVE_UNIFORMS_ARB as GL_OBJECT_ACTIVE_UNIFORMS
 from OpenGL.raw.GL.ARB.shader_objects import GL_OBJECT_ACTIVE_UNIFORM_MAX_LENGTH_ARB as GL_OBJECT_ACTIVE_UNIFORM_MAX_LENGTH
+from OpenGL.raw.GL.ARB.vertex_shader import GL_OBJECT_ACTIVE_ATTRIBUTES_ARB as GL_OBJECT_ACTIVE_ATTRIBUTES
+from OpenGL.raw.GL.ARB.vertex_shader import GL_OBJECT_ACTIVE_ATTRIBUTE_MAX_LENGTH_ARB as GL_OBJECT_ACTIVE_ATTRIBUTE_MAX_LENGTH
 from OpenGL.GL.ARB.shader_objects import glGetInfoLogARB as glGetInfoLog
 from OpenGL.lazywrapper import lazy
 
@@ -220,6 +222,23 @@
     elif name[-1] != _NULL_8_BYTE:
         name = name + _NULL_8_BYTE
     return baseOperation( program, name )
+
+@lazy( glGetActiveAttrib )
+def glGetActiveAttrib(baseOperation, program, index):
+    """Retrieve the name, size and type of the attribute of the index in the program"""
+    max_index = int(glGetProgramiv( program, GL_OBJECT_ACTIVE_ATTRIBUTES ))
+    length = int(glGetProgramiv( program, GL_OBJECT_ACTIVE_ATTRIBUTE_MAX_LENGTH))
+    if index < max_index and index >= 0:
+        if length > 0:
+            name = ctypes.create_string_buffer(length)
+            size = arrays.GLintArray.zeros( (1,))
+            gl_type = arrays.GLenumArray.zeros( (1,))
+            namelen = arrays.GLsizeiArray.zeros( (1,))
+            baseOperation(program, index, length, namelen, size, gl_type, name)
+            return name.value[:int(namelen[0])], size[0], gl_type[0]
+        raise ValueError( """No currently specified attribute names""" )
+    raise IndexError( 'Index %s out of range 0 to %i' % (index, max_index - 1, ) )
+
 @lazy( glGetAttribLocation )
 def glGetAttribLocation( baseOperation, program, name ):
     """Check that name is a string with a null byte at the end of it"""
