*** OpenGL/raw/WGL/_types.py.orig	Tue Feb 25 07:31:33 2014
--- OpenGL/raw/WGL/_types.py	Wed Dec 20 22:52:10 2017
***************
*** 1,13 ****
  from ctypes import *
  from OpenGL import extensions
  from OpenGL.raw.GL._types import *
  from OpenGL._opaque import opaque_pointer_cls as _opaque_pointer_cls
  c_void = None
  
  class _WGLQuerier( extensions.ExtensionQuerier ):
!     prefix = 'WGL_'
      assumed_version = [1,0]
!     version_prefix = 'WGL_VERSION_WGL_'
      def pullVersion( self ):
          # only one version...
          return [1,0]
--- 1,14 ----
  from ctypes import *
  from OpenGL import extensions
  from OpenGL.raw.GL._types import *
+ from OpenGL._bytes import as_8_bit
  from OpenGL._opaque import opaque_pointer_cls as _opaque_pointer_cls
  c_void = None
  
  class _WGLQuerier( extensions.ExtensionQuerier ):
!     prefix = as_8_bit('WGL_')
      assumed_version = [1,0]
!     version_prefix = as_8_bit('WGL_VERSION_WGL_')
      def pullVersion( self ):
          # only one version...
          return [1,0]
***************
*** 17,23 ****
          wglGetCurrentDC.restyle = HDC
          try:
              dc = wglGetCurrentDC()
!             proc_address = PLATFORM.getExtensionProcedure( 'wglGetExtensionsStringARB' )
              wglGetExtensionStringARB = PLATFORM.functionTypeFor( PLATFORM.WGL )(
                  c_char_p,
                  HDC,
--- 18,24 ----
          wglGetCurrentDC.restyle = HDC
          try:
              dc = wglGetCurrentDC()
!             proc_address = PLATFORM.getExtensionProcedure( as_8_bit('wglGetExtensionsStringARB') )
              wglGetExtensionStringARB = PLATFORM.functionTypeFor( PLATFORM.WGL )(
                  c_char_p,
                  HDC,
