*** OpenGL/raw/WGL/_types.py~	Tue Jan  9 14:23:20 2018
--- OpenGL/raw/WGL/_types.py	Thu Feb 15 18:46:50 2018
***************
*** 72,78 ****
  
  HANDLE = POINTER(None) 	# /home/mcfletch/pylive/OpenGL-ctypes/src/wgl.h:60
  # TODO: figure out how to make the handle not appear as a void_p within the code...
! HANDLE.final = True
  
  HGLRC = HANDLE 	# /home/mcfletch/pylive/OpenGL-ctypes/src/wgl.h:62
  HDC = HANDLE 	# /home/mcfletch/pylive/OpenGL-ctypes/src/wgl.h:63
--- 72,78 ----
  
  HANDLE = POINTER(None) 	# /home/mcfletch/pylive/OpenGL-ctypes/src/wgl.h:60
  # TODO: figure out how to make the handle not appear as a void_p within the code...
! #HANDLE.final = True
  
  HGLRC = HANDLE 	# /home/mcfletch/pylive/OpenGL-ctypes/src/wgl.h:62
  HDC = HANDLE 	# /home/mcfletch/pylive/OpenGL-ctypes/src/wgl.h:63
