*** OpenGL/platform/osmesa.py~	Mon Feb 24 05:45:30 2014
--- OpenGL/platform/osmesa.py	Wed Jul 29 18:41:49 2015
***************
*** 18,27 ****
      ]
      @baseplatform.lazy_property
      def GL(self):
          try:
              return ctypesloader.loadLibrary(
                  ctypes.cdll,
!                 'OSMesa', 
                  mode=ctypes.RTLD_GLOBAL 
              ) 
          except OSError as err:
--- 18,29 ----
      ]
      @baseplatform.lazy_property
      def GL(self):
+         import os
+         osmesa_lib_path = os.environ.get('PYOPENGL_OSMESA_LIB_PATH', 'OSMesa')
          try:
              return ctypesloader.loadLibrary(
                  ctypes.cdll,
!                 osmesa_lib_path, 
                  mode=ctypes.RTLD_GLOBAL 
              ) 
          except OSError as err:
--- OpenGL/raw/osmesa/mesa.py.orig	2018-11-15 15:21:53.023014161 -0800
+++ OpenGL/raw/osmesa/mesa.py	2018-11-15 15:32:20.134247007 -0800
@@ -27,6 +27,15 @@
 OSMESA_MAX_WIDTH = _C('OSMESA_MAX_WIDTH', 0x24)
 OSMESA_MAX_HEIGHT = _C('OSMESA_MAX_HEIGHT', 0x25)
 
+OSMESA_DEPTH_BITS = _C('OSMESA_DEPTH_BITS', 0x30)
+OSMESA_STENCIL_BITS = _C('OSMESA_STENCIL_BITS', 0x31)
+OSMESA_ACCUM_BITS = _C('OSMESA_ACCUM_BITS', 0x32)
+OSMESA_PROFILE = _C('OSMESA_PROFILE', 0x33)
+OSMESA_CORE_PROFILE = _C('OSMESA_CORE_PROFILE', 0x34)
+OSMESA_COMPAT_PROFILE = _C('OSMESA_COMPAT_PROFILE', 0x35)
+OSMESA_CONTEXT_MAJOR_VERSION = _C('OSMESA_CONTEXT_MAJOR_VERSION', 0x36)
+OSMESA_CONTEXT_MINOR_VERSION = _C('OSMESA_CONTEXT_MINOR_VERSION', 0x37)
+
 OSMesaGetCurrentContext = _p.GetCurrentContext
 
 @_f
@@ -38,6 +47,10 @@
 def OSMesaCreateContextExt(format, depthBits, stencilBits,accumBits,sharelist ): pass
 
 @_f
+@_p.types(OSMesaContext, ctypes.POINTER(ctypes.c_int), OSMesaContext)
+def OSMesaCreateContextAttribs(attribList, sharelist): pass
+
+@_f
 @_p.types(None, OSMesaContext)
 def OSMesaDestroyContext(ctx): pass
 
@@ -83,12 +96,15 @@
 
 
 __all__ = [
-    'OSMesaCreateContext',
-    'OSMesaCreateContextExt', 'OSMesaMakeCurrent', 'OSMesaGetIntegerv',
+    'OSMesaCreateContext', 'OSMesaCreateContextExt', 'OSMesaCreateContextAttribs',
+    'OSMesaMakeCurrent', 'OSMesaGetIntegerv',
     'OSMesaGetCurrentContext', 'OSMesaDestroyContext', 'OSMesaPixelStore',
     'OSMesaGetDepthBuffer', 'OSMesaGetColorBuffer',
     'OSMESA_COLOR_INDEX', 'OSMESA_RGBA', 'OSMESA_BGRA', 'OSMESA_ARGB',
     'OSMESA_RGB', 'OSMESA_BGR', 'OSMESA_BGR', 'OSMESA_ROW_LENGTH',
     'OSMESA_Y_UP', 'OSMESA_WIDTH', 'OSMESA_HEIGHT', 'OSMESA_FORMAT',
     'OSMESA_TYPE', 'OSMESA_MAX_WIDTH', 'OSMESA_MAX_HEIGHT',
+    'OSMESA_DEPTH_BITS', 'OSMESA_STENCIL_BITS', 'OSMESA_ACCUM_BITS',
+    'OSMESA_PROFILE', 'OSMESA_CORE_PROFILE', 'OSMESA_COMPAT_PROFILE',
+    'OSMESA_CONTEXT_MAJOR_VERSION', 'OSMESA_CONTEXT_MINOR_VERSION',
 ]
