<html>

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Tool: AlphaFold</title>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRight" title="User Guide Index"/></a>

<h3><a href="../index.html#tools">Tool</a>: AlphaFold</h3>
<p>
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold</a>
is an artificial intelligence method for predicting protein structures
that has been highly successful in recent tests.
The ChimeraX <b>AlphaFold</b> tool:
<ul>
<li><a href="#search">finds and retrieves existing predictions</a> from the
from the <a href="https://alphafold.ebi.ac.uk/"
target="_blank">AlphaFold Database</a>
<br>&ndash; freely available for academic and commercial use under 
<a href="https://creativecommons.org/licenses/by/4.0/" 
target="_blank">CC BY 4.0</a>
<br><br>
<li><a href="#predict">runs new predictions</a> with
<a href="https://alphafold.ebi.ac.uk/"
target="_blank">AlphaFold</a> using
<a href="https://colab.research.google.com/signup" 
target="_blank">Google Colab</a>
<br>&ndash; for noncommercial use only, as the
<a href="https://github.com/deepmind/alphafold/#model-parameters-license"
target="_blank">AlphaFold-learned parameters</a> 
needed to run the calculation are licensed under 
<a href="https://creativecommons.org/licenses/by-nc/4.0/legalcode"
target="_blank">CC BY NC 4.0</a>
</ul>
Users should cite:
</p>
<blockquote>
<a href="https://www.nature.com/articles/s41586-021-03819-2" target="_blank">
Highly accurate protein structure prediction with AlphaFold.</a>
Jumper J, Evans R, Pritzel A, <i>et al.</i>
<i>Nature.</i> 2021 Jul 15. doi: 10.1038/s41586-021-03819-2.
</blockquote>
<p>
The
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>
contains predictions for
<a href="https://alphafold.ebi.ac.uk/download" target="_blank">21 species</a>,
including humans. It does not cover all of 
<a href="https://www.uniprot.org" target="_blank">UniProt</a>.
The predicted structures vary in
confidence levels and should be interpreted with caution.
The <a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold</a>
system predicts structures for single chains, not complexes; 
assembling the individual predictions into a complex may give 
unphysical results where parts of the chains intersect or 
interact poorly with one another.
</p><p>
The <b>AlphaFold</b> tool is also implemented as the
<a href="../commands/alphafold.html"><b>alphafold</b></a> command.
See also: <a href="blastprotein.html"><b>Blast Protein</b></a>,
<a href="modeller.html"><b>Modeller Comparative</b></a>,
<a href="https://www.rbvi.ucsf.edu/chimerax/data/alphafold-aug2021/alphafold.html"
target="_blank">comparing AlphaFold and experimental structures</a>, and
<a href="https://www.rbvi.ucsf.edu/chimerax/docs/videos/"
target="_blank">ChimeraX videos</a>:
<a href="https://youtu.be/x-9urTQ1rJY" target="_blank">searching by sequence</a>
for AlphaFold predictions (first 5 minutes),
<a href="https://youtu.be/W6KsBfNwDgs" target="_blank">fetching</a> an
AlphaFold prediction,
<a href="https://youtu.be/vj3sQfvWfDU" target="_blank">matching</a>
an assembly,
<a href="https://youtu.be/le7NatFo8vI" target="_blank">predicting</a>
a new structure
</p>

<p class="nav">
[<a href="#top">back to top: AlphaFold</a>]
</p>
<h3>Dialog and Buttons</h3>
<p>
The <b>AlphaFold</b> tool can be opened from the <b>Structure Prediction</b>
section of the <b>Tools</b> menu and manipulated like other panels
(<a href="../window.html#panels">more...</a>).
</p><p>
The <b>Sequence</b> can be specified by
<a href="https://www.uniprot.org/" target="_blank">UniProt</a>
name or accession number, pasted in as plain text, or chosen
from the menu of currently open protein structure chains.
</p>
<ul>
<li>
<a name="match"></a>
<b>Fetch</b> gets the most sequence-similar prediction available from the
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>
for each specified chain. 
Specifying a whole model specifies all of its protein chains.
For each chain, a prediction is obtained for the exact UniProt entry 
if available, otherwise the single top hit identified by
<a href="https://en.wikipedia.org/wiki/BLAT_(bioinformatics)"
target="_blank">BLAT</a>-searching the 
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>
(<a href="../commands/alphafold.html#search-option">details...</a>).
The corresponding command is
<a href="../commands/alphafold.html#match"><b>alphafold match</b></a>.
If the sequence was specified by structure chain, then:
</p>
<ol>
<li>the chain ID of the predicted structure is made the same as the
corresponding chain of the existing model
<li>the predicted structure is superimposed onto the existing chain using
<a href="../commands/matchmaker.html"><b>matchmaker</b></a> defaults,
and the following are reported in a table
in the <a href="log.html"><b>Log</b></a>:
  <ul>
  <li><b>Chain</b> &ndash; chain ID
  <li><a href="https://www.uniprot.org/help/entry_name" 
  target="_blank"><b>UniProt Name</b></a>
  and <a href="https://www.uniprot.org/help/accession_numbers"
  target="_blank"><b>UniProt Id</b></a> (accession number)
  <li><b>RMSD</b> &ndash; C&alpha; root-mean-square deviation between
  the predicted and experimental structures, over all residues of the latter
  <li><b>Length</b> &ndash; number of residues in the predicted structure
  <li><b>Seen</b> &ndash; number of residues with atomic coordinates 
  in the experimental structure
  <li><b>% Id</b> &ndash; percent identity in the sequence alignment generated 
  by <a href="../commands/matchmaker.html"><b>matchmaker</b></a> 
  for superposition; the number of positions with identical residues
  divided by the length of the shorter sequence
  </ul>
<li>the predicted structure is trimmed to the same residue range as the 
existing chain (<a href="../commands/alphafold.html#trim">details...</a>)
<li>the following <a href="../attributes.html">attributes</a>
 are assigned to the residues of the predicted structure:
  <ul>
  <li><b>c_alpha_distance</b>
&ndash; C&alpha; distance between corresponding positions of the predicted 
and existing chains after their superposition (step 2 above) 
  <li><b>missing_structure</b>
&ndash; positions missing from the coordinates of the existing chain
  <li><b>same_sequence</b>
&ndash; positions with different residue types than the existing chain
  </ul>
These <a href="../attributes.html">attributes</a>
can be used for <a href="#coloring">coloring</a> and other purposes.
</ol>
<p>
The fetched predictions are stored locally in 
<b>~/Downloads/ChimeraX/AlphaFold/</b>,
where ~ indicates a user's home directory. 
If a file specified for opening is not found in this local cache,
the file will be fetched and cached. 
</p>
<li>
<a name="search"></a>
<b>Search</b>
uses a <a href="../tools/blastprotein.html#notes">BLAST</a> web service
hosted by the <a href="https://www.cgl.ucsf.edu/Overview/index.html" 
target="_blank">UCSF RBVI</a> to search the 
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>
using default parameters: <b>BLOSUM62</b>
amino acid similarity matrix for scoring the hits,
similarity score cutoff e-value <b>1e-3</b>,
returning a maximum of <b>100</b> of unique sequences
(however, more than 100 hits may be obtained because multiple structures 
or other sequence-database entries may have the same sequence).
The corresponding command is
<a href="../commands/alphafold.html#search"><b>alphafold search</b></a>.
<b>Search</b> differs from <b>Fetch</b> in that it uses BLAST instead of BLAT,
accepts only a single chain or sequence as input,
and returns a list of hits for the user to inspect,
rather than fetching the single top hit per chain automatically.
When results are returned, the hits are listed in a
<a href="blastprotein.html#results"><b>Blast Protein</b></a> window.
Clicking the name of a hit uses
<a href="../commands/alphafold.html#fetch"><b>alphafold fetch</b></a> 
to retrieve the prediction.
<br><br>
<li>
<a name="predict"></a>
<b>Predict</b>
runs an <a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold</a>
calculation using <a href="https://colab.research.google.com/signup" 
target="_blank">Google Colab</a>. 
The corresponding command is
<a href="../commands/alphafold.html#predict"><b>alphafold predict</b></a>.
A warning will appear saying that this Colab notebook is from github 
(was not authored by Google), with a button to click to run anyway.
Users will need to have a Google account and to sign into it via
a browser. Once that is done, the sign-in may be remembered depending
on the user's browser settings; it is not kept in the 
<a href="../preferences.html">ChimeraX preferences</a>.
A single prediction run generally takes on the order of an hour or more.
The process includes installing various software packages on a virtual machine,
searching sequence databases, generating a multiple sequence alignment,
predicting atomic coordinates, and energy-minimizing the best structure.
</p><p>
The free version of Google Colab does not allow for much run time.
Those who want to run longer and/or more frequent calculations 
may wish to sign up for one of the paid
<a href="https://colab.research.google.com/signup" 
target="_blank">Colab plans</a>.
</p><p>
<a name="caveats"></a>
Please note the following caveats:
</p>
<ul>
<li>Regardless of a user's 
<a href="https://colab.research.google.com/signup" 
target="_blank">Colab plan</a>, <b>Predict</b> is for 
<b>noncommercial use only</b>. Paid 
<a href="https://colab.research.google.com/signup" 
target="_blank">Colab plans</a> provide more computational resources,
but they do not cover access to the 
<a href="https://github.com/deepmind/alphafold/#model-parameters-license"
target="_blank">AlphaFold-learned parameters</a>
needed to run the calculation. The parameters are licensed under 
<a href="https://creativecommons.org/licenses/by-nc/4.0/legalcode"
target="_blank">CC BY NC 4.0</a>; the Colab itself is provided under the 
<a href="https://www.apache.org/licenses/LICENSE-2.0" target="_blank">Apache 
2.0</a> license.
<br><br>
<li>Prediction calculations run a <b>simplified implementation</b> of AlphaFold 
that does not use structural templates and that
searches a smaller set of protein sequences 
(<a href="https://colab.research.google.com/github/deepmind/alphafold/blob/main/notebooks/AlphaFold.ipynb" 
target="_blank">details...</a>).
The instance used by ChimeraX also limits the sequence length
and the number of sequences in the multiple sequence alignment.
For best reliability, it is recommended to use the 
<a href="https://github.com/deepmind/alphafold/" target="_blank">full 
open-source AlphaFold</a> or the <a href="https://alphafold.ebi.ac.uk/"
target="_blank">AlphaFold Database</a> (<i>e.g.</i>, with 
<a href="#search"><b>alphafold search</b></a>) instead.
</ul>
<br>
<li><a name="coloring"></a>
Clicking the <b>Coloring</b> button 
on the <a href="#top"><b>AlphaFold</b></a> tool 
shows the <b>Alphafold Coloring</b> dialog 
for applying different color schemes to the predicted structures,
as well as hiding, showing, and 
<a href="../selection.html">selecting</b></a> their residues based on 
<a href="../attributes.html">attribute</a> value.
<p>
<a name="confidence"></a>
When first opened, AlphaFold-predicted structures are automatically colored by 
the <a href="https://alphafold.ebi.ac.uk/faq#faq-5" target="_blank">pLDDT
confidence measure</a> in the B-factor field:
</p>
     <ul>
     <li>90
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=cornflowerblue width="10px" title="cornflower blue"></td></tr>
</table>
to 100
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=blue width="10px" title="blue"></td></tr>
</table>
&ndash; high accuracy expected
     <li>70
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=yellow width="10px" title="yellow"></td></tr>
</table>
to 90
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=cornflowerblue width="10px" title="cornflower blue"></td></tr>
</table>
&ndash; expected to be modeled well (a generally good backbone prediction)
     <li>50
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=orange width="10px" title="orange"></td></tr>
</table>
to 70
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=yellow width="10px" title="yellow"></td></tr>
</table>
&ndash; low confidence, treat with caution
     <li>0
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=red width="10px" title="red"></td></tr>
</table>
to 50
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=orange width="10px" title="orange"></td></tr>
</table>
&ndash; should not be interpreted, may be disordered
     </ul>
<p>
...in other words, using
</p>
<blockquote><b>
<a href="../commands/color.html#bfactor">color bfactor</a>
<a href="../commands/color.html#palette-options">palette</a> alphafold
<!-- 0,red:50,orange:70,yellow:90,cornflowerblue:100,blue -->
<!-- 100,blue:90,cornflowerblue:70,yellow:50,orange:0,red -->
</b></blockquote>
<p>
The <a href="colorkey.html"><b>Color Key</b></a> graphical interface
or a command can be used to draw a corresponding color key, for example:
</p>
  <blockquote>
  <a href="../commands/key.html"><b>key</b></a> <b>red:low orange: yellow: cornflowerblue: blue:high</b>
  &nbsp;[<a href="../commands/key.html#options"><i>other-key-options</i></a>]
<!--
<br>
  <b>key</b> <b>blue:100 cornflowerblue:90 yellow:70 orange:50 red:0</b>
  &nbsp;<b>numericLabelSpacing equal</b>
  &nbsp;[<a href="../commands/key.html#options"><i>other-key-options</i></a>]
-->
  </blockquote>
<p>
In the <b>AlphaFold Coloring</b> dialog,
the <b>Residues</b> to act on are specified by using the menus to choose an
AlphaFold-predicted model and one of the following:
</p>
  <ul>
  <li><b>all</b>
&ndash; all residues
  <li><b>confidence below</b> [<i>N</i>]
&ndash; based on the <b>bfactor</b>
<a href="../attributes.html#atom">atom attribute</a>
(the <a href="#confidence">confidence value</a> is read from the
B-factor field of the PDB file)
  <li><b>C-alpha distance greater than</b> [<i>d</i>]
&ndash; based on the <b>c_alpha_distance</b>
residue attribute of AlphaFold predictions 
<a href="#match">fetched by existing structure chain</a>;
C&alpha; distance between corresponding positions of the predicted and
existing chains after their automatic superposition
  <li><b>missing structure</b>
&ndash; based on the <b>missing_structure</b> 
residue attribute of AlphaFold predictions 
<a href="#match">fetched by existing structure chain</a>;
positions missing from the coordinates of the existing chain
  <li><b>different sequence</b>
&ndash; based on the <b>same_sequence</b> 
residue attribute of AlphaFold predictions 
<a href="#match">fetched by existing structure chain</a>;
positions with different residue types than the existing chain
  <li><b>confidence above</b> [<i>N</i>]
&ndash; based on the <b>bfactor</b>
<a href="../attributes.html#atom">atom attribute</a>
(the <a href="#confidence">confidence value</a> is read from the
B-factor field of the PDB file)
  <li><b>C-alpha distance less than</b> [<i>d</i>]
&ndash; based on the <b>c_alpha_distance</b>
residue attribute of AlphaFold predictions 
<a href="#match">fetched by existing structure chain</a>;
C&alpha; distance between corresponding positions of the predicted and
existing chains after their automatic superposition
  <li><b>paired structure</b>
&ndash; based on the <b>missing_structure</b> 
residue attribute of AlphaFold predictions 
<a href="#match">fetched by existing structure chain</a>;
positions present in the coordinates of the existing structure chain
  <li><b>same sequence</b>
&ndash; based on the <b>same_sequence</b> 
residue attribute of AlphaFold predictions 
<a href="#match">fetched by existing structure chain</a>;
positions with identical residue types as the existing chain
  </ul>
<p>
Buttons act on the designated residues:
</p>
  <ul>
  <li><b>Color</b> buttons:
    <ul>
    <li><b>Custom</b> for choosing a color interactively
    using the system color picker
    <li>a series of square buttons for specific colors:
<table cellpadding="8px" cellspacing="5x"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor=lightgray width="1px" title="lightgray"></td>
<td bgcolor=red width="1px" title="red"></td>
<td bgcolor=magenta width="1px" title="magenta"></td>
<td bgcolor=yellow width="1px" title="yellow"></td>
<td bgcolor=lime width="1px" title="lime"></td>
<td bgcolor=cyan width="1px" title="cyan"></td>
</tr></table>
    </ul>
  <li><b>Hide</b> &ndash; hide the specified residues
  <li><b>Show</b> &ndash; show the specified residues
  <li><b>Select</b> &ndash; <a href="../selection.html">select</a>
the specified residues
  </ul>
<p>
The <b>AlphaFold Coloring</b> dialog does not color continuously 
along a gradient to show the attribute values. 
For coloring along a gradient, see the commands
<a href="../commands/color.html#bfactor"><b>color bfactor</b></a>
(for the <a href="#confidence">confidence value</a>, which is read from the
B-factor field of the PDB file) and
<a href="../commands/color.html#byattribute"><b>color byattribute</b></a>
(for other numerical <a href="../attributes.html">attributes</a>).
</p>
</ul>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics / 
September 2021</address>
</body></html>
