<html>

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Tool: Blast Protein</title>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRight" title="User Guide Index"/></a>

<h3><a href="../index.html#tools">Tool</a>: Blast Protein</h3>
<p>
<b>Blast Protein</b> runs a protein sequence similarity search
using a <a href="#notes">BLAST</a> web service hosted by the
<a href="https://www.cgl.ucsf.edu/Overview/index.html" target="_blank">UCSF
Resource for Biocomputing, Visualization, and Informatics (RBVI)</a>.
See also:
<a href="alphafold.html"><b>AlphaFold</b></a>,
<a href="../commands/matchmaker.html"><b>matchmaker</b></a>
</p><p>
The <b>Blast Protein</b> tool
can be opened from the <b>Sequence</b> section of the
<b>Tools</b> menu and manipulated like other panels
(<a href="../window.html#panels">more...</a>).
It is also implemented as the
<a href="../commands/blastprotein.html"><b>blastprotein</b></a> command.
</p>

<a name="parameters"></a>
<p class="nav">
[<a href="#top">back to top: Blast Protein</a>]
</p>
<h3>Search Parameters</h3>

<ul class="none">
<li>the first menu allows choosing a protein chain from the currently open 
atomic structures to use as the query
<br>
*Note: the <a href="../commands/blastprotein.html"><b>blastprotein</b></a>
command additionally allows using a sequence
(for which no structure may be available) as the query,
such as to find templates for <a href="modeller.html">comparative modeling</a>
<li><b>Database</b> &ndash; protein sequence database to search:
   <ul>
   <li><b>PDB</b> (default) &ndash; experimentally determined structures in the
<a href="https://www.rcsb.org" target="_blank">Protein Data Bank (PDB)</a>
   <li><b>NR</b> &ndash; NCBI &ldquo;non-redundant&rdquo; database containing
<a href="https://www.ncbi.nlm.nih.gov/genbank/" target="_blank">GenBank</a>
CDS translations + <a href="https://www.rcsb.org" target="_blank">PDB</a>
+ <a href="https://www.uniprot.org/" target="_blank">SwissProt</a> +
<a href="https://proteininformationresource.org/"  target="_blank">PIR</a>
+ <a href="https://www.prf.or.jp/index-e.html" target="_blank">PRF</a>
excluding environmental samples from whole-genome sequencing; this database
is much larger than <b>PDB</b> alone and takes much longer to search
  <li><b>AlphaFold</b> 
&ndash; artificial-intelligence-predicted structures in the
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>
(<a href="alphafold.html">more...</a>)
   </ul>
<li><b># Sequences</b> (default <b>100</b>)
&ndash; maximum number of unique sequences to return; more hits than this number
may be obtained because multiple structures or other sequence-database entries
may have the same sequence
<li><b>Matrix</b>
&ndash; amino acid substitution matrix to use for alignment scoring:
  <ul>
  <li><b>BLOSUM45</b>
  <li><b>BLOSUM50</b>
  <li><b>BLOSUM62</b> (default)
  <li><b>BLOSUM80</b>
  <li><b>BLOSUM90</b>
  <li><b>PAM30</b>
  <li><b>PAM70</b>
  <li><b>PAM250</b>
  <li><b>IDENTITY</b>
  </ul>
<li><b>Cutoff</b> (default <b>1e-3</b>) &ndash; significance cutoff;
only hits with 
<a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=FAQ#expect" target="_blank">E-value</a>
no larger than the specified value will be returned
</ul>
<p>
Clicking <b>Apply</b> (or <b>OK</b>, which also dismisses the dialog)
runs the search, whereas <b>Close</b> dismisses the dialog without 
starting a search. <b>Help</b> opens this page in the 
<a href="helpviewer.html"><b>Help Viewer</b></a>, and
<b>Reset</b> restores the parameters to factory default settings.
</p>

<a name="results"></a>
<p class="nav">
[<a href="#top">back to top: Blast Protein</a>]
</p>
<h3>BLAST Protein Results</h3>
<p>
When results are returned, the table of hits is shown in a separate window.
These results are saved in 
<a href="../commands/save.html#session">ChimeraX sessions</a>.
</p><p>
Checkboxes in the bottom section of the panel control which columns of 
information are shown in the table of hits, with buttons:
</p>
<ul>
<li><b>All</b> &ndash; show all columns
<li><b>Default</b> &ndash; restore the previously saved
<a href="../preferences.html#other">preference</a> for which columns are shown
<li><b>Standard</b> &ndash; use &ldquo;factory default&rdquo; set of 
initially displayed columns
<li><b>Set Default</b> &ndash; save the currently shown set as a 
user <a href="../preferences.html#other">preference</a>
<li><b>Toggle Controls</b> &ndash; show/hide the checkbox section
</ul>
<p>
Clicking a column header sorts by the values in that column.
</p><p>
Double-clicking a row with a corresponding structure 
<a href="../commands/open.html#fetch">fetches</a> it,
and if a structure chain was used as the query,
automatically superimposes the hit onto the query
using <a href="../commands/matchmaker.html"><b>matchmaker</b></a>.
AlphaFold-predicted structures are
<a href="../commands/alphafold.html#colorConfidence">colored by
confidence value</a>.
</p><p>
One or more hits can be chosen (highlighted) in the list
by clicking and dragging with the left mouse button;
<b>Ctrl</b>-click (or <b>command</b>-click if using a Mac)
toggles whether a row is chosen.
The result panel's <a href="../window.html#context-menu">context menu</a> 
can be used to fetch and superimpose all of the corresponding structures, or to
show their <a href="#notes">multiple sequence alignment</a> with the query
in the <a href="sequenceviewer.html"><b>Sequence Viewer</b></a>.
</p><p>
Some columns are available no matter which database is searched:
</p>
<ul>
<li><b>Rank</b> &ndash; 
<a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=FAQ#expect" target="_blank">E-value</a>
rank (from most significant to least)
<li><b>Name</b>
  <ul>
  <li>for PDB sequences, the corresponding PDB identifier, including the chain 
  <li>for AlphaFold predictions, the corresponding
<a href="https://www.uniprot.org/help/entry_name"
target="_blank">UniProt entry name</a>
  <li>otherwise, sequence GI number
<!-- clicking it opens the web page for
that sequence in the <a href="https://www.ncbi.nlm.nih.gov/protein"
target="_blank">NCBI Protein</a> database -->
  </ul>
<li><b>E-Value</b> &ndash; significance of the hit; BLAST
<a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=FAQ#expect" target="_blank">Expect value</a>
<li><b>Score</b> &ndash; alignment score
<li><b>Title</b> &ndash; a brief description of the sequence
<li><b>Species</b> &ndash; source organism
<!--
<li><b>Description</b> 
(or for AlphaFold, separate <b>Title</b> and <b>Species</b> columns)
&ndash; sequence title and source species
<li><b>URL</b> &ndash; URL used by ChimeraX for the <b>Name</b> click action
-->
</ul>
<p>
Additional columns for AlphaFold entries:
</p>
<ul>
<li><b>Gene</b> &ndash; 
<a href="https://www.uniprot.org/help/gene%5Fname" target="_blank">gene name</a>
<li><b>Protein Existence</b> &ndash; 
type of evidence that the protein exists
<li><b>Sequence Version</b> &ndash; 
<a href="https://www.uniprot.org/help/entry%5Fhistory" 
target="_blank">version number</a>
<li><b>Taxonomic ID</b> 
&ndash; <a href="https://www.uniprot.org/help/taxonomic%5Fidentifier" 
target="_blank">NCBI taxid</a> of the source organism
</ul>
<p>
Additional columns for PDB entries (from searching PDB or NR):
</p>
<ul>
<li><b>Date</b> &ndash; structure deposition date
<!-- <li><b>Title</b> &ndash; structure title -->
<li><b>Method</b> &ndash; method of structure determination
<li><b>Resolution</b> &ndash; crystallographic resolution
<li><b>Authors</b> &ndash; structure authors
<li><b>PubMed</b> &ndash; <a href="https://www.ncbi.nlm.nih.gov/pubmed/"
target="_blank">PubMed</a> identifier of literature reference, if any
<li><b>Total Atoms</b>
&ndash; total number of atoms in the structure (all chains)
<li><b>Total Residues</b>
&ndash; total number of residues in the structure (all chains)
<li><b>Chain Names</b>
&ndash; chain identifiers and descriptions of polymer chains in the structure
<li><b>Copies</b> &ndash; number of copies of the hit chain in the structure
<li><b>Polymers</b> &ndash; number of different polymer chains in the structure
(not counting multiple copies of the same sequence)
<li><b>Residues</b> &ndash; number of residues in the hit chain
<!-- <li><b>Species</b> &ndash; scientific name of source organism -->
<li><b>UniProt</b>
&ndash; <a href="https://www.uniprot.org/" target="_blank">UniProt</a>
identifier, if any, for hit chain
<li><b>Weight</b> &ndash; molecular weight of hit chain
<li><b>Ligand Formulas</b> &ndash; chemical formulae of ligand chemical components
<li><b>Ligand Names</b> &ndash; names of ligand chemical components
<li><b>Ligand Smiles</b> &ndash;
<a href="https://www.daylight.com/dayhtml/doc/theory/theory.smiles.html"
target="_blank"><b>SMILES</b></a> strings of ligand chemical components
<li><b>Ligand Symbols</b> &ndash; residue names of ligand chemical components
<li><b>Ligand Weights</b>
&ndash; molecular weights of ligand chemical components
</ul>

<a name="notes"></a>
<p class="nav">
[<a href="#top">back to top: Blast Protein</a>]
</p>
<h3>Notes</h3>
<p>
<b>Pseudo-multiple alignment</b>.
The pseudo-multiple alignment from BLAST is not a true multiple alignment,
but a consolidation of the pairwise alignments of individual hits to the query.
This output corresponds to the BLAST formatting option
(<a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=BlastHelp#alignment_view" target="_blank">alignment view</a>)
&ldquo;flat query-anchored with letters for identities.&rdquo;
</p><p>
<b>Basic Local Alignment Search Tool (BLAST)</b>.
The <a href="https://blast.ncbi.nlm.nih.gov/" target="_blank">BLAST</a>
software is provided by the
<a href="https://www.ncbi.nlm.nih.gov/" target="_blank">NCBI</a> and described in:
</p>
<blockquote>
<a href="https://nar.oxfordjournals.org/content/25/17/3389.full"
target="_blank">Gapped BLAST and PSI-BLAST:
a new generation of protein database search programs.</a>
Altschul SF, Madden TL, Sch&auml;ffer AA, Zhang J, Zhang Z, Miller W, Lipman DJ.
<i>Nucleic Acids Res.</i> 1997 Sep 1;25(17):3389-402.
<br><br>
<a href="https://www.ncbi.nlm.nih.gov/pubmed/2231712" target="_blank">
Basic local alignment search tool.</a>
Altschul SF, Gish W, Miller W, Myers EW, Lipman DJ.
<i>J Mol Biol.</i> 1990 Oct 5;215(3):403-10.
</blockquote>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics /
October 2021</address>
</body></html>
