<html>

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Tool: ESMFold</title>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRight" title="User Guide Index"/></a>

<h3><a href="../index.html#tools">Tool</a>: ESMFold</h3>
<p>
<a href="https://github.com/facebookresearch/esm" target="_blank">ESMFold</a>
(Evolutionary Scale Modeling)
is an artificial intelligence method for predicting protein structures.
The method is described in:
</p>
<blockquote>
<a href="https://pubmed.ncbi.nlm.nih.gov/36927031/" target="_blank">
Evolutionary-scale prediction of atomic-level protein structure with a language model.</a>
Lin Z, Akin H, Rao R, Hie B, Zhu Z, Lu W, Smetanin N, Verkuil R, Kabeli O, Shmueli Y, Dos Santos Costa A, Fazel-Zarandi M, Sercu T, Candido S, Rives A.
<i>Science.</i> 2023 Mar 17;379(6637):1123-1130.
</blockquote>
<p>
The ChimeraX <b>ESMFold</b> tool:
</p>
<ul>
<li><a href="#match">finds and retrieves existing models</a>
from the <a href="https://esmatlas.com/about#about_this_website"
target="_blank">ESM Metagenomic Atlas</a>, which contains
over 600 million predicted protein structures
<br><br>
<li><a href="#predict">runs new ESMFold predictions</a>
on the <a href="https://esmatlas.com/about#fold" target="_blank">prediction
server</a> provided by the
<a href="https://esmatlas.com/about#about_this_website"
target="_blank">ESM Metagenomic Atlas</a>
</ul>
<p>
ESMFold-predicted structures vary in confidence levels
(see <a href="#colorConfidence">coloring</a>)
and should be interpreted with caution. 
The related tool <a href="#pae"><b>ESMFold Error Plot</b></a>
plots residue-residue alignment errors for ESMFold structures.
The <b>ESMFold</b> tool is also implemented as the
<a href="../commands/esmfold.html"><b>esmfold</b></a> command.
See the
<a href="https://www.rbvi.ucsf.edu/chimerax/data/esmfold-nov2022/esmfold.html" target="_blank">ChimeraX ESMFold example</a>.
See also:
<a href="alphafold.html"><b>AlphaFold</b></a>,
<a href="blastprotein.html"><b>Blast Protein</b></a>,
<a href="modeller.html"><b>Modeller Comparative</b></a>,
<a href="modelloops.html"><b>Model Loops</b></a>,
<a href="rotamers.html"><b>Rotamers</b></a>
</p>

<blockquote>
<a href="#dialog"><b>ESMFold Dialog</b></a>
<br><a href="#coloring"><b>ESMFold Coloring Dialog</b></a>
<br><a href="#pae"><b>ESM Fold Error Plot</b></a>
</blockquote>

<a name="dialog"></a>
<p class="nav">
[<a href="#top">back to top: ESMFold</a>]
</p>
<h3>ESMFold Dialog</h3>
<p>
The <b>ESMFold</b> tool can be opened from the <b>Structure Prediction</b>
section of the <b>Tools</b> menu and manipulated like other panels
(<a href="../window.html#panels">more...</a>).
</p><p>
The <b>Sequence</b> can be specified by
<a href="https://www.uniprot.org/" target="_blank">UniProt</a>
name or accession number, pasted in as plain text, or chosen
from the menu of currently open protein structure chains.
</p>
<blockquote>
<a href="#match"><b>Fetch</b></a>
<br><a href="#search"><b>Search</b></a> 
<br><a href="#predict"><b>Predict</b></a>
<br><a href="#options"><b>Options</b></a>
</blockquote>
<p>
<a href="#dialog" class="nounder">&larr;</a>
<a name="match"></a>
<b>Fetch</b> gets the most sequence-similar model available from the
<a href="https://esmatlas.com/about#about_this_website"
target="_blank">ESM Metagenomic Atlas</a> for each specified chain. 
Specifying a whole model specifies all of its protein chains.
For each chain, a model is obtained for the single top hit identified by
<a href="https://www.rbvi.ucsf.edu/chimerax/data/kmer-aug2022/kmer_search.html"
target="_blank">K-mer search</a> of the
<a href="https://esmatlas.com/about#about_this_website"
target="_blank">ESM Metagenomic Atlas</a>.
The corresponding command is
<a href="../commands/esmfold.html#match"><b>esmfold match</b></a>.
If the sequence was specified by structure chain, then:
</p>
<ol>
<li>the chain ID of the predicted structure is made the same as the
corresponding chain of the existing model
<li>the predicted structure is superimposed onto the existing chain using
<a href="../commands/matchmaker.html"><b>matchmaker</b></a>,
and the following are reported in a table
in the <a href="log.html"><b>Log</b></a>:
  <ul>
  <li><b>Chain</b> &ndash; chain ID in ChimeraX
  <li><b>MGnify Id</b></a> &ndash; sequence ID used by the
<a href="https://esmatlas.com/about#about_this_website"
target="_blank">ESM Metagenomic Atlas</a>
  <li><b>RMSD</b> &ndash; C&alpha; root-mean-square deviation between
  the predicted and experimental structures, over all residues of the latter
  <li><b>Length</b> &ndash; number of residues in the predicted structure
  <li><b>Seen</b> &ndash; number of residues with atomic coordinates
  in the experimental structure
  <li><b>% Id</b> &ndash; percent identity in the sequence alignment generated
  by <a href="matchmaker.html"><b>matchmaker</b></a> for superposition;
  the number of positions with identical residues
  divided by the length of the shorter sequence
  </ul>
<li>by <a href="#options">default</a>,
the predicted structure is trimmed to the same residue range as the 
existing chain (<a href="../commands/alphafold.html#trim">details...</a>)
<li>the following <a href="../attributes.html">attributes</a>
 are assigned to the residues of the predicted structure:
  <ul>
  <li><b>c_alpha_distance</b>
&ndash; C&alpha; distance between corresponding positions of the predicted 
and existing chains after their superposition (step 2 above) 
  <li><b>missing_structure</b>
&ndash; positions missing from the coordinates of the existing chain
  <li><b>same_sequence</b>
&ndash; positions with different residue types than the existing chain
  </ul>
These <a href="../attributes.html">attributes</a>
can be used for <a href="#coloring">coloring</a> and other purposes.
</ol>
<p>
The fetched models are stored locally in 
<b>~/Downloads/ChimeraX/ESMFold/</b>,
where ~ indicates a user's home directory. 
If a file specified for opening is not found in this local cache,
the file will be fetched and cached. 
</p>
<p>
<a href="#dialog" class="nounder">&larr;</a>
<a name="search"></a>
<b>Search</b>
uses a <a href="../tools/blastprotein.html#notes">BLAST</a> web service
hosted by the <a href="https://www.cgl.ucsf.edu/Overview/index.html" 
target="_blank">UCSF RBVI</a> to search the 
<a href="https://esmatlas.com/about#about_this_website"
target="_blank">ESM Metagenomic Atlas</a>
using default parameters: <b>BLOSUM62</b>
amino acid similarity matrix for scoring the hits,
similarity score cutoff e-value <b>1e-3</b>,
returning a maximum of <b>100</b> of unique sequences.
However, different values of these parameters can be specified using
the corresponding command,
<a href="../commands/esmfold.html#search"><b>esmfold search</b></a>.
<b>Search</b> differs from <b>Fetch</b> in that it uses BLAST instead
of fast (but low-sensitivity)
<a href="https://www.rbvi.ucsf.edu/chimerax/data/kmer-aug2022/kmer_search.html"
target="_blank">K-mer searching</a>,
accepts only a single chain or sequence as input,
and returns a list of hits for the user to inspect,
rather than fetching the single top hit per chain automatically.
When results are returned, the hits are listed in a
<a href="blastprotein.html#results"><b>Blast Protein</b></a> window.
Double-clicking a hit uses
<a href="../commands/esmfold.html#fetch"><b>esmfold fetch</b></a> 
to retrieve the model,
or multiple chosen hits can be retrieved at once by using the results panel
<a href="../window.html#context-menu">context menu</a>
or <b>Load Structures</b> button
(<a href="blastprotein.html#results">details...</a>).
</p><p>
<a href="#dialog" class="nounder">&larr;</a>
<a name="predict"></a>
<b>Predict</b> runs a calculation
on the <a href="https://esmatlas.com/about#fold" target="_blank">prediction
server</a> provided by the
from the <a href="https://esmatlas.com/about#about_this_website"
target="_blank">ESM Metagenomic Atlas</a>.
The corresponding command is
<a href="../commands/esmfold.html#predict"><b>esmfold predict</b></a>.
</p><p>
<a href="#dialog" class="nounder">&larr;</a>
<a name="options"></a>
The <b>Options</b> button shows/hides additional options:
</p>
<ul>
<li><b>Results directory</b> 
(default <b>~/Downloads/ChimeraX/ESMFold/</b>)
&ndash; the pathname (name and location) of a folder or directory
in which to store prediction results. Clicking <b>Browse</b>
brings up a file browser window for choosing it interactively.
The directory does not need to exist already, as it will be
created by running the prediction.
<li><b>Trim fetched structure to the aligned structure sequence</b>
(default <b>on</b>)
&ndash; whether to trim a fetched structure to the same residue range 
as the structure to which it is aligned 
(<a href="../commands/esmfold.html#trim">details...</a>)
</ul>
<p>
The model for a sequence that was specified by structure chain will
be superimposed on that chain and assigned structure-comparison attributes 
for further analysis
(<a href="../commands/esmfold.html#structmatch">details...</a>).
</p><p>
<a name="caveats"></a>
Please note the following caveats of running a prediction:
</p>
<ul>
<li>ESMFold is faster but often <b>less accurate</b> than AlphaFold.
<li>The prediction server has a <b>maximum sequence length</b> of 400 residues.
(The atlas contains longer predictions, up to 1024 residues.)
<li><b>No multimer prediction</b>.
Only single-chain structures (not multimers) are predicted,
although structures of the individual protein chains in a complex
can be predicted separately.
<li><b>PAE not available from server</b>.
Although ESMFold computes <a href="#pae">predicted aligned error (PAE)</a>,
the prediction server does not provide it. The PAE is available for
entries <a href="#fetch">fetched</a> from the atlas, however.
<li>The server developers request that users run
<b>only one prediction at a time</b> due to capacity limitations.
<li>The server <b>may time out</b> during a prediction.
</ul>
<p>
<b>Coloring</b> 
shows the <a href="#coloring"><b>ESMfold Coloring</b> dialog</a>
for applying different color schemes to the predicted structures,
as well as hiding, showing, and 
<a href="../selection.html">selecting</b></a> their residues based on 
<a href="../attributes.html">attribute</a> value.
</p><p>
<b>Error plot</b> draws the <a href="#pae">ESMFold Error Plot</a>,
in which color gradations show (for each pairwise combination of residues)
the expected error in position of one residue when the
true and predicted structures are aligned based on the other residue.
</p>

<a name="coloring"></a>
<p class="nav">
[<a href="#top">back to top: ESMFold</a>]
</p>
<h3>ESMFold Coloring Dialog</h3>
<p>
Clicking the <b>Coloring</b> button on the main
<a href="#dialog"><b>ESMFold</b> tool</a> shows the
<b>ESMFold Coloring</b> dialog for applying different color schemes
to the predicted structures, as well as hiding, showing, and 
<a href="../selection.html">selecting</b></a> their residues based on 
<a href="../attributes.html">attribute</a> value.
</p><p>
<a name="confidence"></a>
When first opened, ESMFold-predicted structures are automatically colored by 
the pLDDT confidence measure
(same as <a href="https://alphafold.ebi.ac.uk/faq#faq-5"
target="_blank">for AlphaFold</a> except mapped to 0-1 instead of 0-100)
in the B-factor field:
</p>
   <ul>
     <li>1.0
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=blue width="10px" title="blue"></td></tr>
</table>
to 0.9
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=cornflowerblue width="10px" title="cornflower blue"></td></tr>
</table>
&ndash; high accuracy expected
     <li>0.9
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=cornflowerblue width="10px" title="cornflower blue"></td></tr>
</table>
to 0.7
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=yellow width="10px" title="yellow"></td></tr>
</table>
&ndash; backbone expected to be modeled well
     <li>0.7
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=yellow width="10px" title="yellow"></td></tr>
</table>
to 0.5
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=orange width="10px" title="orange"></td></tr>
</table>
&ndash; low confidence, caution
     <li>0.5
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=orange width="10px" title="orange"></td></tr>
</table>
to 0.0
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle">
<tr><td bgcolor=red width="10px" title="red"></td></tr>
</table>
&ndash; should not be interpreted, may be disordered
     </ul>
     </ul>
<p>
...in other words, using
</p>
<blockquote><b>
<a href="color.html#bfactor">color bfactor</a>
<a href="color.html#palette-options">palette</a> esmfold
<!-- 0,red:0.5,orange:0.7,yellow:0.9,cornflowerblue:1,blue -->
<!-- 1,blue:0.9,cornflowerblue:0.7,yellow:0.5,orange:0,red -->
</b></blockquote
<p>
The <a href="colorkey.html"><b>Color Key</b></a> graphical interface
or a command can be used to draw a corresponding color key, for example:
</p>
  <blockquote>
  <a href="../commands/key.html"><b>key</b></a> <b>red:low orange: yellow: cornflowerblue: blue:high</b>
  &nbsp;[<a href="../commands/key.html#options"><i>other-key-options</i></a>]
<!--
<br>
  <b>key</b> <b>blue:1 cornflowerblue:0.9 yellow:0.7 orange:0.5 red:0</b>
  &nbsp;<b>numericLabelSpacing equal</b>
  &nbsp;[<a href="key.html#options"><i>other-key-options</i></a>]
-->
  </blockquote>
<p>
In the <b>ESMFold Coloring</b> dialog,
the <b>Residues</b> to act on are specified by using the menus to choose an
ESMFold-predicted model and one of the following:
</p>
  <ul>
  <li><b>all</b>
&ndash; all residues
  <li><b>confidence below</b> [<i>N</i>]
&ndash; based on the <b>bfactor</b>
<a href="../attributes.html#atom">atom attribute</a>
(the <a href="#confidence">confidence value</a> is read from the
B-factor field of the PDB file)
  <li><b>C-alpha distance greater than</b> [<i>d</i>]
&ndash; based on the <b>c_alpha_distance</b>
residue attribute of ESMFold models 
<a href="#match">fetched by existing structure chain</a>;
C&alpha; distance between corresponding positions of the predicted and
existing chains after their automatic superposition
  <li><b>missing structure</b>
&ndash; based on the <b>missing_structure</b> 
residue attribute of ESMFold models 
<a href="#match">fetched by existing structure chain</a>;
positions missing from the coordinates of the existing chain
  <li><b>different sequence</b>
&ndash; based on the <b>same_sequence</b> 
residue attribute of ESMFold models 
<a href="#match">fetched by existing structure chain</a>;
positions with different residue types than the existing chain
  <li><b>confidence above</b> [<i>N</i>]
&ndash; based on the <b>bfactor</b>
<a href="../attributes.html#atom">atom attribute</a>
(the <a href="#confidence">confidence value</a> is read from the
B-factor field of the PDB file)
  <li><b>C-alpha distance less than</b> [<i>d</i>]
&ndash; based on the <b>c_alpha_distance</b>
residue attribute of ESMFold models 
<a href="#match">fetched by existing structure chain</a>;
C&alpha; distance between corresponding positions of the predicted and
existing chains after their automatic superposition
  <li><b>paired structure</b>
&ndash; based on the <b>missing_structure</b> 
residue attribute of ESMFold models 
<a href="#match">fetched by existing structure chain</a>;
positions present in the coordinates of the existing structure chain
  <li><b>same sequence</b>
&ndash; based on the <b>same_sequence</b> 
residue attribute of ESMFold models 
<a href="#match">fetched by existing structure chain</a>;
positions with identical residue types as the existing chain
  </ul>
<p>
Buttons act on the designated residues:
</p>
  <ul>
  <li><b>Color</b> buttons:
    <ul>
    <li><b>Custom</b> for choosing a color interactively
    using the system color picker
    <li>a series of square buttons for specific colors:
<table cellpadding="8px" cellspacing="5x"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor=lightgray width="1px" title="lightgray"></td>
<td bgcolor=red width="1px" title="red"></td>
<td bgcolor=magenta width="1px" title="magenta"></td>
<td bgcolor=yellow width="1px" title="yellow"></td>
<td bgcolor=lime width="1px" title="lime"></td>
<td bgcolor=cyan width="1px" title="cyan"></td>
</tr></table>
    </ul>
  <li><b>Hide</b> &ndash; hide the specified residues
  <li><b>Show</b> &ndash; show the specified residues
  <li><b>Select</b> &ndash; <a href="../selection.html">select</a>
the specified residues
  </ul>
<p>
The <b>ESMFold Coloring</b> dialog does not color continuously 
along a gradient to show the attribute values. 
For coloring along a gradient, see 
<a href="render.html"><b>Render by Attribute</b></a> and/or the commands
<a href="../commands/color.html#bfactor"><b>color bfactor</b></a>
(for the <a href="#confidence">confidence value</a>, which is read from the
B-factor field of the PDB file) and
<a href="../commands/color.html#byattribute"><b>color byattribute</b></a>
(for other numerical <a href="../attributes.html">attributes</a>).
</p>

<a name="pae"></a>
<p class="nav">
[<a href="#top">back to top: ESMFold</a>]
</p>
<h3>ESMFold Error Plot</h3>
<p>
Besides the per-residue
<a href="#confidence">pLDDT confidence measure</a>, ESMFold gives
for each pair of residues (X,Y) the expected position error
at residue X if the predicted and true structures were aligned on residue Y.
These residue-residue &ldquo;predicted aligned error&rdquo; or PAE values
are not provided by the <a href="#predict">prediction server</a>,
but are available for structures already in the
<a href="https://esmatlas.com/about#about_this_website"
target="_blank">ESM Metagenomic Atlas</a> and
can be shown with <b>ESMFold Error Plot</b>, which can be
opened from the <b>Structure Prediction</b>
section of the <b>Tools</b> menu and manipulated like other panels
(<a href="../window.html#panels">more...</a>).
See also:
<a href="../commands/esmfold.html#contacts"><b>esmfold contacts</b></a>
</p><p>
Choosing the corresponding ESMFold structure from the menu of open 
atomic models associates it with the plot. This association allows
coloring the structure as described below, and for selections
on the plot to highlight the corresponding parts of the structure.
</p><p>
The PAE values can be either:
</p>
<ul>
<li>fetched from the
<a href="https://esmatlas.com/about#about_this_website"
target="_blank">ESM Metagenomic Atlas</a>
by giving the MGnify identifier of the sequence (<i>e.g.</i>, as evident from
the name of the corresponding previously <a href="#match">fetched</a>
ESMfold structure model)
<li>a json file from ESMFold
(<i>e.g.</i>, obtained by URL such as
<a href="https://api.esmatlas.com/fetchConfidencePrediction/MGYP002537940442"
target="_blank">https://api.esmatlas.com/fetchConfidencePrediction/MGYP002537940442</a>)
</ul>
<p>
The PAE plot can also be shown by clicking the <b>Error plot</b> button
on the <a href="#dialog"><b>ESMFold</b> dialog</a> or by using 
the command <a href="../commands/esmfold.html#pae"><b>esmfold pae</b></a>, 
or the command 
<a href="../commands/esmfold.html#fetch"><b>esmfold fetch</b></a>
or <a href="../commands/esmfold.html#fetch"><b>esmfold match</b></a> 
with the option <b>pae true</b>.
</p><p>
When the mouse cursor is over the plot, the residue pair and PAE value at 
its current position are reported in the bottom right corner of the window.
</p><p>
Clicking <b>Color PAE Domains</b> clusters the residues into
coherent domains (sets of residues with relatively low PAE values)
and uses randomly chosen colors to distinguish these domains in the structure 
(<a href="../commands/esmfold.html#colorDomains">details...</a>).
Clicking <b>Color pLDDT</b> returns the structure to
the default <a href="#confidence">confidence coloring</a>.
</p><p>
<a name="context"></a>
The plot's <a href="../window.html#context-menu">context menu</a> includes:
</p>
<ul>
<li><b>Dragging box colors structure</b> (initial default checked <b>on</b>) 
&ndash; whether dragging a box on the plot highlights the corresponding parts
of the 3D structure with bright colors and makes everything else gray; 
if this option is unchecked, highlighting will be done with
<a href="../selection.html">selection</a> instead of coloring
<li><b>Color plot from structure</b>
&ndash; color the plot to match the 3D structure where the pair of residues
represented by an X,Y point have the same ribbon color; show the rest of the
plot in shades of gray
<li><b>Color plot rainbow</b>
&ndash; use the <b>pae</b>
<a href="../commands/color.html#palette-options">palette</a> (default)
to color the plot, with colors assigned to values as follows:
<table border cellpadding="0" cellspacing="0"
style="display:inline-block; vertical-align:top"><tr>
<td bgcolor="blue" width="30px" title="blue"
align="center"><font size="-1" color="white">0</font></td>
<td bgcolor="cornflowerblue" width="30px" title="cornflowerblue"
align="center"><font size="-1" color="white">5</font></td>
<td bgcolor="yellow" width="30px" title="yellow"
align="center"><font size="-1">10</font></td>
<td bgcolor="orange" width="30px" title="orange"
align="center"><font size="-1">15</font></td>
<td bgcolor="gray" width="30px" title="gray"
align="center"><font size="-1" color="white">20</font></td>
<td bgcolor="lightgray" width="30px" title="lightgray"
align="center"><font size="-1">25</font></td>
<td bgcolor="white" width="30px" title="white"
align="center"><font size="-1">30</font></td>
</tr></table>
<li><b>Color plot green</b>
&ndash; use the <b>paegreen</b>
<a href="../commands/color.html#palette-options">palette</a> to color the plot:
<table border cellpadding="0" cellspacing="0"
style="display:inline-block; vertical-align:top"><tr>
<td bgcolor="#1e3546661e35" width="30px" title="#1e3546661e35"
align="center"><font size="-1" color="white">0</font></td>
<td bgcolor="#245a922c245a" width="30px" title="#245a922c245a"
align="center"><font size="-1" color="white">5</font></td>
<td bgcolor="#374cb168374c" width="30px" title="#374cb168374c"
align="center"><font size="-1" color="white">10</font></td>
<td bgcolor="#5687c9ba5687" width="30px" title="#5687c9ba5687"
align="center"><font size="-1">15</font></td>
<td bgcolor="#828fddf3828f" width="30px" title="#828fddf3828f"
align="center"><font size="-1">20</font></td>
<td bgcolor="#bae1efdebae1" width="30px" title="#bae1efdebae1"
align="center"><font size="-1">25</font></td>
<td bgcolor=white width="30px" title="white"
align="center"><font size="-1">30</font></td>
</tr></table>
<li><b>Show chain divider lines</b> (initial default checked <b>on</b>)
&ndash; for multimer predictions, draw lines on the plot demarcating the
end of one chain and the start of another; the lines may obscure a few 
chain-terminal residues in the plot, and can be hidden if this is problematic
<li><b>Save image</b> &ndash; save the plot as a PNG file 
</ul>
<p>
The <a href="colorkey.html"><b>Color Key</b></a> graphical interface
or a command can be used to draw (in the main graphics window)
a color key for the PAE plot. For example, to make a color key that
matches the <b>pae</b> or <b>paegreen</b> scheme, respectively:
</p>
  <blockquote>
  <a href="../commands/key.html"><b>key</b></a>&nbsp;<b>pae&nbsp;:0&nbsp;:&nbsp;:&nbsp;:15&nbsp;:&nbsp;:&nbsp;:30</b>
  &nbsp;<a href="../commands/key.html#options"><b>showTool true</b></a>
<br>
  <a href="../commands/key.html"><b>key</b></a>&nbsp;<b>paegreen&nbsp;:0&nbsp;:&nbsp;:&nbsp;:15&nbsp;:&nbsp;:&nbsp;:30</b>
  &nbsp;<a href="../commands/key.html#options"><b>showTool true</b></a>
  </blockquote>
<p>
A title for the color key (<i>e.g.</i>, &ldquo;<b>Predicted Aligned
Error (&Aring;)</b>&rdquo;) would need to be created separately with
<a href="../commands/2dlabels.html"><b>2dlabels</b></a>.
</p>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics / 
March 2023</address>
</body></html>
