<html lang="en">

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Tool: Foldseek (Similar Structures)</title>
<style>@media (prefers-color-scheme: dark) { :root { color-scheme: dark; } }</style>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRighticon" title="User Guide Index"/></a>

<h3><a href="../index.html#tools">Tool</a>: Foldseek (Similar Structures)</h3>
<p>
The <b>Foldseek</b> tool (also called <b>Similar Structures</b>)
searches the <a href="https://www.rcsb.org" target="_blank">PDB</a> or
<a href="https://alphafold.ebi.ac.uk/" target="_blank">AlphaFold Database</a>
for structures similar to a protein chain already open in ChimeraX.
The tool facilitates exploring up to hundreds of protein single-chain 
structures by efficiently showing them in 3D as backbone traces, 
potentially with ligands, and in 2D as sequence alignments or 
reduced-dimensionality scatter plots based on backbone conformation.
</p>
<ul>
<li>
The <a href="#notes"><b>Foldseek</b></a> search method
finds similar 3D structures (regardless of sequence similarity)
using extremely fast approaches that were developed for sequence searches.
It does so by describing the 3D interactions along a chain of amino acids
as a linear sequence of characters. 
<br><br>
<li>
Alternatively, <a href="#notes"><b>MMseqs2</b></a> (very fast) or
<a href="blastprotein.html#notes"><b>BLAST</b></a> can be used to 
search for protein structures by sequence similarity. 
</ul>
<p>
The tool can be started from the
<b>Structure Analysis</b> section of the <b>Tools</b> menu
and manipulated like other panels
(<a href="../window.html#panels">more...</a>).
It is also implemented as the commands
<a href="../commands/foldseek.html"><b>foldseek</b></a>,
<a href="../commands/similarstructures.html"><b>similarstructures</b></a>, and
<a href="../commands/sequence.html#search"><b>sequence search</b></a>.
See also:
<a href="alphafold.html"><b>AlphaFold</b></a>,
<a href="esmfold.html"><b>ESMFold</b></a>,
<a href="blastprotein.html"><b>Blast Protein</b></a>
</p>
<blockquote><b>
<a href="#input">Search Setup</a>
<br>
<a href="#results">Similar Structures List</a>
<br>
<a href="#options">Options</a>
<br>
<a href="#sequences">Sequence Plot and Residue Attributes</a>
<br>
<a href="#traces">Traces</a>
<br>
<a href="#clusters">Cluster Plot</a>
<br>
<a href="#ligands">Ligands</a>
<br>
<a href="#notes">References</a>
</b></blockquote>

<a name="input"></a>
<p class="nav">
[<a href="#top">back to top: Foldseek</a>]
</p>
<h3>Search Setup</h3>
<p>
The query should be chosen from the pulldown menu
of protein chains in structures already open in ChimeraX.
Choices of <b>database</b> to search:
</p>
<ul>
<li><b>PDB</b> &ndash;
<a href="https://www.rcsb.org/" target="_blank">Protein Data Bank</a>
<li><b>Alphafold DB</b> &ndash; <a href="https://alphafold.ebi.ac.uk/"
target="_blank">AlphaFold Database</a>
</ul>
<p>
Choices of search method:
</p>
<ul>
<li><a href="#notes"><b>Foldseek</b></a> &ndash; fast 3D structure search
using the <a href="https://search.foldseek.com/search" 
target="_blank">Foldseek Search Service</a> provided by the
<a href="https://www.mpinat.mpg.de/soeding" target="_blank">S&ouml;ding</a>
and <a href="https://steineggerlab.com/en/" target="_blank">Steinegger</a>
groups, with maximum number of hits 1000.
If this method is used, <b>PDB</b> refers to the
&ldquo;pdb100&rdquo; redundancy-filtered version of the PDB 
created with Foldseek (one chain per cluster with 100% sequence identity 
and &ge;95% sequence overlap, reducing ~1 million chains to 340,000)
<!--
see https://www.ebi.ac.uk/training/online/courses/alphafold/glossary-of-terms/
-->
and <b>Alphafold DB</b> to the &ldquo;afdb50&rdquo; UniProt50 subset of the 
database version 4.
<!--
UniRef50??  https://www.uniprot.org/help/uniref
-->
<li><a href="#notes"><b>MMseqs2</b></a> &ndash; very fast sequence search
using the <a href="https://search.rcsb.org" target="_blank">RCSB web service</a>
<li><a href="blastprotein.html#notes"><b>BLAST</b></a> 
&ndash; sequence search using the BLAST web service hosted by the
<a href="https://www.cgl.ucsf.edu/Overview/index.html" 
target="_blank">UCSF RBVI</a>
</ul>
<p>
Clicking <b>Search</b> sends the input parameters and structure to the
web service. When results are returned, a <a href="#results">table 
of similar structures</a> is shown in the tool window.
</p>

<a name="results"></a>
<p class="nav">
[<a href="#top">back to top: Foldseek</a>]
</p>
<h3>Similar Structures List</h3>
<p>
Searching with the <b>Foldseek</b> (<b>Similar Structures</b>) tool
or the commands
<a href="../commands/foldseek.html"><b>foldseek</b></a>,
<a href="../commands/sequence.html#search"><b>sequence search</b></a>, and
<a href="../commands/similarstructures.html#blast"><b>similarstructures
blast</b></a> shows a table or list of hits in the tool window.
Because this list is relatively large, the ChimeraX graphics and/or
overall window may be resized; to avoid this, the tool can be 
<a href="../window.html#panels">undocked</a> from the main window
beforehand. See also the <b>Tool windows start undocked</b> setting in the
<a href="../preferences.html#window"><b>Window</b> preferences</a>.
</p><p>
Columns in the <b>Similar Structures</b> table:
</p>
<ul>
<li><b>PDB</b> &ndash; <i>PDB-identifier</i>_<i>chain-identifier</i>
<br>&ndash; <i>OR</i> &ndash; (depending on which database was searched)<br>
<li><b>AFDB</b>
&ndash; <a href="https://www.uniprot.org/help/accession_numbers" 
target="_blank">UniProt accession number</a>
<li><b>Identity</b> &ndash; % sequence identity compared to the query
<!-- across aligned positions? -->
<li><b>E-value</b> &ndash; significance value according to the search method
<li><b>% Close</b>
&ndash; percentage of residues in the <a href="#options">trimmed</a> hit
that are close to the paired residue of the query (within the 
<a href="#options"><b>Alignment pruning C-alpha atom distance</b></a>,
default <b>2.0</b> &Aring;)
<li><b>% Cover</b>
&ndash; percentage of query residues paired with hit residues
<li><b>Species</b> &ndash; source organism
<li><b>Description</b>
&ndash; description of the protein or (if <b>PDB</b>)
the overall structure containing the protein
</ul>
<p>
The <b>% Close</b> and <b>% Cover</b> values are only filled in automatically
by the <a href="#notes">Foldseek</a> search method, 
which uses and returns 3D coordinates.
For the other search methods (which are based on sequence only),
these columns can be filled in by using 
<a href="../commands/similarstructures.html#fetchcoords"><b>similarstructures 
fetchcoords</b></a> to get &alpha;-carbon coordinates for the
corresponding structures.
</p><p>
<a name="chosen"></a>
One or more hits can be <b><i>chosen</i></b> (highlighted) in the table
by clicking and dragging with the left mouse button;
<b>Ctrl</b>-click (or <b>command</b>-click if using a Mac)
toggles whether a row is chosen. 
</p><p>
<a name="buttons"></a>
Buttons across the bottom of the dialog:
</p>
<ul>
<li><b>Open</b>
&ndash; fetch the <a href="#chosen">chosen</a>
structures from the respective database
(<a href="https://www.rcsb.org/" target="_blank">Protein Data Bank</a>
or <a href="https://alphafold.ebi.ac.uk/"
target="_blank">AlphaFold Database</a>) and process them
according to the <a href="#options">options</a>. For each structure,
the hit chain is superimposed onto the query chain by least-squares fitting
the &alpha;-carbons of the paired residues and iteratively pruning 
far-apart pairs as described for the
<a href="../commands/align.html"><b>align</b></a> command.
<li><b>Sequences</b> 
&ndash; show an interactive heatmap of the sequence alignment of all hits
(<a href="#sequences">details...</a>)
<!-- 7/25 green is same residue type as query, red is same residue type
as query and >N% conserved in column  ... 50? 70? -->
<li><b>Traces</b> 
&ndash; show hit structures as thin tube backbones
(<a href="#traces">details...</a>)
<li><b>Clusters</b> 
&ndash; show an interactive 2D scatter plot of the hits based on 
their backbone conformations
(<a href="#clusters">details...</a>)
<li><b>Ligands</b>
&ndash; show any solvent, ligands, and ions from the hit structures
(<a href="#ligands">ligands...</a>)
<li><b>Options</b> 
&ndash; show <a href="#options">options</a> for handling structures and 
for setting whether buttons should apply to all hits or only the 
<a href="#chosen">chosen</a> hits
<li><b>Help</b> 
&ndash; show this page in the <a href="helpviewer.html"><b>Help Viewer</b></a>
</ul>
<p>
Choosing <b>Save CSV or TSV File...</b> from the tool's
<a href="../window.html#context-menu">context menu</a> opens
a separate dialog for exporting the data (all rows and columns) to a
comma-separated or tab-separated values file, retaining the current sort order.
The current contents of the tool are also saved in
<a href="../commands/save.html#session">ChimeraX sessions</a>.
</p><p>
<a name="sms"></a>
In addition, search results are saved in ~/Downloads/ChimeraX 
under subdirectories Foldseek, MMseqs2, or BLAST, with filenames 
based on the query name, the database searched, and the search method,
ending with the suffix .sms.
These <a href="../formats/sms.html">similar structures files</a>
use a JSON file format specific to ChimeraX and are listed in the 
<a href="filehistory.html"><b>File History</b></a> for easy access.
Simply opening an .sms file loads the set of results into the 
<b>Similar Structures</b> interface.
</p><p>
Doing another search or opening a file of previously saved results replaces the
contents of the <b>Similar Structures</b> table, since (currently) the tool 
only allows showing one set of results at a time.
Sets of results are assigned names such as fs1, fs2, mm1, mm2, bl1, and bl2 
that can be used in analysis commands even if the corresponding results 
are not shown. However, the only way to get a set of results that is open
but not shown in the table is to use the <b>showTable false</b> option of 
the search command.
The names of currently open sets can be listed with the command 
<a href="../commands/similarstructures.html#list"><b>similarstructures 
list</b></a>.
</p>

<a name="options"></a>
<p class="nav">
[<a href="#top">back to top: Foldseek</a>]
</p>
<h3>Options</h3>
<p>
Clicking <b>Options</b> shows/hides the following settings:
</p>
<ul>
<li><b>Trim</b> &ndash; delete any/all (default) of the following from
the retrieved structure:
  <ul>
  <li><b>extra chains</b> &ndash; for PDB entries, chains other than
the hit chain
  <li><b>sequence ends</b> &ndash; N- and C-terminal segments of the hit chain
that were not included in the sequence alignment returned by the search method
  <li><b>far ligands</b> &ndash; ligands, solvent, and ions &gt; 3 &Aring;
from the hit chain
  </ul>
<li><b>Alignment pruning C-alpha atom distance</b> (default <b>2.0</b> &Aring;)
&ndash; iterate the fit over the sequence-aligned pairs of CA atoms
so that only pairs within the specified distance are used in the final fit,
as described for the <a href="../commands/align.html"><b>align</b></a> command
<li><b>Traces, clusters and ligands for selected rows only</b>
&ndash; whether the <b>Traces</b>, <b>Clusters</b>, and <b>Ligands</b>
<a href="#buttons">buttons</a> should show data for the 
<a href="#chosen">chosen</a> hits only; otherwise,
show data for all hits regardless of which are <a href="#chosen">chosen</a>
</ul>

<a name="sequences"></a>
<p class="nav">
[<a href="#top">back to top: Foldseek</a>]
</p>
<h3>Sequence Plot and Residue Attributes</h3>
<table class="clRight">
<tr><td><a href="sequences.png"><img class="outline" src="sequences.png" 
alt="Foldseek sequence plot" width="400px"
title="click to enlarge..."></a></td></tr>
</table>
<p>
Clicking the <b>Sequences</b> button displays a high-level (without
amino acid codes) plot of the sequence alignment of all of the hits to 
the query. The plot gives an overview of which parts of the query sequence
are matched by the hits, and the depth of coverage.
</p><p>
Each row of the image is one sequence, so 200 hits would 
produce an image 200 pixels tall. The columns of the image correspond 
to the residues of the query structure. Initially, pixels in the plot are
colored as follows:
</p>
<ul class="none">
<li><font style="background:white;border:solid black 1px; 
padding:0px 1px">&nbsp;&nbsp;&nbsp;</font>
&ndash; no aligned residue
<li><font style="background:#E1BE6A;padding:0px 2px">&nbsp;&nbsp;&nbsp;</font>
&ndash; residue of the same amino acid type as the query
in a column where &ge;0.5 of the residues have that same type
(and the column contains at least 10 residues)
<li><font style="background:#40B0A6;padding:0px 2px">&nbsp;&nbsp;&nbsp;</font>
&ndash; residue of the same amino acid type as the query
but not meeting the column criteria above
<li><font style="background:black;padding:0px 2px">&nbsp;&nbsp;&nbsp;</font>
&ndash; residue of a different amino acid type than the query
</ul>
<p>
Different coloring can be applied with the
<a href="../commands/similarstructures.html#sequences"><b>similarstructures
sequences</b></a> command.
</p><p>
<a name="context"></a>
Hovering the mouse over the sequence plot shows pop-up labels to indicate
the underlying row (hit structure) and column (query residue number).
Left- or right-clicking the plot raises a <b><i>context menu</i></b>,
in which some entries reflect the row or column position of the click:
<!--
(This differs from other tools, in which
<a href="../window.html#context-menu">right- or Ctrl-click</a> is used.)
-->
<ul>
<li><b>Open structure [<i>hit</i>]</b>
&ndash; fetch the structure and superimpose it on the query
as described for the <a href="#buttons"><b>Open</b></a> button
<li><b>Show [<i>hit</i>] in table</b>
&ndash; scroll the table of results to the corresponding row
<li><b>Select query residue [<i>query residue</i>]</b>
&ndash; <a href="../selection">select</a> the corresponding residue in the query 
<li><b>Order sequences by:</b>
  <ul>
  <li><b>e-value</b> &ndash; lowest to highest E-value (default,
if clustering by coverage only gives one cluster, see below)
  <li><b>cluster</b> &ndash; grouping the sequences by which part of the 
query they cover (default, if clustering by coverage gives &gt;1 cluster)
  <li><b>identity</b>
&ndash; percent sequence identity compared to the query
  <li><a name="lddt"></a><b>mean LDDT</b>
&ndash; average <b><i>local distance difference test</i></b> (LDDT)
over all residues in a hit structure.
The LDDT indicates the similarity of a hit residue to the aligned
query residue in a neighborhood of 15 &Aring; from the query residue 
&alpha;-carbon
(see <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC3799472/" 
target="_blank">Mariani <i>et al.</i></a>, 
Bioinformatics <b>29</b>:2722 (2013)).
  </ul>
<li><b>Color conserved</b> (default)
&ndash; by conservation of the query residue type
as described <a href="#sequences">above</a>
<li><b>Color by LDDT</b> 
&ndash; by <a href="#lddt">LDDT</a> of each aligned residue in each structure:
<table border cellpadding="0" cellspacing="0"
style="display:inline-block; vertical-align:top"><tr>
<td bgcolor=red width="30px" title="red"
align="center"><font size="-1" color="white">0</font></td>
<td bgcolor=orange width="30px" title="orange"
align="center"><font size="-1">0.2</font></td>
<td bgcolor=yellow width="30px" title="yellow"
align="center"><font size="-1" color="black">0.4</font></td>
<td bgcolor=cornflowerblue width="30px" title="cornflowerblue"
align="center"><font size="-1" color="white">0.6</font></td>
<td bgcolor=blue width="30px" title="blue"
align="center"><font size="-1" color="white">0.8</font></td>
</tr></table>
<br>
(if both of the above <b>Color...</b> options are turned on, only the
positions that would otherwise be shown in black are colored by
<a href="#lddt">LDDT</a> instead)
<li><b>Color query structure by:</b>
  <ul>
  <li><b>coverage</b>
&ndash; number of residues (non-gap characters) in the aligned column 
(query residue <a href="../attributes.html">attribute</a> <b>coverage</b>):
<table border cellpadding="0" cellspacing="0"
style="display:inline-block; vertical-align:top"><tr>
<td bgcolor=red width="30px" title="red"
align="center"><font size="-1" color="white">0</font></td>
<td bgcolor=white width="30px" title="white"
align="center"><font size="-1">N/2</font></td>
<td bgcolor=blue width="30px" title="blue"
align="center"><font size="-1" color="white">N</font></td>
</tr></table>
where N is the number in the most highly populated column of the alignment
  <li><b>conservation</b>
&ndash; fraction of hits in the aligned column with the most prevalent 
residue type in that column, not necessarily the same residue type 
as in the query
(query residue <a href="../attributes.html">attribute</a> <b>conservation</b>):
<table border cellpadding="0" cellspacing="0"
style="display:inline-block; vertical-align:top"><tr>
<td bgcolor=blue width="30px" title="blue"
align="center"><font size="-1" color="white">0</font></td>
<td bgcolor=white width="30px" title="white"
align="center"><font size="-1">0.25</font></td>
<td bgcolor=red width="30px" title="red"
align="center"><font size="-1" color="white">0.5</font></td>
</tr></table>
  <li><b>highly conserved</b>
&ndash; red where 
<!-- query residue <a href="../attributes.html">attribute</a> -->
<b>conservation</b> &ge; 0.5, otherwise gray
  <li><b>local alignment</b>
&ndash; average <a href="#lddt">LDDT</a> at that position across all 
aligned structures
(query residue <a href="../attributes.html">attribute</a> <b>lddt</b>):
<!-- 0 red, 0.2 orange, 0.4 yellow, 0.6 cornflowerblue, 0.8,blue -->
<table border cellpadding="0" cellspacing="0"
style="display:inline-block; vertical-align:top"><tr>
<td bgcolor=red width="30px" title="red"
align="center"><font size="-1" color="white">0</font></td>
<td bgcolor=orange width="30px" title="orange"
align="center"><font size="-1">0.2</font></td>
<td bgcolor=yellow width="30px" title="yellow"
align="center"><font size="-1" color="black">0.4</font></td>
<td bgcolor=cornflowerblue width="30px" title="cornflowerblue"
align="center"><font size="-1" color="white">0.6</font></td>
<td bgcolor=blue width="30px" title="blue"
align="center"><font size="-1" color="white">0.8</font></td>
</tr></table>
  </ul>
<li><b>Save image</b> &ndash; save the sequence plot as a PNG image file
</ul>
<p>
The query residue <a href="../attributes.html">attributes</a>
<b>coverage</b>, <b>conservation</b>, and <b>lddt</b>
are assigned when the sequence plot is shown.
Different coloring schemes to show these 
<a href="../attributes.html">attributes</a>
can be applied to the query structure with
<a href="render.html"><b>Render by Attribute</b></a> or the command
<a href="../commands/color.html#byattribute"><b>color byattribute</b></a>.
</p>

<a name="traces"></a>
<p class="nav">
[<a href="#top">back to top: Foldseek</a>]
</p>
<h3>Traces</h3>
<p>
Clicking the <b>Traces</b> button displays hit structures as
&ldquo;licorice&rdquo; (spaghetti-like) ribbons superimposed on the query,
for either all hits or just the <a href="#chosen">chosen</a> ones,
as per the <a href="#options">options</a>.
These traces are meant to give an overview of the variability
of a large number of stuctures and their coverage of the query,
and <a href="../commands/lighting.html#preset">soft lighting</a>
is recommended to better reveal their shapes.
</p><p>
Only backbone &alpha;-carbons are included in the condensed structural
information returned by a <b>Foldseek</b> search, not secondary structure
information, so the ribbons do not vary in style to show helix and strand.
<b>MMseqs2</b> and <b>Blast</b> search results do not automatically
include &alpha;-carbon coordinates, but clicking the
<b>Traces</b> button will raise a dialog asking the user whether to
fetch them, since it may take several minutes to do so.
</p><p>
All of the hit structure &alpha;-carbons are loaded as a single atomic model,
one chain per structure, with chain ID set to the database ID of the structure.
The residue types of the hit are retained, but the residues are renumbered
according to the paired residues of the query structure.
</p><p>
The traces are initially displayed as follows:
</p>
<ol>
<li>the ribbon is broken into segments where 
two consecutive aligned &alpha;-carbons are &gt;5 &Aring; apart
<li>ribbons are shown for &ge;5 contiguous &alpha;-carbons within a segment 
and within 4 &Aring; of the corresponding query &alpha;-carbons
<li>ribbons are shown for entire segments in which every &alpha;-carbon
is within 10 &Aring; of its counterpart
</ol>
<p>
Different parameters can be applied with the
<a href="../commands/similarstructures.html#traces"><b>similarstructures
traces</b></a> command.
</p><p>
Ctrl-double-clicking a trace shows a 
<a href="../selection.html#context">selection context menu</a>
for the corresponding hit, with entries including:
</p>
<ul>
<li><b>Open similar structure [<i>hit</i>]</b>
&ndash; fetch the structure and superimpose it on the query
as described for the <a href="#buttons"><b>Open</b></a> button
<li><b>Show [<i>hit</i>] in similar structures table</b>
&ndash; scroll the table of results to the corresponding row 
<li><b>Show all traces</b> &ndash; show trace ribbons for all hits
<li><b>Show full traces</b> &ndash; show entire traces
regardless of the distance and length criteria for initial display
<li><b>Show only close traces</b> &ndash; go back to showing
only the parts that meet the distance and length criteria for initial display
<li><b>Show only trace [<i>hit</i>]</b>
</ul>
</p>
The trace ribbons can be shown/hidden or colored selectively with the 
menu above and the <a href="#clusters">cluster plot</a> context menu.
</p>

<a name="clusters"></a>
<p class="nav">
[<a href="#top">back to top: Foldseek</a>]
</p>
<h3>Cluster Plot</h3>
<table class="clRight">
<tr><td><img class="outline" src="clusters-1gvz-af.png" 
alt="Foldseek clusters plot" width="150px"></td></tr>
</table>
<p>
Clicking the <b>Clusters</b> button displays a scatter plot of the hits 
clustered by backbone conformation, for either all hits or just the
<a href="#chosen">chosen</a> ones, as per the
<a href="#options">options</a>. Each structure is represented by a circle
labeled with its name. Clicking the button generates the plot as follows:
</p>
<ol>
<a name="reference-atoms"></a>
<li>the five residues in the query most conserved in the sequence
alignment of hits are identified and their &alpha;-carbons used as the
<b><i>reference atoms</i></b>
<li>for each hit,
the &alpha;-carbon (x,y,z) coordinates of the corresponding five residues
are concatenated to give a vector of length 15; hit structures
without a residue in any of the five alignment columns are omitted
<li>the vector is projected to a point in two dimensions with UMAP
(<a href="https://umap-learn.readthedocs.io/en/latest/" 
target="_blank">Uniform Manifold Approximation and Projection</a>)
<li>the points in 2D are clustered by distance,
and the clusters are assigned random colors
</ol>
<p>
Different parameters such as a different number of reference residues
can be specified with the
<a href="../commands/similarstructures.html#cluster"><b>similarstructures 
cluster</b></a> command.
</p><p>
The plot can be zoomed by scrolling and translated with the middle mouse
button or trackpad equivalent. Clicking the plot raises a context menu.
Menu items acting on traces will generate them as needed (if not already
present) as described <a href="#traces">above</a>, and those referring to
a specific hit only appear when the click is on a circle:
</p>
<ul>
<li><b>Show traces for cluster [<i>hit</i>]</b>
&ndash; show traces for all hits in the same cluster as the clicked one
<li><b>Show only traces for cluster [<i>hit</i>]</b>
&ndash; show traces for all hits in the same cluster as the clicked one,
hide all other traces
<li><b>Hide traces for cluster [<i>hit</i>]</b>
&ndash; hide traces for all hits in the same cluster as the clicked one
<li><b>Show all traces</b>
<li><b>Show one trace per cluster</b>
&ndash; per cluster, show only the trace of structure closest to the
average for that structure (minimum RMSD to average &alpha;-carbon positions)
<li><b>Show traces not on plot</b>
&ndash; show traces for the hits that were omitted from the plot
due to not having a residue at one or more reference positions
<li><b>Hide traces not on plot</b>
&ndash; hide traces for the hits that were omitted from the plot
due to not having a residue at one or more reference positions
<li><b>Color traces to match plot</b>
<li><b>Change cluster [<i>hit</i>] color</b></b>
&ndash; use system color editor to interactively change the color of all
circles for hits in the same cluster (or if coloring by species, from
the same species) as the clicked one
<li><b>Color by cluster</b> (default)
<li><b>Color by species</b>
&ndash; different random colors for different source species
<li><b>Show table row for [<i>hit</i>]</b>
&ndash; scroll the table of results to the corresponding row 
<li><b>Select rows for cluster [<i>hit</i>]</b>
&ndash; <a href="#chosen">choose</a> table rows for all hits in the same 
cluster (or if coloring by species, from the same species) as the clicked one,
and report their structure IDs in the <a href="log.html"><b>Log</b></a>
<li><b>Log structure names for cluster [<i>hit</i>]</b>
&ndash; report structure IDs in the <a href="log.html"><b>Log</b></a>
for all hits in the same cluster as the clicked one
<li><b>Show reference atoms</b>
&ndash; display the reference &alpha;-carbons as spheres
and <a href="../selection.html">select</a> them
<li><b>Select reference atoms</b>
&ndash; <a href="../selection.html">select</a> the reference &alpha;-carbons
</ul>

<a name="ligands"></a>
<p class="nav">
[<a href="#top">back to top: Foldseek</a>]
</p>
<h3>Ligands</h3>
<p>
Clicking the <b>Ligands</b> button copies the
ligands, ions, and solvent molecules (nonpolymer residues) from the hits
onto corresponding locations on the query structure,
for either all hits or just the <a href="#chosen">chosen</a> ones,
as per the <a href="#options">options</a>. A dialog will appear
to ask the user whether the structures should be fetched,
since it may take several minutes to do so.
</p><p>
Each ligand (ion, solvent) residue is evaluated for mapping onto the query 
structure, as follows:
</p>
<ol>
<li>protein residues within 5 &Aring; of the ligand are identified
<li>if at least half of those nearby protein residues are paired with query
residues, the &alpha;-carbons of those pairs are fitted
<li>if the resulting RMSD is &le;3 &Aring;, the ligand is copied to 
corresponding position relative to the query structure
</ol>
<p>
Different parameters can be applied with the
<a href="../commands/similarstructures.html"><b>similarstructures
ligands</b></a> command.
</p><p>
How many residues were copied and their residue types
are reported in the <a href="log.html"><b>Log</b></a>.
Often thousands of water molecules, and ions, and crystallization adjuvants
are found, and they can be hidden to get a better view of more interesting 
ligands. For example, commands: 
</p>
<blockquote>
<b><a href="../commands/show.html">hide</a> solvent</b>
<br>
<b><a href="../commands/show.html">hide</a> ions</b>
<br>
<b><a href="../commands/show.html">hide</a> :SO4</b>
</blockquote>
<p>
By default, the copied ligand, ion, and solvent residues are loaded as a 
single atomic model, in which the chain ID of a residue is generated from the
PDB ID and chain ID of its source structure (<i>e.g.</i>, <b>2cml_B</b>).
Pausing the cursor over a residue in the graphics window shows
its name and chain ID in a pop-up balloon.
</p><p>
See also: <a href="https://alphafill.eu/" target="_blank">AlphaFill</a>
</p>

<a name="notes"></a>
<p class="nav">
[<a href="#top">back to top: Foldseek</a>]
</p>
<h3>References</h3>
<p>
<b>Foldseek</b>.
The <a href="https://github.com/steineggerlab/foldseek" 
target="_blank">Foldseek</a> method is described in:
</p>
<blockquote>
<a href="https://www.nature.com/articles/s41587-023-01773-0" target="_blank">
Fast and accurate protein structure search with Foldseek.</a>
van Kempen M, Kim SS, Tumescheit C, Mirdita M, Lee J, Gilchrist CLM, SÃ¶ding J, Steinegger M.
<i>Nat Biotechnol.</i> 2024 Feb;42(2):243-246.
</blockquote>
<p>
<b>Many-against-Many sequence searching (MMseqs2)</b>.
The <a href="https://github.com/soedinglab/MMseqs2" 
target="_blank">MMseqs2</a> method is described in:
<blockquote>
<a href="https://www.ncbi.nlm.nih.gov/pubmed/29035372" target="_blank">
MMseqs2 enables sensitive protein sequence searching for the analysis 
of massive data sets.</a>
Steinegger M, S&ouml;ding J.
<i>Nat Biotechnol.</i> 2017 Nov;35(11):1026-1028.
<br><br>
<a href="https://www.ncbi.nlm.nih.gov/pubmed/30615063" target="_blank">
MMseqs2 desktop and local web server app for fast, interactive sequence 
searches.</a>
Mirdita M, Steinegger M, S&ouml;ding J.
<i>Bioinformatics.</i> 2019 Aug 15;35(16):2856-2858.
</blockquote>
<b>Local distance difference test</i> (LDDT)</b>:
<blockquote>
<a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC3799472/" target="_blank">
lDDT: a local superposition-free score for comparing protein structures and models using distance difference tests.</a>
Mariani V, Biasini M, Barbato A, Schwede T.
<i>Bioinformatics.</i> 2013 Nov 1;29(21):2722-8.
</blockquote>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics /
November 2025</address>
</body></html>
