<html>

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>Tool: Sequence Viewer</title>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRight" title="User Guide Index"/></a>

<table class="clRight">
<tr><td>
<img class="outline" src="1bzm-others-screenshot.png"
alt="1bzm, 1urt, sequence alignment" width="630px"></td></tr>
<tr><td width="620px" align="center">
<font size=-1>
Models #1 <font style="background:tan;padding:0px 2px">1bzm</font>
and #2 <font style="background:skyblue;padding:0px 2px">1urt</font>
associated with sequences in an alignment, with
<font style="background:lightgreen;padding:0px 2px">ChimeraX selection</font>
sequence region.
</font>
</td></tr>
</table>

<h3><a href="../index.html#tools">Tool</a>: Sequence Viewer</h3>
<p>
The <b>Sequence Viewer</b> displays individual sequences and 
sequence alignments of amino acids and nucleotides, with crosstalk to any
<a href="#association">associated</a> structures.
To <a href="../findseq.html">find and select</a>
a specified sequence or sequence pattern, use
<a href="../menu.html#selectseq"><b>Select... Sequence</b></a> in the menu
or the <a href="../commands/select.html#new"><b>select</b></a> command.
See also:
<a href="../commands/modeller.html"><b>modeller</b></a>,
<a href="https://www.rbvi.ucsf.edu/chimerax/data/conservation-coloring/conservation-coloring.html"
target="_blank">Coloring by Sequence Conservation</a>
</p><p>
Sequences for biopolymer structures in ChimeraX can be shown:
</p>
<ul>
<li>with the menu: <b>Tools... Sequence... Show Sequence Viewer</b>
and choosing which chain sequence(s) to show
<li>with the
<a href="../commands/sequence.html#chain"><b>sequence chain</b></a> command
<li>by clicking the 
<a href="../tools/moldisplay.html"><b>Molecule Display</b></a> icon
<a href="../tools/moldisplay.html" title="Molecule Display Toolbar...">
<img class="iconbar" src="../tools/shortcut-icons/sequence.png"></a>
<li>by clicking the <a href="../modelinfo.html">chain description</a> 
link that appears in the <a href="log.html"><b>Log</b></a> 
when a structure is opened 
</ul>
<p>
Independent of structure, sequence alignments and individual sequences 
can also be <a href="../commands/open.html#local">opened</a> 
from <a href="../commands/open.html#sequence">files</a>
or <a href="../commands/open.html#fetch">fetched</a>
from <a href="https://www.uniprot.org/" target="_blank">UniProt</a>.
Other tools or commands may generate new sequence alignments
(<i>e.g.</i>, <a href="blastprotein.html#results"><b>Blast Protein</b> 
results</a>, <a href="matchmaker.html"><b>Matchmaker</b></a>,
<a href="../commands/sequence.html#align">sequence realignment</a>).
</p><p>
Multiple <b>Sequence Viewer</b> windows can be shown at the same time, 
and they can be manipulated like other panels in ChimeraX 
(<a href="../window.html#panels">more...</a>).
</p>

<blockquote><b>
<a href="#context">Context Menu</a>
<br><a href="#headers">Headers</a>
<br><a href="#regions">Regions</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;<a href="#uniprot">UniProt Sequence Features</a>
<br><a href="#association">Sequence-Structure Association</a>
<br><a href="#settings">Settings</a>
</b></blockquote>

<a name="context"></a>
<p class="nav">
[<a href="#top">back to top: Sequence Viewer</a>]
</p>
<h3>Sequence Viewer Context Menu</h3>
<p>
Each <b>Sequence Viewer</b> window has a 
<a href="../window.html#context-menu">context menu</a> 
with choices including:
<p>
<ul>
<li><b>File</b>
  <ul>
  <li><b>Save As</b> &nbsp;<font size="-1" color="dimgray">&#9654;</font>
  [various <a href="../commands/save.html#seqformats">sequence file formats</a>]
  &ndash; save to file; 
  see also <a href="../commands/save.html#sequence"><b>save</b></a>
  <li><b>Load Sequence Coloring File...</b> open a
  <a href="../formats/scf.html">sequence coloring file</a>
  to create <a href="#regions">regions</a> and color 
  <a href="#association">associated</a> structure to match
  </ul>
<li><b>Edit</b>
  <ul>
  <li><b>Realign Sequences with Clustal Omega</b>
(<a href="../commands/sequence.html#clustal">details...</a>)
  <li><b>Realign Sequences with MUSCLE</b>
(<a href="../commands/sequence.html#muscle">details...</a>)
  </ul>
<li><b>Structure</b>
  <ul>
  <li><b>Associations...</b> show dialog for controlling 
  <a href="#association">sequence-structure association</a>
  </ul>
<li><b>Headers</b> 
&ndash; which <a href="#headers">headers</a> are shown currently
  <ul>
  <li><b>Consensus</b>
  <li><b>Conservation</b> 
  <li><b>RMSD</b> (<b>C&alpha;</b> or <b>backbone</b> as per
   the <a href="#settings-headers"><b>Headers</b> settings</a>)
  <li><b>Save</b> &nbsp;<font size="-1" color="dimgray">&#9654;</font>
  [choice of available headers]
  &ndash; save header information to file
  </ul>
<li><b>Tools</b>
  <ul>
  <li><b>Modeller Comparative Modeling...</b> 
start <a href="modeller.html"><b>Modeller Comparative Modeling</b></a>
(requires a structure to be <a href="#association">associated</a> 
with any sequence in the alignment, for use as a template in modeling)
  <li><b>Blast Protein...</b> 
run <a href="blastprotein.html"><b>Blast Protein</b></a>
with default parameters on the specified sequence
  </ul>
<li><b>Sequence Features...</b> for sequences
<a href="../commands/open.html#fetch">fetched</a> from
<a href="https://www.uniprot.org" target="_blank">UniProt</a>,
show the <a href="#uniprot">features dialog</a>
<li><b>Settings...</b> show the <a href="#settings"><b>Sequence Viewer</b>
preferences</a> to control wrapping, spacing, coloring, <i>etc.</i>
</ul>

<a name="headers"></a>
<p class="nav">
[<a href="#top">back to top: Sequence Viewer</a>]
</p>
<h3>Headers</h3>
<p>
Sequence <b><i>headers</i></b> are rows of information above the sequences,
such as a <b>Consensus</b> sequence, <b>Conservation</b> histogram, 
or histogram of the <b>RMSD</b> (root-mean-square displacement)
among multiple <a href="#association">associated</a> structures.
Which headers are shown initially and how they are calculated 
can be controlled in the
<a href="#settings-headers"><b>Headers</b> settings</a>.
Headers for a specific window can be shown, hidden, or saved to a file 
with the <a href="#context"><b>Sequence Viewer</b> context menu</a> or the
<a href="../commands/sequence.html#header"><b>sequence header</b></a> command.
</p><p>
Per-column header values are assigned as 
<a href="../attributes.html">attributes</a> of the structure residues
<a href="#association">associated</a> with sequences in the alignment.
The numerical attributes (those from headers displayed as histograms)
are named &ldquo;seq_<i>name</i>,&ldquo;
where <i>name</i> is the name of the header, for example,
<a href="../attributes.html#residue">seq_conservation</a> or
<a href="../attributes.html#residue">seq_rmsd</a>.
The structures can be
<a href="../commands/color.html#byattribute">colored</a>
to show the values of such numerical attributes, as in the tutorial:
<a href="https://www.rbvi.ucsf.edu/chimerax/data/conservation-coloring/conservation-coloring.html"
target="_blank">Coloring by Sequence Conservation</a>.
</p>

<a name="regions"></a>
<p class="nav">
[<a href="#top">back to top: Sequence Viewer</a>]
</p>
<h3>Regions (Colored Boxes)</h3>
<p>
Sequence <b><i>regions</i></b> are colored boxes or outlines
that enclose one or more residue symbols.
A single region may contain any number of disjoint and/or abutting boxes.
Pausing the cursor over a region (but not directly over a residue symbol)
shows its name in a pop-up balloon.
</p><p>
Some regions are created automatically, such as by 
<a href="#association">association</a> with a structure.
Regions can also be created manually by dragging within the sequence window.
Manually created regions are shown in
<font style="background:rgb(233,218,198);padding:1px 3px">this&nbsp;color</font>
(default, see <a href="#settings-regions"><b>Regions</b> settings</a>).
</p><p>
<a name="active"></a>
The <b><i>active region</i></b> is the region most recently clicked or created
by dragging, and is indicated with a dashed outline.
<b>Shift</b>-dragging with the left mouse button <i>adds to</i> 
the active region, whereas <b>Ctrl</b>-dragging <i>creates</i> a new region
and makes it the active region. (On Mac, command key replaces Ctrl.)
Backspace/delete deletes the active region. Clicking the active region
deactivates it, and clicking a different region deactivates the
former active region and makes the new region active.
A region with no interior color is only responsive to clicks on its borders.
<!--
Where regions <a href="#overlap">overlap</a>,
only the highest is responsive to clicks.
-->
</p>

<table class="clRight">
<tr><td><img class="outline" src="uniprot-features.png"
alt="cah1_human sequence features" width="400px"></td></tr>
</table>
<a name="uniprot"></a>
<p><a href="#regions" class="nounder"><b>&larr;</b></a>
<b>UniProt Sequence Features</b>
</p><p>
Sequences <a href="../commands/open.html#fetch">fetched</a> from
<a href="https://www.uniprot.org" target="_blank">UniProt</a> may include
several annotations. These &ldquo;sequence features&rdquo; are automatically
assigned as <a href="#regions">regions</a> in the sequence window.
The features are hidden by default, but a dialog for browsing and 
displaying them appears when the sequence is first opened. This dialog
can be reshown at any time by choosing <b>Sequence Features...</b>
from the <a href="#context">context menu</a>.
</p><p>
Choosing a feature type on the left side of the dialog shows the regions
for all features of that type; if multiple features of that type are
listed on the right, clicking (choosing) any one shows just
the region for the individual feature. 
When the dialog has mouse focus, starting to type the name of a 
feature type instantly chooses that entry.
The up and down arrow keys can also be used to navigate the dialog.
Feature listings may include links to evidence code descriptions in the
<a href="https://evidenceontology.org/" target="_blank">Evidence & Conclusion
Ontology</a> and/or sequence variant entries in 
<a href="https://www.ncbi.nlm.nih.gov/snp/" target="_blank">dbSNP</a>.  
</p><p>
The border and interior colors of the currently shown feature region(s)
can be changed by clicking the color wells and using the system color editor,
and checkboxes allow leaving either the border or the interior uncolored.
The initially assigned colors may vary between different instances 
of opening the same data.
</p><p>
If a structure is <a href="#association">associated</a> with the sequence, 
the corresponding structure residues are listed in the lower right area 
of the dialog. 
For example, the figure shows all <b>metal ion-binding site</b> regions in
UniProt sequence <a href="https://www.uniprot.org/uniprot/P00915"
target="_blank">cah1_human</a>, corresponding to residues
64,67,94,96,119, and 200 in chain A of the associated structure, PDB
<a href="http://www.rcsb.org/structure/1BZM" target="_blank">1bzm</a>.
A checkbox option allows automatically 
<a href="../selection.html">selecting</a> the associated structure residues
(if any) as features are chosen.
</p>

<a name="association"></a>
<p class="nav">
[<a href="#top">back to top: Sequence Viewer</a>]
</p>
<h3>Sequence-Structure Association</h3>
<p>
A sequence that arose from a structure's chain information
is automatically associated with that chain,
with <a href="#regions">regions</a> for protein
<font style="background:rgb(255,255,204);
border:solid goldenrod 1px;padding:1px 3px">
<b>structure&nbsp;helices</b></font> and
<font style="background:lightcyan;
border:solid rgb(168,191,191) 1px;padding:1px 3px">
<b>structure&nbsp;strands</b></font>.
Such sequences are closed when their structure chains are deleted or closed,
and they are not automatically associated with other structure chains.
</p><p>
When sequences and structures are opened independently,
a structure chain will be associated automatically with a sequence 
if their sequences can be aligned without mismatches 
exceeding 1/10 of the residues in the structure chain.
The association and number of mismatches will be reported
in the <a href="log.html"><b>Log</b></a>. For automatic association,
<ul>
<li>Gaps in the structure sequence 
(relative to the sequence opened independently)
are not allowed except where residues are missing from the structure,
such as in disordered regions with poorly defined density.
<li>Insertions in the structure sequence relative to the 
sequence opened independently are not allowed.
</ul>
<p>
For manual association, both of the above are allowed, 
but only the &ldquo;extra&rdquo; residues in the structure not associated 
with any positions in the sequence (minus the length difference
if the structure sequence is longer) are counted along with mismatches 
and reported in the <a href="log.html"><b>Log</b></a>.
</p><p>
Associations can be controlled manually (forced or removed):
</p>
<ul>
<li>with the
<a href="../commands/sequence.html#associate"><b>sequence</b></a> command
<li>by choosing <b>Associations...</b> from the 
<a href="#context">context menu</a>.
The resulting dialog lists structure chains and allows changing 
the association of each to none, best-matching, or a specific sequence.
</ul>
<p>
In the sequence window, <a href="#regions">regions</a> may be created to show
<font style="background:rgb(255,76,76);padding:1px 3px">
<b>mismatches</b></font> and
<font style="border:solid black 1px;padding:1px 3px">
<b>missing&nbsp;residues</b></font>
according to the association.
If a mismatch or missing residue only occurs in a subset of multiple
associated chains, a 
<font style="background:pink;padding:1px 3px">
pink box</font> or 
<font style="border:solid rgb(190,190,190) 1px;padding:1px 3px">
gray outline</font> is shown instead
(default, see <a href="#settings-regions"><b>Regions</b> settings</a>).
</p><p>
Further effects of association:
</p>
<ul>
<li>a color swatch matching the associated model is shown around the 
sequence name, or a 
<font style="border:dashed darkgreen 1px;padding:1px 3px">
dark-green dashed outline</font> if chains from multiple models are associated
<li>pausing the cursor over the sequence name reports the association, and 
over residues in the sequence window reports the corresponding structure residue(s)
<li>clicking or dragging to update the <a href="#active">active region</a> will
<a href="../selection.html">select</a> the corresponding structure residues
<li><a href="../selection.html">selections</a> (by any means) of the structure
are shown as the <a href="#regions">region</a> named
<font style="background:lightgreen;padding:1px 3px">
<b>ChimeraX&nbsp;selection</b></font>
(default, see <a href="#settings-regions"><b>Regions</b> settings</a>)
</ul>

<a name="settings"></a>
<p class="nav">
[<a href="#top">back to top: Sequence Viewer</a>]
</p>
<h3>Sequence Viewer Settings (Preferences)</h3>
<p>
Choosing <b>Settings...</b> from the 
<a href="#context"><b>Sequence Viewer</b> context menu</a>
shows its settings in a separate window, with sections:
</p>
<ul>
<li><a href="#settings-appearance"><b>Appearance</b></a>
<li><a href="#settings-headers"><b>Headers</b></a>
<li><a href="#settings-regions"><b>Regions</b></a>
</ul>
<p>
The settings window can be manipulated like other panels in the
ChimeraX interface (<a href="../window.html#panels">more...</a>).
</p><p>
<b>Save</b> saves the current settings as
<a href="../preferences.html#other">preferences</a>,
<b>Reset</b> replaces the current settings with 
the initial &ldquo;factory&rdquo; defaults (values shown in <b>bold</b> below),
and <b>Restore</b> restores values that were saved previously.
The <b>Buttons below...</b> option indicates
whether these buttons should apply only
to the currently shown section (<i>e.g.</i>, <b>Appearance</b>)
or to all of the <b>Sequence Viewer</b> preferences.
Although there can be multiple settings windows with different
values for multiple <b>Sequence Viewer</b> windows,
there can be only one set of saved preference values.
</p>
<a name="settings-appearance"></a>
<p><a href="#settings" class="nounder"><b>&larr;</b></a>
<b>Appearance</b> &ndash; how text is displayed in the 
<b>Sequence Viewer</b> window
</p>
<ul class="none">
<li><b>Single Sequence</b>
  <ul class="none">
  <li><b>Horizontal spacing (pixels)</b> [<i>S</i>]
  &ndash; space between characters within a row
  (initial default <b>&minus;2</b>)
  <li><b>Wrap</b> &ndash; whether to wrap lines within the window (<b>on</b>)
  <li><b>Wrap lines at multiple of</b> [<i>N</i>]
  &ndash; when wrapping, allow breaks at multiples of <i>N</i> residues
  (<b>5</b>)
  <li><b>Vertically separate wrapped lines</b> (<b>off</b>)
  </ul>
<li><b>Alignment</b>
  <ul class="none">
  <li><b>Horizontal spacing (pixels)</b> [<i>S</i>]
  &ndash; space between characters within a row (initial default <b>0</b>)
  <li><b>Wrap if</b> [<i>M</i>] <b>sequences or fewer</b>
  &ndash; do not wrap if the alignment contains 
  &gt;<i>M</i> sequences (<b>8</b>); set to &le;1 for no wrapping
  <li><b>Wrap lines at multiple of</b> [<i>N</i>]
  &ndash; when wrapping, allow breaks at multiples of 
  <i>N</i> alignment positions (<b>5</b>)
  <li><b>Vertically separate wrapped blocks</b> (<b>on</b>)
  <li><b>Show numbering at startup</b> 
  &ndash; whether to show alignment numbering (1, 11, <i>etc.</i>)
  for subsequently opened alignments (<b>on</b>)
  </ul>
</ul>

<a name="settings-headers"></a>
<p><a href="#settings" class="nounder"><b>&larr;</b></a>
<b>Headers</b> &ndash; rows of information above the sequences in an alignment
(<a href="#headers">more...</a>)
</p>
<ul class="none">
<li><b>Consensus</b>
  <ul>
  <li><b>Show initially</b> (initial default <b>off</b>)
  <li><b>Capitalization threshold</b> (initial default <b>0.8</b>)
  <li><b>Ignore gap characters</b> (initial default <b>off</b>)
  </ul>
The consensus sequence shows the residue type that is most common
at each position in the alignment, even if it appears in less than 0.5
of the sequences; if a consensus type appears in at least the 
<b>capitalization threshold</b> fraction of sequences, 
it will be capitalized and shown in purple,
or in red if all of the sequences have the same type.
If several residue types are equally the most common,
one is chosen at random. The consensus will show a gap at positions 
where a gap is the most prevalent or equally the most prevalent
as one or more residue types, unless <b>ignore gap characters</b>
is turned on. Regardless of this option, however,
gap positions are always included in the denominator
for calculating the consensus fraction.
</p>
<li><b>Conservation</b>
  <ul>
  <li><b>Show initially</b> (initial default <b>on</b>)
  <li><b>Style</b>
    <ul>
    <li><b>identity histogram</b> &ndash; what proportion of the sequences
have the most prevalent nongap character per alignment column.
Histogram bar heights are proportional to (M-1)/(N-1), where
M is the number of occurrences of the most prevalent residue type at
a position in the alignment and N is the number of sequences in the
alignment. Thus, if every sequence has a different residue at a
given position, the bar height is zero, not 1/N.
However, M/N values (not the values used for display purposes) are
used for the <a href="#headers">seq_conservation attribute</a>.
    <li><b>Clustal characters</b> &ndash; as in Clustal format, 
     <ul class="none"> 
     <li>* (asterisk) for complete identity
     <li>: (colon) for 
     <a href="../formats/clustgroups.html">strong group conservation</a> 
     <li>. (period) for 
     <a href="../formats/clustgroups.html">weak group conservation</a>
     <li> blank otherwise
     </ul>
    <li><b>AL2CO</b> (<b>AL</b>ignment <b>2</b> <b>CO</b>nservation,
  initial default) &ndash; more sophisticated options
  for calculating conservation. Users should cite:
  <a name="al2coref"></a>
    <blockquote>
    <a href="https://www.ncbi.nlm.nih.gov/pubmed/11524371" target="_blank">
    AL2CO: calculation of positional conservation in a
    protein sequence alignment.</a>
    Pei J, Grishin NV.
    <i>Bioinformatics.</i> 2001 Aug;17(8):700-12.
   </blockquote>
  Conservation values from AL2CO are in standard deviations from the mean 
  (Z-scores) and can range from &ndash;&infin; (least conserved) to
  +&infin; (most conserved). For purposes of histogram display only,
  the values are mapped to bar heights 0-1, with conservation values
  of zero giving a histogram bar height of 0.5.
</p><p>
<a name="al2co-parameters"></a>
<b>AL2CO parameters</b>:
</p>
<ul>
<li><b>Averaging window</b> (initial default <b>1</b>)
&ndash; window width, or
how many alignment positions to average in a sliding window.
The result is assigned to the central column of the window
when the width is odd, the column left of center when the width is even
(<i>e.g.</i>, the fourth position in a window of 8),
with various adjustments at the termini.
The authors of AL2CO recommend a width of 3 for motif analyses.
<li><a name="al2comeas"><b>Conservation measure</b></a>
&ndash; which AL2CO conservation equation to use:
  <ul>
  <li><b>entropy-based</b> (initial default)
  <li><b>variance-based</b>
  <li><b>sum of pairs</b>
  </ul>
<li><b>Frequency estimation method</b> 
&ndash; whether/how to weight the sequences
  <ul>
  <li><b>unweighted</b>
  <li><b>modified Henikoff & Henikoff</b>
  <li><b>independent counts</b> (initial default)
  </ul>
<li><a name="gapfract"><b>Gap fraction</b></a> (initial default <b>0.5</b>)
 &ndash; maximum fraction of sequences with gaps in a column
for conservation to still be calculated; for example, if the
<b>gap fraction</b> is 0.25, conservation values will only be calculated
for columns in which at least three-fourths of the sequences have residues
</ul><br>
Additional settings if the AL2CO
<a href="#al2comeas"><b>Conservation measure</b></a> is <b>sum of pairs</b>:
<ul>
<li><b>Matrix</b> [choices: BLOSUM-30, BLOSUM-35, BLOSUM-40, BLOSUM-45, 
BLOSUM-50, BLOSUM-55, BLOSUM-60, <b>BLOSUM-62</b> (initial default), 
BLOSUM-65, BLOSUM-70, BLOSUM-75, BLOSUM-80, BLOSUM-85, BLOSUM-90, 
BLOSUM-100, BLOSUM-N, Grantham, HSDM, identity, Nucleic, PAM-40, PAM-120, 
PAM-150, PAM-250, SDM] &ndash; which residue similarity matrix to use
<li><b>Matrix transformation</b> 
&ndash; whether/how to modify the matrix
   <ul>
   <li><b>none</b> (initial default)
   <li><b>normalization</b>
   &nbsp;&nbsp;&nbsp; <i>S'<sub>ab</sub></i> = <i>S<sub>ab</sub></i> / (<i>S<sub>aa</sub>S<sub>bb</sub></i>)<sup>&frac12;</sup>
   <li><b>adjustment</b>
   &nbsp;&nbsp;&nbsp; <i>S''<sub>ab</sub></i> = 2<i>S<sub>ab</sub></i> &ndash; &frac12;(<i>S<sub>aa</sub></i> + <i>S<sub>bb</sub></i>)
   </ul>
</ul>
</ul>
  </ul>
<br>
<li><b>RMSD</b> &ndash; per-column root-mean-square displacement among 
<a href="#association">associated</a> structure residues
  <ul>
  <li><b>Show initially</b> (initial default <b>off</b>)
  <li><b>Atoms used for RMSD</b>
    <ul>
    <li><b>C&alpha;</b> (initial default)
    <li><b>backbone</b>
    </ul>
  </ul>
<p>
When more than one chain per model is <a href="#association">associated</a>
with the same alignment, the RMSD
calculation uses only the chain that gives the lowest overall RMSD.
</p>
</ul>

<a name="settings-regions"></a>
<p><a href="#settings" class="nounder"><b>&larr;</b></a>
<b>Regions</b>
&ndash; creation and display of <a href="#regions">regions</a> (colored boxes)
</p><p>
The <b>Regions</b> section of <a href="#settings">settings</a> controls
whether different types of <a href="#regions">regions</a> should be shown,
and if so, their initial colors.
Each color well can be clicked to choose a color interactively.
On/off checkboxes indicate whether to color the region 
<b>border</b> (box outline) and/or <b>interior</b> (box fill),
whereas <b>full</b> and <b>partial</b> refer to whether the
region applies to all structures <a href="#association">associated</a>
with a position or to only some fraction of those structures.
</p>
<ul class="none">
<li><b>New-region border</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="#BEBEBE" width="10px"></td>
</tr></table>
(initial default <b>off</b>, no outline)
<li><b>New-region interior</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="#E9DAC6" width="10px"></td>
</tr></table>
(<b>on</b>)
<li><b>Show selection region</b> (<b>on</b>)
<li><b>Selected-structure border</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="#BEBEBE" width="10px"></td>
</tr></table>
(<b>off</b>, no outline)
<li><b>Selected-structure interior</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="#90EE90" width="10px"></td>
</tr></table>
(<b>on</b>)
<li><b>Show missing-structure regions</b> (<b>on</b>)
<li><b>Missing-structure border &nbsp;full</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="black" width="10px"></td>
</tr></table>
&nbsp;<b>partial</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="#BEBEBE" width="10px"></td>
</tr></table>
(both <b>on</b>)
<li><b>Missing-structure interior &nbsp;full</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="#BEBEBE" width="10px"></td>
</tr></table>
&nbsp;<b>partial</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="#BEBEBE" width="10px"></td>
</tr></table>
(both <b>off</b>)
<li><b>Show structure-mismatch regions</b> (<b>on</b>)
<li><b>Structure-mismatch border &nbsp;full</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="#BEBEBE" width="10px"></td>
</tr></table>
&nbsp;<b>partial</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="#BEBEBE" width="10px"></td>
</tr></table>
(both <b>off</b>)
<li><b>Structure-mismatch interior &nbsp;full</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="#FF4C4C" width="10px"></td>
</tr></table>
&nbsp;<b>partial</b>
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor="pink" width="10px"></td>
</tr></table>
(both <b>on</b>)
</ul>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics / 
February 2022</address>
</body></html>
