<html>

<!--
=== UCSF ChimeraX Copyright ===
Copyright 2016 Regents of the University of California.
All rights reserved.  This software provided pursuant to a
license agreement containing restrictions on its disclosure,
duplication and use.  For details see:
http://www.rbvi.ucsf.edu/chimerax/docs/licensing.html
This notice must be embedded in or attached to all copies,
including partial copies, of the software or any revisions
or derivations thereof.
=== UCSF ChimeraX Copyright ===
-->

<head>
<link rel="stylesheet" type="text/css" href="../userdocs.css" />
<title>ChimeraX Tutorial: Protein-Ligand Binding Sites</title>
<script type="text/javascript" src="../../js/cxlinks.js"></script>
</head><body>

<a name="top"></a>
<a href="../index.html">
<img width="60px" src="../ChimeraX-docs-icon.svg" alt="ChimeraX docs icon"
class="clRight" title="User Guide Index"/></a>

<h3><a href="https://www.rbvi.ucsf.edu/chimerax/tutorials.html"
target="_blank">ChimeraX Tutorial</a>: Protein-Ligand Binding Sites</h3>

<table class="clRight" cellpadding="0px">
<tr><td>
<a href="#hbonds">
<img class="outline" src="bind2.png" alt="H-bonds" width="100px"></a>
</td></tr>
<tr><td>
<a href="#mlp">
<img class="outline" src="bind3.png" alt="lipophilicity" width="100px"></a>
</td></tr>
<tr><td>
<a href="#esp">
<img class="outline" src="bind6.png" alt="electrostatics" width="100px"></a>
</td></tr>
</table>

<ul>
<!-- Kinase and inhibitor -->
<li><a href="#opening"><b>Opening a Structure and Surveying Its Parts</b></a>
<li><a href="#manipulation"><b>Manipulation</b></a>
<img class="iconbar" src="../tools/mouse-icons/rotate.png">
<img class="iconbar" src="../tools/mouse-icons/translate.png">
<img class="iconbar" src="../tools/mouse-icons/zoom.png">
<img class="iconbar" src="../tools/mouse-icons/select.png">
<li><a href="#display"><b>Atomic Display Styles</b></a>
<img class="iconbar" src="../tools/shortcut-icons/stick.png">
<img class="iconbar" src="../tools/shortcut-icons/sphere.png">
<img class="iconbar" src="../tools/shortcut-icons/ball.png">
<li><a href="#hbonds"><b>H-Bonds and Contacts</b></a>
<li><a href="#resist"><b>Resistance Mutations and Sequence Viewer</b></a>
<!-- Enzyme with hydrophobic substrate; membrane protein(s) -->
<li><a href="#mlp"><b>Surface Coloring by Lipophilicity</b></a>
<img class="iconbar" src="../tools/shortcut-icons/hydrophobicity.png">
  <ul>
  <li><a href="#cap-example"><b>Coloring a Clipped Surface</b></a>
  </ul>
<!-- Bacterial gene regulator with DNA and small-molecule binding sites -->
<li><a href="#esp"><b>Surface Coloring by Electrostatic Potential</b></a>
<img class="iconbar" src="../tools/shortcut-icons/esp.png">
  <ul>
  <li><a href="#key"><b>Color Key and 2D Label</b></a>
  </ul>
</ul>

<p>
This tutorial explores some ways to examine the complementarity
between proteins and their ligands, including drug molecules.
Related tutorials:
<a href="/chimerax/data/conservation-coloring/conservation-coloring.html"
target="_blank">Coloring by Sequence Conservation</a>,
<a href="/chimerax/data/delta-opioid-jan2020/binding.html"
target="_blank">Delta-Opioid Receptor</a>,
<a href="https://rbvi.github.io/chimera-tutorials/presentations/modules/chimerax-comp-structures/index.html#/"
target="_blank">Structure Analysis and Comparison</a>,
<a href="https://www.rbvi.ucsf.edu/chimerax/data/electrostatics-apr2021/crm1.html" target="_blank">Coloring CRM1 by Electrostatic Potential</a>
</p><p>
** ChimeraX 1.2 (production release May 2021) or newer should be used 
to follow this tutorial. **
</p>

<a name="opening"></a>
<h3><a href="#top" class="nounder">&larr;</a>
Opening a Structure and Surveying Its Parts</h3>
<p>
Start ChimeraX.
If you want to use the click-to-execute links in this tutorial, view this
page in the <a href="../tools/helpviewer.html"><b>ChimeraX Browser</b></a>.
One way to do so is by entering the ChimeraX command:
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:open help:user/tutorials/binding-sites.html">open help:user/tutorials/binding-sites.html</a></b>
</blockquote>
All of the click-to-execute links in this
tutorial are indented and appear after &ldquo;<i>Command</i>:&rdquo;
Other links are simply informative and can be clicked or not, as you prefer.
</p><p>
ChimeraX can fetch a structure directly from the
<a href="https://www.rcsb.org" target="_blank">Protein Data Bank</a> (PDB)
given its 4-character ID. Open
<a href="https://www.rcsb.org/structure/2HYY" target="_blank">2hyy</a>,
a structure of the human Abl kinase domain in complex with the anticancer drug 
<a href="https://en.wikipedia.org/wiki/Imatinib"
target="_blank">imatinib</a>:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:open 2hyy">open 2hyy</a></b>
</blockquote>
<p>
The <a href="../tools/modelpanel.html"><b>Model Panel</b></a>
(lower right side of the ChimeraX window) lists 2hyy as model 1, and the 
&nbsp;<img class="iconbar" src="../tools/ui-icons/shown.png" title="eye icon" 
alt="eye icon">&nbsp; checkbox indicates that the model is displayed.
</p><p>
The <a href="../tools/log.html"><b>Log</b></a> (right side of window)
shows several kinds of information associated with the structure:
</p>
<blockquote>
<p>
In the <b>Chain information</b> table, try clicking the chain IDs 
(<b>A</b>, <i>etc.</i>); see that the corresponding
part of the structure becomes <a href="../selection.html">selected</a>
in the graphics window, as indicated with green outlines.  
The structure includes four copies of the kinase as chains A, B, C, and D.
</p>
<blockquote style="background-color:#f0f0f0; margin:15px; 
border:solid black 1px; padding:5px">
Note:
<a href="../selection.html">Selections</a>
are used to limit the scope of using the <b>Actions</b> menu
or clicking certain icons.
The &nbsp;<img class="iconbar" src="../tools/ui-icons/select.png"
title="hand icon" alt="hand icon"> checkbox
in the <a href="../tools/modelpanel.html"><b>Model Panel</b></a>
shows when a model is selected, with three possible states: 
checked, unchecked, and partially checked (shown with a dash on Mac).
</blockquote>
<p>
In the <b>Non-standard residues</b> table,
click the residue name <b>STI</b> (imatinib)
to see where it is in the structure.
In this context, &ldquo;nonstandard&rdquo; means residues other than water
and the types of amino acids and nucleic acids typically found in biomolecules. 
Imatinib is a kinase inhibitor, and each copy of the kinase
has its own imatinib binding site.
</p><p>
In the <b>2hyy mmCIF Assemblies</b> table, click <b>1</b> to generate
&ldquo;assembly 1,&rdquo; a single copy of the protein-ligand complex.
As can be seen in the <a href="../tools/modelpanel.html"><b>Model Panel</b></a>,
this opens the assembly as model 2 and hides the original structure (model 1,
<img class="iconbar" src="../tools/ui-icons/shown.png"
title="eye icon" alt="eye icon">&nbsp; box unchecked).
Close the original structure:
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:close #1">close #1</a></b>
</blockquote>
<blockquote style="background-color:#f0f0f0; margin:15px; 
border:solid black 1px; padding:5px">
Note:
In this case, the assembly has fewer chains than the original structure.
For assemblies with more than 12 copies of the original structure,
clicking a link in the <b>Assemblies</b> table adds copies as graphical clones
for efficient visualization.
However, if the plan is to analyze atomic interactions between the copies
or to display them differently from one another,
the assembly should be generated instead as full atomic copies with the
<a href="../commands/sym.html#assembly"><b>sym</b></a> command and
<b>copies true</b>, for example:
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<b>sym #1 assembly 1 copies true</b>
<br>
Alternatively, entire assemblies can be fetched directly from the PDB with the
<a href="../commands/open.html#fetch"><b>open</b></a> command, 
as shown in the section on 
<a href="#esp"><b>Surface Coloring by Electrostatic Potential</b></a>.
</blockquote>
</blockquote>
<p>
If the model information becomes inconvenient to access by scrolling,
or gets removed by clearing the <a href="../tools/log.html"><b>Log</b></a>, 
it can be re-shown with commands (<a href="../modelinfo.html">details...</a>).
</p>

<a name="manipulation"></a>
<h3><a href="#top" class="nounder">&larr;</a>
Manipulation</h3>
<p>
Try moving the structure around. Default mouse/trackpad assignments:
</p>
<table border cellpadding="4" cellspacing="0"><tr>
<td align="center" rowspan="2">rotate
<br><img src="../tools/mouse-icons/rotate.png" class="icon3"></td>
<td>left mouse button</td>
</tr><tr>
<td class="text">trackpad click-drag</td>
</tr><tr>
<td align="center" rowspan="2">translate in 2D
<br><img src="../tools/mouse-icons/translate.png" class="icon3"></td>
<td>middle or right mouse button</td>
</tr><tr>
<td class="text">
trackpad + <b>Alt</b>
</td>
</tr><tr>
<td colspan="2" align="center">
+ <b>Shift</b> switches from rotate to translate and <i>vice versa</i>
</td>
</tr><tr>
<td align="center" rowspan="2">zoom
<br><img src="../tools/mouse-icons/zoom.png" class="icon3"></td>
<td>mouse scroll wheel</td>
</tr><tr>
<td class="text">
trackpad 2-finger drag (except pinch on Mac as per
<a href="../preferences.html#trackpad">Trackpad preferences</a>)</td>
</tr><tr>
<td align="center" rowspan="2"><a href="../selection.html">select</a>
<br><img src="../tools/mouse-icons/select.png" class="icon3"></td>
<td>left mouse button + <b>Ctrl</b>, + <b>Shift</b> to toggle
or add to selection</td>
</tr><tr>
<td class="text">
trackpad + <b>Ctrl</b>, + <b>Shift</b> to toggle
or add to selection</td>
</td>
</tr></table>
<p>
To list the current assignments in the <b>Log</b>:
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:mousemode">mousemode</a></b>
</blockquote>
<p>
If anything is <a href="../selection.html">selected</a>,
Ctrl-click in &ldquo;empty space&rdquo; in the graphics window
(away from any atoms) to clear the selection.
Alternatively, you can use the menu: <b>Select... Clear</b> or the command:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:select clear">select clear</a></b>
</blockquote>

<a name="display"></a>
<h3><a href="#top" class="nounder">&larr;</a>
Atomic Display Styles</h3>
<p>
The initial display has cartoons (ribbons) for the protein, 
with most atoms hidden.
Only the nonstandard residue and nearby sidechains are shown in atomic detail.
</p><p>
Many actions can be done in several different ways.
One way to change the display is with presets. 
For example, all atoms as sticks can be shown with menu:
<b>Presets... Sticks</b> or the command:
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:preset sticks">preset sticks</a></b>
</blockquote>
<p>
...and the original appearance or at least something similar
(in this case, the color is different)
can be restored with menu: <b>Presets... Original Look</b> or the command:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:preset original">preset original</a></b>
</blockquote>
<p>
Of course, you can change the style of any subset of the atoms.
Atoms to change can be
<a href="../commands/atomspec.html">specified</a> by chain ID, 
residue number, residue name, atom name, <i>etc.</i>, or by 
<a href="../selection.html">selection</a> 
(indicated as <b>sel</b> in the command line).
If nothing is specified, all atoms are changed, even though they might not
be displayed at the time.
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:style ball">style ball</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:style :STI sphere">style :STI sphere</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:style stick">style stick</a></b>
</blockquote>
<p>
The style icons <img class="iconbar" src="../tools/shortcut-icons/stick.png">
<img class="iconbar" src="../tools/shortcut-icons/sphere.png">
<img class="iconbar" src="../tools/shortcut-icons/ball.png">
in the <a href="../tools/moldisplay.html"><b>Molecule Display</b></a> tab
of the <a href="../tools/toolbar.html"><b>Toolbar</b></a>
apply to the currently <a href="../selection.html">selected</a> atoms, or
if none are selected, all atoms.
</p>
<table class="clRight">
<tr><td>
<img class="outline" src="bind1.png" alt="imatinib binding site" width="300px">
</td></tr>
</table>
<p>
Make imatinib more obvious by changing the color of its carbon atoms:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:color :sti & C salmon">
color :sti & C salmon</a></b>
</blockquote>
<p>
Capitalization of residue names in the command line is optional, 
but element symbols always need to be capitalized.
To list color names in the <b>Log</b>:
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:color list">
color list</a></b>
</blockquote>
<p>
(A more organized
<a href="../commands/colornames.html#builtin"
target="_blank">table</a> is also available.)
</p>

<a name="hbonds"></a>
<h3><a href="#top" class="nounder">&larr;</a>
H-Bonds and Contacts</h3>
<p>
To find only the H-bonds involving imatinib,
Ctrl-click to select any atom in that residue, then
press the keyboard up arrow to promote the selection to the whole residue.
If the initial selection is a bond rather than an atom, 
the up-arrow key will need to be pressed twice.
If you overshoot and select more than the drug molecule, 
simply press the keyboard down arrow to go back.
</p><p>
Alternatively, the imatinib residue could be selected
with the menu: <b>Select... Residues... STI</b> or a command:
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:select :sti">select :sti</a></b>
</blockquote>
<p>
(You could also scroll back up in the <b>Log</b> to the
<b>Non-standard residues</b> table and click its name there.)
Focus the view on the selection and hide all other atoms:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:view sel">view sel</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:hide ~sel">hide ~sel</a></b>
</blockquote>
<p>
Use the <b><a href="../tools/hbonds.html">H-Bonds</a></b> tool
(menu: <b>Tools... Structure Analysis... H-Bonds</b>). 
Keep the default settings, except turn on (check)
the following if they are not already:
<ul>
<li><b>Limit by selection: with at least one end selected</b>
&ndash; only find H-bonds involving the
currently <a href="../selection.html">selected</a> atoms
<li><b>Reveal atoms of H-bonding residues</b> &ndash;
the &ldquo;pseudobonds&rdquo; representing the H-bonds are only
displayed when both of their endpoint atoms are displayed; 
this option displays atoms as needed to show all of the H-bonds
<li><b>Log</b> &ndash; show information about each H-bond in
the <a href="../tools/log.html"><b>Log</b></a>
</ul>
Make sure that the imatinib residue is still selected, then run the calculation
by clicking <b>Apply</b> or <b>OK</b> (the difference between those two
is that <b>OK</b> closes the dialog).
</p>
<table class="clRight">
<tr><td>
<img class="outline" src="bind2.png" alt="imatinib H-bonds" width="300px">
</td></tr>
</table>
<p>
The <b>Log</b> lists five H-bonds involving imatinib.
If the wrong settings were used by mistake, no worries: just change the
settings and run the calculation again.
The same calculation could be done with the following command:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:hbonds :sti reveal true log true">
hbonds :sti reveal true log true</a></b>
</blockquote>
<p>
The structure does not include hydrogens, but their inferred positions
were used in the calculation.
</p><p>
The H-bonds are shown in the graphics window as dashed lines. 
Some of the dashed lines just go to the ribbon. By default, ribbon display
suppresses the display of backbone atoms. Enable showing backbone atoms at the 
same time as the ribbon with command:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:cartoon suppress false">cartoon suppress false</a></b>
</blockquote>
<p>
Now you can see two backbone nitrogen atoms donating H-bonds to imatinib,
and a backbone carbonyl oxygen accepting an H-bond from imatinib.
The backbone atoms can still be shown and hidden individually,
and most backbone atoms are still hidden.
The backbone atoms of these specific residues were displayed 
by the <b>Reveal atoms...</b> option in the H-bonds calculation.
</p><p>
Use the <b><a href="../tools/clashes.html">Contacts</a></b> tool
(menu: <b>Tools... Structure Analysis... Contacts</b>).
Keep the default settings, except turn on (check)
the following if they are not already:
<ul>
<li><b>Limit by selection: with at least one end selected</b>
<li><b>Reveal atoms of interacting residues</b>
<li><b>Log</b>
</ul>
...and turn off (uncheck) <b>Display as pseudobonds</b>,
Make sure that the imatinib residue
is selected as described <a href="#hbonds">above</a>, 
then click <b>Apply</b> or <b>OK</b>.
</p><p>
The <b>Log</b> lists 87 atomic contacts involving imatinib.
The same calculation could be done with the following command:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:contacts :sti makePseudobonds false reveal true log true">contacts :sti makePseudobonds false reveal true log true</a></b>
</blockquote>
<p>
Label the residues with displayed atoms:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:label @@display">label @@display</a></b>
</blockquote>
<p>
A couple of the displayed residues are water (residue name HOH).
Hide water and clear the selection:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:hide solvent">hide solvent</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:select clear">select clear</a></b>
</blockquote>

<a name="resist"></a>
<h3><a href="#top" class="nounder">&larr;</a>
Resistance Mutations and Sequence Viewer</h3>
<p>
Several mutations in kinase domain of the Bcr-Abl fusion protein
have been implicated in cancer resistance to imatinib treatment.
According to <a href="https://www.nature.com/articles/s41598-019-38672-x"
target="_blank">Chandrasekhar <i>et al.</i></a> Sci Rep <b>9</b>:2412 (2019),
these include 
G250E, Y253H, V256G, L301I, T315I, F317L, Y320H, M351T, E373D, and D381N.
</p><p>
Coloring the carbon atoms and residue labels at these 10 positions
shows that five of them are already displayed due to their interactions 
with imatinib:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:name resist :250,253,256,301,315,317,320,351,373,381">name resist :250,253,256,301,315,317,320,351,373,381</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:color resist & C gold target al">color resist & C gold target al</a></b>
</blockquote>
<p>
The five residues 253, 256, 315, 317, and 381 are among those listed in the 
<b>Log</b> as forming H-bonds and/or contact interactions with imatinib.
Atoms at the other five positions implicated in resistance to this drug
are not yet displayed. To see them all:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:show resist">show resist</a></b>
</blockquote>
<p>
The newly displayed residues have gold carbons but are not labeled.
One way to make the 10 residues more obvious is to change their style,
for example: 
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:style resist ball">style resist ball</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:view resist clip false">view resist clip false</a></b>
</blockquote>
<p>
The latter command zooms out to make the view contain all of the specified
residues and turns off clipping. Clipping had been turned on by the use of
<a href="../commands/view.html#initial"><b>view</b></a>
in the <a href="#hbonds">previous section</a>.
If you like, make the labels bigger (default height 0.7 &Aring;):
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:label height 1.0">label height 1.0</a></b>
</blockquote>
<p>
Residues not directly involved in binding might contribute to drug resistance
by changing in the conformation, flexibility, or stability of the protein.
</p><p>
Show the protein sequence in the 
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a>:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:seq chain /A">seq chain /A</a></b>
</blockquote>
<p>
If the sequence window seems vertically scrunched, you can click and drag 
its top margin upwards to increase its size within the overall window.
</p><p>
A selection in the 3D structure is shown as green boxes in the sequence,
for example:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:select resist">select resist</a></b>
</blockquote>
<p>
Conversely, you can change the 3D structure selection
by clicking an existing box or dragging a new box in the
<a href="../tools/sequenceviewer.html"><b>Sequence Viewer</b></a> window.
The other colored boxes indicate information from the structure:
<span class="nowrap"><font style="background:rgb(255,255,204);
border:solid goldenrod 1px;padding:1px 3px">
<b>structure&nbsp;helices</b></font></span>,
<span class="nowrap"><font style="background:lightcyan;
border:solid rgb(168,191,191) 1px;padding:1px 3px">
<b>structure&nbsp;strands</b></font></span>,
<span class="nowrap"><font style="border:solid black 1px;padding:1px 3px">
<b>missing&nbsp;from structure</b></font></span>.
</p><p>
When done with this structure, close the session (both structure and sequence)
and clear the <b>Log</b>:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:close session">close session</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:log clear">log clear</a></b>
</blockquote>

<a name="mlp"></a>
<h3><a href="#top" class="nounder">&larr;</a>
Surface Coloring by Lipophilicity</h3>
<p>
Coloring a protein surface by lipophilicity (hydrophobicity) may help 
to identify interaction sites for nonpolar molecules or the lipid membrane.
</p><p>
Open <a href="https://www.rcsb.org/structure/1FPP" target="_blank">1fpp</a>,
a structure of the protein farnesyltransferase in complex with 
farnesyl diphosphate (residue name FPP):
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:open 1fpp">open 1fpp</a></b>
</blockquote>
<p>
From the interactive tables of information in the <b>Log</b>, 
you can see that the structure is a heterodimer
(two different protein chains) bound to a single molecule of FPP.
Remember that clicking a residue name in the <b>Non-standard residues</b>
table selects the corresponding residue(s). Clear the selection by 
Ctrl-clicking in empty space or with the command:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:select clear">select clear</a></b>
</blockquote>
<p>
A &ldquo;molecular lipophilicity potential&rdquo; (MLP) is calculated using
atomic coordinates, atomic lipophilicity parameters, and a distance-dependent
function. ChimeraX can calculate MLP maps for proteins and color their 
surfaces by the values. This can be done with the command:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:mlp">mlp</a></b>
</blockquote>
<table class="clRight">
<tr><td>
<img class="outline" src="bind3.png" alt="hydrophobic pocket" width="300px">
</td></tr>
</table>
<p>
...or by clicking the
<img class="iconbar" src="../tools/shortcut-icons/hydrophobicity.png"> icon
in the <a href="../tools/moldisplay.html"><b>Molecule Display</b></a> tab
of the <a href="../tools/toolbar.html"><b>Toolbar</b></a>.
The default coloring is from <b>dark cyan</b> for most hydrophilic 
through <b>white</b> to <b>dark goldenrod</b> for most hydrophobic:
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor=darkcyan width="10px" title="darkcyan"></td>
<td bgcolor=white width="10px" title="white"></td>
<td bgcolor=darkgoldenrod width="10px" title="darkgoldenrod"></td>
</tr></table>.
Different options for the functional form of the calculation and the coloring 
scheme are available with the <a href="../commands/mlp.html"><b>mlp</b></a>
command.
</p><p>
If you rotate the structure, you may be able to see that FPP
occupies a deep pocket that is mostly hydrophobic.
The pocket is so deep, however, it may be difficult to see the coloring.
A better view is obtained by limiting the surface to a zone around FPP
and hiding ribbons and atoms other than FPP:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:surface zone #1 near :fpp dist 8 max 1">surface zone #1 near :fpp dist 8 max 1</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:hide ~:fpp target ar">hide ~:fpp target ar</a></b>
</blockquote>
<p>
The pocket surface can now be viewed from the protein side as well as the
ligand side.
</p><p>
In the surface command, <b>max 1</b> limits the display to the largest 
surface component, hiding any disconnected bits. Both protein chains
contribute to the pocket. Using the disclosure triangle for <b>1fpp</b>
in the <a href="../tools/modelpanel.html"><b>Model Panel</b></a>
(lower right side of the ChimeraX window) shows that there are 
separate models for the surfaces of chain A and chain B.
You can hide/show them individually with the checkboxes under the eye icon
&nbsp;<img class="iconbar" src="../tools/ui-icons/shown.png" title="eye icon" 
alt="eye icon">.
</p><p>
When finished viewing this pocket, close the structure:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:close">close</a></b>
</blockquote>
<p>
Coloring by lipophilicity is also used to identify or
highlight membrane-embedded regions.
Optionally, try using the icon 
<img class="iconbar" src="../tools/shortcut-icons/hydrophobicity.png">
or the command <a href="../commands/mlp.html"><b>mlp</b></a> 
on a transmembrane protein (many examples:
<a href="https://www.rcsb.org/structure/1PHO" target="_blank">1pho</a>,
<a href="https://www.rcsb.org/structure/4XT3" target="_blank">4xt3</a>,
<a href="https://www.rcsb.org/structure/1EYS" target="_blank">1eys</a>,
<i>etc.</i>). A &ldquo;belt&rdquo; of hydrophobicity may be visible 
where a protein sits in the membrane. For example:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:open 4xt3">open 4xt3</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:mlp">mlp</a></b>
</blockquote>
<table class="clRight">
<tr><th>4xt3: MLP vs. Kyte-Doolittle coloring</th></tr>
<tr><td>
<img class="outline" src="bind4.png" alt="4xt3 MLP vs. kdHydrophobicity" 
width="250px">
</td></tr>
</table>
<p>
An alternative to calculating the MLP is to use a constant lookup value for 
each type of amino acid residue. 
A ChimeraX command script to assign Kyte-Doolittle hydrophobicity values
and color the protein surface from <b>dodger blue</b> for the most
hydrophilic through <b>white</b> to <b>orange red</b> for the most hydrophobic
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor=dodgerblue width="10px" title="dodgerblue"></td>
<td bgcolor=white width="10px" title="white"></td>
<td bgcolor=orangered width="10px" title="orangered"></td>
</tr></table>
is provided here:
<a href="https://www.rbvi.ucsf.edu/chimerax/docs/user/kyte-doolittle_hydrophobicity.cxc">kyte-doolittle_hydrophobicity.cxc</a>. 
Saving the file as plain text and opening it in ChimeraX 
will execute the commands on any proteins being viewed at the time.
The script can also be used to define a 
<a href="../preferences.html#startup">custom preset</a> in ChimeraX
similar to the preset that was built into Chimera.
(Of course, a different coloring scheme could be used, including the
same as used by <a href="../commands/mlp.html"><b>mlp</b></a>.)
</p><p>
Coloring by a lookup value for each residue type often gives a similar 
impression to MLP coloring, but can be quite different in specific cases
because all types of amino acids contain both polar and nonpolar atoms.
In other words, even the most hydrophobic residues include some 
hydrophilic atoms, and <i>vice versa</i>.
</p><p>
If any structures are open, close them:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:close">close</a></b>
</blockquote>

<a name="cap-example"></a>
<h3><a href="#top" class="nounder">&larr;</a>
Coloring a Clipped Surface</h3>
<table class="clRight">
<tr><td>
<img class="outline" src="mlpclip.png" alt="clipped surface colored by MLP" 
width="220px">
</td></tr>
</table>
<p>
The protein in structure <a href="https://www.rcsb.org/structure/1G74"
target="_blank">1g74</a> has an oleic acid residue <b>OLA</b> in an
interior pocket. 
The following commands open the structure (model #1), show its
<a href="../commands/surface.html">molecular surface</a> (model #1.1),
and turn on <a href="../commands/clip.html">clipping</a> to slice the structure
(creating a planar surface cap, model #1.1.1):
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:open 1g74">open 1g74</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:surface">surface</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:clip">clip</a></b>
</blockquote>
<p>
The model hierarchy can be viewed in the 
<a href="../tools/modelpanel.html"><b>Model Panel</b></a>.
</p><p>
Next, turn off clipping of the atomic model so that only the surface is clipped
(<a href="../commands/clip.html#model">per-model clipping</a>),
hide ribbons and atoms other than oleic acid, and color 
the carbon atoms in oleic acid green:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:clip model #!1 off">clip model #!1 off</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:hide ~:ola targ ar">hide ~:ola targ ar</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:color :ola & C green">color :ola & C green</a></b>
</blockquote>
<p>
The clipping plane can be translated and rotated
interactively <a href="../commands/ui.html#mousemode">with the mouse</a>
<a href="../tools/mousemodes.html" title="Right Mouse Toolbar...">
<img class="icon" border=1 src="../tools/mouse-icons/clip.png"></a>
<a href="../tools/mousemodes.html" title="Right Mouse Toolbar...">
<img class="icon" border=1 src="../tools/mouse-icons/cliprot.png"></a>.
</p><p>
Color the molecular surface by
<a href="../commands/mlp.html">molecular lipophilicity potential</a> (MLP).
The planar surface cap is not automatically colored 
along with the outer surface, but can be colored with an additional command
in which the atoms to use (here, all
<a href="../commands/atomspec.html#builtin"><b>protein</b></a> atoms)
are specified separately from the surface to color:
</p>
<blockquote> 
<i>Command</i>:&nbsp;<b><a href="cxcmd:mlp">mlp</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:mlp protein surface #1.1.1">mlp protein surface #1.1.1</a></b>
</blockquote>
<p>
The cap can be made transparent with another command, for example:
</p>
<blockquote> 
<i>Command</i>:&nbsp;<b><a href="cxcmd:transparency #1.1.1 30">transparency #1.1.1 30</a></b>
</blockquote>
Analogously, to color the surface cap by 
<a href="../commands/coulombic.html">Coulombic electrostatic potential</a>
(more about electrostatic potential in the <a href="#esp">next section</a>): 
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:coulombic protein">coulombic protein</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:coulombic protein surf #1.1.1 offset 0">coulombic protein surf #1.1.1 offset 0</a></b>
</blockquote>
<p>
The Coulombic cap coloring would need to be reapplied if the clipping plane 
were subsequently moved, however (unless you are using ChimeraX 1.3 from
5/16/2021 or newer).
</p><p>
When finished with this structure, close the session:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:close session">close session</a></b>
</blockquote>

<a name="esp"></a>
<h3><a href="#top" class="nounder">&larr;</a>
Surface Coloring by Electrostatic Potential</h3>
<p>
Coloring a protein surface by electrostatic potential (ESP)
may help to identify binding sites for charged or polar molecules.
Regions of positive potential are complementary to negative charges,
whereas regions of negative potential complement positive charges.
</p><p>
ESP is calculated from atomic coordinates and partial charges.
ChimeraX can calculate a simple 
<a href="../commands/coulombic.html">Coulombic ESP</a> on the fly,
or it can read in an ESP map file calculated with another program, such as APBS.
This tutorial uses the first (simple) method, as it does not require
a separate calculation.
</p><p>
Open an assembly of
<a href="https://www.rcsb.org/structure/3D6Y" target="_blank">3d6y</a>,
the bacterial gene regulator BmrR:
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:open 3d6y from rcsb_bio maxAssemblies 1">open 3d6y from rcsb_bio maxAssemblies 1</a></b>
</blockquote>
<table class="clRight">
<tr><td>
<img class="outline" src="bind5.png" alt="3d6y assembly" width="300px">
</td></tr>
</table>
<p>
BmrR regulates multidrug resistance. In this structure, it is bound to DNA 
and the positively charged alkaloid
<a href="https://en.wikipedia.org/wiki/Berberine"
target="_blank">berberine</a>. The <b>Non-standard residues</b> table 
in the <b>Log</b> lists berberine and glycerol; click the residue names
<b>BER</b> and <b>GOL</b> to see where they are in the structure.
Hide glycerol, water, and protein atoms: 
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:hide :gol">hide :gol</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:hide solvent">hide solvent</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:hide protein">hide protein</a></b>
</blockquote>
<p>
To avoid any confusion from a hidden selection, clear it by Ctrl-clicking
in empty space or using the command:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:select clear">select clear</a></b>
</blockquote>
<p>
The following shows surfaces for the protein chains
and colors them by Coulombic ESP calculated on the fly:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:coulombic protein">coulombic protein</a></b>
</blockquote>
<p>
The default color scheme is red for negative potential, white at zero,
and blue for positive:
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr>
<td bgcolor=red width="10px" title="red"></td>
<td bgcolor=white width="10px" title="white"></td>
<td bgcolor=blue width="10px" title="blue"></td>
</tr></table>.
There are positive patches for binding the negatively charged DNA, 
and a deep pocket of negative potential for binding 
the positively charged small molecule berberine.
</p><p>
The <a href="../tools/moldisplay.html"><b>Molecule Display</b></a> tab
of the <a href="../tools/toolbar.html"><b>Toolbar</b></a>
includes an <img class="iconbar" src="../tools/shortcut-icons/esp.png">
icon for Coulombic coloring,
but it will also apply to the DNA unless protein (only) is 
<a href="../selection.html">selected</a> before the icon is clicked.
</p><p>
The remaining commands in this section are just examples of display settings
that might be useful for figures. 
</p><p>
The nucleotides are color-coded by base type:
<table border cellpadding="4" cellspacing="0"
style="display:inline-block; vertical-align:middle"><tr style="line-height:8px">
<td style="background:#ff4040;color:white">A</td>
<td style="background:#4040ff;color:white">T</td>
<td bgcolor="#40ff40">G</td>
<td bgcolor="#ffff40">C</td>
</tr></table>.
Make them a single color instead, so as not to distract from the surface 
coloring. You can also change the DNA ribbon to a tube shape
and show nucleotide sidechains as stubs (broken ladder rungs): 
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:color nucleic tan">color nucleic tan</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:cartoon style nucleic xsection round width 1.5 thick 1.5">cartoon style nucleic xsection round width 1.5 thick 1.5</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:nuc stubs">nuc stubs</a></b>
</blockquote>
<p>
The nucleotide change to stubs can also be done by clicking
<img class="iconbar" src="../tools/nuc-icons/nuc-stubs.png">
in the <a href="../tools/nucdisplay.html"><b>Nucleotides</b></a> tab
of the <a href="../tools/toolbar.html"><b>Toolbar</b></a>.
</p><p>
Make the berberine stand out 
with fatter sticks and bright green carbons:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:size :ber stickrad .7">size :ber stickrad .7</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:color C & :ber limegreen">color C & :ber limegreen</a></b>
</blockquote>
<p>
Use icons in the <a href="../tools/graphics.html"><b>Graphics</b></a> tab
of the <a href="../tools/toolbar.html"><b>Toolbar</b></a> to:
</p>
<ul>
<li>change the background color
<img class="iconbar" src="../tools/shortcut-icons/whitebg.png">
<img class="iconbar" src="../tools/shortcut-icons/graybg.png">
<img class="iconbar" src="../tools/shortcut-icons/blackbg.png">
<li>toggle silhouettes
<img class="iconbar" src="../tools/shortcut-icons/silhouette.png">
or shadows
<img class="iconbar" src="../tools/shortcut-icons/shadow.png">
<li>try different lighting modes
<img class="iconbar" src="../tools/shortcut-icons/simplelight.png">
<img class="iconbar" src="../tools/shortcut-icons/softlight.png">
<img class="iconbar" src="../tools/shortcut-icons/fulllight.png">
<img class="iconbar" src="../tools/shortcut-icons/flat.png">
</ul>
<table class="clRight">
<tr><td>
<img class="outline" src="bind6.png" alt="Coulombic ESP flat lighting" 
width="300px">
</td></tr>
</table>
<a name="key"></a>
<h4><a href="#esp" class="nounder">&larr;</a>
Color Key and 2D Label</h4>
<p>
When making a figure, it is usually best to compose the rest of the image
before adding any <a href="../commands/key.html">color key</a> and/or
<a href="../commands/2dlabels.html">2D labels</a>, since their
sizes and positions will need to be adjusted to fit the image layout.
</p><p>
A color key can be added with the 
<b>key</b> option of the coloring command, in this case,
<a href="../commands/coulombic.html"><b>coulombic</b></a>:
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:coulombic protein key true">coulombic protein key true</a></b>
</blockquote>
<p>
(If you already did the coloring, just repeat the same command except with
<b>key true</b> appended.)
This draws an initial color key, but also opens the
<a href="../tools/colorkey.html"><b>Color Key</b></a> tool and
enables redrawing and moving the key with the mouse 
(the <b>Adjust key...</b> option in the tool). 
The <b>Adjust key...</b> option can be unchecked
to restore the mouse to moving the structure instead of the key.
Closing the tool also restores the mouse function to moving the structure.
</p><p>
When the color key is in place, 
a text label can be added with a command, for example:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:2dlab text 'electrostatic potential'">2dlab text 'electrostatic potential'</a></b>
</blockquote>
<p>
...and then dragged interactively to the desired location
with the <b>move label</b> mouse mode:
in the <a href="../tools/mousemodes.html"><b>Right Mouse</b></a> tab
of the <a href="../tools/toolbar.html"><b>Toolbar</b></a>, click the icon 
<img class="icon" border=1 src="../tools/mouse-icons/movelabel.png">
and then use the right mouse button or trackpad + <b>Alt</b> on Windows,
trackpad + <img src="../icon-clover.png" width="12px"> on Mac to drag the label.
</p><p>
Although it is usually most convenient 
to position the key and label interactively as described above, 
and to set color-key colors and values
by using the <b>key</b> option of the coloring command, 
these could be done explicitly with commands instead. For example:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:2dlab text 'electrostatic potential' xpos .53 ypos .15">2dlab text 'electrostatic potential' xpos .53 ypos .15</a></b>
<br>
<i>Command</i>:&nbsp;<b><a href="cxcmd:key red:-10 white:0 blue:10 pos .53,.09 size .32,.04">key red:-10 white:0 blue:10 pos .51,.09 size .32,.04</a></b>
</blockquote>
<p>
If you now have multiple copies of the label, you can hide or delete the extras.
Each label is added as a submodel of the <b>2D labels</b> model. 
One way to show/hide/close them is by using the 
<a href="../tools/modelpanel.html"><b>Model Panel</b></a>; in that tool,
use the disclosure triangle next to a model's name to show its submodels.
</p><p>
To quit from ChimeraX, just close its main window or use a command:
</p>
<blockquote>
<i>Command</i>:&nbsp;<b><a href="cxcmd:exit">exit</a></b>
</blockquote>

<hr>
<address>UCSF Resource for Biocomputing, Visualization, and Informatics /
July 2021</address>
</body></html>
