--- ms/ntdll.mak.orig	2013-04-05 17:15:32.816631600 -0700
+++ ms/ntdll.mak	2013-04-05 17:25:04.585258300 -0700
@@ -1415,9 +1415,9 @@
 $(OBJ_D)\fips_ers.obj: $(SRC_D)\crypto\fips_ers.c
 	$(CC) /Fo$(OBJ_D)\fips_ers.obj  $(SHLIB_CFLAGS) -DOPENSSL_BUILD_SHLIBCRYPTO -c $(SRC_D)\crypto\fips_ers.c
 
-$(TMP_D)\x86_64cpuid.asm: crypto\x86_64cpuid.pl
+$(TMP_D)/x86_64cpuid.asm: crypto/x86_64cpuid.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\x86_64cpuid.pl $@
+	$(PERL) crypto/x86_64cpuid.pl $@
 
 $(OBJ_D)\x86_64cpuid.obj: $(TMP_D)\x86_64cpuid.asm
 	$(ASM) /Fo$@ $(TMP_D)\x86_64cpuid.asm
@@ -1434,9 +1434,9 @@
 $(OBJ_D)\md5_one.obj: $(SRC_D)\crypto\md5\md5_one.c
 	$(CC) /Fo$(OBJ_D)\md5_one.obj  $(SHLIB_CFLAGS) -DOPENSSL_BUILD_SHLIBCRYPTO -c $(SRC_D)\crypto\md5\md5_one.c
 
-$(TMP_D)\md5-x86_64.asm: crypto\md5\asm\md5-x86_64.pl
+$(TMP_D)/md5-x86_64.asm: crypto/md5/asm/md5-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\md5\asm\md5-x86_64.pl $@
+	$(PERL) crypto/md5/asm/md5-x86_64.pl $@
 
 $(OBJ_D)\md5-x86_64.obj: $(TMP_D)\md5-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\md5-x86_64.asm
@@ -1459,23 +1459,23 @@
 $(OBJ_D)\sha512.obj: $(SRC_D)\crypto\sha\sha512.c
 	$(CC) /Fo$(OBJ_D)\sha512.obj  $(SHLIB_CFLAGS) -DOPENSSL_BUILD_SHLIBCRYPTO -c $(SRC_D)\crypto\sha\sha512.c
 
-$(TMP_D)\sha1-x86_64.asm: crypto\sha\asm\sha1-x86_64.pl
+$(TMP_D)/sha1-x86_64.asm: crypto/sha/asm/sha1-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\sha\asm\sha1-x86_64.pl $@
+	$(PERL) crypto/sha/asm/sha1-x86_64.pl $@
 
 $(OBJ_D)\sha1-x86_64.obj: $(TMP_D)\sha1-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\sha1-x86_64.asm
 
-$(TMP_D)\sha256-x86_64.asm: crypto\sha\asm\sha512-x86_64.pl
+$(TMP_D)/sha256-x86_64.asm: crypto/sha/asm/sha512-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\sha\asm\sha512-x86_64.pl $@
+	$(PERL) crypto/sha/asm/sha512-x86_64.pl $@
 
 $(OBJ_D)\sha256-x86_64.obj: $(TMP_D)\sha256-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\sha256-x86_64.asm
 
-$(TMP_D)\sha512-x86_64.asm: crypto\sha\asm\sha512-x86_64.pl
+$(TMP_D)/sha512-x86_64.asm: crypto/sha/asm/sha512-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\sha\asm\sha512-x86_64.pl $@
+	$(PERL) crypto/sha/asm/sha512-x86_64.pl $@
 
 $(OBJ_D)\sha512-x86_64.obj: $(TMP_D)\sha512-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\sha512-x86_64.asm
@@ -1603,16 +1603,16 @@
 $(OBJ_D)\rc2ofb64.obj: $(SRC_D)\crypto\rc2\rc2ofb64.c
 	$(CC) /Fo$(OBJ_D)\rc2ofb64.obj  $(SHLIB_CFLAGS) -DOPENSSL_BUILD_SHLIBCRYPTO -c $(SRC_D)\crypto\rc2\rc2ofb64.c
 
-$(TMP_D)\rc4-x86_64.asm: crypto\rc4\asm\rc4-x86_64.pl
+$(TMP_D)/rc4-x86_64.asm: crypto/rc4/asm/rc4-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\rc4\asm\rc4-x86_64.pl $@
+	$(PERL) crypto/rc4/asm/rc4-x86_64.pl $@
 
-$(OBJ_D)\rc4-x86_64.obj: $(TMP_D)\rc4-x86_64.asm
+$(OBJ_D)/rc4-x86_64.obj: $(TMP_D)/rc4-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\rc4-x86_64.asm
 
-$(TMP_D)\rc4-md5-x86_64.asm: crypto\rc4\asm\rc4-md5-x86_64.pl
+$(TMP_D)/rc4-md5-x86_64.asm: crypto/rc4/asm/rc4-md5-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\rc4\asm\rc4-md5-x86_64.pl $@
+	$(PERL) crypto/rc4/asm/rc4-md5-x86_64.pl $@
 
 $(OBJ_D)\rc4-md5-x86_64.obj: $(TMP_D)\rc4-md5-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\rc4-md5-x86_64.asm
@@ -1686,37 +1686,37 @@
 $(OBJ_D)\aes_wrap.obj: $(SRC_D)\crypto\aes\aes_wrap.c
 	$(CC) /Fo$(OBJ_D)\aes_wrap.obj  $(SHLIB_CFLAGS) -DOPENSSL_BUILD_SHLIBCRYPTO -c $(SRC_D)\crypto\aes\aes_wrap.c
 
-$(TMP_D)\aes-x86_64.asm: crypto\aes\asm\aes-x86_64.pl
+$(TMP_D)/aes-x86_64.asm: crypto/aes/asm/aes-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\aes\asm\aes-x86_64.pl $@
+	$(PERL) crypto/aes/asm/aes-x86_64.pl $@
 
 $(OBJ_D)\aes-x86_64.obj: $(TMP_D)\aes-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\aes-x86_64.asm
 
-$(TMP_D)\vpaes-x86_64.asm: crypto\aes\asm\vpaes-x86_64.pl
+$(TMP_D)/vpaes-x86_64.asm: crypto/aes/asm/vpaes-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\aes\asm\vpaes-x86_64.pl $@
+	$(PERL) crypto/aes/asm/vpaes-x86_64.pl $@
 
 $(OBJ_D)\vpaes-x86_64.obj: $(TMP_D)\vpaes-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\vpaes-x86_64.asm
 
-$(TMP_D)\bsaes-x86_64.asm: crypto\aes\asm\bsaes-x86_64.pl
+$(TMP_D)/bsaes-x86_64.asm: crypto/aes/asm/bsaes-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\aes\asm\bsaes-x86_64.pl $@
+	$(PERL) crypto/aes/asm/bsaes-x86_64.pl $@
 
 $(OBJ_D)\bsaes-x86_64.obj: $(TMP_D)\bsaes-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\bsaes-x86_64.asm
 
-$(TMP_D)\aesni-x86_64.asm: crypto\aes\asm\aesni-x86_64.pl
+$(TMP_D)/aesni-x86_64.asm: crypto/aes/asm/aesni-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\aes\asm\aesni-x86_64.pl $@
+	$(PERL) crypto/aes/asm/aesni-x86_64.pl $@
 
 $(OBJ_D)\aesni-x86_64.obj: $(TMP_D)\aesni-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\aesni-x86_64.asm
 
-$(TMP_D)\aesni-sha1-x86_64.asm: crypto\aes\asm\aesni-sha1-x86_64.pl
+$(TMP_D)/aesni-sha1-x86_64.asm: crypto/aes/asm/aesni-sha1-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\aes\asm\aesni-sha1-x86_64.pl $@
+	$(PERL) crypto/aes/asm/aesni-sha1-x86_64.pl $@
 
 $(OBJ_D)\aesni-sha1-x86_64.obj: $(TMP_D)\aesni-sha1-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\aesni-sha1-x86_64.asm
@@ -1736,9 +1736,9 @@
 $(OBJ_D)\cmll_utl.obj: $(SRC_D)\crypto\camellia\cmll_utl.c
 	$(CC) /Fo$(OBJ_D)\cmll_utl.obj  $(SHLIB_CFLAGS) -DOPENSSL_BUILD_SHLIBCRYPTO -c $(SRC_D)\crypto\camellia\cmll_utl.c
 
-$(TMP_D)\cmll-x86_64.asm: crypto\camellia\asm\cmll-x86_64.pl
+$(TMP_D)/cmll-x86_64.asm: crypto/camellia/asm/cmll-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\camellia\asm\cmll-x86_64.pl $@
+	$(PERL) crypto/camellia/asm/cmll-x86_64.pl $@
 
 $(OBJ_D)\cmll-x86_64.obj: $(TMP_D)\cmll-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\cmll-x86_64.asm
@@ -1785,9 +1785,9 @@
 $(OBJ_D)\xts128.obj: $(SRC_D)\crypto\modes\xts128.c
 	$(CC) /Fo$(OBJ_D)\xts128.obj  $(SHLIB_CFLAGS) -DOPENSSL_BUILD_SHLIBCRYPTO -c $(SRC_D)\crypto\modes\xts128.c
 
-$(TMP_D)\ghash-x86_64.asm: crypto\modes\asm\ghash-x86_64.pl
+$(TMP_D)/ghash-x86_64.asm: crypto/modes/asm/ghash-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\modes\asm\ghash-x86_64.pl $@
+	$(PERL) crypto/modes/asm/ghash-x86_64.pl $@
 
 $(OBJ_D)\ghash-x86_64.obj: $(TMP_D)\ghash-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\ghash-x86_64.asm
@@ -1849,30 +1849,30 @@
 $(OBJ_D)\bn_asm.obj: $(SRC_D)\crypto\bn\bn_asm.c
 	$(CC) /Fo$(OBJ_D)\bn_asm.obj  $(SHLIB_CFLAGS) -DOPENSSL_BUILD_SHLIBCRYPTO -c $(SRC_D)\crypto\bn\bn_asm.c
 
-$(TMP_D)\x86_64-mont.asm: crypto\bn\asm\x86_64-mont.pl
+$(TMP_D)/x86_64-mont.asm: crypto/bn/asm/x86_64-mont.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\bn\asm\x86_64-mont.pl $@
+	$(PERL) crypto/bn/asm/x86_64-mont.pl $@
 
 $(OBJ_D)\x86_64-mont.obj: $(TMP_D)\x86_64-mont.asm
 	$(ASM) /Fo$@ $(TMP_D)\x86_64-mont.asm
 
-$(TMP_D)\x86_64-mont5.asm: crypto\bn\asm\x86_64-mont5.pl
+$(TMP_D)/x86_64-mont5.asm: crypto/bn/asm/x86_64-mont5.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\bn\asm\x86_64-mont5.pl $@
+	$(PERL) crypto/bn/asm/x86_64-mont5.pl $@
 
 $(OBJ_D)\x86_64-mont5.obj: $(TMP_D)\x86_64-mont5.asm
 	$(ASM) /Fo$@ $(TMP_D)\x86_64-mont5.asm
 
-$(TMP_D)\x86_64-gf2m.asm: crypto\bn\asm\x86_64-gf2m.pl
+$(TMP_D)/x86_64-gf2m.asm: crypto/bn/asm/x86_64-gf2m.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\bn\asm\x86_64-gf2m.pl $@
+	$(PERL) crypto/bn/asm/x86_64-gf2m.pl $@
 
 $(OBJ_D)\x86_64-gf2m.obj: $(TMP_D)\x86_64-gf2m.asm
 	$(ASM) /Fo$@ $(TMP_D)\x86_64-gf2m.asm
 
-$(TMP_D)\modexp512-x86_64.asm: crypto\bn\asm\modexp512-x86_64.pl
+$(TMP_D)/modexp512-x86_64.asm: crypto/bn/asm/modexp512-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\bn\asm\modexp512-x86_64.pl $@
+	$(PERL) crypto/bn/asm/modexp512-x86_64.pl $@
 
 $(OBJ_D)\modexp512-x86_64.obj: $(TMP_D)\modexp512-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\modexp512-x86_64.asm
@@ -3182,9 +3182,9 @@
 $(OBJ_D)\wp_dgst.obj: $(SRC_D)\crypto\whrlpool\wp_dgst.c
 	$(CC) /Fo$(OBJ_D)\wp_dgst.obj  $(SHLIB_CFLAGS) -DOPENSSL_BUILD_SHLIBCRYPTO -c $(SRC_D)\crypto\whrlpool\wp_dgst.c
 
-$(TMP_D)\wp-x86_64.asm: crypto\whrlpool\asm\wp-x86_64.pl
+$(TMP_D)/wp-x86_64.asm: crypto/whrlpool/asm/wp-x86_64.pl
 	set ASM=$(ASM)
-	$(PERL) crypto\whrlpool\asm\wp-x86_64.pl $@
+	$(PERL) crypto/whrlpool/asm/wp-x86_64.pl $@
 
 $(OBJ_D)\wp-x86_64.obj: $(TMP_D)\wp-x86_64.asm
 	$(ASM) /Fo$@ $(TMP_D)\wp-x86_64.asm
