diff -ru wxPython_Phoenix-3.0.1.dev75822/ext/wxWidgets/include/wx/glcanvas.h wxPython_Phoenix-3.0.1.dev75822.unchanged/ext/wxWidgets/include/wx/glcanvas.h
--- wxPython_Phoenix-3.0.1.dev75822/ext/wxWidgets/include/wx/glcanvas.h	2013-07-26 09:02:46.000000000 -0700
+++ wxPython_Phoenix-3.0.1.dev75822.unchanged/ext/wxWidgets/include/wx/glcanvas.h	2014-02-21 09:25:01.000000000 -0800
@@ -48,7 +48,13 @@
     WX_GL_MIN_ACCUM_BLUE,  // use blue buffer with most bits (> MIN_ACCUM_BLUE bits)
     WX_GL_MIN_ACCUM_ALPHA, // use alpha buffer with most bits (> MIN_ACCUM_ALPHA bits)
     WX_GL_SAMPLE_BUFFERS,  // 1 for multisampling support (antialiasing)
-    WX_GL_SAMPLES          // 4 for 2x2 antialiasing supersampling on most graphics cards
+    WX_GL_SAMPLES,         // 4 for 2x2 antialiasing supersampling on most graphics cards
+    WX_GL_OPENGL_PROFILE,  // opengl profile selector for Mac OS X
+};
+
+enum {
+    WX_GL_OPENGL_PROFILE_LEGACY,
+    WX_GL_OPENGL_PROFILE_3_2CORE,
 };
 
 #define wxGLCanvasName wxT("GLCanvas")
diff -ru wxPython_Phoenix-3.0.1.dev75822/ext/wxWidgets/interface/wx/glcanvas.h wxPython_Phoenix-3.0.1.dev75822.unchanged/ext/wxWidgets/interface/wx/glcanvas.h
--- wxPython_Phoenix-3.0.1.dev75822/ext/wxWidgets/interface/wx/glcanvas.h	2013-07-26 09:02:46.000000000 -0700
+++ wxPython_Phoenix-3.0.1.dev75822.unchanged/ext/wxWidgets/interface/wx/glcanvas.h	2014-02-21 09:25:10.000000000 -0800
@@ -137,8 +137,18 @@
     WX_GL_SAMPLE_BUFFERS,
 
     /// 4 for 2x2 antialiasing supersampling on most graphics cards
-    WX_GL_SAMPLES
+    WX_GL_SAMPLES,
 
+    /// OpenGL profile selector for Mac OS X
+    WX_GL_OPENGL_PROFILE,
+};
+
+enum {
+    /// Legacy OpenGL 1.x/2.x profile
+    WX_GL_OPENGL_PROFILE_LEGACY,
+
+    /// OpenGL 3.2+ profile (no legacy support)
+    WX_GL_OPENGL_PROFILE_3_2CORE,
 };
 
 /**
diff -ru wxPython_Phoenix-3.0.1.dev75822/ext/wxWidgets/src/osx/cocoa/glcanvas.mm wxPython_Phoenix-3.0.1.dev75822.unchanged/ext/wxWidgets/src/osx/cocoa/glcanvas.mm
--- wxPython_Phoenix-3.0.1.dev75822/ext/wxWidgets/src/osx/cocoa/glcanvas.mm	2013-09-25 23:31:20.000000000 -0700
+++ wxPython_Phoenix-3.0.1.dev75822.unchanged/ext/wxWidgets/src/osx/cocoa/glcanvas.mm	2014-02-21 09:25:30.000000000 -0800
@@ -213,6 +213,19 @@
                     data[p++] = NSOpenGLPFASamples;
                     data[p++] = (NSOpenGLPixelFormatAttribute) attribList[arg++];
                     break;
+
+                case WX_GL_OPENGL_PROFILE:
+                    data[p++] = NSOpenGLPFAOpenGLProfile;
+                    switch ( attribList[arg++] ) {
+                        case WX_GL_OPENGL_PROFILE_LEGACY:
+                        default:
+                            data[p++] = NSOpenGLProfileVersionLegacy;
+                            break;
+                        case WX_GL_OPENGL_PROFILE_3_2CORE:
+                            data[p++] = NSOpenGLProfileVersion3_2Core;
+                            break;
+                    }
+                    break;
             }
         }

*** wxPython_Phoenix-3.0.3.dev1719+62a2221/ext/wxWidgets/src/unix/glx11.cpp.orig	Mon Mar  2 05:39:45 2015
--- wxPython_Phoenix-3.0.3.dev1719+62a2221/ext/wxWidgets/src/unix/glx11.cpp	Wed Sep  2 17:41:08 2015
***************
*** 209,215 ****
                  return false;
  
              // indicates whether we have a boolean attribute
!             bool isBoolAttr = false;
  
              switch ( wxattrs[arg++] )
              {
--- 209,215 ----
                  return false;
  
              // indicates whether we have a boolean attribute
!             bool isBoolAttr = false, noAttr = false;
  
              switch ( wxattrs[arg++] )
              {
***************
*** 315,320 ****
--- 315,329 ----
  
                      return false;
  
+ 	        case WX_GL_OPENGL_PROFILE:
+ 		    if (wxattrs[arg++] == WX_GL_OPENGL_PROFILE_3_2CORE)
+ 		    {
+ 		      glattrs[p++] = GLX_CONTEXT_PROFILE_MASK_ARB;
+ 		      glattrs[p++] = GLX_CONTEXT_CORE_PROFILE_BIT_ARB;
+ 		    }
+ 		    noAttr = true;
+ 		    break;
+ 
                  default:
                      wxLogDebug(wxT("Unsupported OpenGL attribute %d"),
                                 wxattrs[arg - 1]);
***************
*** 328,334 ****
                  if ( GetGLXVersion() >= 13 )
                      glattrs[p++] = True;
              }
!             else // attribute with real (non-boolean) value
              {
                  // copy attribute value as is
                  glattrs[p++] = wxattrs[arg++];
--- 337,343 ----
                  if ( GetGLXVersion() >= 13 )
                      glattrs[p++] = True;
              }
!             else if ( !noAttr ) // attribute with real (non-boolean) value
              {
                  // copy attribute value as is
                  glattrs[p++] = wxattrs[arg++];
