*** wxPython_Phoenix-3.0.3.dev1719+62a2221/wx/lib/agw/aui/framemanager.py.orig	Tue Aug 25 12:40:03 2015
--- wxPython_Phoenix-3.0.3.dev1719+62a2221/wx/lib/agw/aui/framemanager.py	Wed Sep  2 15:23:17 2015
***************
*** 8476,8481 ****
--- 8476,8482 ----
  
          pane = self.GetPane(wnd)
          if not pane.IsOk():
+             return
              raise Exception("Pane window not found")
  
          if self.GetAGWFlags() & AUI_MGR_ALLOW_ACTIVE_PANE:
***************
*** 8494,8499 ****
--- 8495,8501 ----
  
          pane = self.GetPane(wnd)
          if not pane.IsOk():
+             return
              raise Exception("Pane window not found")
  
          if not pane.IsSnappable():
***************
*** 8655,8660 ****
--- 8657,8665 ----
          if not self._frame.GetSizer():
              return
  
+         if not self._frame.IsShown():
+             return
+ 
          mouse = wx.GetMouseState()
          mousePos = wx.Point(mouse.GetX(), mouse.GetY())
          point = self._frame.ScreenToClient(mousePos)
