--- wxPython_Phoenix-3.0.3.dev1839+4ecd949/wx/py/shell.py.orig	2016-03-15 17:13:48.000000000 -0700
+++ wxPython_Phoenix-3.0.3.dev1839+4ecd949/wx/py/shell.py	2016-03-15 20:22:53.000000000 -0700
@@ -750,10 +750,10 @@
         
         #unique (no duplicate words
         #oneliner from german python forum => unique list
-        unlist = [thlist[i] for i in xrange(len(thlist)) if thlist[i] not in thlist[:i]]
+        unlist = [thlist[i] for i in range(len(thlist)) if thlist[i] not in thlist[:i]]
             
         #sort lowercase
-        unlist.sort(lambda a, b: cmp(a.lower(), b.lower()))
+        unlist.sort(key=lambda s: s.lower())
         
         #this is more convenient, isn't it?
         self.AutoCompSetIgnoreCase(True)
