--- wxPython_Phoenix-3.0.1.dev75822/Phoenix/build.py.orig	2014-02-06 00:05:32.000000000 -0800
+++ wxPython_Phoenix-3.0.1.dev75822/Phoenix/build.py	2014-03-24 16:58:37.650278772 -0700
@@ -61,8 +61,8 @@
     'linux'  : 'b9fa64b1f6f5a6407777a0dda0de5778', 
 }
 
-wafCurrentVersion = '1.7.10'
-wafMD5 = '6825d465baf2968c1d20a58dd65a0e9d'
+wafCurrentVersion = '1.7.15'
+wafMD5 = '2ba0e10baf44db334e3baa39e59688db'
 
 doxygenCurrentVersion = '1.8.2'
 doxygenMD5 = {
