--- wxPython_Phoenix-3.0.3.dev1830+0b5f910/wx/py/introspect.py.orig	2015-08-11 08:40:03.000000000 -0700
+++ wxPython_Phoenix-3.0.3.dev1830+0b5f910/wx/py/introspect.py	2015-11-10 18:34:24.056707716 -0800
@@ -175,7 +175,7 @@
         pass
     elif inspect.isfunction(obj):
         # tip1 is a string like: "getCallTip(command='', locals=None)"
-        argspec = inspect.getargspec(obj)
+        argspec = inspect.getfullargspec(obj)
         argspec = inspect.formatargspec(*argspec)
         if dropSelf:
             # The first parameter to a method is a reference to an
